
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../git/git/">
      
      
        <link rel="next" href="../../var_detect_advanced/var_detect_advanced_background/">
      
      
        
      
      
      <link rel="icon" href="../../../img/favicon-16x16.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Workflow Translation With Janis - MBITE Tutorials</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<script async src="https://www.googletagmanager.com/gtag/js?id=G-EMNLJD0DWP"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }
    gtag('js', new Date());
    gtag('config', 'G-EMNLJD0DWP');
</script>

    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#janis-translate-migrating-cwl-galaxy-to-nextflow" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="MBITE Tutorials" class="md-header__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MBITE Tutorials
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Workflow Translation With Janis
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="MBITE Tutorials" class="md-nav__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    MBITE Tutorials
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Popular
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Popular
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Git for Researchers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/nectar-instances/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to the Nectar Cloud
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to R for Biologists
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/rna-seq-counts-to-genes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq: counts to genes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-proteomics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Proteomics
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Tools and skill development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Tools and skill development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/bioinfo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    What is Bioinformatics?
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/galaxy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to the Galaxy Platform
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galaxy-workflows/galaxy-workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Galaxy Workflows
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_overview/python_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Python
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Unix
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro_R_biologists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to R for Biologists
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Genome_browsers/GenomeBrowsers_Intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Genome Browsers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl_2022/cwl_2022/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to CWL
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq Experimental Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../docker/docker/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Containerized Bioinformatics
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../bionitio/bionitio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Command-line software development
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../git/git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Version Control with Git
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Workflow Translation With Janis
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Workflow Translation With Janis
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#topic" class="md-nav__link">
    <span class="md-ellipsis">
      
        Topic
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#skill-level" class="md-nav__link">
    <span class="md-ellipsis">
      
        Skill level
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#description" class="md-nav__link">
    <span class="md-ellipsis">
      
        Description
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      
        Learning Objectives
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#required-software" class="md-nav__link">
    <span class="md-ellipsis">
      
        Required Software
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#required-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Required Data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    <span class="md-ellipsis">
      
        Author Information
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contents" class="md-nav__link">
    <span class="md-ellipsis">
      
        Contents
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-11-samtools-flagstat-tool-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.1: Samtools Flagstat Tool (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.1: Samtools Flagstat Tool (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-janis-translate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-samtools-flagstat-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Samtools Flagstat as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-12-gatk-haplotypecaller-tool-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.2: GATK HaplotypeCaller Tool (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.2: GATK HaplotypeCaller Tool (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-gatk-haplotypecaller-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running GATK HaplotypeCaller as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-13-align-sort-markdup-workflow-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.3: Align Sort Markdup Workflow (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.3: Align Sort Markdup Workflow (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translation-output" class="md-nav__link">
    <span class="md-ellipsis">
      
        Translation Output
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-translated-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running the Translated Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-1-process-order" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 1: Process Order
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-2-unquoted-strings" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 2: Unquoted Strings
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-3-filename-clashes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 3: Filename Clashes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#completed-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Completed Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-21-samtools-flagstat-tool-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.1: Samtools Flagstat Tool (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.1: Samtools Flagstat Tool (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-samtools-flagstat-as-a-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Samtools Flagstat as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-22-limma-voom-tool-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.2: Limma Voom Tool (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.2: Limma Voom Tool (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-limma-voom-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Limma Voom as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-23-rna-seq-reads-to-counts-workflow-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.3: RNA-Seq Reads to Counts Workflow (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.3: RNA-Seq Reads to Counts Workflow (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-translated-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running the Translated Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-1-fastqc-outdir" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 1: FASTQC outdir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-2-cutadapt-output-collection" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 2: CUTADAPT output collection
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-3-hisat2-index" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 3: HISAT2 index
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-4-featurecounts-command" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 4: FEATURECOUNTS command
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-5-picard_mark_duplicates-command" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 5: PICARD_MARK_DUPLICATES command
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-6-multiqc-input-filename-collision" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 6: MULTIQC input filename collision
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#final-error-multiqc-script" class="md-nav__link">
    <span class="md-ellipsis">
      
        Final Error: MULTIQC script
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#completed-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Completed Workflow
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Genomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Genomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Variant detection
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Variant detection
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Variant detection
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_galaxy_1/variant_calling_galaxy_1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant Calling part 1 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant Calling part 2 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../longread_sv_calling/longread_sv_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Long-read Structural Variant Calling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_gatk1/variant_calling_gatk1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant calling using command-line tools
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    de novo assembly
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to de novo assembly with Velvet
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-protocol/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to de novo genome assembly for Illumina reads
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of Illumina reads using Velvet (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of Illumina reads using Spades (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pacbio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of PacBio reads (unix)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../hybrid_assembly/nanopore_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Hybrid genome assembly - Nanopore and Illumina
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Metabarcoding
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Metabarcoding
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2/qiime2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Metabarcoding using QIIME2: *Marine dataset*
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2_mammal/qiime2_mammal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Metabarcoding using QIIME2: *Mammal dataset*
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Transcriptomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Transcriptomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq Experimental Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna-seq-lcg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq: counts to genes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Bulk RNA-seq
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Bulk RNA-seq
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to bulk RNAseq analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq analysis using HISAT2 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tuxedo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq analysis using Tuxedo (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_advanced/rna_seq_advanced_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq differential expression tool comparision (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_in_r/rna_seq_r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq differential expression in R
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Single-cell RNA-seq
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Single-cell RNA-seq
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../singlecell/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    10X scRNA-seq analysis in R
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-de/seurat-de/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    scRNA-seq Integration and Differential Expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-go/seurat-go/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Gene Ontology (GO) and Data Integration for scRNA-seq
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Proteomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Proteomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proteomics_basic/proteomics_basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Identifying proteins from mass spectrometry data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Statistics and Visualisation
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    Statistics and Visualisation
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_viz/python_viz/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Data visualisation with Python
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Structural Modelling
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Structural Modelling
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_101/molecular_dynamics_101/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Molecular Dynamics - Introduction to cluster computing
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_201/molecular_dynamics_201/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Molecular Dynamics - Building input files, visualising the trajectory
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Workshop Delivery Mode Information
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Workshop Delivery Mode Information
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workshop_delivery_mode_info/workshops_nectar/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Workshops on Nectar Instances
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
        
          
          <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Archive
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_10">
            <span class="md-nav__icon md-icon"></span>
            
  
    Archive
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix_archive_2022/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Unix-Archive
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl/cwl/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to CWL
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../using_git/Using_Git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Git and Github (deprecated)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq Experimental Design (deprecated)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#topic" class="md-nav__link">
    <span class="md-ellipsis">
      
        Topic
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#skill-level" class="md-nav__link">
    <span class="md-ellipsis">
      
        Skill level
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#description" class="md-nav__link">
    <span class="md-ellipsis">
      
        Description
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      
        Learning Objectives
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#required-software" class="md-nav__link">
    <span class="md-ellipsis">
      
        Required Software
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#required-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Required Data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    <span class="md-ellipsis">
      
        Author Information
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contents" class="md-nav__link">
    <span class="md-ellipsis">
      
        Contents
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-11-samtools-flagstat-tool-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.1: Samtools Flagstat Tool (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.1: Samtools Flagstat Tool (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-janis-translate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-samtools-flagstat-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Samtools Flagstat as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-12-gatk-haplotypecaller-tool-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.2: GATK HaplotypeCaller Tool (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.2: GATK HaplotypeCaller Tool (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-gatk-haplotypecaller-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running GATK HaplotypeCaller as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-13-align-sort-markdup-workflow-cwl" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1.3: Align Sort Markdup Workflow (CWL)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1.3: Align Sort Markdup Workflow (CWL)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translation-output" class="md-nav__link">
    <span class="md-ellipsis">
      
        Translation Output
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-translated-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running the Translated Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-1-process-order" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 1: Process Order
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-2-unquoted-strings" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 2: Unquoted Strings
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-3-filename-clashes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 3: Filename Clashes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#completed-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Completed Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-21-samtools-flagstat-tool-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.1: Samtools Flagstat Tool (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.1: Samtools Flagstat Tool (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-samtools-flagstat-as-a-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Samtools Flagstat as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-22-limma-voom-tool-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.2: Limma Voom Tool (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.2: Limma Voom Tool (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-limma-voom-as-a-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running Limma Voom as a Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conclusion_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-23-rna-seq-reads-to-counts-workflow-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2.3: RNA-Seq Reads to Counts Workflow (Galaxy)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2.3: RNA-Seq Reads to Counts Workflow (Galaxy)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#janis-translate_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Janis Translate
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-translated-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running the Translated Workflow
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#manual-adjustments_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Manual Adjustments
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-1-fastqc-outdir" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 1: FASTQC outdir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-2-cutadapt-output-collection" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 2: CUTADAPT output collection
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-3-hisat2-index" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 3: HISAT2 index
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-4-featurecounts-command" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 4: FEATURECOUNTS command
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-5-picard_mark_duplicates-command" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 5: PICARD_MARK_DUPLICATES command
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#error-6-multiqc-input-filename-collision" class="md-nav__link">
    <span class="md-ellipsis">
      
        Error 6: MULTIQC input filename collision
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#final-error-multiqc-script" class="md-nav__link">
    <span class="md-ellipsis">
      
        Final Error: MULTIQC script
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#completed-workflow_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Completed Workflow
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conclusion
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<p><img alt="alt text" src="../media/logos_banner.png" /></p>
<h1 id="janis-translate-migrating-cwl-galaxy-to-nextflow">Janis Translate - Migrating CWL / Galaxy to Nextflow<a class="headerlink" href="#janis-translate-migrating-cwl-galaxy-to-nextflow" title="Permanent link">&para;</a></h1>
<p>Anticipated workshop duration when delivered to a group of participants is <strong>3.5 hours</strong>.  </p>
<p>For queries relating to this workshop, contact Melbourne Bioinformatics (<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#98;&#105;&#111;&#105;&#110;&#102;&#111;&#114;&#109;&#97;&#116;&#105;&#99;&#115;&#45;&#116;&#114;&#97;&#105;&#110;&#105;&#110;&#103;&#64;&#117;&#110;&#105;&#109;&#101;&#108;&#98;&#46;&#101;&#100;&#117;&#46;&#97;&#117;">&#98;&#105;&#111;&#105;&#110;&#102;&#111;&#114;&#109;&#97;&#116;&#105;&#99;&#115;&#45;&#116;&#114;&#97;&#105;&#110;&#105;&#110;&#103;&#64;&#117;&#110;&#105;&#109;&#101;&#108;&#98;&#46;&#101;&#100;&#117;&#46;&#97;&#117;</a>).</p>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<h3 id="topic">Topic<a class="headerlink" href="#topic" title="Permanent link">&para;</a></h3>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Genomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Transcriptomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Proteomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Metabolomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Statistics and visualisation</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Structural Modelling</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Basic skills</li>
</ul>
<h3 id="skill-level">Skill level<a class="headerlink" href="#skill-level" title="Permanent link">&para;</a></h3>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Beginner  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Intermediate  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Advanced  </li>
</ul>
<p>The workshop is conducted in a Unix environment.<br>
Command line experience is required. <br>
Prior experience with developing, running, and troubleshooting Nextflow workflows is strongly recommended.</p>
<h3 id="description">Description<a class="headerlink" href="#description" title="Permanent link">&para;</a></h3>
<p>Bioinformatics workflows are critical for reproducibly transferring methodologies between research groups and for scaling between computational infrastructures. Research groups currently invest a lot of time and effort in creating and updating workflows; the ability to translate from one workflow language into another can make them easier to share, and maintain with minimal effort. For example, research groups that would like to run an existing Galaxy workflow on HPC, or extend it for their use, might find translating the workflow to Nextflow more suitable for their ongoing use-cases. </p>
<p><strong><em>Janis</em></strong> is a framework that provides an abstraction layer for describing workflows, and a tool that can translate workflows between existing languages such as CWL, WDL, Galaxy and Nextflow. Janis aims to translate as much as it can, leaving the user to validate the workflow and make small manual adjustments where direct translations are not possible. Originating from the Portable Pipelines Project between Melbourne Bioinformatics, the Peter MacCallum Cancer Centre, and the Walter and Eliza Hall Institute of Medical Research, this tool is now available for everyone to use.</p>
<p>This workshop provides an introduction to Janis and how it can be used to translate Galaxy and CWL based tools and workflows into Nextflow. Using hands-on examples well step you through the process and demonstrate how to optimise, troubleshoot and test the translated workflows.</p>
<p><strong><em>Section 1</em></strong> covers migration of CWL tools / workflows to Nextflow. <br>
<strong><em>Section 2</em></strong> covers migration of Galaxy tool wrappers / workflows to Nextflow.</p>
<hr />
<h3 id="learning-objectives">Learning Objectives<a class="headerlink" href="#learning-objectives" title="Permanent link">&para;</a></h3>
<p>By the end of the workshop you should be able to:
- Recognise the main aspects and benefits of workflow translation
- Use Janis to translated Galaxy / CWL tools &amp; workflows to Nextflow
- Configure Nextflow to run translated tools &amp; workflows
- Troubleshoot translated Nextflow tools &amp; workflow 
- Adjust the translated Nextflow tools / workflows &amp; complete missing translations manually</p>
<hr />
<h3 id="required-software">Required Software<a class="headerlink" href="#required-software" title="Permanent link">&para;</a></h3>
<ul>
<li>IDE of your choosing (we use VS Code in this workshop).</li>
<li>Refer back to the <a href="https://docs.google.com/document/d/1m4MNPNGM0qdMPYqT10uUwk9bYwC5EG0PrUfmuivhJCg/edit?usp=sharing">setup instructions document</a> provided if needed. </li>
</ul>
<h3 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h3>
<ul>
<li>Sample data will be provided on the compute resource.</li>
</ul>
<hr />
<h3 id="author-information">Author Information<a class="headerlink" href="#author-information" title="Permanent link">&para;</a></h3>
<p>Written by: Grace Hall<br />
Melbourne Bioinformatics, University of Melbourne</p>
<p>Created/Reviewed: May 2023</p>
<p><br></p>
<hr />
<h2 id="contents">Contents<a class="headerlink" href="#contents" title="Permanent link">&para;</a></h2>
<p>This workshop has 2 main sections. </p>
<p>We will translate CWL and Galaxy tools and workflows using janis-translate, then will do manual adjustments &amp; run on sample data to check validity.</p>
<p>Section 1: CWL  Nextflow</p>
<ul>
<li><a href="#section-11-samtools-flagstat-tool-cwl">Section 1.1: Samtools Flagstat Tool</a></li>
<li><a href="#section-12-gatk-haplotypecaller-tool-cwl">Section 1.2: GATK HaplotypeCaller Tool</a></li>
<li><a href="#section-13-align-sort-markdup-workflow-cwl">Section 1.3: Align Sort Markdup Workflow</a></li>
</ul>
<p>Section 2: Galaxy  Nextflow</p>
<ul>
<li><a href="#section-21-samtools-flagstat-tool-galaxy">Section 2.1: Samtools Flagstat Tool</a></li>
<li><a href="#section-22-limma-voom-tool-galaxy">Section 2.2: Limma Voom Tool</a></li>
<li><a href="#section-23-rna-seq-reads-to-counts-workflow-galaxy">Section 2.3: RNA-Seq Reads to Counts Workflow</a></li>
</ul>
<p><br></p>
<hr />
<h2 id="section-11-samtools-flagstat-tool-cwl">Section 1.1: Samtools Flagstat Tool (CWL)<a class="headerlink" href="#section-11-samtools-flagstat-tool-cwl" title="Permanent link">&para;</a></h2>
<h3 id="introduction">Introduction<a class="headerlink" href="#introduction" title="Permanent link">&para;</a></h3>
<p>This section demonstrates translation of a basic <code>samtools flagstat</code> tool from CWL to Nextflow using <code>janis translate</code>. </p>
<p>The CWL tool used in this section - <a href="https://github.com/genome/analysis-workflows/blob/master/definitions/tools/samtools_flagstat.cwl">samtools_flagstat.cwl</a> -  is taken from the <a href="https://www.genome.wustl.edu/">McDonnell Genome Institute</a> (MGI) <a href="https://github.com/genome/analysis-workflows">analysis-workflows</a> repository. </p>
<p>This resource stores publically available analysis pipelines for genomics data. <br>
It is a fantastic piece of research software, and the authors thank MGI for their contribution to open-source research software. </p>
<p>The underlying software run by this tool - <a href="http://www.htslib.org/doc/samtools-flagstat.html">Samtools Flagstat</a> - displays summary information for an alignment file.</p>
<p><br></p>
<h3 id="janis-translate">Janis Translate<a class="headerlink" href="#janis-translate" title="Permanent link">&para;</a></h3>
<p><strong>Downloading Janis Translate</strong></p>
<p>In this workshop we will use a singularity container to run <code>janis translate</code>. </p>
<p>Containers are great because they remove the need for package managers, and guarantee that the software can run on any machine. </p>
<p>Singularity is already set up on your VM.</p>
<p>Run the following command to pull the janis image:
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>singularity pull janis.sif docker://pppjanistranslate/janis-translate:0.13.0
</span></code></pre></div></p>
<p>Check your image by running the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>singularity exec ~/janis.sif janis translate
</span></code></pre></div></p>
<p>If the image is working, you should see the janis translate helptext.</p>
<p><br></p>
<p><strong>Downloading Training Data and Tool / Workflow Source Files</strong></p>
<p>For this workshop we will fetch all needed data from zenodo using wget.  </p>
<p>This archive contains source CWL / Galaxy files, sample data, and finished translations as a reference.</p>
<p>Run the following commands to download &amp; uncompress the zenodo archive:
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>wget https://zenodo.org/record/8052348/files/data.tar.gz
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>tar -xvf data.tar.gz
</span></code></pre></div></p>
<p>Inside the data folder we have the following structure: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>data
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a> final               # finalised translations for your reference
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>    cwl
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>    galaxy
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a> sample_data         # sample data to test nextflow processes / workflows
</span><span id="__span-3-6"><a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>    cwl
</span><span id="__span-3-7"><a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>    galaxy
</span><span id="__span-3-8"><a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a> source              # CWL / Galaxy files we will translate
</span><span id="__span-3-9"><a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a>     cwl
</span><span id="__span-3-10"><a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>     galaxy
</span></code></pre></div></p>
<p>The test data provided in <code>data/sample_data</code> is also available at <code>/cvmfs/data.biocommons.aarnet.edu.au/training_materials/MelbBio_training/Janis_0723/sample_data/</code></p>
<p><br></p>
<h3 id="running-janis-translate">Running Janis Translate<a class="headerlink" href="#running-janis-translate" title="Permanent link">&para;</a></h3>
<p>To translate a tool / workflow,  we use <code>janis translate</code>.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>janis translate --from &lt;src&gt; --to &lt;dest&gt; &lt;filepath&gt;
</span></code></pre></div>
<p>The <code>--from</code> argument specifies the workflow language of the source file(s), and <code>--to</code> specifies the destination we want to translate to. </p>
<p>The <code>&lt;filepath&gt;</code> argument is the source file we will translate. </p>
<p>In this section we are translating a single CWL CommandLineTool, so we will only produce a single nextflow file. </p>
<p><br></p>
<p>In this section, we want to translate CWL -&gt; Nextflow, and our source CWL file is in the <code>data</code> folder we just downloaded from zenodo. </p>
<p>It&rsquo;s path is <code>data/source/cwl/samtools_flagstat.cwl</code>.</p>
<p>Run the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>singularity exec ~/janis.sif janis translate --from cwl --to nextflow data/source/cwl/samtools_flagstat.cwl
</span></code></pre></div></p>
<p><br></p>
<p>You will see a folder called <code>translated/</code> appear, and a nextflow process called <code>samtools_flagstat.nf</code> will be present inside. </p>
<p><br></p>
<h3 id="manual-adjustments">Manual Adjustments<a class="headerlink" href="#manual-adjustments" title="Permanent link">&para;</a></h3>
<p>The <code>translated/samtools_flagstat.nf</code> file should be similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>process SAMTOOLS_FLAGSTAT {
</span><span id="__span-6-4"><a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>
</span><span id="__span-6-5"><a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>    container &quot;quay.io/biocontainers/samtools:1.11--h6270b1f_0&quot;
</span><span id="__span-6-6"><a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>
</span><span id="__span-6-7"><a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>    input:
</span><span id="__span-6-8"><a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>    path bam
</span><span id="__span-6-9"><a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>
</span><span id="__span-6-10"><a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>    output:
</span><span id="__span-6-11"><a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>    path &quot;${bam[0]}.flagstat&quot;, emit: flagstats
</span><span id="__span-6-12"><a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>
</span><span id="__span-6-13"><a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>    script:
</span><span id="__span-6-14"><a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>    def bam = bam[0]
</span><span id="__span-6-15"><a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a>    &quot;&quot;&quot;
</span><span id="__span-6-16"><a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a>    /usr/local/bin/samtools flagstat \
</span><span id="__span-6-17"><a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a>    ${bam} \
</span><span id="__span-6-18"><a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a>    &gt; ${bam}.flagstat \
</span><span id="__span-6-19"><a id="__codelineno-6-19" name="__codelineno-6-19" href="#__codelineno-6-19"></a>    &quot;&quot;&quot;
</span><span id="__span-6-20"><a id="__codelineno-6-20" name="__codelineno-6-20" href="#__codelineno-6-20"></a>
</span><span id="__span-6-21"><a id="__codelineno-6-21" name="__codelineno-6-21" href="#__codelineno-6-21"></a>}
</span></code></pre></div>
<p>We can see that this nextflow process has a single input, a single output, and calls <code>samtools flagstat</code> on the input <code>bam</code>. </p>
<p>We can also see that a container image is available for this tool. In the next section we will run this process using some sample data and the specified container. </p>
<p>This translation is correct for the <code>samtools_flagstat.cwl</code> file and needs no adjusting. <br>
Have a look at the source CWL file to see how they match up. </p>
<p><br></p>
<blockquote>
<p>Note: <br>
<code>def bam = bam[0]</code> in the script block is used to handle datatypes with secondary files. <br>
The <code>bam</code> input is an indexed bam type, so requires a <code>.bai</code> file to also be present in the working directory alongside the <code>.bam</code> file. <br><br>
For this reason the <code>bam</code> input is supplied as an Array with 2 files - the <code>.bam</code> and the <code>.bai</code>. <br>
Here the <code>def bam = bam[0]</code> is used so that <code>bam</code> refers to the <code>.bam</code> file in that Array. </p>
</blockquote>
<p><br></p>
<h3 id="running-samtools-flagstat-as-a-workflow">Running Samtools Flagstat as a Workflow<a class="headerlink" href="#running-samtools-flagstat-as-a-workflow" title="Permanent link">&para;</a></h3>
<p><strong>Collecting Process Outputs</strong></p>
<p>Let&rsquo;s add a <code>publishDir</code> directive to our translated process so that we can capture the outputs of this process.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>process SAMTOOLS_FLAGSTAT {
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>    container &quot;quay.io/biocontainers/samtools:1.11--h6270b1f_0&quot;
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>    publishDir &quot;./outputs&quot;                                          &lt;-
</span><span id="__span-7-5"><a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>    ....
</span><span id="__span-7-6"><a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a>
</span><span id="__span-7-7"><a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a>}                          
</span></code></pre></div>
<p>Nextflow allows us to capture the outputs created by a process using the <code>publishDir</code> directive seen above. </p>
<p><br></p>
<p><strong>Setting up nextflow.config</strong></p>
<p>To run this process, we will set up a <code>nextflow.config</code> file and add some lines to the top of our process definition to turn it into a workflow.</p>
<p>Create a new file called <code>nextflow.config</code> in the <code>translated/</code> folder alongside <code>samtools_flagstat.nf</code>. </p>
<p>For this workshop, we are using sample data located at <code>/home2/training/data/sample_data/</code>. </p>
<p>Copy and paste the following code into your <code>nextflow.config</code> file: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>singularity.enabled = true
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>params {
</span><span id="__span-8-6"><a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>
</span><span id="__span-8-7"><a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a>    bam = [
</span><span id="__span-8-8"><a id="__codelineno-8-8" name="__codelineno-8-8" href="#__codelineno-8-8"></a>        &#39;/home2/training/data/sample_data/cwl/2895499223_sorted.bam&#39;,
</span><span id="__span-8-9"><a id="__codelineno-8-9" name="__codelineno-8-9" href="#__codelineno-8-9"></a>        &#39;/home2/training/data/sample_data/cwl/2895499223_sorted.bam.bai&#39;,
</span><span id="__span-8-10"><a id="__codelineno-8-10" name="__codelineno-8-10" href="#__codelineno-8-10"></a>    ]
</span><span id="__span-8-11"><a id="__codelineno-8-11" name="__codelineno-8-11" href="#__codelineno-8-11"></a>
</span><span id="__span-8-12"><a id="__codelineno-8-12" name="__codelineno-8-12" href="#__codelineno-8-12"></a>}
</span></code></pre></div>
<p>This tells nextflow how to run, and sets up an input parameter for our indexed bam input.</p>
<p>The <code>bam</code> parameter is a list which provides paths to the <code>.bam</code> and <code>.bai</code> sample data we will use to test the nextflow translation. From here, we can refer to the indexed bam input as <code>params.bam</code> in other files.</p>
<p><br></p>
<blockquote>
<p>NOTE<br>
<code>nextflow.enable.dsl = 2</code> ensures that we are using the dsl2 nextflow syntax which is the current standard. <br><br>
<code>singularity.enabled = true</code> tells nextflow to run processes using singularity. Our <code>samtools_flagstat.nf</code> has a directive with the form <code>container "quay.io/biocontainers/samtools:1.11--h6270b1f_0"</code> provided, so it will use the specified image when running this process. <br> <br>
<code>singularity.cacheDir = "$HOME/.singularity/cache"</code> tells nextflow where singularity images are stored. <br> <br>
Nextflow will handle the singularity image download and stored it in the cache specified above. If you&rsquo;d like to use an already available container, you can modify <code>samtools_flagstat.nf</code> container directive to <code>container "/cvmfs/singularity.galaxyproject.org/all/samtools:1.11--h6270b1f_0"</code>. <br> <br>
The test data used above is also available at <code>/cvmfs</code>, and can be accessed as follow: <code>bam=['/cvmfs/data.biocommons.aarnet.edu.au/training_materials/MelbBio_training/Janis_0723/sample_data/cwl/2895499223_sorted.bam', '/cvmfs/data.biocommons.aarnet.edu.au/training_materials/MelbBio_training/Janis_0723/sample_data/cwl/2895499223_sorted.bam.bai']</code>. </p>
</blockquote>
<p><br></p>
<p><strong>Creating Workflow &amp; Passing Data</strong> </p>
<p>Now that we have the <code>nextflow.config</code> file set up, we will add a few lines to <code>samtools_flagstat.nf</code> to turn it into a workflow. </p>
<p>Copy and paste the following lines at the top of <code>samtools_flagstat.nf</code>:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>ch_bam = Channel.fromPath( params.bam ).toList()
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>workflow {
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>    SAMTOOLS_FLAGSTAT(ch_bam)
</span><span id="__span-9-5"><a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a>}
</span></code></pre></div>
<p>The first line creates a nextflow <code>Channel</code> for our <code>bam</code> input and ensures it is a list. <br>
The <code>Channel.toList()</code> part collects our files into a list, as both the <code>.bam</code> and <code>.bai</code> files must be passed together. <br>
The <code>params.bam</code> global variable we set up previously is used to supply the paths to our sample data.</p>
<p>The new <code>workflow {}</code> section declares the main workflow entry point. <br>
When we run this file, nextflow will look for this section and run the workflow contained within. </p>
<p>In our case, the workflow only contains a single task, which runs the <code>SAMTOOLS_FLAGSTAT</code> process defined below the workflow section. The single <code>SAMTOOLS_FLAGSTAT</code> input is being passed data from our <code>ch_bam</code> channel we declared at the top of the file. </p>
<p><br></p>
<p><strong>Running Our Workflow</strong></p>
<p>Ensure you are in the <code>translated/</code> working directory, where <code>nextflow.config</code> and <code>samtools_flagstat.nf</code> reside. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>cd translated
</span></code></pre></div>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>nextflow run samtools_flagstat.nf
</span></code></pre></div></p>
<p>Nextflow will automatically check if there is a <code>nextflow.config</code> file in the working directory, and if so will use that to configure itself. Our inputs are supplied in <code>nextflow.config</code> alongside the dsl2 &amp; singularity config, so it should run without issue. </p>
<p>Once completed, we can check the <code>outputs/</code> folder to view our results. </p>
<p>If everything went well, there should be a file called <code>2895499223_sorted.bam.flagstat</code> with the following contents:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>2495 + 0 in total (QC-passed reads + QC-failed reads)
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>0 + 0 secondary
</span><span id="__span-12-3"><a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>1 + 0 supplementary
</span><span id="__span-12-4"><a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a>0 + 0 duplicates
</span><span id="__span-12-5"><a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a>2480 + 0 mapped (99.40% : N/A)
</span><span id="__span-12-6"><a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a>2494 + 0 paired in sequencing
</span><span id="__span-12-7"><a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a>1247 + 0 read1
</span><span id="__span-12-8"><a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a>1247 + 0 read2
</span><span id="__span-12-9"><a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a>2460 + 0 properly paired (98.64% : N/A)
</span><span id="__span-12-10"><a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a>2464 + 0 with itself and mate mapped
</span><span id="__span-12-11"><a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a>15 + 0 singletons (0.60% : N/A)
</span><span id="__span-12-12"><a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a>0 + 0 with mate mapped to a different chr
</span><span id="__span-12-13"><a id="__codelineno-12-13" name="__codelineno-12-13" href="#__codelineno-12-13"></a>0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</span></code></pre></div>
<p><br></p>
<h3 id="conclusion">Conclusion<a class="headerlink" href="#conclusion" title="Permanent link">&para;</a></h3>
<p>In section 1.1 we explored how to translate a simple CWL tool to a Nextflow process. </p>
<p>If needed, you can check the <code>final/cwl/samtools_flagstat</code> folder in the zenodo <code>data/</code> directory. 
This contains the files we created in this section for your reference. </p>
<p>We will build on our knowledge in section 1.2 by translating a more complex CWL CommandLineTool definition: <em>GATK HaplotypeCaller</em>.</p>
<p><br></p>
<hr />
<h2 id="section-12-gatk-haplotypecaller-tool-cwl">Section 1.2: GATK HaplotypeCaller Tool (CWL)<a class="headerlink" href="#section-12-gatk-haplotypecaller-tool-cwl" title="Permanent link">&para;</a></h2>
<h3 id="introduction_1">Introduction<a class="headerlink" href="#introduction_1" title="Permanent link">&para;</a></h3>
<p>Now that we have translated a simple CWL CommandLineTool, let&rsquo;s translate a more real-world CommandLineTool. </p>
<p>This section demonstrates translation of a <code>gatk HaplotypeCaller</code> tool from CWL to Nextflow using <code>janis translate</code>. <br></p>
<p>The source CWL file used in this section is taken from the <a href="https://www.genome.wustl.edu/">McDonnell Genome Institute</a> (MGI) <a href="https://github.com/genome/analysis-workflows">analysis-workflows</a> repository. </p>
<p>This repository stores publically available analysis pipelines for genomics data. <br>
It is a fantastic piece of research software, and the authors thank MGI for their contribution to open-source research software. </p>
<p>The software tool encapsulated by the source CWL - <a href="https://github.com/genome/analysis-workflows/blob/master/definitions/tools/gatk_haplotype_caller.cwl">gatk_haplotype_caller.cwl</a> - displays summary information for an alignment file. </p>
<p><br></p>
<h3 id="janis-translate_1">Janis Translate<a class="headerlink" href="#janis-translate_1" title="Permanent link">&para;</a></h3>
<p>As in section 1.1, we will use janis-translate to translate our CWL CommandLineTool to a Nextflow process. </p>
<p>First, let&rsquo;s make sure we&rsquo;re back in the main training directory
<div class="language-text highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>cd /home2/training
</span></code></pre></div></p>
<p>This time, since we already have a <code>translated/</code> folder, we will supply the <code>-o</code> argument to janis-translate to specify an output directory. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>singularity exec ~/janis.sif janis translate -o haplotype_caller --from cwl --to nextflow data/source/cwl/gatk_haplotype_caller.cwl
</span></code></pre></div>
<p>You will see a folder called <code>haplotype_caller/</code> appear, and a nextflow process called <code>gatk_haplotype_caller.nf</code> will be present inside. </p>
<p><br></p>
<h3 id="manual-adjustments_1">Manual Adjustments<a class="headerlink" href="#manual-adjustments_1" title="Permanent link">&para;</a></h3>
<p>The <code>haplotype_caller/gatk_haplotype_caller.nf</code> file should be similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>process GATK_HAPLOTYPE_CALLER {
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a>
</span><span id="__span-15-5"><a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a>    container &quot;broadinstitute/gatk:4.1.8.1&quot;
</span><span id="__span-15-6"><a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>
</span><span id="__span-15-7"><a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>    input:
</span><span id="__span-15-8"><a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a>    path bam
</span><span id="__span-15-9"><a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a>    path reference
</span><span id="__span-15-10"><a id="__codelineno-15-10" name="__codelineno-15-10" href="#__codelineno-15-10"></a>    path dbsnp_vcf, stageAs: &#39;dbsnp_vcf/*&#39;
</span><span id="__span-15-11"><a id="__codelineno-15-11" name="__codelineno-15-11" href="#__codelineno-15-11"></a>    val intervals
</span><span id="__span-15-12"><a id="__codelineno-15-12" name="__codelineno-15-12" href="#__codelineno-15-12"></a>    val gvcf_gq_bands
</span><span id="__span-15-13"><a id="__codelineno-15-13" name="__codelineno-15-13" href="#__codelineno-15-13"></a>    val emit_reference_confidence
</span><span id="__span-15-14"><a id="__codelineno-15-14" name="__codelineno-15-14" href="#__codelineno-15-14"></a>    val contamination_fraction
</span><span id="__span-15-15"><a id="__codelineno-15-15" name="__codelineno-15-15" href="#__codelineno-15-15"></a>    val max_alternate_alleles
</span><span id="__span-15-16"><a id="__codelineno-15-16" name="__codelineno-15-16" href="#__codelineno-15-16"></a>    val ploidy
</span><span id="__span-15-17"><a id="__codelineno-15-17" name="__codelineno-15-17" href="#__codelineno-15-17"></a>    val read_filter
</span><span id="__span-15-18"><a id="__codelineno-15-18" name="__codelineno-15-18" href="#__codelineno-15-18"></a>
</span><span id="__span-15-19"><a id="__codelineno-15-19" name="__codelineno-15-19" href="#__codelineno-15-19"></a>    output:
</span><span id="__span-15-20"><a id="__codelineno-15-20" name="__codelineno-15-20" href="#__codelineno-15-20"></a>    tuple path(&quot;output.g.vcf.gz&quot;), path(&quot;*.tbi&quot;), emit: gvcf
</span><span id="__span-15-21"><a id="__codelineno-15-21" name="__codelineno-15-21" href="#__codelineno-15-21"></a>
</span><span id="__span-15-22"><a id="__codelineno-15-22" name="__codelineno-15-22" href="#__codelineno-15-22"></a>    script:
</span><span id="__span-15-23"><a id="__codelineno-15-23" name="__codelineno-15-23" href="#__codelineno-15-23"></a>    def bam = bam[0]
</span><span id="__span-15-24"><a id="__codelineno-15-24" name="__codelineno-15-24" href="#__codelineno-15-24"></a>    def dbsnp_vcf = dbsnp_vcf[0] != null ? &quot;--dbsnp ${dbsnp_vcf[0]}&quot; : &quot;&quot;
</span><span id="__span-15-25"><a id="__codelineno-15-25" name="__codelineno-15-25" href="#__codelineno-15-25"></a>    def reference = reference[0]
</span><span id="__span-15-26"><a id="__codelineno-15-26" name="__codelineno-15-26" href="#__codelineno-15-26"></a>    def gvcf_gq_bands_joined = gvcf_gq_bands != params.NULL_VALUE ? &quot;-GQB &quot; + gvcf_gq_bands.join(&#39; &#39;) : &quot;&quot;
</span><span id="__span-15-27"><a id="__codelineno-15-27" name="__codelineno-15-27" href="#__codelineno-15-27"></a>    def intervals_joined = intervals.join(&#39; &#39;)
</span><span id="__span-15-28"><a id="__codelineno-15-28" name="__codelineno-15-28" href="#__codelineno-15-28"></a>    def contamination_fraction = contamination_fraction != params.NULL_VALUE ? &quot;-contamination ${contamination_fraction}&quot; : &quot;&quot;
</span><span id="__span-15-29"><a id="__codelineno-15-29" name="__codelineno-15-29" href="#__codelineno-15-29"></a>    def max_alternate_alleles = max_alternate_alleles != params.NULL_VALUE ? &quot;--max_alternate_alleles ${max_alternate_alleles}&quot; : &quot;&quot;
</span><span id="__span-15-30"><a id="__codelineno-15-30" name="__codelineno-15-30" href="#__codelineno-15-30"></a>    def ploidy = ploidy != params.NULL_VALUE ? &quot;-ploidy ${ploidy}&quot; : &quot;&quot;
</span><span id="__span-15-31"><a id="__codelineno-15-31" name="__codelineno-15-31" href="#__codelineno-15-31"></a>    def read_filter = read_filter != params.NULL_VALUE ? &quot;--read_filter ${read_filter}&quot; : &quot;&quot;
</span><span id="__span-15-32"><a id="__codelineno-15-32" name="__codelineno-15-32" href="#__codelineno-15-32"></a>    &quot;&quot;&quot;
</span><span id="__span-15-33"><a id="__codelineno-15-33" name="__codelineno-15-33" href="#__codelineno-15-33"></a>    /gatk/gatk --java-options -Xmx16g HaplotypeCaller \
</span><span id="__span-15-34"><a id="__codelineno-15-34" name="__codelineno-15-34" href="#__codelineno-15-34"></a>    -R ${reference} \
</span><span id="__span-15-35"><a id="__codelineno-15-35" name="__codelineno-15-35" href="#__codelineno-15-35"></a>    -I ${bam} \
</span><span id="__span-15-36"><a id="__codelineno-15-36" name="__codelineno-15-36" href="#__codelineno-15-36"></a>    -ERC ${emit_reference_confidence} \
</span><span id="__span-15-37"><a id="__codelineno-15-37" name="__codelineno-15-37" href="#__codelineno-15-37"></a>    ${gvcf_gq_bands_joined} \
</span><span id="__span-15-38"><a id="__codelineno-15-38" name="__codelineno-15-38" href="#__codelineno-15-38"></a>    -L ${intervals_joined} \
</span><span id="__span-15-39"><a id="__codelineno-15-39" name="__codelineno-15-39" href="#__codelineno-15-39"></a>    ${dbsnp_vcf} \
</span><span id="__span-15-40"><a id="__codelineno-15-40" name="__codelineno-15-40" href="#__codelineno-15-40"></a>    ${contamination_fraction} \
</span><span id="__span-15-41"><a id="__codelineno-15-41" name="__codelineno-15-41" href="#__codelineno-15-41"></a>    ${max_alternate_alleles} \
</span><span id="__span-15-42"><a id="__codelineno-15-42" name="__codelineno-15-42" href="#__codelineno-15-42"></a>    ${ploidy} \
</span><span id="__span-15-43"><a id="__codelineno-15-43" name="__codelineno-15-43" href="#__codelineno-15-43"></a>    ${read_filter} \
</span><span id="__span-15-44"><a id="__codelineno-15-44" name="__codelineno-15-44" href="#__codelineno-15-44"></a>    -O &quot;output.g.vcf.gz&quot;
</span><span id="__span-15-45"><a id="__codelineno-15-45" name="__codelineno-15-45" href="#__codelineno-15-45"></a>    &quot;&quot;&quot;
</span><span id="__span-15-46"><a id="__codelineno-15-46" name="__codelineno-15-46" href="#__codelineno-15-46"></a>
</span><span id="__span-15-47"><a id="__codelineno-15-47" name="__codelineno-15-47" href="#__codelineno-15-47"></a>}
</span></code></pre></div>
<p>We can see that this nextflow process has a multiple inputs, single output, and calls <code>gatk HaplotypeCaller</code> using the input data we supply to the process.  </p>
<p><br></p>
<blockquote>
<p>Notes on translation: <br><br>
<strong>(1)</strong> <code>def bam = bam[0]</code><br><br>
This pattern is used in the script block to handle datatypes with secondary files. <br>
The <code>bam</code> input is an indexed bam type, so requires a <code>.bai</code> file to also be present in the working directory. <br><br>
To facilitate this, we supply the <code>bam</code> input as a tuple of 2 files:<br> 
<code>[filename.bam, filename.bam.bai]</code>. <br><br>
<code>def bam = bam[0]</code> is used so when we reference <code>${bam}</code> in the script body, we are refering to the <code>.bam</code> file in that list. <br><br>
<strong>(2)</strong> <code>def ploidy = ploidy != params.NULL_VALUE ? "-ploidy ${ploidy}" : ""</code> <br><br>
This is how we handle optional val inputs in nextflow. <br>
Nextflow doesn&rsquo;t like <strong><em>null</em></strong> values to be passed to process inputs, so we use 2 different tricks to make <strong><em>optional</em></strong> inputs possible.<br><br>
For <code>val</code> inputs we set up a <code>NULL_VALUE</code> param in <code>nextflow.config</code> which we use as a placeholder.  <br>
For <code>path</code> inputs (ie files and directories) we set up a <strong><em>null</em></strong> file rather than passing <strong><em>null</em></strong> directly. <br>
This ensures that the file is staged correctly in the working directory when an actual filepath is provided. <br><br>
The format above is a ternary operator of the form <code>def myvar = cond_check ? cond_true : cond_false</code>.<br>
We are redefining the <code>ploidy</code> string variable so that it will be correctly formatted when used in the script block.<br>
If we supply an actual value, it will take the form <code>"-ploidy ${ploidy}"</code>. <br>
If we supply the placeholder <code>params.NULL_VALUE</code> value, it will be a blank string <code>""</code>. <br><br>
<strong>(3)</strong> <code>def intervals_joined = intervals.join(' ')</code> <br><br>
Templates our <code>intervals</code> list of strings to a single string.<br>
Each item is joined by a space: eg <code>["hello", "there"]</code> -&gt; <code>"hello there"</code>. <br><br>
<strong>(4)</strong> <code>def dbsnp_vcf = dbsnp_vcf[0] != null ? "--dbsnp ${dbsnp_vcf[0]}" : ""</code><br><br>
Same as <strong>(2)</strong> except uses a different check for null value, and selects the primary <code>.vcf.gz</code> file. <br><br>
As the <code>path dbsnp_vcf</code> input is optional and consists of a <code>.vcf.gz</code> and <code>.vcf.gz.tbi</code> file, <code>dbsnp_vcf[0] != null</code> checks if the primary <code>.vcf.gz</code> file is present. <br><br>
If true, it templates a string we can use in the script body to provide the required argument: <code>--dbsnp ${dbsnp_vcf[0]}</code>.<br>
If false, it templates an empty string.<br><br>
This ensures that when we use <code>${dbsnp_vcf}</code> in the script body, it will be formatted correctly for either case.<br><br></p>
</blockquote>
<p><br></p>
<p>This translation is correct for the <code>gatk_haplotype_caller.cwl</code> file and needs no adjusting. <br>
Have a look at the source CWL file to see how they match up. </p>
<p><br></p>
<h3 id="running-gatk-haplotypecaller-as-a-workflow">Running GATK HaplotypeCaller as a Workflow<a class="headerlink" href="#running-gatk-haplotypecaller-as-a-workflow" title="Permanent link">&para;</a></h3>
<p><strong>Collecting Process Outputs</strong></p>
<p>Let&rsquo;s add a <code>publishDir</code> directive to our translated process so that we can capture the outputs of this process.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>process GATK_HAPLOTYPE_CALLER {
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a>
</span><span id="__span-16-3"><a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a>    container &quot;broadinstitute/gatk:4.1.8.1&quot;
</span><span id="__span-16-4"><a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a>    publishDir &quot;./outputs&quot;    
</span><span id="__span-16-5"><a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a>    ...
</span><span id="__span-16-6"><a id="__codelineno-16-6" name="__codelineno-16-6" href="#__codelineno-16-6"></a>}                               
</span></code></pre></div>
<p>Nextflow allows us to capture the outputs created by a process using the <code>publishDir</code> directive seen above. </p>
<p><br></p>
<blockquote>
<p>NOTE<br>
Our <code>gatk_haplotype_caller.nf</code> has a container directive with the form <code>container "broadinstitute/gatk:4.1.8.1"</code> provided, so Nextflow will handle this singularity image download and will use the specified image when running this process (provided that <code>singularity.enabled</code> is set to <code>true</code> in the nextflow config). <br>
If you&rsquo;d like to use an already available container, you can modify this container directive to <code>container "/cvmfs/singularity.galaxyproject.org/all/gatk4:4.1.8.1--py38_0"</code>.</p>
</blockquote>
<p><br></p>
<p><strong>Setting up nextflow.config</strong></p>
<p>To run this process, we will set up a <code>nextflow.config</code> file and add some lines to the top of our process definition to turn it into a workflow.</p>
<p>Create a new file called <code>nextflow.config</code> in the <code>haplotype_caller/</code> folder alongside <code>gatk_haplotype_caller.nf</code>. </p>
<p>Copy and paste the following code into your <code>nextflow.config</code> file: </p>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>singularity.enabled = true
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-17-4"><a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>
</span><span id="__span-17-5"><a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a>params {
</span><span id="__span-17-6"><a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a>    NULL_VALUE = &#39;NULL&#39;
</span><span id="__span-17-7"><a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a>
</span><span id="__span-17-8"><a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a>    bam = [
</span><span id="__span-17-9"><a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a>        &#39;/home2/training/data/sample_data/cwl/2895499223_sorted.bam&#39;,
</span><span id="__span-17-10"><a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a>        &#39;/home2/training/data/sample_data/cwl/2895499223_sorted.bam.bai&#39;,
</span><span id="__span-17-11"><a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a>    ]
</span><span id="__span-17-12"><a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a>    dbsnp = [
</span><span id="__span-17-13"><a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a>        &#39;/home2/training/data/sample_data/cwl/chr17_test_dbsnp.vcf.gz&#39;,
</span><span id="__span-17-14"><a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a>        &#39;/home2/training/data/sample_data/cwl/chr17_test_dbsnp.vcf.gz.tbi&#39;,
</span><span id="__span-17-15"><a id="__codelineno-17-15" name="__codelineno-17-15" href="#__codelineno-17-15"></a>    ]
</span><span id="__span-17-16"><a id="__codelineno-17-16" name="__codelineno-17-16" href="#__codelineno-17-16"></a>    reference = [
</span><span id="__span-17-17"><a id="__codelineno-17-17" name="__codelineno-17-17" href="#__codelineno-17-17"></a>        &#39;/home2/training/data/sample_data/cwl/chr17_test.fa&#39;,
</span><span id="__span-17-18"><a id="__codelineno-17-18" name="__codelineno-17-18" href="#__codelineno-17-18"></a>        &#39;/home2/training/data/sample_data/cwl/chr17_test.dict&#39;,
</span><span id="__span-17-19"><a id="__codelineno-17-19" name="__codelineno-17-19" href="#__codelineno-17-19"></a>        &#39;/home2/training/data/sample_data/cwl/chr17_test.fa.fai&#39;,
</span><span id="__span-17-20"><a id="__codelineno-17-20" name="__codelineno-17-20" href="#__codelineno-17-20"></a>    ]
</span><span id="__span-17-21"><a id="__codelineno-17-21" name="__codelineno-17-21" href="#__codelineno-17-21"></a>    gvcf_gq_bands = NULL_VALUE
</span><span id="__span-17-22"><a id="__codelineno-17-22" name="__codelineno-17-22" href="#__codelineno-17-22"></a>    intervals = [&quot;chr17&quot;]
</span><span id="__span-17-23"><a id="__codelineno-17-23" name="__codelineno-17-23" href="#__codelineno-17-23"></a>    emit_reference_confidence = &#39;GVCF&#39;
</span><span id="__span-17-24"><a id="__codelineno-17-24" name="__codelineno-17-24" href="#__codelineno-17-24"></a>    contamination_fraction = NULL_VALUE
</span><span id="__span-17-25"><a id="__codelineno-17-25" name="__codelineno-17-25" href="#__codelineno-17-25"></a>    max_alternate_alleles = NULL_VALUE
</span><span id="__span-17-26"><a id="__codelineno-17-26" name="__codelineno-17-26" href="#__codelineno-17-26"></a>    ploidy = NULL_VALUE
</span><span id="__span-17-27"><a id="__codelineno-17-27" name="__codelineno-17-27" href="#__codelineno-17-27"></a>    read_filter = NULL_VALUE
</span><span id="__span-17-28"><a id="__codelineno-17-28" name="__codelineno-17-28" href="#__codelineno-17-28"></a>}
</span></code></pre></div>
This tells nextflow how to run, and sets up the sample data as inputs.</p>
<p><em>file inputs</em></p>
<p>The <code>bam</code> parameter is a list which provides paths to the <code>.bam</code> and <code>.bai</code> sample data we will use to test the nextflow translation. From here, we can refer to the indexed bam input as <code>params.bam</code> in other files. The <code>dbsnp</code> and <code>reference</code> params follow this same pattern. </p>
<p><em>non-file inputs</em></p>
<p>We also set up a <code>NULL_VALUE</code> param which we use as a <em>placeholder</em> for a null value. <br> 
In this case we are providing null values for the <code>gvcf_gq_bands</code>, <code>contamination_fraction</code>, <code>max_alternate_alleles</code>, <code>ploidy</code> and <code>read_filter</code> inputs as they are all optional.</p>
<p><br></p>
<p><strong>Creating Workflow &amp; Passing Data</strong> </p>
<p>Now that we have the <code>nextflow.config</code> file set up, we will add a few lines to <code>gatk_haplotype_caller.nf</code> to turn it into a workflow. </p>
<p>Copy and paste the following lines at the top of <code>gatk_haplotype_caller.nf</code>:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>ch_bam = Channel.fromPath( params.bam ).toList()
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a>ch_dbsnp = Channel.fromPath( params.dbsnp ).toList()
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a>ch_reference = Channel.fromPath( params.reference ).toList()
</span><span id="__span-18-4"><a id="__codelineno-18-4" name="__codelineno-18-4" href="#__codelineno-18-4"></a>
</span><span id="__span-18-5"><a id="__codelineno-18-5" name="__codelineno-18-5" href="#__codelineno-18-5"></a>workflow {
</span><span id="__span-18-6"><a id="__codelineno-18-6" name="__codelineno-18-6" href="#__codelineno-18-6"></a>
</span><span id="__span-18-7"><a id="__codelineno-18-7" name="__codelineno-18-7" href="#__codelineno-18-7"></a>    GATK_HAPLOTYPE_CALLER(
</span><span id="__span-18-8"><a id="__codelineno-18-8" name="__codelineno-18-8" href="#__codelineno-18-8"></a>        ch_bam,
</span><span id="__span-18-9"><a id="__codelineno-18-9" name="__codelineno-18-9" href="#__codelineno-18-9"></a>        ch_reference,
</span><span id="__span-18-10"><a id="__codelineno-18-10" name="__codelineno-18-10" href="#__codelineno-18-10"></a>        ch_dbsnp,
</span><span id="__span-18-11"><a id="__codelineno-18-11" name="__codelineno-18-11" href="#__codelineno-18-11"></a>        params.intervals,
</span><span id="__span-18-12"><a id="__codelineno-18-12" name="__codelineno-18-12" href="#__codelineno-18-12"></a>        params.gvcf_gq_bands,
</span><span id="__span-18-13"><a id="__codelineno-18-13" name="__codelineno-18-13" href="#__codelineno-18-13"></a>        params.emit_reference_confidence,
</span><span id="__span-18-14"><a id="__codelineno-18-14" name="__codelineno-18-14" href="#__codelineno-18-14"></a>        params.contamination_fraction,
</span><span id="__span-18-15"><a id="__codelineno-18-15" name="__codelineno-18-15" href="#__codelineno-18-15"></a>        params.max_alternate_alleles,
</span><span id="__span-18-16"><a id="__codelineno-18-16" name="__codelineno-18-16" href="#__codelineno-18-16"></a>        params.ploidy,
</span><span id="__span-18-17"><a id="__codelineno-18-17" name="__codelineno-18-17" href="#__codelineno-18-17"></a>        params.read_filter,
</span><span id="__span-18-18"><a id="__codelineno-18-18" name="__codelineno-18-18" href="#__codelineno-18-18"></a>    )
</span><span id="__span-18-19"><a id="__codelineno-18-19" name="__codelineno-18-19" href="#__codelineno-18-19"></a>
</span><span id="__span-18-20"><a id="__codelineno-18-20" name="__codelineno-18-20" href="#__codelineno-18-20"></a>}
</span></code></pre></div>
<p>The first 3 lines create nextflow <code>Channels</code> for our <code>bam</code>, <code>dbsnp</code>, and <code>reference</code> inputs and ensures they are lists.</p>
<p>The <code>Channel.toList()</code> aspect collects our files into a list, as the primary &amp; secondary files for these datatypes must be passed together as a tuple.</p>
<p>The <code>params.bam</code>, <code>params.dbsnp</code> and <code>params.reference</code> global variables we set up previously are used to supply the paths to our sample data for these channels.</p>
<p>The new <code>workflow {}</code> section declares the main workflow entry point. <br>
When we run this file, nextflow will look for this section and run the workflow contained within. </p>
<p>In our case, the workflow only contains a single task, which runs the <code>GATK_HAPLOTYPE_CALLER</code> process defined below the workflow section. We call <code>GATK_HAPLOTYPE_CALLER</code> by feeding inputs in the correct order, using the channels we declared at the top of the file, and variables we set up in the global <code>params</code> object. </p>
<p><br></p>
<p><strong>Running Our Workflow</strong></p>
<p>Ensure you are in the <code>haplotype_caller/</code> working directory, where <code>nextflow.config</code> and <code>gatk_haplotype_caller.nf</code> reside. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>cd haplotype_caller
</span></code></pre></div></p>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>nextflow run gatk_haplotype_caller.nf
</span></code></pre></div></p>
<p>Nextflow will automatically check if there is a <code>nextflow.config</code> file in the working directory, and if so will use that to configure itself. Our inputs are supplied in <code>nextflow.config</code> alongside the dsl2 &amp; singularity config, so it should run without issue. </p>
<p>Once completed, we can check the <code>outputs/</code> folder to view our results. <br>
If everything went well, the <code>outputs/</code> folder should contain 2 files: </p>
<ul>
<li><code>output.g.vcf.gz</code></li>
<li><code>output.g.vcf.gz.tbi</code></li>
</ul>
<p><br></p>
<h3 id="conclusion_1">Conclusion<a class="headerlink" href="#conclusion_1" title="Permanent link">&para;</a></h3>
<p>In this section we explored how to translate the <code>gatk_haplotype_caller</code> CWL CommandLineTool to a Nextflow process. </p>
<p>This is a more real-world situation, where the CommandLineTool has multiple inputs, secondary files, and optionality. </p>
<p>If needed, you can check the <code>final/cwl/gatk_haplotype_caller/</code> folder inside our <code>data/</code> directory which contains the files we created in this tutorial as reference.  </p>
<p>Now that we have translated two CWL CommandLineTools, we will translate a full CWL Workflow into Nextflow.</p>
<p><br></p>
<hr />
<h2 id="section-13-align-sort-markdup-workflow-cwl">Section 1.3: Align Sort Markdup Workflow (CWL)<a class="headerlink" href="#section-13-align-sort-markdup-workflow-cwl" title="Permanent link">&para;</a></h2>
<h3 id="introduction_2">Introduction<a class="headerlink" href="#introduction_2" title="Permanent link">&para;</a></h3>
<p>This section demonstrates translation of a CWL Workflow to Nextflow using <code>janis translate</code>. </p>
<p>The workflow used in this tutorial is taken from the <a href="https://www.genome.wustl.edu/">McDonnell Genome Institute</a> (MGI) <a href="https://github.com/genome/analysis-workflows">analysis-workflows</a> repository. </p>
<p>This resource stores publically available analysis pipelines for genomics data. <br>
It is a fantastic piece of research software, and the authors thank MGI for their contribution to open-source research software. </p>
<p>The workflow using in this tutorial - <a href="https://github.com/genome/analysis-workflows/blob/master/definitions/subworkflows/align_sort_markdup.cwl">align_sort_markdup.cwl</a> - accepts multiple unaligned readsets as input and produces a single polished alignment bam file. </p>
<p><strong><em>Main Inputs</em></strong> </p>
<ul>
<li>Unaligned reads stored across multiple BAM files</li>
<li>Reference genome index</li>
</ul>
<p><strong><em>Main Outputs</em></strong></p>
<ul>
<li>Single BAM file storing alignment for all readsets</li>
</ul>
<p><strong><em>Steps</em></strong></p>
<ul>
<li>Read alignment (run in parallel across all readsets) - <code>bwa mem</code></li>
<li>Merging alignment BAM files to single file - <code>samtools merge</code></li>
<li>Sorting merged BAM by coordinate - <code>sambamba sort</code></li>
<li>Tagging duplicate reads in alignment - <code>picard MarkDuplicates</code> </li>
<li>Indexing final BAM - <code>samtools index</code></li>
</ul>
<p><br></p>
<h3 id="janis-translate_2">Janis Translate<a class="headerlink" href="#janis-translate_2" title="Permanent link">&para;</a></h3>
<p>Previously we were translating single CWL CommandLineTools, but in this section we are translating a full CWL workflow. </p>
<p>To translate a workflow, we supply the main CWL workflow file to janis translate. <br>
In addition to the main CWL Workflow, all Subworkflows / CommandLineTools will be detected &amp; translated.</p>
<p>First, let&rsquo;s make sure we&rsquo;re back in the main training directory
<div class="language-text highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>cd /home2/training
</span></code></pre></div></p>
<p>In this section, we are translating CWL -&gt; Nextflow, and our source CWL file is located at <code>data/source/cwl/align_sort_markdup/subworkflows/align_sort_markdup.cwl</code> relative to this document.</p>
<p>We will add the <code>-o align_sort_markdup</code> argument to specify this as the output directory. </p>
<p>To translate, run the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>singularity exec ~/janis.sif janis translate -o align_sort_markdup --from cwl --to nextflow data/source/cwl/align_sort_markdup/subworkflows/align_sort_markdup.cwl
</span></code></pre></div></p>
<p><br></p>
<h3 id="translation-output">Translation Output<a class="headerlink" href="#translation-output" title="Permanent link">&para;</a></h3>
<p>The output translation will contain multiple files and directories.<br>
You will see a folder called <code>align_sort_markdup/</code> appear - inside this folder, we should see the following structure:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>align_sort_markdup
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a> main.nf                             # main workflow (align_sort_markdup)
</span><span id="__span-23-3"><a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a> modules                             # folder containing nextflow processes
</span><span id="__span-23-4"><a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a>    align_and_tag.nf
</span><span id="__span-23-5"><a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a>    index_bam.nf
</span><span id="__span-23-6"><a id="__codelineno-23-6" name="__codelineno-23-6" href="#__codelineno-23-6"></a>    mark_duplicates_and_sort.nf
</span><span id="__span-23-7"><a id="__codelineno-23-7" name="__codelineno-23-7" href="#__codelineno-23-7"></a>    merge_bams_samtools.nf
</span><span id="__span-23-8"><a id="__codelineno-23-8" name="__codelineno-23-8" href="#__codelineno-23-8"></a>    name_sort.nf
</span><span id="__span-23-9"><a id="__codelineno-23-9" name="__codelineno-23-9" href="#__codelineno-23-9"></a> nextflow.config                     # config file to supply input information
</span><span id="__span-23-10"><a id="__codelineno-23-10" name="__codelineno-23-10" href="#__codelineno-23-10"></a> subworkflows                        # folder containing nextflow subworkflows
</span><span id="__span-23-11"><a id="__codelineno-23-11" name="__codelineno-23-11" href="#__codelineno-23-11"></a>    align.nf    
</span><span id="__span-23-12"><a id="__codelineno-23-12" name="__codelineno-23-12" href="#__codelineno-23-12"></a> templates                           # folder containing any scripts used by processes
</span><span id="__span-23-13"><a id="__codelineno-23-13" name="__codelineno-23-13" href="#__codelineno-23-13"></a>     markduplicates_helper.sh
</span></code></pre></div>
<p>Now that we have performed translation using <code>janis translate</code>, we need to check the translated workflow for correctness.  </p>
<p>From here, we will do a test-run of the workflow using sample data, and make manual adjustments to the translated workflow where needed. </p>
<p><br></p>
<h3 id="running-the-translated-workflow">Running the Translated Workflow<a class="headerlink" href="#running-the-translated-workflow" title="Permanent link">&para;</a></h3>
<p><strong>Inspect main.nf</strong></p>
<p>The main workflow translation appears as <code>main.nf</code> in the <code>align_sort_markdup/</code> folder. <br></p>
<p>This filename is just a convention, and we use it to provide clarity about the main entry point of the workflow. <br>
In our case <code>main.nf</code> is equivalent to the main CWL Workflow file we translated: <code>align_sort_markdup.cwl</code>. </p>
<p><br></p>
<blockquote>
<p>NOTE: <br>
Before continuing, feel free to have a look at the other nextflow files which have been generated during translation:<br>
Each CWL Subworkflow appears as a nextflow <code>workflow</code> in the <code>subworkflows/</code> directory.<br>
Each CWL CommandLineTool appears as a nextflow <code>process</code> in the <code>modules/</code> directory. </p>
</blockquote>
<p><br></p>
<p>In <code>main.nf</code> we see the nextflow workflows / processes called by the main workflow:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>include { ALIGN } from &#39;./subworkflows/align&#39;
</span><span id="__span-24-2"><a id="__codelineno-24-2" name="__codelineno-24-2" href="#__codelineno-24-2"></a>include { INDEX_BAM } from &#39;./modules/index_bam&#39;
</span><span id="__span-24-3"><a id="__codelineno-24-3" name="__codelineno-24-3" href="#__codelineno-24-3"></a>include { MARK_DUPLICATES_AND_SORT } from &#39;./modules/mark_duplicates_and_sort&#39;
</span><span id="__span-24-4"><a id="__codelineno-24-4" name="__codelineno-24-4" href="#__codelineno-24-4"></a>include { MERGE_BAMS_SAMTOOLS as MERGE } from &#39;./modules/merge_bams_samtools&#39;
</span><span id="__span-24-5"><a id="__codelineno-24-5" name="__codelineno-24-5" href="#__codelineno-24-5"></a>include { NAME_SORT } from &#39;./modules/name_sort&#39;
</span></code></pre></div>
<p>We also see that some nextflow <code>Channels</code> and a single <code>variable</code> have been set up. <br>
These are used to supply data according to nextflow&rsquo;s adoption of the <em>dataflow</em> programming model.
<div class="language-text highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>// data which will be passed as channels
</span><span id="__span-25-2"><a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a>ch_bams        = Channel.fromPath( params.bams ).toList()
</span><span id="__span-25-3"><a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a>ch_readgroups  = Channel.of( params.readgroups ).toList()
</span><span id="__span-25-4"><a id="__codelineno-25-4" name="__codelineno-25-4" href="#__codelineno-25-4"></a>
</span><span id="__span-25-5"><a id="__codelineno-25-5" name="__codelineno-25-5" href="#__codelineno-25-5"></a>// data which will be passed as variables
</span><span id="__span-25-6"><a id="__codelineno-25-6" name="__codelineno-25-6" href="#__codelineno-25-6"></a>reference  = params.reference.collect{ file(it) }
</span></code></pre></div></p>
<p>Focusing on the channel declarations, we want to note a few things:</p>
<ul>
<li>
<p><code>ch_bams</code> is analygous to the <em>&lsquo;bams&rsquo;</em> input in <code>align_sort_markdup.cwl</code>. <br>
It declares a queue channel which expects the data supplied via <code>params.bams</code> are <code>path</code> types. <br>
It then groups the bams together as a sole emission using <code>.toList()</code>. <br>
We will need to set up <code>params.bams</code> to supply this data.</p>
</li>
<li>
<p><code>ch_readgroups</code> is analygous to the <em>&lsquo;readgroups&rsquo;</em> input in <code>align_sort_markdup.cwl</code>. <br>
It is the same as <code>ch_bams</code>, except it requires <code>val</code> types rather than <code>path</code> types. <br>
We will need to set up <code>params.readgroups</code> to supply this data.</p>
</li>
</ul>
<p>We also see a new variable called <code>reference</code> being created. </p>
<p>This <code>reference</code> var is analygous to the <em>&lsquo;reference&rsquo;</em> input in <code>align_sort_markdup.cwl</code>. 
It collects the <em>&lsquo;reference&rsquo;</em> primary &amp; secondary files together in a list. </p>
<p>We will need to set up <code>params.reference</code> to supply this data. </p>
<p><br></p>
<blockquote>
<p>Note: Why does <code>reference</code> appear as a variable, rather than a <code>Channel</code>? <br><br>
In Nextflow, we find it is best to create Channels for data that <strong><em>moves</em></strong> through the pipeline. <br>
The <code>ch_bams</code> channel is created because we want to consume / transform the input bams during the pipeline. <br> <br>
On the other hand, the <code>reference</code> genome variable is not something we really want to consume or transform.  <br>
It&rsquo;s more like static helper data which can be used by processes, but it doesn&rsquo;t <strong><em>move</em></strong> through our pipeline, and definitely shouldn&rsquo;t be consumed or transformed in any way. </p>
</blockquote>
<p><br></p>
<blockquote>
<p>Note: <code>.toList()</code> <br><br>
Nextflow queue channels work differently to lists. <br>
Instead of supplying all items together, queue channels emit each item separately. <br> 
This results in a separate task being spawned for each item in the queue when the channel is used. <br>
As the CWL workflow input specifies that <code>bams</code> is a list, we use <code>.toList()</code> to group all items as a sole emission. <br>
This mimics a CWL array which is the datatype of the <code>bams</code> inputs. <br><br>
As it turns out, the CWL workflow ends up running the <code>align</code> step in parallel across the <code>bams</code> &amp; <code>readgroups</code> inputs. <br><br>
Parallelisation in nextflow happens by default. <br>
To facilitate this, the <code>.flatten()</code> method is called on  <code>ch_bams</code> and <code>ch_readgroups</code> when used in the <code>ALIGN</code> task. <br>
This emits items in <code>ch_bams</code> and <code>ch_readgroups</code> individually, spawning a new <code>ALIGN</code> task for each pair. <br><br>
We&rsquo;re kinda doing redundant work by calling <code>.toList()</code>, then <code>.flatten()</code> when <code>ch_bams</code> and <code>ch_readgroups</code> are used.<br>
<code>janis translate</code> isn&rsquo;t smart enough yet to detect this yet, but may do so in future. </p>
</blockquote>
<p><br></p>
<p>Continuing down <code>main.nf</code>, we see the main <code>workflow {}</code> has 5 tasks. </p>
<p>Each task has been supplied values according to the source workflow.
Comments display the name of the process/workflow input which is being fed a particular value. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>workflow {
</span><span id="__span-26-2"><a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a>
</span><span id="__span-26-3"><a id="__codelineno-26-3" name="__codelineno-26-3" href="#__codelineno-26-3"></a>    ALIGN(
</span><span id="__span-26-4"><a id="__codelineno-26-4" name="__codelineno-26-4" href="#__codelineno-26-4"></a>        ch_bams.flatten(),       // bam
</span><span id="__span-26-5"><a id="__codelineno-26-5" name="__codelineno-26-5" href="#__codelineno-26-5"></a>        ch_reference,            // reference
</span><span id="__span-26-6"><a id="__codelineno-26-6" name="__codelineno-26-6" href="#__codelineno-26-6"></a>        ch_readgroups.flatten()  // readgroup
</span><span id="__span-26-7"><a id="__codelineno-26-7" name="__codelineno-26-7" href="#__codelineno-26-7"></a>    )
</span><span id="__span-26-8"><a id="__codelineno-26-8" name="__codelineno-26-8" href="#__codelineno-26-8"></a>
</span><span id="__span-26-9"><a id="__codelineno-26-9" name="__codelineno-26-9" href="#__codelineno-26-9"></a>    INDEX_BAM(
</span><span id="__span-26-10"><a id="__codelineno-26-10" name="__codelineno-26-10" href="#__codelineno-26-10"></a>        MARK_DUPLICATES_AND_SORT.out.sorted_bam.map{ tuple -&gt; tuple[0] }  // bam
</span><span id="__span-26-11"><a id="__codelineno-26-11" name="__codelineno-26-11" href="#__codelineno-26-11"></a>    )
</span><span id="__span-26-12"><a id="__codelineno-26-12" name="__codelineno-26-12" href="#__codelineno-26-12"></a>
</span><span id="__span-26-13"><a id="__codelineno-26-13" name="__codelineno-26-13" href="#__codelineno-26-13"></a>    MARK_DUPLICATES_AND_SORT(
</span><span id="__span-26-14"><a id="__codelineno-26-14" name="__codelineno-26-14" href="#__codelineno-26-14"></a>        params.mark_duplicates_and_sort.script,  // script
</span><span id="__span-26-15"><a id="__codelineno-26-15" name="__codelineno-26-15" href="#__codelineno-26-15"></a>        NAME_SORT.out.name_sorted_bam,           // bam
</span><span id="__span-26-16"><a id="__codelineno-26-16" name="__codelineno-26-16" href="#__codelineno-26-16"></a>        params.NULL_VALUE,                       // input_sort_order
</span><span id="__span-26-17"><a id="__codelineno-26-17" name="__codelineno-26-17" href="#__codelineno-26-17"></a>        params.final_name                        // output_name
</span><span id="__span-26-18"><a id="__codelineno-26-18" name="__codelineno-26-18" href="#__codelineno-26-18"></a>    )
</span><span id="__span-26-19"><a id="__codelineno-26-19" name="__codelineno-26-19" href="#__codelineno-26-19"></a>
</span><span id="__span-26-20"><a id="__codelineno-26-20" name="__codelineno-26-20" href="#__codelineno-26-20"></a>    MERGE(
</span><span id="__span-26-21"><a id="__codelineno-26-21" name="__codelineno-26-21" href="#__codelineno-26-21"></a>        ALIGN.out.tagged_bam.toList(),  // bams
</span><span id="__span-26-22"><a id="__codelineno-26-22" name="__codelineno-26-22" href="#__codelineno-26-22"></a>        params.final_name               // name
</span><span id="__span-26-23"><a id="__codelineno-26-23" name="__codelineno-26-23" href="#__codelineno-26-23"></a>    )
</span><span id="__span-26-24"><a id="__codelineno-26-24" name="__codelineno-26-24" href="#__codelineno-26-24"></a>
</span><span id="__span-26-25"><a id="__codelineno-26-25" name="__codelineno-26-25" href="#__codelineno-26-25"></a>    NAME_SORT(
</span><span id="__span-26-26"><a id="__codelineno-26-26" name="__codelineno-26-26" href="#__codelineno-26-26"></a>        MERGE.out.merged_bam  // bam
</span><span id="__span-26-27"><a id="__codelineno-26-27" name="__codelineno-26-27" href="#__codelineno-26-27"></a>    )
</span><span id="__span-26-28"><a id="__codelineno-26-28" name="__codelineno-26-28" href="#__codelineno-26-28"></a>
</span><span id="__span-26-29"><a id="__codelineno-26-29" name="__codelineno-26-29" href="#__codelineno-26-29"></a>}
</span></code></pre></div>
<p>Before <code>main.nf</code> can be run, we will need to supply values for the <code>params</code> variables. 
This is done in <code>nextflow.config</code>. </p>
<p><br></p>
<p><strong>Inspect nextflow.config</strong></p>
<p>To test the translated workflow, we will first set up workflow inputs in <code>nextflow.config</code>. </p>
<p>Before running a workflow, nextflow will attempt to open <code>nextflow.config</code> and read in config information and global <em>param</em> variables from this file. 
We use this file to tell nextflow how to run and to supply workflow inputs.</p>
<p>Inside the <code>align_sort_markdup/</code> folder you will see that <code>nextflow.config</code> is already provided. </p>
<p>Janis translate creates this file to provide clarity about the necessary workflow inputs, and to set some other config variables. </p>
<p>Open <code>nextflow.config</code> and have a look at the contents. It should look similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-27-2"><a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a>singularity.enabled = true
</span><span id="__span-27-3"><a id="__codelineno-27-3" name="__codelineno-27-3" href="#__codelineno-27-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-27-4"><a id="__codelineno-27-4" name="__codelineno-27-4" href="#__codelineno-27-4"></a>
</span><span id="__span-27-5"><a id="__codelineno-27-5" name="__codelineno-27-5" href="#__codelineno-27-5"></a>params {
</span><span id="__span-27-6"><a id="__codelineno-27-6" name="__codelineno-27-6" href="#__codelineno-27-6"></a>
</span><span id="__span-27-7"><a id="__codelineno-27-7" name="__codelineno-27-7" href="#__codelineno-27-7"></a>    // Placeholder for null values.
</span><span id="__span-27-8"><a id="__codelineno-27-8" name="__codelineno-27-8" href="#__codelineno-27-8"></a>    // Do not alter unless you know what you are doing.
</span><span id="__span-27-9"><a id="__codelineno-27-9" name="__codelineno-27-9" href="#__codelineno-27-9"></a>    NULL_VALUE = &#39;NULL&#39;
</span><span id="__span-27-10"><a id="__codelineno-27-10" name="__codelineno-27-10" href="#__codelineno-27-10"></a>
</span><span id="__span-27-11"><a id="__codelineno-27-11" name="__codelineno-27-11" href="#__codelineno-27-11"></a>    // WORKFLOW OUTPUT DIRECTORY
</span><span id="__span-27-12"><a id="__codelineno-27-12" name="__codelineno-27-12" href="#__codelineno-27-12"></a>    outdir  = &#39;./outputs&#39;
</span><span id="__span-27-13"><a id="__codelineno-27-13" name="__codelineno-27-13" href="#__codelineno-27-13"></a>
</span><span id="__span-27-14"><a id="__codelineno-27-14" name="__codelineno-27-14" href="#__codelineno-27-14"></a>    // INPUTS (MANDATORY)
</span><span id="__span-27-15"><a id="__codelineno-27-15" name="__codelineno-27-15" href="#__codelineno-27-15"></a>    bams        = []          // (MANDATORY array)             eg. [file1, ...]
</span><span id="__span-27-16"><a id="__codelineno-27-16" name="__codelineno-27-16" href="#__codelineno-27-16"></a>    reference   = []          // (MANDATORY fastawithindexes)  eg. [fasta, amb, ann, bwt, dict, fai, pac, sa]
</span><span id="__span-27-17"><a id="__codelineno-27-17" name="__codelineno-27-17" href="#__codelineno-27-17"></a>    readgroups  = NULL_VALUE  // (MANDATORY array)             eg. [string1, ...]
</span><span id="__span-27-18"><a id="__codelineno-27-18" name="__codelineno-27-18" href="#__codelineno-27-18"></a>
</span><span id="__span-27-19"><a id="__codelineno-27-19" name="__codelineno-27-19" href="#__codelineno-27-19"></a>    // INPUTS (OPTIONAL)
</span><span id="__span-27-20"><a id="__codelineno-27-20" name="__codelineno-27-20" href="#__codelineno-27-20"></a>    final_name  = &quot;final.bam&quot; 
</span><span id="__span-27-21"><a id="__codelineno-27-21" name="__codelineno-27-21" href="#__codelineno-27-21"></a>
</span><span id="__span-27-22"><a id="__codelineno-27-22" name="__codelineno-27-22" href="#__codelineno-27-22"></a>    // PROCESS: ALIGN_AND_TAG
</span><span id="__span-27-23"><a id="__codelineno-27-23" name="__codelineno-27-23" href="#__codelineno-27-23"></a>    align_and_tag.cpus    = 8     
</span><span id="__span-27-24"><a id="__codelineno-27-24" name="__codelineno-27-24" href="#__codelineno-27-24"></a>    align_and_tag.memory  = 20000 
</span><span id="__span-27-25"><a id="__codelineno-27-25" name="__codelineno-27-25" href="#__codelineno-27-25"></a>
</span><span id="__span-27-26"><a id="__codelineno-27-26" name="__codelineno-27-26" href="#__codelineno-27-26"></a>    // PROCESS: INDEX_BAM
</span><span id="__span-27-27"><a id="__codelineno-27-27" name="__codelineno-27-27" href="#__codelineno-27-27"></a>    index_bam.memory  = 4000 
</span><span id="__span-27-28"><a id="__codelineno-27-28" name="__codelineno-27-28" href="#__codelineno-27-28"></a>
</span><span id="__span-27-29"><a id="__codelineno-27-29" name="__codelineno-27-29" href="#__codelineno-27-29"></a>    // PROCESS: MARK_DUPLICATES_AND_SORT
</span><span id="__span-27-30"><a id="__codelineno-27-30" name="__codelineno-27-30" href="#__codelineno-27-30"></a>    mark_duplicates_and_sort.script  = &quot;/home2/training/align_sort_markdup/templates/markduplicates_helper.sh&quot; 
</span><span id="__span-27-31"><a id="__codelineno-27-31" name="__codelineno-27-31" href="#__codelineno-27-31"></a>    mark_duplicates_and_sort.cpus    = 8                                                                       
</span><span id="__span-27-32"><a id="__codelineno-27-32" name="__codelineno-27-32" href="#__codelineno-27-32"></a>    mark_duplicates_and_sort.memory  = 40000                                                                   
</span><span id="__span-27-33"><a id="__codelineno-27-33" name="__codelineno-27-33" href="#__codelineno-27-33"></a>
</span><span id="__span-27-34"><a id="__codelineno-27-34" name="__codelineno-27-34" href="#__codelineno-27-34"></a>    // PROCESS: MERGE_BAMS_SAMTOOLS
</span><span id="__span-27-35"><a id="__codelineno-27-35" name="__codelineno-27-35" href="#__codelineno-27-35"></a>    merge_bams_samtools.cpus    = 4    
</span><span id="__span-27-36"><a id="__codelineno-27-36" name="__codelineno-27-36" href="#__codelineno-27-36"></a>    merge_bams_samtools.memory  = 8000 
</span><span id="__span-27-37"><a id="__codelineno-27-37" name="__codelineno-27-37" href="#__codelineno-27-37"></a>
</span><span id="__span-27-38"><a id="__codelineno-27-38" name="__codelineno-27-38" href="#__codelineno-27-38"></a>    // PROCESS: NAME_SORT
</span><span id="__span-27-39"><a id="__codelineno-27-39" name="__codelineno-27-39" href="#__codelineno-27-39"></a>    name_sort.cpus    = 8     
</span><span id="__span-27-40"><a id="__codelineno-27-40" name="__codelineno-27-40" href="#__codelineno-27-40"></a>    name_sort.memory  = 26000 
</span><span id="__span-27-41"><a id="__codelineno-27-41" name="__codelineno-27-41" href="#__codelineno-27-41"></a>
</span><span id="__span-27-42"><a id="__codelineno-27-42" name="__codelineno-27-42" href="#__codelineno-27-42"></a>}
</span></code></pre></div>
<p><br></p>
<blockquote>
<p>NOTE: <code>NULL_VALUE = 'NULL'</code><br><br>
Nextflow doesn&rsquo;t like <code>null</code> values to be passed to process inputs. <br>
This is a challenge for translation as other languages allow <code>optional</code> inputs. <br>
To get around this, Janis Translate sets the <code>params.NULL_VALUE</code> variable as a <code>null</code> placeholder for <code>val</code> type inputs. <br>
You will see this being used in nextflow processes to do optionality checking.</p>
</blockquote>
<p><br></p>
<p>The auto-generated <code>nextflow.config</code> splits up workflow inputs using some headings. </p>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>// INPUTS (MANDATORY)
</span></code></pre></div>
Workflow inputs which are required to run the workflow. We must provide a value for these. </p>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>// INPUTS (OPTIONAL)
</span></code></pre></div>
Workflow inputs which are not required to run the workflow. These are optional. </p>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>// PROCESS: ALIGN_AND_TAG
</span></code></pre></div>
Inputs which are specific to a particular process. These are usually static values rather than files.<br />
May be <em>mandatory</em> or <em>optional</em>. </p>
<p><br></p>
<p><strong>Setting up Workflow Inputs</strong></p>
<p>Janis Translate will enter values for workflow inputs where possible. 
Others need to be manually supplied as they are specific to the input data you wish to use. </p>
<p>In our case, we need to supply values for those under the <code>// INPUTS (MANDATORY)</code> heading. 
Specifically, we need to provide sample data for the <code>bams</code>, <code>reference</code>, and <code>readgroups</code> inputs. </p>
<p>Additionally, the <code>cpu</code> and <code>memory</code> params need to be changed. </p>
<p>The source CWL Workflow has compute requirements set greater than what is available on your VM. </p>
<p>Copy and paste the following text, to override the current <code>nextflow.config</code> file:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-31-2"><a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a>singularity.enabled = true
</span><span id="__span-31-3"><a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-31-4"><a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a>
</span><span id="__span-31-5"><a id="__codelineno-31-5" name="__codelineno-31-5" href="#__codelineno-31-5"></a>params {
</span><span id="__span-31-6"><a id="__codelineno-31-6" name="__codelineno-31-6" href="#__codelineno-31-6"></a>
</span><span id="__span-31-7"><a id="__codelineno-31-7" name="__codelineno-31-7" href="#__codelineno-31-7"></a>    // Placeholder for null values.
</span><span id="__span-31-8"><a id="__codelineno-31-8" name="__codelineno-31-8" href="#__codelineno-31-8"></a>    // Do not alter unless you know what you are doing.
</span><span id="__span-31-9"><a id="__codelineno-31-9" name="__codelineno-31-9" href="#__codelineno-31-9"></a>    NULL_VALUE = &#39;NULL&#39;
</span><span id="__span-31-10"><a id="__codelineno-31-10" name="__codelineno-31-10" href="#__codelineno-31-10"></a>
</span><span id="__span-31-11"><a id="__codelineno-31-11" name="__codelineno-31-11" href="#__codelineno-31-11"></a>    // WORKFLOW OUTPUT DIRECTORY
</span><span id="__span-31-12"><a id="__codelineno-31-12" name="__codelineno-31-12" href="#__codelineno-31-12"></a>    outdir  = &#39;./outputs&#39;
</span><span id="__span-31-13"><a id="__codelineno-31-13" name="__codelineno-31-13" href="#__codelineno-31-13"></a>
</span><span id="__span-31-14"><a id="__codelineno-31-14" name="__codelineno-31-14" href="#__codelineno-31-14"></a>    // INPUTS (MANDATORY)
</span><span id="__span-31-15"><a id="__codelineno-31-15" name="__codelineno-31-15" href="#__codelineno-31-15"></a>    bams        = [
</span><span id="__span-31-16"><a id="__codelineno-31-16" name="__codelineno-31-16" href="#__codelineno-31-16"></a>        &quot;/home2/training/data/sample_data/cwl/2895499223.bam&quot;,
</span><span id="__span-31-17"><a id="__codelineno-31-17" name="__codelineno-31-17" href="#__codelineno-31-17"></a>        &quot;/home2/training/data/sample_data/cwl/2895499237.bam&quot;,
</span><span id="__span-31-18"><a id="__codelineno-31-18" name="__codelineno-31-18" href="#__codelineno-31-18"></a>    ]
</span><span id="__span-31-19"><a id="__codelineno-31-19" name="__codelineno-31-19" href="#__codelineno-31-19"></a>    reference   = [
</span><span id="__span-31-20"><a id="__codelineno-31-20" name="__codelineno-31-20" href="#__codelineno-31-20"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa&quot;,
</span><span id="__span-31-21"><a id="__codelineno-31-21" name="__codelineno-31-21" href="#__codelineno-31-21"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.amb&quot;,
</span><span id="__span-31-22"><a id="__codelineno-31-22" name="__codelineno-31-22" href="#__codelineno-31-22"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.ann&quot;,
</span><span id="__span-31-23"><a id="__codelineno-31-23" name="__codelineno-31-23" href="#__codelineno-31-23"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.bwt&quot;,
</span><span id="__span-31-24"><a id="__codelineno-31-24" name="__codelineno-31-24" href="#__codelineno-31-24"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.fai&quot;,
</span><span id="__span-31-25"><a id="__codelineno-31-25" name="__codelineno-31-25" href="#__codelineno-31-25"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.dict&quot;,
</span><span id="__span-31-26"><a id="__codelineno-31-26" name="__codelineno-31-26" href="#__codelineno-31-26"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.pac&quot;,
</span><span id="__span-31-27"><a id="__codelineno-31-27" name="__codelineno-31-27" href="#__codelineno-31-27"></a>        &quot;/home2/training/data/sample_data/cwl/chr17_test.fa.sa&quot;,
</span><span id="__span-31-28"><a id="__codelineno-31-28" name="__codelineno-31-28" href="#__codelineno-31-28"></a>    ] 
</span><span id="__span-31-29"><a id="__codelineno-31-29" name="__codelineno-31-29" href="#__codelineno-31-29"></a>    readgroups  = [
</span><span id="__span-31-30"><a id="__codelineno-31-30" name="__codelineno-31-30" href="#__codelineno-31-30"></a>        &#39;@RG\tID:2895499223\tPU:H7HY2CCXX.3.ATCACGGT\tSM:H_NJ-HCC1395-HCC1395\tLB:H_NJ-HCC1395-HCC1395-lg24-lib1\tPL:Illumina\tCN:WUGSC&#39;,
</span><span id="__span-31-31"><a id="__codelineno-31-31" name="__codelineno-31-31" href="#__codelineno-31-31"></a>        &#39;@RG\tID:2895499237\tPU:H7HY2CCXX.4.ATCACGGT\tSM:H_NJ-HCC1395-HCC1395\tLB:H_NJ-HCC1395-HCC1395-lg24-lib1\tPL:Illumina\tCN:WUGSC&#39;
</span><span id="__span-31-32"><a id="__codelineno-31-32" name="__codelineno-31-32" href="#__codelineno-31-32"></a>    ]
</span><span id="__span-31-33"><a id="__codelineno-31-33" name="__codelineno-31-33" href="#__codelineno-31-33"></a>
</span><span id="__span-31-34"><a id="__codelineno-31-34" name="__codelineno-31-34" href="#__codelineno-31-34"></a>    // INPUTS (OPTIONAL)
</span><span id="__span-31-35"><a id="__codelineno-31-35" name="__codelineno-31-35" href="#__codelineno-31-35"></a>    final_name  = &quot;final.bam&quot; 
</span><span id="__span-31-36"><a id="__codelineno-31-36" name="__codelineno-31-36" href="#__codelineno-31-36"></a>
</span><span id="__span-31-37"><a id="__codelineno-31-37" name="__codelineno-31-37" href="#__codelineno-31-37"></a>    // PROCESS: ALIGN_AND_TAG
</span><span id="__span-31-38"><a id="__codelineno-31-38" name="__codelineno-31-38" href="#__codelineno-31-38"></a>    align_and_tag.cpus    = 4     
</span><span id="__span-31-39"><a id="__codelineno-31-39" name="__codelineno-31-39" href="#__codelineno-31-39"></a>    align_and_tag.memory  = 10000 
</span><span id="__span-31-40"><a id="__codelineno-31-40" name="__codelineno-31-40" href="#__codelineno-31-40"></a>
</span><span id="__span-31-41"><a id="__codelineno-31-41" name="__codelineno-31-41" href="#__codelineno-31-41"></a>    // PROCESS: INDEX_BAM
</span><span id="__span-31-42"><a id="__codelineno-31-42" name="__codelineno-31-42" href="#__codelineno-31-42"></a>    index_bam.memory  = 10000 
</span><span id="__span-31-43"><a id="__codelineno-31-43" name="__codelineno-31-43" href="#__codelineno-31-43"></a>
</span><span id="__span-31-44"><a id="__codelineno-31-44" name="__codelineno-31-44" href="#__codelineno-31-44"></a>    // PROCESS: MARK_DUPLICATES_AND_SORT
</span><span id="__span-31-45"><a id="__codelineno-31-45" name="__codelineno-31-45" href="#__codelineno-31-45"></a>    mark_duplicates_and_sort.script  = &quot;/home2/training/align_sort_markdup/templates/markduplicates_helper.sh&quot; 
</span><span id="__span-31-46"><a id="__codelineno-31-46" name="__codelineno-31-46" href="#__codelineno-31-46"></a>    mark_duplicates_and_sort.cpus    = 4
</span><span id="__span-31-47"><a id="__codelineno-31-47" name="__codelineno-31-47" href="#__codelineno-31-47"></a>    mark_duplicates_and_sort.memory  = 10000                         
</span><span id="__span-31-48"><a id="__codelineno-31-48" name="__codelineno-31-48" href="#__codelineno-31-48"></a>
</span><span id="__span-31-49"><a id="__codelineno-31-49" name="__codelineno-31-49" href="#__codelineno-31-49"></a>    // PROCESS: MERGE_BAMS_SAMTOOLS
</span><span id="__span-31-50"><a id="__codelineno-31-50" name="__codelineno-31-50" href="#__codelineno-31-50"></a>    merge_bams_samtools.cpus    = 4    
</span><span id="__span-31-51"><a id="__codelineno-31-51" name="__codelineno-31-51" href="#__codelineno-31-51"></a>    merge_bams_samtools.memory  = 10000 
</span><span id="__span-31-52"><a id="__codelineno-31-52" name="__codelineno-31-52" href="#__codelineno-31-52"></a>
</span><span id="__span-31-53"><a id="__codelineno-31-53" name="__codelineno-31-53" href="#__codelineno-31-53"></a>    // PROCESS: NAME_SORT
</span><span id="__span-31-54"><a id="__codelineno-31-54" name="__codelineno-31-54" href="#__codelineno-31-54"></a>    name_sort.cpus    = 4     
</span><span id="__span-31-55"><a id="__codelineno-31-55" name="__codelineno-31-55" href="#__codelineno-31-55"></a>    name_sort.memory  = 10000 
</span><span id="__span-31-56"><a id="__codelineno-31-56" name="__codelineno-31-56" href="#__codelineno-31-56"></a>
</span><span id="__span-31-57"><a id="__codelineno-31-57" name="__codelineno-31-57" href="#__codelineno-31-57"></a>}
</span></code></pre></div>
<p><br></p>
<p><strong>Run the Workflow</strong></p>
<p>Ensure you are in the <code>align_sort_markdup/</code> working directory, where <code>nextflow.config</code> and <code>main.nf</code> reside. </p>
<p>If not, use the following to change directory. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>cd align_sort_markdup
</span></code></pre></div></p>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p>While the workflow runs, you will encounter this error:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-34-1"><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>Access to &#39;MARK_DUPLICATES_AND_SORT.out&#39; is undefined since the process &#39;MARK_DUPLICATES_AND_SORT&#39; has not been invoked before accessing the output attribute
</span></code></pre></div>
<p>This is somewhat expected. Janis translate doesn&rsquo;t produce perfect translations - just the best it can do. </p>
<p>This is the first of <strong><em>3</em></strong> errors we will encounter and fix while making this workflow runnable. </p>
<p><br></p>
<h3 id="manual-adjustments_2">Manual Adjustments<a class="headerlink" href="#manual-adjustments_2" title="Permanent link">&para;</a></h3>
<p>Translations performed by <code>janis translate</code> often require manual changes due to the difficulty of translating between languages with non-overlapping feature sets. </p>
<p>In this section we will fix <strong><em>3 errors</em></strong> to bring the translation to a finished state. </p>
<p><br></p>
<blockquote>
<p>NOTE <br>
If having trouble during this section, the finished workflow is available in the <code>data/final/cwl/align_sort_markdup</code> folder for reference.</p>
</blockquote>
<p><br></p>
<h3 id="error-1-process-order">Error 1: Process Order<a class="headerlink" href="#error-1-process-order" title="Permanent link">&para;</a></h3>
<p><strong>Error message</strong></p>
<p>The first issue we need to address is caused by tasks being in the wrong order. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-35-1"><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a>Access to &#39;MARK_DUPLICATES_AND_SORT.out&#39; is undefined since the process &#39;MARK_DUPLICATES_AND_SORT&#39; has not been invoked before accessing the output attribute
</span></code></pre></div>
<p>This nextflow error message is quite informative, and tells us that a task is trying to access the output of <code>MARK_DUPLICATES_AND_SORT</code> before it has run. </p>
<p>The offending task is <code>INDEX_BAM</code> as it uses <code>MARK_DUPLICATES_AND_SORT.out.sorted_bam.map{ tuple -&gt; tuple[0] }</code> as an input value to the process. </p>
<p><br></p>
<p><strong>Troubleshooting</strong></p>
<p>It seems that a nextflow process is trying to run before a prior process has finished. </p>
<p>This is most likely due to the processes in our <code>main.nf</code> workflow being in the wrong order. </p>
<p>Let&rsquo;s look at the source CWL <code>align_sort_markdup.cwl</code> workflow to view the correct order of steps:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-36-1"><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a>steps:
</span><span id="__span-36-2"><a id="__codelineno-36-2" name="__codelineno-36-2" href="#__codelineno-36-2"></a>    align:
</span><span id="__span-36-3"><a id="__codelineno-36-3" name="__codelineno-36-3" href="#__codelineno-36-3"></a>        ...
</span><span id="__span-36-4"><a id="__codelineno-36-4" name="__codelineno-36-4" href="#__codelineno-36-4"></a>    merge:
</span><span id="__span-36-5"><a id="__codelineno-36-5" name="__codelineno-36-5" href="#__codelineno-36-5"></a>        ...
</span><span id="__span-36-6"><a id="__codelineno-36-6" name="__codelineno-36-6" href="#__codelineno-36-6"></a>    name_sort:
</span><span id="__span-36-7"><a id="__codelineno-36-7" name="__codelineno-36-7" href="#__codelineno-36-7"></a>        ...
</span><span id="__span-36-8"><a id="__codelineno-36-8" name="__codelineno-36-8" href="#__codelineno-36-8"></a>    mark_duplicates_and_sort:
</span><span id="__span-36-9"><a id="__codelineno-36-9" name="__codelineno-36-9" href="#__codelineno-36-9"></a>        ...
</span><span id="__span-36-10"><a id="__codelineno-36-10" name="__codelineno-36-10" href="#__codelineno-36-10"></a>    index_bam:
</span><span id="__span-36-11"><a id="__codelineno-36-11" name="__codelineno-36-11" href="#__codelineno-36-11"></a>        ...
</span></code></pre></div>
<p>This differs from our translated <code>main.nf</code> which has the following order:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-37-1"><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a>workflow {
</span><span id="__span-37-2"><a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a>
</span><span id="__span-37-3"><a id="__codelineno-37-3" name="__codelineno-37-3" href="#__codelineno-37-3"></a>    ALIGN(
</span><span id="__span-37-4"><a id="__codelineno-37-4" name="__codelineno-37-4" href="#__codelineno-37-4"></a>        ...
</span><span id="__span-37-5"><a id="__codelineno-37-5" name="__codelineno-37-5" href="#__codelineno-37-5"></a>    )
</span><span id="__span-37-6"><a id="__codelineno-37-6" name="__codelineno-37-6" href="#__codelineno-37-6"></a>
</span><span id="__span-37-7"><a id="__codelineno-37-7" name="__codelineno-37-7" href="#__codelineno-37-7"></a>    INDEX_BAM(
</span><span id="__span-37-8"><a id="__codelineno-37-8" name="__codelineno-37-8" href="#__codelineno-37-8"></a>        ...
</span><span id="__span-37-9"><a id="__codelineno-37-9" name="__codelineno-37-9" href="#__codelineno-37-9"></a>    )
</span><span id="__span-37-10"><a id="__codelineno-37-10" name="__codelineno-37-10" href="#__codelineno-37-10"></a>
</span><span id="__span-37-11"><a id="__codelineno-37-11" name="__codelineno-37-11" href="#__codelineno-37-11"></a>    MARK_DUPLICATES_AND_SORT(
</span><span id="__span-37-12"><a id="__codelineno-37-12" name="__codelineno-37-12" href="#__codelineno-37-12"></a>        ...
</span><span id="__span-37-13"><a id="__codelineno-37-13" name="__codelineno-37-13" href="#__codelineno-37-13"></a>    )
</span><span id="__span-37-14"><a id="__codelineno-37-14" name="__codelineno-37-14" href="#__codelineno-37-14"></a>
</span><span id="__span-37-15"><a id="__codelineno-37-15" name="__codelineno-37-15" href="#__codelineno-37-15"></a>    MERGE(
</span><span id="__span-37-16"><a id="__codelineno-37-16" name="__codelineno-37-16" href="#__codelineno-37-16"></a>        ...
</span><span id="__span-37-17"><a id="__codelineno-37-17" name="__codelineno-37-17" href="#__codelineno-37-17"></a>    )
</span><span id="__span-37-18"><a id="__codelineno-37-18" name="__codelineno-37-18" href="#__codelineno-37-18"></a>
</span><span id="__span-37-19"><a id="__codelineno-37-19" name="__codelineno-37-19" href="#__codelineno-37-19"></a>    NAME_SORT(
</span><span id="__span-37-20"><a id="__codelineno-37-20" name="__codelineno-37-20" href="#__codelineno-37-20"></a>        ...
</span><span id="__span-37-21"><a id="__codelineno-37-21" name="__codelineno-37-21" href="#__codelineno-37-21"></a>    )
</span><span id="__span-37-22"><a id="__codelineno-37-22" name="__codelineno-37-22" href="#__codelineno-37-22"></a>
</span><span id="__span-37-23"><a id="__codelineno-37-23" name="__codelineno-37-23" href="#__codelineno-37-23"></a>}
</span></code></pre></div>
<p><br></p>
<p><strong>Solution</strong></p>
<p>We will need to rearrange the process calls in <code>main.nf</code> so they mirror the order seen in the source CWL. </p>
<div class="admonition question">
<p class="admonition-title">Rerranging process calls</p>
<p>Cut-and-paste the process calls to be in the correct order in the main workflow.</p>
<details class="hint">
<summary>Correct Order</summary>
<ol>
<li>ALIGN</li>
<li>MERGE</li>
<li>NAME_SORT</li>
<li>MARK_DUPLICATES_AND_SORT</li>
<li>INDEX_BAM</li>
</ol>
</details>
<details class="hint">
<summary>Show Workflow</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-38-1"><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a>workflow {
</span><span id="__span-38-2"><a id="__codelineno-38-2" name="__codelineno-38-2" href="#__codelineno-38-2"></a>
</span><span id="__span-38-3"><a id="__codelineno-38-3" name="__codelineno-38-3" href="#__codelineno-38-3"></a>    ALIGN(
</span><span id="__span-38-4"><a id="__codelineno-38-4" name="__codelineno-38-4" href="#__codelineno-38-4"></a>        ch_bams.flatten(),       // ch_bam
</span><span id="__span-38-5"><a id="__codelineno-38-5" name="__codelineno-38-5" href="#__codelineno-38-5"></a>        reference,               // ch_reference
</span><span id="__span-38-6"><a id="__codelineno-38-6" name="__codelineno-38-6" href="#__codelineno-38-6"></a>        ch_readgroups.flatten()  // ch_readgroup
</span><span id="__span-38-7"><a id="__codelineno-38-7" name="__codelineno-38-7" href="#__codelineno-38-7"></a>    )
</span><span id="__span-38-8"><a id="__codelineno-38-8" name="__codelineno-38-8" href="#__codelineno-38-8"></a>
</span><span id="__span-38-9"><a id="__codelineno-38-9" name="__codelineno-38-9" href="#__codelineno-38-9"></a>    MERGE(
</span><span id="__span-38-10"><a id="__codelineno-38-10" name="__codelineno-38-10" href="#__codelineno-38-10"></a>        ALIGN.out.tagged_bam.toList(),  // bams
</span><span id="__span-38-11"><a id="__codelineno-38-11" name="__codelineno-38-11" href="#__codelineno-38-11"></a>        params.final_name               // name
</span><span id="__span-38-12"><a id="__codelineno-38-12" name="__codelineno-38-12" href="#__codelineno-38-12"></a>    )
</span><span id="__span-38-13"><a id="__codelineno-38-13" name="__codelineno-38-13" href="#__codelineno-38-13"></a>
</span><span id="__span-38-14"><a id="__codelineno-38-14" name="__codelineno-38-14" href="#__codelineno-38-14"></a>    NAME_SORT(
</span><span id="__span-38-15"><a id="__codelineno-38-15" name="__codelineno-38-15" href="#__codelineno-38-15"></a>        MERGE.out.merged_bam  // bam
</span><span id="__span-38-16"><a id="__codelineno-38-16" name="__codelineno-38-16" href="#__codelineno-38-16"></a>    )
</span><span id="__span-38-17"><a id="__codelineno-38-17" name="__codelineno-38-17" href="#__codelineno-38-17"></a>
</span><span id="__span-38-18"><a id="__codelineno-38-18" name="__codelineno-38-18" href="#__codelineno-38-18"></a>    MARK_DUPLICATES_AND_SORT(
</span><span id="__span-38-19"><a id="__codelineno-38-19" name="__codelineno-38-19" href="#__codelineno-38-19"></a>        params.mark_duplicates_and_sort.script,  // script
</span><span id="__span-38-20"><a id="__codelineno-38-20" name="__codelineno-38-20" href="#__codelineno-38-20"></a>        NAME_SORT.out.name_sorted_bam,           // bam
</span><span id="__span-38-21"><a id="__codelineno-38-21" name="__codelineno-38-21" href="#__codelineno-38-21"></a>        params.NULL_VALUE,                       // input_sort_order
</span><span id="__span-38-22"><a id="__codelineno-38-22" name="__codelineno-38-22" href="#__codelineno-38-22"></a>        params.final_name                        // output_name
</span><span id="__span-38-23"><a id="__codelineno-38-23" name="__codelineno-38-23" href="#__codelineno-38-23"></a>    )
</span><span id="__span-38-24"><a id="__codelineno-38-24" name="__codelineno-38-24" href="#__codelineno-38-24"></a>
</span><span id="__span-38-25"><a id="__codelineno-38-25" name="__codelineno-38-25" href="#__codelineno-38-25"></a>    INDEX_BAM(
</span><span id="__span-38-26"><a id="__codelineno-38-26" name="__codelineno-38-26" href="#__codelineno-38-26"></a>        MARK_DUPLICATES_AND_SORT.out.sorted_bam.map{ tuple -&gt; tuple[0] }  // bam
</span><span id="__span-38-27"><a id="__codelineno-38-27" name="__codelineno-38-27" href="#__codelineno-38-27"></a>    )
</span><span id="__span-38-28"><a id="__codelineno-38-28" name="__codelineno-38-28" href="#__codelineno-38-28"></a>
</span><span id="__span-38-29"><a id="__codelineno-38-29" name="__codelineno-38-29" href="#__codelineno-38-29"></a>}
</span></code></pre></div>
</details>
</div>
<p>After you are done, rerun the workflow by using the same command as before.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-39-1"><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a>nextflow run main.nf
</span></code></pre></div>
<p><br></p>
<h3 id="error-2-unquoted-strings">Error 2: Unquoted Strings<a class="headerlink" href="#error-2-unquoted-strings" title="Permanent link">&para;</a></h3>
<p><strong>Error message</strong></p>
<p>The second error is due to the <code>readgroup</code> input of the <code>ALIGN_AND_TAG</code> process being used without enclosing quotes. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-40-1"><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a>Caused by:
</span><span id="__span-40-2"><a id="__codelineno-40-2" name="__codelineno-40-2" href="#__codelineno-40-2"></a>Process `ALIGN:ALIGN_AND_TAG (1)` terminated with an error exit status (1)
</span><span id="__span-40-3"><a id="__codelineno-40-3" name="__codelineno-40-3" href="#__codelineno-40-3"></a>
</span><span id="__span-40-4"><a id="__codelineno-40-4" name="__codelineno-40-4" href="#__codelineno-40-4"></a>Command executed:
</span><span id="__span-40-5"><a id="__codelineno-40-5" name="__codelineno-40-5" href="#__codelineno-40-5"></a>/bin/bash /usr/bin/alignment_helper.sh     2895499223.bam     @RG INCLID:2895499223RY_PU:H7HY2CCXX.3.ATCACGGTITY=ISM:H_NJ-HCC1395-HCC1395ILB:H_NJ-HCC1395-HCC1395-lg24-lib1LEVEL=5PL:IlluminaSCN:WUGSC     chr17_test.fa          8     &gt; refAlign.bam
</span></code></pre></div>
<p>The issue can be seen in the command above. </p>
<p>We see <code>/bin/bash /usr/bin/alignment_helper.sh   2895499223.bam</code> which is expected. <br>
This is then followed by what looks like 2 string arguments: <code>@RG</code> and <code>INCLID...</code>.<br>
This is causing the problem.</p>
<p><br></p>
<p><strong>Troubleshooting</strong></p>
<p>The <code>Command executed</code> section in nextflow error messages is particularly useful. <br> 
This message is printed to the shell, but can also be seen by navigating to the process working directory &amp; viewing <code>.command.sh</code>. </p>
<p>Let&rsquo;s look at the nextflow <code>ALIGN_AND_TAG</code> process so we can match up which process input is causing the error. <br>
Indiviual arguments have been marked with their (position) to aid our investigation. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-41-1"><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a>process ALIGN_AND_TAG {
</span><span id="__span-41-2"><a id="__codelineno-41-2" name="__codelineno-41-2" href="#__codelineno-41-2"></a>
</span><span id="__span-41-3"><a id="__codelineno-41-3" name="__codelineno-41-3" href="#__codelineno-41-3"></a>    container &quot;mgibio/alignment_helper-cwl:1.0.0&quot;
</span><span id="__span-41-4"><a id="__codelineno-41-4" name="__codelineno-41-4" href="#__codelineno-41-4"></a>    publishDir &quot;${params.outdir}/align_and_tag&quot;
</span><span id="__span-41-5"><a id="__codelineno-41-5" name="__codelineno-41-5" href="#__codelineno-41-5"></a>    cpus &quot;${params.align_and_tag.cpus}&quot;
</span><span id="__span-41-6"><a id="__codelineno-41-6" name="__codelineno-41-6" href="#__codelineno-41-6"></a>    memory &quot;${params.align_and_tag.memory}&quot;
</span><span id="__span-41-7"><a id="__codelineno-41-7" name="__codelineno-41-7" href="#__codelineno-41-7"></a>
</span><span id="__span-41-8"><a id="__codelineno-41-8" name="__codelineno-41-8" href="#__codelineno-41-8"></a>    input:
</span><span id="__span-41-9"><a id="__codelineno-41-9" name="__codelineno-41-9" href="#__codelineno-41-9"></a>    path reference
</span><span id="__span-41-10"><a id="__codelineno-41-10" name="__codelineno-41-10" href="#__codelineno-41-10"></a>    path bam
</span><span id="__span-41-11"><a id="__codelineno-41-11" name="__codelineno-41-11" href="#__codelineno-41-11"></a>    val readgroup
</span><span id="__span-41-12"><a id="__codelineno-41-12" name="__codelineno-41-12" href="#__codelineno-41-12"></a>
</span><span id="__span-41-13"><a id="__codelineno-41-13" name="__codelineno-41-13" href="#__codelineno-41-13"></a>    output:
</span><span id="__span-41-14"><a id="__codelineno-41-14" name="__codelineno-41-14" href="#__codelineno-41-14"></a>    path &quot;refAlign.bam&quot;, emit: aligned_bam
</span><span id="__span-41-15"><a id="__codelineno-41-15" name="__codelineno-41-15" href="#__codelineno-41-15"></a>
</span><span id="__span-41-16"><a id="__codelineno-41-16" name="__codelineno-41-16" href="#__codelineno-41-16"></a>    script:
</span><span id="__span-41-17"><a id="__codelineno-41-17" name="__codelineno-41-17" href="#__codelineno-41-17"></a>    def reference = reference[0]
</span><span id="__span-41-18"><a id="__codelineno-41-18" name="__codelineno-41-18" href="#__codelineno-41-18"></a>    &quot;&quot;&quot;
</span><span id="__span-41-19"><a id="__codelineno-41-19" name="__codelineno-41-19" href="#__codelineno-41-19"></a>    (1) /bin/bash (2) /usr/bin/alignment_helper.sh \
</span><span id="__span-41-20"><a id="__codelineno-41-20" name="__codelineno-41-20" href="#__codelineno-41-20"></a>    (3) ${bam} \
</span><span id="__span-41-21"><a id="__codelineno-41-21" name="__codelineno-41-21" href="#__codelineno-41-21"></a>    (4) ${readgroup} \
</span><span id="__span-41-22"><a id="__codelineno-41-22" name="__codelineno-41-22" href="#__codelineno-41-22"></a>    (5) ${reference} \
</span><span id="__span-41-23"><a id="__codelineno-41-23" name="__codelineno-41-23" href="#__codelineno-41-23"></a>    (6) 8 \
</span><span id="__span-41-24"><a id="__codelineno-41-24" name="__codelineno-41-24" href="#__codelineno-41-24"></a>    (7) &gt; (8) refAlign.bam  
</span><span id="__span-41-25"><a id="__codelineno-41-25" name="__codelineno-41-25" href="#__codelineno-41-25"></a>    &quot;&quot;&quot;
</span><span id="__span-41-26"><a id="__codelineno-41-26" name="__codelineno-41-26" href="#__codelineno-41-26"></a>
</span><span id="__span-41-27"><a id="__codelineno-41-27" name="__codelineno-41-27" href="#__codelineno-41-27"></a>}
</span></code></pre></div>
<p>Here is the command executed with the same numbering:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-42-1"><a id="__codelineno-42-1" name="__codelineno-42-1" href="#__codelineno-42-1"></a>(1) /bin/bash (2) /usr/bin/alignment_helper.sh  (3) 2895499223.bam     (4) @RG (5) ID:2895499223    PU:H7HY2CCXX.3.ATCACGGT SM:H_NJ-HCC1395-HCC1395 LB:H_NJ-HCC1395-HCC1395-lg24-lib1   PL:Illumina CN:WUGSC   (6) chr17_test.fa        (7) 8  (8) &gt;  (9) refAlign.bam 
</span></code></pre></div>
<p>Matching up the two, we can see that arguments <code>(1-3)</code> match their expected values. <br>
Argument <code>(4)</code> starts out right, as we expect the <code>readgroup</code> input. </p>
<p>Looking in the <code>script:</code> section of the nextflow process, we expect the <code>${reference}</code> input to appear as argument <code>(5)</code>, but in the actual command it appears as argument <code>(6)</code> <code>chr17_test.fa</code>.</p>
<p>The issue seems to be that the <code>readgroup</code> input has been split into 2 strings, instead of 1 single string. <br>
By tracing back through the workflow, we can track that <code>params.readgroups</code> supplies the value for <code>readgroup</code> in this nextflow process:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-43-1"><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a>modules/align_and_tag.nf:  readgroup
</span><span id="__span-43-2"><a id="__codelineno-43-2" name="__codelineno-43-2" href="#__codelineno-43-2"></a>subworkflows/align.nf:     ch_readgroup
</span><span id="__span-43-3"><a id="__codelineno-43-3" name="__codelineno-43-3" href="#__codelineno-43-3"></a>main.nf:                   ch_readgroups.flatten()
</span><span id="__span-43-4"><a id="__codelineno-43-4" name="__codelineno-43-4" href="#__codelineno-43-4"></a>main.nf:                   ch_readgroups  = Channel.of( params.readgroups ).toList()
</span></code></pre></div>
<p>Looking at <code>nextflow.config</code> we see that this particular readgroup value is supplied as follows:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-44-1"><a id="__codelineno-44-1" name="__codelineno-44-1" href="#__codelineno-44-1"></a>@RG\tID:2895499223\tPU:H7HY2CCXX.3.ATCACGGT\tSM:H_NJ-HCC1395-HCC1395\tLB:H_NJ-HCC1395-HCC1395-lg24-lib1\tPL:Illumina\tCN:WUGSC
</span></code></pre></div>
<p>The <code>@RG</code> being split from the rest of the readgroup value, and the text which follows is incorrect in the command.  </p>
<p>The issue here is that this value in <code>params.readgroups</code> contains spaces and tabs (<code>\t</code>). <br>
When used in a nextflow process, string arugments should be enclosed  using <code>""</code>. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>Back in the <code>modules/align_and_tag.nf</code> file, let&rsquo;s properly enclose the <code>readgroup</code> input in quotes. </p>
<div class="admonition question">
<p class="admonition-title">align_and_tag.nf</p>
<p>In the <code>ALIGN_AND_TAG</code> process script enclose the <code>${readgroup}</code> reference in quotes.</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-45-1"><a id="__codelineno-45-1" name="__codelineno-45-1" href="#__codelineno-45-1"></a>script:
</span><span id="__span-45-2"><a id="__codelineno-45-2" name="__codelineno-45-2" href="#__codelineno-45-2"></a>def reference = reference[0]
</span><span id="__span-45-3"><a id="__codelineno-45-3" name="__codelineno-45-3" href="#__codelineno-45-3"></a>&quot;&quot;&quot;
</span><span id="__span-45-4"><a id="__codelineno-45-4" name="__codelineno-45-4" href="#__codelineno-45-4"></a>/bin/bash /usr/bin/alignment_helper.sh \
</span><span id="__span-45-5"><a id="__codelineno-45-5" name="__codelineno-45-5" href="#__codelineno-45-5"></a>${bam} \
</span><span id="__span-45-6"><a id="__codelineno-45-6" name="__codelineno-45-6" href="#__codelineno-45-6"></a>&quot;${readgroup}&quot; \              &lt;- quotes added
</span><span id="__span-45-7"><a id="__codelineno-45-7" name="__codelineno-45-7" href="#__codelineno-45-7"></a>${reference} \
</span><span id="__span-45-8"><a id="__codelineno-45-8" name="__codelineno-45-8" href="#__codelineno-45-8"></a>8 \
</span><span id="__span-45-9"><a id="__codelineno-45-9" name="__codelineno-45-9" href="#__codelineno-45-9"></a>&gt; refAlign.bam 
</span><span id="__span-45-10"><a id="__codelineno-45-10" name="__codelineno-45-10" href="#__codelineno-45-10"></a>&quot;&quot;&quot;
</span></code></pre></div>
</details>
</div>
<p>After you are have made the change, re-run the workflow by using the same command as before:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-46-1"><a id="__codelineno-46-1" name="__codelineno-46-1" href="#__codelineno-46-1"></a>nextflow run main.nf
</span></code></pre></div>
<p><br></p>
<h3 id="error-3-filename-clashes">Error 3: Filename Clashes<a class="headerlink" href="#error-3-filename-clashes" title="Permanent link">&para;</a></h3>
<p><strong>Error Message</strong></p>
<p>The final error is flagged by the nextflow workflow engine. </p>
<p>Upon re-running the workflow, you will encounter the following message:
<div class="language-text highlight"><pre><span></span><code><span id="__span-47-1"><a id="__codelineno-47-1" name="__codelineno-47-1" href="#__codelineno-47-1"></a>Process `MERGE` input file name collision -- There are multiple input files for each of the following file names: refAlign.bam
</span></code></pre></div></p>
<p>This informs us that more than 1 file with the name <code>refAlign.bam</code> have been inputs to <code>MERGE</code> process. </p>
<p>Nextflow does not allow this behaviour. <br></p>
<p>Other workflow engines use temporary names for files so that name clashes are impossible. <br> 
Nextflow requires us to be more specific with our filenames so we can track which file is which. </p>
<p>This is not a hard-and-fast rule of workflow engines, but nextflow enforces using unique names to encourage best-practises. </p>
<p><br></p>
<p><strong>Troubleshooting</strong></p>
<p>To track the cause, let&rsquo;s look at the data which feeds the <code>MERGE</code> process. </p>
<p><code>main.nf</code>
<div class="language-text highlight"><pre><span></span><code><span id="__span-48-1"><a id="__codelineno-48-1" name="__codelineno-48-1" href="#__codelineno-48-1"></a>MERGE(
</span><span id="__span-48-2"><a id="__codelineno-48-2" name="__codelineno-48-2" href="#__codelineno-48-2"></a>    ALIGN.out.tagged_bam.toList(),  // bams
</span><span id="__span-48-3"><a id="__codelineno-48-3" name="__codelineno-48-3" href="#__codelineno-48-3"></a>    params.final_name               // name
</span><span id="__span-48-4"><a id="__codelineno-48-4" name="__codelineno-48-4" href="#__codelineno-48-4"></a>)
</span></code></pre></div></p>
<p>We can see the translated nextflow workflow is collecting the <code>tagged_bam</code> output of all <code>ALIGN</code> tasks as a list using <code>.toList()</code>. <br>
This value feeds the <code>bams</code> input of the <code>MERGE</code> process. </p>
<p>The issue is that 2+ files in this list must have the same filename: &ldquo;refAlign.bam&rdquo;.<br>
We need to see how files end up being placed into <code>ALIGN.out.tagged_bam</code> to track where the error is occuring. </p>
<p>Looking in the <code>ALIGN</code> subworkflow in <code>subworkflows/align.nf</code>, we see the offending <code>tagged_bam</code> output emit. <br>
This gets its data from the <code>aligned_bam</code> output of the <code>ALIGN_AND_TAG</code> process in <code>modules/align_and_tag.nf</code>. </p>
<p>The source of the issue must be that the <code>aligned_bam</code> output is producing the same filename each time the task is run. </p>
<p>Opening <code>modules/align_and_tag.nf</code> we can see how the <code>align_bam</code> output is created &amp; collected:
<div class="language-text highlight"><pre><span></span><code><span id="__span-49-1"><a id="__codelineno-49-1" name="__codelineno-49-1" href="#__codelineno-49-1"></a>    output:
</span><span id="__span-49-2"><a id="__codelineno-49-2" name="__codelineno-49-2" href="#__codelineno-49-2"></a>    path &quot;refAlign.bam&quot;, emit: aligned_bam          &lt;-
</span><span id="__span-49-3"><a id="__codelineno-49-3" name="__codelineno-49-3" href="#__codelineno-49-3"></a>
</span><span id="__span-49-4"><a id="__codelineno-49-4" name="__codelineno-49-4" href="#__codelineno-49-4"></a>    script:
</span><span id="__span-49-5"><a id="__codelineno-49-5" name="__codelineno-49-5" href="#__codelineno-49-5"></a>    def reference = reference[0]
</span><span id="__span-49-6"><a id="__codelineno-49-6" name="__codelineno-49-6" href="#__codelineno-49-6"></a>    &quot;&quot;&quot;
</span><span id="__span-49-7"><a id="__codelineno-49-7" name="__codelineno-49-7" href="#__codelineno-49-7"></a>    /bin/bash /usr/bin/alignment_helper.sh \
</span><span id="__span-49-8"><a id="__codelineno-49-8" name="__codelineno-49-8" href="#__codelineno-49-8"></a>    ${bam} \
</span><span id="__span-49-9"><a id="__codelineno-49-9" name="__codelineno-49-9" href="#__codelineno-49-9"></a>    &quot;${readgroup}&quot; \
</span><span id="__span-49-10"><a id="__codelineno-49-10" name="__codelineno-49-10" href="#__codelineno-49-10"></a>    ${reference} \
</span><span id="__span-49-11"><a id="__codelineno-49-11" name="__codelineno-49-11" href="#__codelineno-49-11"></a>    8 \
</span><span id="__span-49-12"><a id="__codelineno-49-12" name="__codelineno-49-12" href="#__codelineno-49-12"></a>    &gt; refAlign.bam                                  &lt;-
</span><span id="__span-49-13"><a id="__codelineno-49-13" name="__codelineno-49-13" href="#__codelineno-49-13"></a>    &quot;&quot;&quot;
</span></code></pre></div></p>
<p>From looking at the <code>output</code> and <code>script</code> section of this process, we can see that the <code>aligned_bam</code> output will always have the same filename. </p>
<p>In the script, we generate a file called <code>refAlign.bam</code> by redirecting <code>stdout</code> to this file. <br>
In the output, <code>refAlign.bam</code> is collected as the output. </p>
<p>This would be fine if our workflow was supplied only a single BAM input file, but we want our workflow to run when we have many BAMs. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue, we need to give the output files unique names.  </p>
<p>This can be accomplished using a combination of the <code>path bam</code> process input, and the <code>.simpleName</code> operator.</p>
<p>As our input BAM files will all have unique names, we can use their name as a base for our output alignemnts. 
If we were to use their filename directly, we would get their file extension too, which we don&rsquo;t want!</p>
<p>Luckily the <code>.simpleName</code> nextflow operator will return the basename of a file without directory path or extensions. </p>
<p>For example, imagine we have a process input <code>path reads</code>.<br>
Suppose it gets fed the value <code>"data/SRR1234.fq"</code> at runtime.<br>
Calling <code>${reads.simpleName}</code> in the process script would yield <code>"SRR1234"</code>.</p>
<p>Note that directory path and the extension have been trimmed out!</p>
<div class="admonition question">
<p class="admonition-title">align_and_tag.nf</p>
<p>Script section:<br>
Use <code>.simpleName</code> to generate an output filename based on the <code>path bam</code> process input.<br> 
Add <code>_refAlign.bam</code> onto the end of this filename for clarity.</p>
<p>Output section:<br>
Alter the collection expression for the <code>aligned_bam</code> output to collect this file.</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-50-1"><a id="__codelineno-50-1" name="__codelineno-50-1" href="#__codelineno-50-1"></a>process ALIGN_AND_TAG {
</span><span id="__span-50-2"><a id="__codelineno-50-2" name="__codelineno-50-2" href="#__codelineno-50-2"></a>
</span><span id="__span-50-3"><a id="__codelineno-50-3" name="__codelineno-50-3" href="#__codelineno-50-3"></a>    container &quot;mgibio/alignment_helper-cwl:1.0.0&quot;
</span><span id="__span-50-4"><a id="__codelineno-50-4" name="__codelineno-50-4" href="#__codelineno-50-4"></a>    publishDir &quot;${params.outdir}/align_and_tag&quot;
</span><span id="__span-50-5"><a id="__codelineno-50-5" name="__codelineno-50-5" href="#__codelineno-50-5"></a>    cpus &quot;${params.align_and_tag.cpus}&quot;
</span><span id="__span-50-6"><a id="__codelineno-50-6" name="__codelineno-50-6" href="#__codelineno-50-6"></a>    memory &quot;${params.align_and_tag.memory}&quot;
</span><span id="__span-50-7"><a id="__codelineno-50-7" name="__codelineno-50-7" href="#__codelineno-50-7"></a>
</span><span id="__span-50-8"><a id="__codelineno-50-8" name="__codelineno-50-8" href="#__codelineno-50-8"></a>    input:
</span><span id="__span-50-9"><a id="__codelineno-50-9" name="__codelineno-50-9" href="#__codelineno-50-9"></a>    path reference
</span><span id="__span-50-10"><a id="__codelineno-50-10" name="__codelineno-50-10" href="#__codelineno-50-10"></a>    path bam
</span><span id="__span-50-11"><a id="__codelineno-50-11" name="__codelineno-50-11" href="#__codelineno-50-11"></a>    val readgroup
</span><span id="__span-50-12"><a id="__codelineno-50-12" name="__codelineno-50-12" href="#__codelineno-50-12"></a>
</span><span id="__span-50-13"><a id="__codelineno-50-13" name="__codelineno-50-13" href="#__codelineno-50-13"></a>    output:
</span><span id="__span-50-14"><a id="__codelineno-50-14" name="__codelineno-50-14" href="#__codelineno-50-14"></a>    path &quot;${bam.simpleName}_refAlign.bam&quot;, emit: aligned_bam    &lt;-
</span><span id="__span-50-15"><a id="__codelineno-50-15" name="__codelineno-50-15" href="#__codelineno-50-15"></a>
</span><span id="__span-50-16"><a id="__codelineno-50-16" name="__codelineno-50-16" href="#__codelineno-50-16"></a>    script:
</span><span id="__span-50-17"><a id="__codelineno-50-17" name="__codelineno-50-17" href="#__codelineno-50-17"></a>    def reference = reference[0]
</span><span id="__span-50-18"><a id="__codelineno-50-18" name="__codelineno-50-18" href="#__codelineno-50-18"></a>    &quot;&quot;&quot;
</span><span id="__span-50-19"><a id="__codelineno-50-19" name="__codelineno-50-19" href="#__codelineno-50-19"></a>    /bin/bash /usr/bin/alignment_helper.sh \
</span><span id="__span-50-20"><a id="__codelineno-50-20" name="__codelineno-50-20" href="#__codelineno-50-20"></a>    ${bam} \
</span><span id="__span-50-21"><a id="__codelineno-50-21" name="__codelineno-50-21" href="#__codelineno-50-21"></a>    &quot;${readgroup}&quot; \
</span><span id="__span-50-22"><a id="__codelineno-50-22" name="__codelineno-50-22" href="#__codelineno-50-22"></a>    ${reference} \
</span><span id="__span-50-23"><a id="__codelineno-50-23" name="__codelineno-50-23" href="#__codelineno-50-23"></a>    8 \
</span><span id="__span-50-24"><a id="__codelineno-50-24" name="__codelineno-50-24" href="#__codelineno-50-24"></a>    &gt; ${bam.simpleName}_refAlign.bam                            &lt;-
</span><span id="__span-50-25"><a id="__codelineno-50-25" name="__codelineno-50-25" href="#__codelineno-50-25"></a>    &quot;&quot;&quot;
</span><span id="__span-50-26"><a id="__codelineno-50-26" name="__codelineno-50-26" href="#__codelineno-50-26"></a>
</span><span id="__span-50-27"><a id="__codelineno-50-27" name="__codelineno-50-27" href="#__codelineno-50-27"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p>After you are have made these changes, re-run the workflow by using the same command as before:
<div class="language-text highlight"><pre><span></span><code><span id="__span-51-1"><a id="__codelineno-51-1" name="__codelineno-51-1" href="#__codelineno-51-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p>With any luck, this will fix the remaining issues and the workflow will now run to completion. </p>
<p><br></p>
<h3 id="completed-workflow">Completed Workflow<a class="headerlink" href="#completed-workflow" title="Permanent link">&para;</a></h3>
<p>Once completed, we can check the <code>outputs/</code> folder to view our results. <br>
If everything went well, the <code>outputs/</code> folder should have the following structure:
<div class="language-text highlight"><pre><span></span><code><span id="__span-52-1"><a id="__codelineno-52-1" name="__codelineno-52-1" href="#__codelineno-52-1"></a>outputs
</span><span id="__span-52-2"><a id="__codelineno-52-2" name="__codelineno-52-2" href="#__codelineno-52-2"></a> align_and_tag
</span><span id="__span-52-3"><a id="__codelineno-52-3" name="__codelineno-52-3" href="#__codelineno-52-3"></a>    2895499223_refAlign.bam
</span><span id="__span-52-4"><a id="__codelineno-52-4" name="__codelineno-52-4" href="#__codelineno-52-4"></a>    2895499237_refAlign.bam
</span><span id="__span-52-5"><a id="__codelineno-52-5" name="__codelineno-52-5" href="#__codelineno-52-5"></a> index_bam
</span><span id="__span-52-6"><a id="__codelineno-52-6" name="__codelineno-52-6" href="#__codelineno-52-6"></a>    final.bam
</span><span id="__span-52-7"><a id="__codelineno-52-7" name="__codelineno-52-7" href="#__codelineno-52-7"></a>    final.bam.bai
</span><span id="__span-52-8"><a id="__codelineno-52-8" name="__codelineno-52-8" href="#__codelineno-52-8"></a> mark_duplicates_and_sort
</span><span id="__span-52-9"><a id="__codelineno-52-9" name="__codelineno-52-9" href="#__codelineno-52-9"></a>    final.bam
</span><span id="__span-52-10"><a id="__codelineno-52-10" name="__codelineno-52-10" href="#__codelineno-52-10"></a>    final.bam.bai
</span><span id="__span-52-11"><a id="__codelineno-52-11" name="__codelineno-52-11" href="#__codelineno-52-11"></a>    final.mark_dups_metrics.txt
</span><span id="__span-52-12"><a id="__codelineno-52-12" name="__codelineno-52-12" href="#__codelineno-52-12"></a> merge_bams_samtools
</span><span id="__span-52-13"><a id="__codelineno-52-13" name="__codelineno-52-13" href="#__codelineno-52-13"></a>    final.bam.merged.bam
</span><span id="__span-52-14"><a id="__codelineno-52-14" name="__codelineno-52-14" href="#__codelineno-52-14"></a> name_sort
</span><span id="__span-52-15"><a id="__codelineno-52-15" name="__codelineno-52-15" href="#__codelineno-52-15"></a>     final.NameSorted.bam
</span></code></pre></div></p>
<p><br></p>
<p>If having trouble, the finished workflow is also available in the <code>data/final/cwl/align_sort_markdup/</code> folder for reference.</p>
<p>In addition, the following diffs shows the changes we made to files during manual adjustment of the workflow. </p>
<p><code>main.nf</code></p>
<p><img alt="alt text" src="../media/asm_main.PNG" /></p>
<p><code>align_and_tag.nf</code></p>
<p><img alt="alt text" src="../media/asm_align_and_tag.PNG" /></p>
<p><br></p>
<h3 id="conclusion_2">Conclusion<a class="headerlink" href="#conclusion_2" title="Permanent link">&para;</a></h3>
<p>In this section we explored how to translate the <code>align_sort_markdup</code> CWL Workflow to Nextflow using <code>janis translate</code>. </p>
<p>This is the end of section 1. </p>
<p>Let&rsquo;s take a break!</p>
<p>In section 2 we will be working with Galaxy, and will translate some Galaxy Tool Wrappers and Workflows into Nextflow as seen with CWL.</p>
<p><br></p>
<hr />
<h2 id="section-21-samtools-flagstat-tool-galaxy">Section 2.1: Samtools Flagstat Tool (Galaxy)<a class="headerlink" href="#section-21-samtools-flagstat-tool-galaxy" title="Permanent link">&para;</a></h2>
<h3 id="introduction_3">Introduction<a class="headerlink" href="#introduction_3" title="Permanent link">&para;</a></h3>
<p>This section demonstrates translation of a basic <code>samtools flagstat</code> Galaxy Tool Wrapper to Nextflow using <code>janis translate</code>. <br></p>
<p>The Galaxy Tool Wrapper used in this section was created by contributors to the <a href="https://github.com/galaxyproject/tools-devteam">Galaxy Devteam repository</a> of tools. </p>
<p>The underlying software used in the Galaxy Tool Wrapper - <a href="http://www.htslib.org/doc/samtools-flagstat.html">samtools_flagstat</a> - displays summary information for an alignment file. </p>
<p><br></p>
<h3 id="janis-translate_3">Janis Translate<a class="headerlink" href="#janis-translate_3" title="Permanent link">&para;</a></h3>
<p>As in the previous sections we will use janis-translate. <br>
This time instead of CWL -&gt; Nextflow, we will do a Galaxy -&gt; Nextflow translation. </p>
<p>Aside from local filepaths, <code>janis translate</code> can also access Galaxy Tool Wrappers using a tool ID. 
We will use this method today as it is an easier way to access Tool Wrappers. </p>
<p>To get the <code>samtools_flagstat</code> Tool ID, navigate to the tool using any usegalaxy.org server. <br>
The following is a link to the samtools flagstat tool in Galaxy Australia: <br> 
<a href="https://usegalaxy.org.au/root?tool_id=toolshed.g2.bx.psu.edu/repos/devteam/samtools_flagstat/samtools_flagstat/2.0.4">https://usegalaxy.org.au/root?tool_id=toolshed.g2.bx.psu.edu/repos/devteam/samtools_flagstat/samtools_flagstat/2.0.4</a>. </p>
<p>Once here, we will copy the Tool ID. </p>
<p><img alt="alt text" src="../media/samtools_flagstat_copy_tool_id.PNG" /></p>
<p>At time of writing, the current Tool ID for the <code>samtools_flagstat</code> tool wrapper is <em>toolshed.g2.bx.psu.edu/repos/devteam/samtools_flagstat/samtools_flagstat/2.0.4</em></p>
<p>Now we have the Tool ID, we can access &amp; translate this Galaxy Tool Wrapper to a Nextflow process. </p>
<p>We will add <code>-o samtools_flagstat</code> to our command to set the output directory. </p>
<p>First, let&rsquo;s make sure we&rsquo;re in the training directory 
<div class="language-text highlight"><pre><span></span><code><span id="__span-53-1"><a id="__codelineno-53-1" name="__codelineno-53-1" href="#__codelineno-53-1"></a>cd /home2/training
</span></code></pre></div></p>
<p>To translate the Galaxy Tool, run the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-54-1"><a id="__codelineno-54-1" name="__codelineno-54-1" href="#__codelineno-54-1"></a>singularity exec ~/janis.sif janis translate -o samtools_flagstat --from galaxy --to nextflow toolshed.g2.bx.psu.edu/repos/devteam/samtools_flagstat/samtools_flagstat/2.0.4
</span></code></pre></div></p>
<p><br></p>
<p>Once complete, you will see a folder called <code>samtools_flagstat/</code> appear, and a nextflow process called <code>samtools_flagstat.nf</code> will be present inside. </p>
<p>For your own reference / interest, the actual Galaxy Tool Wrapper files will be downloaded during translation &amp; will be presented to you in <code>samtools_flagstat/source/</code>. </p>
<p><br></p>
<h3 id="manual-adjustments_3">Manual Adjustments<a class="headerlink" href="#manual-adjustments_3" title="Permanent link">&para;</a></h3>
<p>The <code>samtools_flagstat/samtools_flagstat.nf</code> file should be similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-55-1"><a id="__codelineno-55-1" name="__codelineno-55-1" href="#__codelineno-55-1"></a>nextflow.enable.dsl=2
</span><span id="__span-55-2"><a id="__codelineno-55-2" name="__codelineno-55-2" href="#__codelineno-55-2"></a>
</span><span id="__span-55-3"><a id="__codelineno-55-3" name="__codelineno-55-3" href="#__codelineno-55-3"></a>process SAMTOOLS_FLAGSTAT {
</span><span id="__span-55-4"><a id="__codelineno-55-4" name="__codelineno-55-4" href="#__codelineno-55-4"></a>
</span><span id="__span-55-5"><a id="__codelineno-55-5" name="__codelineno-55-5" href="#__codelineno-55-5"></a>    container &quot;quay.io/biocontainers/samtools:1.13--h8c37831_0&quot;
</span><span id="__span-55-6"><a id="__codelineno-55-6" name="__codelineno-55-6" href="#__codelineno-55-6"></a>
</span><span id="__span-55-7"><a id="__codelineno-55-7" name="__codelineno-55-7" href="#__codelineno-55-7"></a>    input:
</span><span id="__span-55-8"><a id="__codelineno-55-8" name="__codelineno-55-8" href="#__codelineno-55-8"></a>    path input1
</span><span id="__span-55-9"><a id="__codelineno-55-9" name="__codelineno-55-9" href="#__codelineno-55-9"></a>    val addthreads
</span><span id="__span-55-10"><a id="__codelineno-55-10" name="__codelineno-55-10" href="#__codelineno-55-10"></a>
</span><span id="__span-55-11"><a id="__codelineno-55-11" name="__codelineno-55-11" href="#__codelineno-55-11"></a>    output:
</span><span id="__span-55-12"><a id="__codelineno-55-12" name="__codelineno-55-12" href="#__codelineno-55-12"></a>    path &quot;output1.txt&quot;, emit: output1
</span><span id="__span-55-13"><a id="__codelineno-55-13" name="__codelineno-55-13" href="#__codelineno-55-13"></a>
</span><span id="__span-55-14"><a id="__codelineno-55-14" name="__codelineno-55-14" href="#__codelineno-55-14"></a>    script:
</span><span id="__span-55-15"><a id="__codelineno-55-15" name="__codelineno-55-15" href="#__codelineno-55-15"></a>    &quot;&quot;&quot;
</span><span id="__span-55-16"><a id="__codelineno-55-16" name="__codelineno-55-16" href="#__codelineno-55-16"></a>    samtools flagstat \
</span><span id="__span-55-17"><a id="__codelineno-55-17" name="__codelineno-55-17" href="#__codelineno-55-17"></a>    --output-fmt &quot;txt&quot; \
</span><span id="__span-55-18"><a id="__codelineno-55-18" name="__codelineno-55-18" href="#__codelineno-55-18"></a>    -@ ${addthreads} \
</span><span id="__span-55-19"><a id="__codelineno-55-19" name="__codelineno-55-19" href="#__codelineno-55-19"></a>    ${input1} \
</span><span id="__span-55-20"><a id="__codelineno-55-20" name="__codelineno-55-20" href="#__codelineno-55-20"></a>    &gt; output1.txt
</span><span id="__span-55-21"><a id="__codelineno-55-21" name="__codelineno-55-21" href="#__codelineno-55-21"></a>    &quot;&quot;&quot;
</span><span id="__span-55-22"><a id="__codelineno-55-22" name="__codelineno-55-22" href="#__codelineno-55-22"></a>
</span><span id="__span-55-23"><a id="__codelineno-55-23" name="__codelineno-55-23" href="#__codelineno-55-23"></a>}
</span></code></pre></div>
<p>We can see that this nextflow process has two inputs, a single output, and calls <code>samtools flagstat</code>.  </p>
<p>Before continuing, let&rsquo;s check the <a href="http://www.htslib.org/doc/samtools-flagstat.html">samtools flagstat</a> documentation. 
In the documentation, we see the following:
<div class="language-text highlight"><pre><span></span><code><span id="__span-56-1"><a id="__codelineno-56-1" name="__codelineno-56-1" href="#__codelineno-56-1"></a>samtools flagstat in.sam|in.bam|in.cram
</span><span id="__span-56-2"><a id="__codelineno-56-2" name="__codelineno-56-2" href="#__codelineno-56-2"></a>
</span><span id="__span-56-3"><a id="__codelineno-56-3" name="__codelineno-56-3" href="#__codelineno-56-3"></a>-@ INT
</span><span id="__span-56-4"><a id="__codelineno-56-4" name="__codelineno-56-4" href="#__codelineno-56-4"></a>Set number of additional threads to use when reading the file.
</span><span id="__span-56-5"><a id="__codelineno-56-5" name="__codelineno-56-5" href="#__codelineno-56-5"></a>
</span><span id="__span-56-6"><a id="__codelineno-56-6" name="__codelineno-56-6" href="#__codelineno-56-6"></a>--output-fmt/-O FORMAT
</span><span id="__span-56-7"><a id="__codelineno-56-7" name="__codelineno-56-7" href="#__codelineno-56-7"></a>Set the output format. FORMAT can be set to `default&#39;, `json&#39; or `tsv&#39; to select the default, JSON or tab-separated values output format. If this option is not used, the default format will be selected.
</span></code></pre></div></p>
<p>By matching up the process <code>inputs:</code> section and the <code>script:</code> section, we can see that:
- <code>path input1</code> will be the input <code>sam | bam | cram</code>
- <code>val addthreads</code> will be the threads argument passed to <code>-@</code>
- the <code>--output-fmt</code> option has been assigned the default value of <code>"txt"</code></p>
<p>We can also see that a container image is available for this tool. </p>
<p>This translation is correct for the <code>samtools_flagstat.cwl</code> file and needs no adjusting. </p>
<p><br></p>
<blockquote>
<p>Note: <br>
If you would like to expose the <code>--output-fmt</code> option as a process input, you can do the following: </p>
<ul>
<li>add a <code>val format</code> input to the process </li>
<li>reference this input in the script, replacing the hardcoded <code>"txt"</code> value<br>
  (e.g. <code>--output-fmt ${format}</code>)</li>
</ul>
</blockquote>
<p><br></p>
<h3 id="running-samtools-flagstat-as-a-workflow_1">Running Samtools Flagstat as a Workflow<a class="headerlink" href="#running-samtools-flagstat-as-a-workflow_1" title="Permanent link">&para;</a></h3>
<p><strong>Setting up nextflow.config</strong></p>
<p>To run this process, we will set up a <code>nextflow.config</code> file and add some lines to the top of our process definition to turn it into a workflow.</p>
<p>Create a new file called <code>nextflow.config</code> in the <code>samtools_flagstat/</code> folder alongside <code>samtools_flagstat.nf</code>. </p>
<p>Copy and paste the following code into your <code>nextflow.config</code> file: </p>
<p><div class="language-text highlight"><pre><span></span><code><span id="__span-57-1"><a id="__codelineno-57-1" name="__codelineno-57-1" href="#__codelineno-57-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-57-2"><a id="__codelineno-57-2" name="__codelineno-57-2" href="#__codelineno-57-2"></a>singularity.enabled = true
</span><span id="__span-57-3"><a id="__codelineno-57-3" name="__codelineno-57-3" href="#__codelineno-57-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-57-4"><a id="__codelineno-57-4" name="__codelineno-57-4" href="#__codelineno-57-4"></a>
</span><span id="__span-57-5"><a id="__codelineno-57-5" name="__codelineno-57-5" href="#__codelineno-57-5"></a>params {
</span><span id="__span-57-6"><a id="__codelineno-57-6" name="__codelineno-57-6" href="#__codelineno-57-6"></a>    bam = &quot;/home2/training/data/sample_data/galaxy/samtools_flagstat/samtools_flagstat_input1.bam&quot;
</span><span id="__span-57-7"><a id="__codelineno-57-7" name="__codelineno-57-7" href="#__codelineno-57-7"></a>    threads = 1
</span><span id="__span-57-8"><a id="__codelineno-57-8" name="__codelineno-57-8" href="#__codelineno-57-8"></a>}
</span></code></pre></div>
<br></p>
<p>This tells nextflow how to run, and sets up inputs parameters we can use to supply values to the <code>SAMTOOLS_FLAGSTAT</code> process:</p>
<ul>
<li>The <code>bam</code> parameter is the input bam file we wish to analyse. </li>
<li>The <code>threads</code> parameter is an integer, and controls how many additional compute threads to use during runtime.</li>
</ul>
<p>From here, we can refer to these inputs as <code>params.bam</code> / <code>params.threads</code> in other files.</p>
<p><br></p>
<p><strong>Creating Workflow &amp; Passing Data</strong> </p>
<p>Now that we have the <code>nextflow.config</code> file set up, we will add a few lines to <code>samtools_flagstat.nf</code> to turn it into a workflow. </p>
<p>Copy and paste the following lines at the top of <code>samtools_flagstat.nf</code>:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-58-1"><a id="__codelineno-58-1" name="__codelineno-58-1" href="#__codelineno-58-1"></a>nextflow.enable.dsl=2
</span><span id="__span-58-2"><a id="__codelineno-58-2" name="__codelineno-58-2" href="#__codelineno-58-2"></a>
</span><span id="__span-58-3"><a id="__codelineno-58-3" name="__codelineno-58-3" href="#__codelineno-58-3"></a>ch_bam = Channel.fromPath( params.bam )
</span><span id="__span-58-4"><a id="__codelineno-58-4" name="__codelineno-58-4" href="#__codelineno-58-4"></a>
</span><span id="__span-58-5"><a id="__codelineno-58-5" name="__codelineno-58-5" href="#__codelineno-58-5"></a>workflow {
</span><span id="__span-58-6"><a id="__codelineno-58-6" name="__codelineno-58-6" href="#__codelineno-58-6"></a>    SAMTOOLS_FLAGSTAT(
</span><span id="__span-58-7"><a id="__codelineno-58-7" name="__codelineno-58-7" href="#__codelineno-58-7"></a>        ch_bam,             // input1 
</span><span id="__span-58-8"><a id="__codelineno-58-8" name="__codelineno-58-8" href="#__codelineno-58-8"></a>        params.threads      // addthreads
</span><span id="__span-58-9"><a id="__codelineno-58-9" name="__codelineno-58-9" href="#__codelineno-58-9"></a>    )
</span><span id="__span-58-10"><a id="__codelineno-58-10" name="__codelineno-58-10" href="#__codelineno-58-10"></a>    SAMTOOLS_FLAGSTAT.out.output1.view()
</span><span id="__span-58-11"><a id="__codelineno-58-11" name="__codelineno-58-11" href="#__codelineno-58-11"></a>}
</span></code></pre></div>
<p>The first line creates a nextflow <code>Channel</code> for our <code>bam</code> input. <br>
The <code>params.bam</code> global variable we set up previously is used to supply the 
path to our sample data.</p>
<p>The new <code>workflow {}</code> section declares the main workflow entry point. <br>
When we run this file, nextflow will look for this section and run the workflow contained within. </p>
<p>In our case, the workflow only contains a single task, which runs the <code>SAMTOOLS_FLAGSTAT</code> process defined below the workflow section. We then supply input values to <code>SAMTOOLS_FLAGSTAT</code> using our <code>ch_bams</code> channel we created for <code>input1</code>, and <code>params.threads</code> for the <code>addthreads</code> input.</p>
<p><br></p>
<p><strong>Adding publishDir directive</strong></p>
<p>So that we can collect the output of <code>SAMTOOLS_FLAGSTAT</code> when it runs, we will add a <code>publishDir</code> directive to the process:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-59-1"><a id="__codelineno-59-1" name="__codelineno-59-1" href="#__codelineno-59-1"></a>process SAMTOOLS_FLAGSTAT {
</span><span id="__span-59-2"><a id="__codelineno-59-2" name="__codelineno-59-2" href="#__codelineno-59-2"></a>
</span><span id="__span-59-3"><a id="__codelineno-59-3" name="__codelineno-59-3" href="#__codelineno-59-3"></a>    container &quot;quay.io/biocontainers/samtools:1.13--h8c37831_0&quot;
</span><span id="__span-59-4"><a id="__codelineno-59-4" name="__codelineno-59-4" href="#__codelineno-59-4"></a>    publishDir &quot;./outputs&quot;
</span><span id="__span-59-5"><a id="__codelineno-59-5" name="__codelineno-59-5" href="#__codelineno-59-5"></a>
</span><span id="__span-59-6"><a id="__codelineno-59-6" name="__codelineno-59-6" href="#__codelineno-59-6"></a>    ...
</span><span id="__span-59-7"><a id="__codelineno-59-7" name="__codelineno-59-7" href="#__codelineno-59-7"></a>
</span><span id="__span-59-8"><a id="__codelineno-59-8" name="__codelineno-59-8" href="#__codelineno-59-8"></a>}
</span></code></pre></div>
<p>Now that we have set up <code>SAMTOOLS_FLAGSTAT</code> as a workflow, we can run it and check the output. </p>
<p><br></p>
<p><strong>Running Our Workflow</strong></p>
<p>Ensure you are in the <code>samtools_flagstat/</code> working directory, where <code>nextflow.config</code> and <code>samtools_flagstat.nf</code> reside. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-60-1"><a id="__codelineno-60-1" name="__codelineno-60-1" href="#__codelineno-60-1"></a>cd samtools_flagstat
</span></code></pre></div>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-61-1"><a id="__codelineno-61-1" name="__codelineno-61-1" href="#__codelineno-61-1"></a>nextflow run samtools_flagstat.nf
</span></code></pre></div></p>
<p>Once completed, the check the <code>./outputs</code> folder inside <code>samtools_flagstat/</code>. </p>
<p>If everything went well, you should see a single file called <code>output1.txt</code> with the following contents: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-62-1"><a id="__codelineno-62-1" name="__codelineno-62-1" href="#__codelineno-62-1"></a>200 + 0 in total (QC-passed reads + QC-failed reads)
</span><span id="__span-62-2"><a id="__codelineno-62-2" name="__codelineno-62-2" href="#__codelineno-62-2"></a>200 + 0 primary
</span><span id="__span-62-3"><a id="__codelineno-62-3" name="__codelineno-62-3" href="#__codelineno-62-3"></a>0 + 0 secondary
</span><span id="__span-62-4"><a id="__codelineno-62-4" name="__codelineno-62-4" href="#__codelineno-62-4"></a>0 + 0 supplementary
</span><span id="__span-62-5"><a id="__codelineno-62-5" name="__codelineno-62-5" href="#__codelineno-62-5"></a>0 + 0 duplicates
</span><span id="__span-62-6"><a id="__codelineno-62-6" name="__codelineno-62-6" href="#__codelineno-62-6"></a>0 + 0 primary duplicates
</span><span id="__span-62-7"><a id="__codelineno-62-7" name="__codelineno-62-7" href="#__codelineno-62-7"></a>25 + 0 mapped (12.50% : N/A)
</span><span id="__span-62-8"><a id="__codelineno-62-8" name="__codelineno-62-8" href="#__codelineno-62-8"></a>25 + 0 primary mapped (12.50% : N/A)
</span><span id="__span-62-9"><a id="__codelineno-62-9" name="__codelineno-62-9" href="#__codelineno-62-9"></a>200 + 0 paired in sequencing
</span><span id="__span-62-10"><a id="__codelineno-62-10" name="__codelineno-62-10" href="#__codelineno-62-10"></a>100 + 0 read1
</span><span id="__span-62-11"><a id="__codelineno-62-11" name="__codelineno-62-11" href="#__codelineno-62-11"></a>100 + 0 read2
</span><span id="__span-62-12"><a id="__codelineno-62-12" name="__codelineno-62-12" href="#__codelineno-62-12"></a>0 + 0 properly paired (0.00% : N/A)
</span><span id="__span-62-13"><a id="__codelineno-62-13" name="__codelineno-62-13" href="#__codelineno-62-13"></a>0 + 0 with itself and mate mapped
</span><span id="__span-62-14"><a id="__codelineno-62-14" name="__codelineno-62-14" href="#__codelineno-62-14"></a>25 + 0 singletons (12.50% : N/A)
</span><span id="__span-62-15"><a id="__codelineno-62-15" name="__codelineno-62-15" href="#__codelineno-62-15"></a>0 + 0 with mate mapped to a different chr
</span><span id="__span-62-16"><a id="__codelineno-62-16" name="__codelineno-62-16" href="#__codelineno-62-16"></a>0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</span></code></pre></div></p>
<p>If needed, you can check the <code>data/final/galaxy/samtools_flagstat</code> folder as a reference. </p>
<p><br></p>
<h3 id="conclusion_3">Conclusion<a class="headerlink" href="#conclusion_3" title="Permanent link">&para;</a></h3>
<p>In this section we explored how to translate a simple Galaxy Tool to a Nextflow process. </p>
<p>Will build on this in the next section where we translate a much more complex Galaxy Tool Wrapper: <strong><em>limma</em></strong>.</p>
<p><br></p>
<hr />
<h2 id="section-22-limma-voom-tool-galaxy">Section 2.2: Limma Voom Tool (Galaxy)<a class="headerlink" href="#section-22-limma-voom-tool-galaxy" title="Permanent link">&para;</a></h2>
<h3 id="introduction_4">Introduction<a class="headerlink" href="#introduction_4" title="Permanent link">&para;</a></h3>
<p>This section demonstrates translation of the <code>limma</code> tool from Galaxy to Nextflow using <code>janis translate</code>.  </p>
<p>The aim of this section is to document a <strong><em>more challenging</em></strong> translation, as this Galaxy Tool Wrapper is complex and uses an Rscript to run its analysis. </p>
<p><strong><em>Shian Su</em></strong> is the original author of the <code>limma</code> Galaxy tool wrapper we are translating today. <br>
<strong><em>Maria Doyle</em></strong> has also contributed multiple times over its many years of use as maintainence and upgrades. </p>
<p>Limma is an R package which analyses gene expression using microarray or RNA-Seq data. <br>
The most common use (at time of writing) is referred to as limma-voom, which performs Differential Expression (DE) analysis on RNA-Seq data between multiple samples. </p>
<p>The <code>voom</code> part of <code>limma-voom</code> is functionality within <code>limma</code> which adapts limma to use RNA-Seq data instead of microarray data. 
Since RNA-Seq data has become more common than microarray data at time of writing, limma-voom is likely the most popular use of the <code>limma</code> package as a workflow.</p>
<p>Under the hood, the Galaxy <code>limma</code> tool contains an R script which configures &amp; runs limma based on command-line inputs. 
This allows Galaxy users to run <code>limma</code> as a complete workflow, rather than as individual R library functions. </p>
<p>The Limma package has had many contributors over the years, including but not limited to:</p>
<ul>
<li>Gordon K. Smyth</li>
<li>Matthew Ritchie</li>
<li>Natalie Thorne</li>
<li>James Wettenhall</li>
<li>Wei Shi</li>
<li>Yifang Hu</li>
</ul>
<p>See the following articles related to Limma:</p>
<ul>
<li>
<p>Limma package capabilities (new and old)<br>
<em>Ritchie, ME, Phipson, B, Wu, D, Hu, Y, Law, CW, Shi, W, and Smyth, GK (2015).
limma powers differential expression analyses for RNA-sequencing and microarray studies.
Nucleic Acids Research 43(7), e47.</em></p>
</li>
<li>
<p>Limma for DE analysis<br>
<em>Phipson, B, Lee, S, Majewski, IJ, Alexander, WS, and Smyth, GK (2016). Robust
hyperparameter estimation protects against hypervariable genes and improves power to
detect differential expression. Annals of Applied Statistics 10(2), 946963.</em></p>
</li>
<li>
<p>Limma-Voom for DE analysis of RNA-Seq data<br>
<em>Law, CW, Chen, Y, Shi, W, and Smyth, GK (2014). Voom: precision weights unlock
linear model analysis tools for RNA-seq read counts. Genome Biology 15, R29.</em></p>
</li>
</ul>
<p><br></p>
<h3 id="janis-translate_4">Janis Translate<a class="headerlink" href="#janis-translate_4" title="Permanent link">&para;</a></h3>
<p>We will run janis-translate in the same manner as section 2.1. </p>
<p>To get the Tool ID you can either:</p>
<ul>
<li>Go to the <a href="https://usegalaxy.org.au">Galaxy Australia</a> server &amp; copy the Tool ID from the limma tool, or</li>
<li>Copy the following ID: <em>toolshed.g2.bx.psu.edu/repos/iuc/limma_voom/limma_voom/3.50.1+galaxy0</em></li>
</ul>
<p>We add <code>-o limma_voom</code> to our command to specify this as the output directory. </p>
<p>As always, let&rsquo;s make sure we&rsquo;re back in the training folder:
<div class="language-text highlight"><pre><span></span><code><span id="__span-63-1"><a id="__codelineno-63-1" name="__codelineno-63-1" href="#__codelineno-63-1"></a>cd /home2/training
</span></code></pre></div></p>
<p>To translate, run the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-64-1"><a id="__codelineno-64-1" name="__codelineno-64-1" href="#__codelineno-64-1"></a>singularity exec ~/janis.sif janis translate -o limma_voom --from galaxy --to nextflow toolshed.g2.bx.psu.edu/repos/iuc/limma_voom/limma_voom/3.50.1+galaxy0
</span></code></pre></div></p>
<p><br></p>
<p>Once complete, you will see a folder called <code>limma_voom</code> appear, and a nextflow process called <code>limma_voom.nf</code> will be present inside. </p>
<p>For your own reference / interest, the actual Galaxy Tool Wrapper files will be downloaded during translation &amp; presented to you in <code>limma_voom/source</code>. </p>
<p><br></p>
<h3 id="manual-adjustments_4">Manual Adjustments<a class="headerlink" href="#manual-adjustments_4" title="Permanent link">&para;</a></h3>
<p>Inside the <code>limma_voom/</code> folder, we see the following: </p>
<ul>
<li>A Nextflow process named <code>limma_voom.nf</code></li>
<li>An Rscript named <code>limma_voom.R</code> </li>
<li>A folder named <code>source/</code> containing the Galaxy Tool Wrapper XML</li>
</ul>
<p>Unlike previous tool translations, the <code>limma_voom.nf</code> process will need adjusting. </p>
<p><br></p>
<p><strong>limma_voom.nf</strong></p>
<p>The <code>limma_voom/limma_voom.nf</code> file is our nextflow process. It should be similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-65-1"><a id="__codelineno-65-1" name="__codelineno-65-1" href="#__codelineno-65-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-65-2"><a id="__codelineno-65-2" name="__codelineno-65-2" href="#__codelineno-65-2"></a>
</span><span id="__span-65-3"><a id="__codelineno-65-3" name="__codelineno-65-3" href="#__codelineno-65-3"></a>process LIMMA_VOOM {
</span><span id="__span-65-4"><a id="__codelineno-65-4" name="__codelineno-65-4" href="#__codelineno-65-4"></a>
</span><span id="__span-65-5"><a id="__codelineno-65-5" name="__codelineno-65-5" href="#__codelineno-65-5"></a>    container &quot;quay.io/biocontainers/quay.io/biocontainers/bioconductor-limma:3.50.1--r41h5c21468_0&quot;
</span><span id="__span-65-6"><a id="__codelineno-65-6" name="__codelineno-65-6" href="#__codelineno-65-6"></a>
</span><span id="__span-65-7"><a id="__codelineno-65-7" name="__codelineno-65-7" href="#__codelineno-65-7"></a>    input:
</span><span id="__span-65-8"><a id="__codelineno-65-8" name="__codelineno-65-8" href="#__codelineno-65-8"></a>    path anno_geneanno
</span><span id="__span-65-9"><a id="__codelineno-65-9" name="__codelineno-65-9" href="#__codelineno-65-9"></a>    path cont_cinfo
</span><span id="__span-65-10"><a id="__codelineno-65-10" name="__codelineno-65-10" href="#__codelineno-65-10"></a>    path input_counts
</span><span id="__span-65-11"><a id="__codelineno-65-11" name="__codelineno-65-11" href="#__codelineno-65-11"></a>    path input_fact_finfo
</span><span id="__span-65-12"><a id="__codelineno-65-12" name="__codelineno-65-12" href="#__codelineno-65-12"></a>    path limma_voom_script
</span><span id="__span-65-13"><a id="__codelineno-65-13" name="__codelineno-65-13" href="#__codelineno-65-13"></a>    path out_report1
</span><span id="__span-65-14"><a id="__codelineno-65-14" name="__codelineno-65-14" href="#__codelineno-65-14"></a>    val out_report_files_path
</span><span id="__span-65-15"><a id="__codelineno-65-15" name="__codelineno-65-15" href="#__codelineno-65-15"></a>
</span><span id="__span-65-16"><a id="__codelineno-65-16" name="__codelineno-65-16" href="#__codelineno-65-16"></a>    output:
</span><span id="__span-65-17"><a id="__codelineno-65-17" name="__codelineno-65-17" href="#__codelineno-65-17"></a>    path &quot;output_dir/*_filtcounts&quot;, emit: outFilt
</span><span id="__span-65-18"><a id="__codelineno-65-18" name="__codelineno-65-18" href="#__codelineno-65-18"></a>    path &quot;output_dir/*_normcounts&quot;, emit: outNorm
</span><span id="__span-65-19"><a id="__codelineno-65-19" name="__codelineno-65-19" href="#__codelineno-65-19"></a>    path &quot;outReport.html&quot;, emit: outReport2
</span><span id="__span-65-20"><a id="__codelineno-65-20" name="__codelineno-65-20" href="#__codelineno-65-20"></a>    path &quot;output_dir/*.tsv?&quot;, emit: outTables
</span><span id="__span-65-21"><a id="__codelineno-65-21" name="__codelineno-65-21" href="#__codelineno-65-21"></a>    path &quot;libsizeinfo&quot;, emit: out_libinfo
</span><span id="__span-65-22"><a id="__codelineno-65-22" name="__codelineno-65-22" href="#__codelineno-65-22"></a>    path &quot;unknown_collection_pattern&quot;, emit: out_rscript
</span><span id="__span-65-23"><a id="__codelineno-65-23" name="__codelineno-65-23" href="#__codelineno-65-23"></a>
</span><span id="__span-65-24"><a id="__codelineno-65-24" name="__codelineno-65-24" href="#__codelineno-65-24"></a>    script:
</span><span id="__span-65-25"><a id="__codelineno-65-25" name="__codelineno-65-25" href="#__codelineno-65-25"></a>    &quot;&quot;&quot;
</span><span id="__span-65-26"><a id="__codelineno-65-26" name="__codelineno-65-26" href="#__codelineno-65-26"></a>    Rscript \
</span><span id="__span-65-27"><a id="__codelineno-65-27" name="__codelineno-65-27" href="#__codelineno-65-27"></a>    ${limma_voom_script} \
</span><span id="__span-65-28"><a id="__codelineno-65-28" name="__codelineno-65-28" href="#__codelineno-65-28"></a>    -C ${cont_cinfo} \
</span><span id="__span-65-29"><a id="__codelineno-65-29" name="__codelineno-65-29" href="#__codelineno-65-29"></a>    -R ${out_report1} \
</span><span id="__span-65-30"><a id="__codelineno-65-30" name="__codelineno-65-30" href="#__codelineno-65-30"></a>    -a ${anno_geneanno} \
</span><span id="__span-65-31"><a id="__codelineno-65-31" name="__codelineno-65-31" href="#__codelineno-65-31"></a>    -f ${input_fact_finfo} \
</span><span id="__span-65-32"><a id="__codelineno-65-32" name="__codelineno-65-32" href="#__codelineno-65-32"></a>    -m ${input_counts} \
</span><span id="__span-65-33"><a id="__codelineno-65-33" name="__codelineno-65-33" href="#__codelineno-65-33"></a>    -G 10 \
</span><span id="__span-65-34"><a id="__codelineno-65-34" name="__codelineno-65-34" href="#__codelineno-65-34"></a>    -P &quot;i&quot; \
</span><span id="__span-65-35"><a id="__codelineno-65-35" name="__codelineno-65-35" href="#__codelineno-65-35"></a>    -c 1 \
</span><span id="__span-65-36"><a id="__codelineno-65-36" name="__codelineno-65-36" href="#__codelineno-65-36"></a>    -d &quot;BH&quot; \
</span><span id="__span-65-37"><a id="__codelineno-65-37" name="__codelineno-65-37" href="#__codelineno-65-37"></a>    -j &quot;&quot; \
</span><span id="__span-65-38"><a id="__codelineno-65-38" name="__codelineno-65-38" href="#__codelineno-65-38"></a>    -l 0 \
</span><span id="__span-65-39"><a id="__codelineno-65-39" name="__codelineno-65-39" href="#__codelineno-65-39"></a>    -n &quot;TMM&quot; \
</span><span id="__span-65-40"><a id="__codelineno-65-40" name="__codelineno-65-40" href="#__codelineno-65-40"></a>    -o ${out_report_files_path} \
</span><span id="__span-65-41"><a id="__codelineno-65-41" name="__codelineno-65-41" href="#__codelineno-65-41"></a>    -p 0.05 \
</span><span id="__span-65-42"><a id="__codelineno-65-42" name="__codelineno-65-42" href="#__codelineno-65-42"></a>    -s 0 \
</span><span id="__span-65-43"><a id="__codelineno-65-43" name="__codelineno-65-43" href="#__codelineno-65-43"></a>    -t 3 \
</span><span id="__span-65-44"><a id="__codelineno-65-44" name="__codelineno-65-44" href="#__codelineno-65-44"></a>    -z 0
</span><span id="__span-65-45"><a id="__codelineno-65-45" name="__codelineno-65-45" href="#__codelineno-65-45"></a>    &quot;&quot;&quot;
</span><span id="__span-65-46"><a id="__codelineno-65-46" name="__codelineno-65-46" href="#__codelineno-65-46"></a>
</span><span id="__span-65-47"><a id="__codelineno-65-47" name="__codelineno-65-47" href="#__codelineno-65-47"></a>}
</span></code></pre></div>
<p>We see that this nextflow process has multiple inputs, many command line arguments, and multiple outputs. </p>
<p>Look at the command line arguments for this process in the <code>script:</code> block.</p>
<p>This Galaxy Tool Wrapper is evidently running an Rscript which we supply to the process via the <code>path limma_voom_script</code> input.
In the script section we see that this Rscript will be run, and has some CLI arguments which follow.</p>
<p>When translating Galaxy Tool Wrappers, we often see this usage of scripts. 
<code>limma</code> is an R library, so we need an Rscript to run the analysis we want. 
The Galaxy Tool Wrapper supplies user inputs to this script, then the script will run the analysis. </p>
<p><br></p>
<blockquote>
<p>NOTE<br>
<code>janis translate</code> will copy across any scripts referenced by a source Tool / Workflow. <br>
The <code>limma_voom/limma_voom.R</code> file is the Rscript which this Galaxy Tool Wrapper uses to run an analysis. </p>
</blockquote>
<p><br></p>
<p><strong>limma_voom.R</strong></p>
<p>From reading the nextflow process, it isn&rsquo;t particularly obvious what each command line argument does. 
We know that the process input are supplied to the <code>limma_voom.R</code> script, but the names aren&rsquo;t very descriptive. 
Luckily, the author of this script has good documentation at the top of the script to help us out. </p>
<p>Open <code>limma_voom.R</code>. </p>
<p>At the top of the file, we see some documentation:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-66-1"><a id="__codelineno-66-1" name="__codelineno-66-1" href="#__codelineno-66-1"></a># This tool takes in a matrix of feature counts as well as gene annotations and
</span><span id="__span-66-2"><a id="__codelineno-66-2" name="__codelineno-66-2" href="#__codelineno-66-2"></a># outputs a table of top expressions as well as various plots for differential
</span><span id="__span-66-3"><a id="__codelineno-66-3" name="__codelineno-66-3" href="#__codelineno-66-3"></a># expression analysis
</span><span id="__span-66-4"><a id="__codelineno-66-4" name="__codelineno-66-4" href="#__codelineno-66-4"></a>#
</span><span id="__span-66-5"><a id="__codelineno-66-5" name="__codelineno-66-5" href="#__codelineno-66-5"></a># ARGS: htmlPath&quot;, &quot;R&quot;, 1, &quot;character&quot;      -Path to html file linking to other outputs
</span><span id="__span-66-6"><a id="__codelineno-66-6" name="__codelineno-66-6" href="#__codelineno-66-6"></a>#       outPath&quot;, &quot;o&quot;, 1, &quot;character&quot;       -Path to folder to write all output to
</span><span id="__span-66-7"><a id="__codelineno-66-7" name="__codelineno-66-7" href="#__codelineno-66-7"></a>#       filesPath&quot;, &quot;j&quot;, 2, &quot;character&quot;     -JSON list object if multiple files input
</span><span id="__span-66-8"><a id="__codelineno-66-8" name="__codelineno-66-8" href="#__codelineno-66-8"></a>#       matrixPath&quot;, &quot;m&quot;, 2, &quot;character&quot;    -Path to count matrix
</span><span id="__span-66-9"><a id="__codelineno-66-9" name="__codelineno-66-9" href="#__codelineno-66-9"></a>#       factFile&quot;, &quot;f&quot;, 2, &quot;character&quot;      -Path to factor information file
</span><span id="__span-66-10"><a id="__codelineno-66-10" name="__codelineno-66-10" href="#__codelineno-66-10"></a>#       factInput&quot;, &quot;i&quot;, 2, &quot;character&quot;     -String containing factors if manually input
</span><span id="__span-66-11"><a id="__codelineno-66-11" name="__codelineno-66-11" href="#__codelineno-66-11"></a>#       annoPath&quot;, &quot;a&quot;, 2, &quot;character&quot;      -Path to input containing gene annotations
</span><span id="__span-66-12"><a id="__codelineno-66-12" name="__codelineno-66-12" href="#__codelineno-66-12"></a>#       contrastFile&quot;, &quot;C&quot;, 1, &quot;character&quot;  -Path to contrasts information file
</span><span id="__span-66-13"><a id="__codelineno-66-13" name="__codelineno-66-13" href="#__codelineno-66-13"></a>#       contrastInput&quot;, &quot;D&quot;, 1, &quot;character&quot; -String containing contrasts of interest
</span><span id="__span-66-14"><a id="__codelineno-66-14" name="__codelineno-66-14" href="#__codelineno-66-14"></a>#       cpmReq&quot;, &quot;c&quot;, 2, &quot;double&quot;           -Float specifying cpm requirement
</span><span id="__span-66-15"><a id="__codelineno-66-15" name="__codelineno-66-15" href="#__codelineno-66-15"></a>#       cntReq&quot;, &quot;z&quot;, 2, &quot;integer&quot;          -Integer specifying minimum total count requirement
</span><span id="__span-66-16"><a id="__codelineno-66-16" name="__codelineno-66-16" href="#__codelineno-66-16"></a>#       sampleReq&quot;, &quot;s&quot;, 2, &quot;integer&quot;       -Integer specifying cpm requirement
</span><span id="__span-66-17"><a id="__codelineno-66-17" name="__codelineno-66-17" href="#__codelineno-66-17"></a>#       normCounts&quot;, &quot;x&quot;, 0, &quot;logical&quot;      -String specifying if normalised counts should be output
</span><span id="__span-66-18"><a id="__codelineno-66-18" name="__codelineno-66-18" href="#__codelineno-66-18"></a>#       rdaOpt&quot;, &quot;r&quot;, 0, &quot;logical&quot;          -String specifying if RData should be output
</span><span id="__span-66-19"><a id="__codelineno-66-19" name="__codelineno-66-19" href="#__codelineno-66-19"></a>#       lfcReq&quot;, &quot;l&quot;, 1, &quot;double&quot;           -Float specifying the log-fold-change requirement
</span><span id="__span-66-20"><a id="__codelineno-66-20" name="__codelineno-66-20" href="#__codelineno-66-20"></a>#       pValReq&quot;, &quot;p&quot;, 1, &quot;double&quot;          -Float specifying the p-value requirement
</span><span id="__span-66-21"><a id="__codelineno-66-21" name="__codelineno-66-21" href="#__codelineno-66-21"></a>#       pAdjOpt&quot;, &quot;d&quot;, 1, &quot;character&quot;       -String specifying the p-value adjustment method
</span><span id="__span-66-22"><a id="__codelineno-66-22" name="__codelineno-66-22" href="#__codelineno-66-22"></a>#       normOpt&quot;, &quot;n&quot;, 1, &quot;character&quot;       -String specifying type of normalisation used
</span><span id="__span-66-23"><a id="__codelineno-66-23" name="__codelineno-66-23" href="#__codelineno-66-23"></a>#       robOpt&quot;, &quot;b&quot;, 0, &quot;logical&quot;          -String specifying if robust options should be used
</span><span id="__span-66-24"><a id="__codelineno-66-24" name="__codelineno-66-24" href="#__codelineno-66-24"></a>#       trend&quot;, &quot;t&quot;, 1, &quot;double&quot;            -Float for prior.count if limma-trend is used instead of voom
</span><span id="__span-66-25"><a id="__codelineno-66-25" name="__codelineno-66-25" href="#__codelineno-66-25"></a>#       weightOpt&quot;, &quot;w&quot;, 0, &quot;logical&quot;       -String specifying if voomWithQualityWeights should be used
</span><span id="__span-66-26"><a id="__codelineno-66-26" name="__codelineno-66-26" href="#__codelineno-66-26"></a>#       topgenes&quot;, &quot;G&quot;, 1, &quot;integer&quot;        -Integer specifying no. of genes to highlight in volcano and heatmap
</span><span id="__span-66-27"><a id="__codelineno-66-27" name="__codelineno-66-27" href="#__codelineno-66-27"></a>#       treatOpt&quot;, &quot;T&quot;, 0, &quot;logical&quot;        -String specifying if TREAT function should be used
</span><span id="__span-66-28"><a id="__codelineno-66-28" name="__codelineno-66-28" href="#__codelineno-66-28"></a>#       plots, &quot;P&quot;, 1, &quot;character&quot;          -String specifying additional plots to be created
</span><span id="__span-66-29"><a id="__codelineno-66-29" name="__codelineno-66-29" href="#__codelineno-66-29"></a>#
</span><span id="__span-66-30"><a id="__codelineno-66-30" name="__codelineno-66-30" href="#__codelineno-66-30"></a># OUT:
</span><span id="__span-66-31"><a id="__codelineno-66-31" name="__codelineno-66-31" href="#__codelineno-66-31"></a>#       Density Plots (if filtering)
</span><span id="__span-66-32"><a id="__codelineno-66-32" name="__codelineno-66-32" href="#__codelineno-66-32"></a>#       Box Plots (if normalising)
</span><span id="__span-66-33"><a id="__codelineno-66-33" name="__codelineno-66-33" href="#__codelineno-66-33"></a>#       MDS Plot
</span><span id="__span-66-34"><a id="__codelineno-66-34" name="__codelineno-66-34" href="#__codelineno-66-34"></a>#       Voom/SA plot
</span><span id="__span-66-35"><a id="__codelineno-66-35" name="__codelineno-66-35" href="#__codelineno-66-35"></a>#       MD Plot
</span><span id="__span-66-36"><a id="__codelineno-66-36" name="__codelineno-66-36" href="#__codelineno-66-36"></a>#       Volcano Plot
</span><span id="__span-66-37"><a id="__codelineno-66-37" name="__codelineno-66-37" href="#__codelineno-66-37"></a>#       Heatmap
</span><span id="__span-66-38"><a id="__codelineno-66-38" name="__codelineno-66-38" href="#__codelineno-66-38"></a>#       Expression Table
</span><span id="__span-66-39"><a id="__codelineno-66-39" name="__codelineno-66-39" href="#__codelineno-66-39"></a>#       HTML file linking to the ouputs
</span><span id="__span-66-40"><a id="__codelineno-66-40" name="__codelineno-66-40" href="#__codelineno-66-40"></a># Optional:
</span><span id="__span-66-41"><a id="__codelineno-66-41" name="__codelineno-66-41" href="#__codelineno-66-41"></a>#       Normalised counts Table
</span><span id="__span-66-42"><a id="__codelineno-66-42" name="__codelineno-66-42" href="#__codelineno-66-42"></a>#       RData file
</span><span id="__span-66-43"><a id="__codelineno-66-43" name="__codelineno-66-43" href="#__codelineno-66-43"></a>#
</span><span id="__span-66-44"><a id="__codelineno-66-44" name="__codelineno-66-44" href="#__codelineno-66-44"></a>#
</span><span id="__span-66-45"><a id="__codelineno-66-45" name="__codelineno-66-45" href="#__codelineno-66-45"></a># Author: Shian Su - registertonysu@gmail.com - Jan 2014
</span><span id="__span-66-46"><a id="__codelineno-66-46" name="__codelineno-66-46" href="#__codelineno-66-46"></a># Modified by: Maria Doyle - Jun 2017, Jan 2018, May 2018
</span></code></pre></div>
<p>For each <strong><em>process input</em></strong>, find the command line argument it feeds, then look up the argument documentation in <code>limma_voom.R</code>. </p>
<p>For example, the <code>anno_geneanno</code> process input feeds the <code>-a</code> argument.<br>
Looking at the documentation, we see that this is the gene annotations file. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-67-1"><a id="__codelineno-67-1" name="__codelineno-67-1" href="#__codelineno-67-1"></a># limma_voom.nf
</span><span id="__span-67-2"><a id="__codelineno-67-2" name="__codelineno-67-2" href="#__codelineno-67-2"></a>-a ${anno_geneanno} \
</span><span id="__span-67-3"><a id="__codelineno-67-3" name="__codelineno-67-3" href="#__codelineno-67-3"></a>
</span><span id="__span-67-4"><a id="__codelineno-67-4" name="__codelineno-67-4" href="#__codelineno-67-4"></a># limma_voom.R
</span><span id="__span-67-5"><a id="__codelineno-67-5" name="__codelineno-67-5" href="#__codelineno-67-5"></a>annoPath&quot;, &quot;a&quot;, 2, &quot;character&quot;      -Path to input containing gene annotations
</span></code></pre></div>
The <code>cont_cinfo</code> process input feeds the <code>-C</code> argument, which is a file containing contrasts of interest:
<div class="language-text highlight"><pre><span></span><code><span id="__span-68-1"><a id="__codelineno-68-1" name="__codelineno-68-1" href="#__codelineno-68-1"></a># limma_voom.nf
</span><span id="__span-68-2"><a id="__codelineno-68-2" name="__codelineno-68-2" href="#__codelineno-68-2"></a>-C ${cont_cinfo} \
</span><span id="__span-68-3"><a id="__codelineno-68-3" name="__codelineno-68-3" href="#__codelineno-68-3"></a>
</span><span id="__span-68-4"><a id="__codelineno-68-4" name="__codelineno-68-4" href="#__codelineno-68-4"></a># limma_voom.R
</span><span id="__span-68-5"><a id="__codelineno-68-5" name="__codelineno-68-5" href="#__codelineno-68-5"></a>contrastFile&quot;, &quot;C&quot;, 1, &quot;character&quot;  -Path to contrasts information file
</span></code></pre></div></p>
<p><br></p>
<p>Now that we have had a little look at how the <code>LIMMA_VOOM</code> nextflow process will run, 
let&rsquo;s make some adjustments so that it functions as intended. </p>
<p><br></p>
<p><strong>Modifying Container</strong></p>
<p>The first thing to do is change the container requirement. </p>
<p>Galaxy uses <code>conda</code> to handle tool requirements. </p>
<p>For Galaxy Tool Wrappers which only have a single requirement, janis-translate will just fetch a container for that single requirement by looking up registries on <code>quay.io</code>. For those which have 2+ requirements, things get complicated. </p>
<p>For the <code>limma</code> Galaxy Tool Wrapper, the following are listed:</p>
<ul>
<li>bioconductor-limma==3.50.1</li>
<li>bioconductor-edger==3.36.0</li>
<li>r-statmod==1.4.36</li>
<li>r-scales==1.1.1</li>
<li>r-rjson==0.2.21</li>
<li>r-getopt==1.20.3</li>
<li>r-gplots==3.1.1</li>
<li>bioconductor-glimma==2.4.0</li>
</ul>
<p>For best-practises pipelines, we want to use containers instead of conda, and want a single image.</p>
<p>To facilitate this, janis-translate has a new test-feature: <code>--build-galaxy-tool-images</code></p>
<p>This feature allows janis-translate to build a single container image during runtime which has all software requirements. </p>
<p>As this feature requires docker (which is not available on our training VMs today), and also because containers can take anywhere from 2-15 minutes to build, we won&rsquo;t use this feature today. </p>
<p>Instead, we have pre-built images for the relevant tools using janis-translate, and placed them on a <code>quay.io</code> repository. </p>
<div class="admonition question">
<p class="admonition-title">Swapping container directive</p>
<p>Change the LIMMA_VOOM process container directive from:<br>
<code>quay.io/biocontainers/bioconductor-limma:3.50.1--r41h5c21468_0</code></p>
<p>To:<br>
<code>quay.io/grace_hall1/limma-voom:3.50.1</code></p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-69-1"><a id="__codelineno-69-1" name="__codelineno-69-1" href="#__codelineno-69-1"></a>process LIMMA_VOOM {
</span><span id="__span-69-2"><a id="__codelineno-69-2" name="__codelineno-69-2" href="#__codelineno-69-2"></a>
</span><span id="__span-69-3"><a id="__codelineno-69-3" name="__codelineno-69-3" href="#__codelineno-69-3"></a>    container &quot;quay.io/grace_hall1/limma-voom:3.50.1&quot;     &lt;- 
</span><span id="__span-69-4"><a id="__codelineno-69-4" name="__codelineno-69-4" href="#__codelineno-69-4"></a>    publishDir &quot;./outputs&quot;
</span><span id="__span-69-5"><a id="__codelineno-69-5" name="__codelineno-69-5" href="#__codelineno-69-5"></a>
</span><span id="__span-69-6"><a id="__codelineno-69-6" name="__codelineno-69-6" href="#__codelineno-69-6"></a>    input:
</span><span id="__span-69-7"><a id="__codelineno-69-7" name="__codelineno-69-7" href="#__codelineno-69-7"></a>    path anno_geneanno
</span><span id="__span-69-8"><a id="__codelineno-69-8" name="__codelineno-69-8" href="#__codelineno-69-8"></a>    path cont_cinfo
</span><span id="__span-69-9"><a id="__codelineno-69-9" name="__codelineno-69-9" href="#__codelineno-69-9"></a>    path input_counts
</span><span id="__span-69-10"><a id="__codelineno-69-10" name="__codelineno-69-10" href="#__codelineno-69-10"></a>    path input_fact_finfo
</span><span id="__span-69-11"><a id="__codelineno-69-11" name="__codelineno-69-11" href="#__codelineno-69-11"></a>    path limma_voom_script
</span><span id="__span-69-12"><a id="__codelineno-69-12" name="__codelineno-69-12" href="#__codelineno-69-12"></a>    path out_report1
</span><span id="__span-69-13"><a id="__codelineno-69-13" name="__codelineno-69-13" href="#__codelineno-69-13"></a>    val out_report_files_path
</span><span id="__span-69-14"><a id="__codelineno-69-14" name="__codelineno-69-14" href="#__codelineno-69-14"></a>    ...
</span><span id="__span-69-15"><a id="__codelineno-69-15" name="__codelineno-69-15" href="#__codelineno-69-15"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p><strong>Add publishDir Directive</strong></p>
<p>While we&rsquo;re modifying directives, let&rsquo;s add a <code>publishDir</code> directive. </p>
<p>This lets us specify a folder where outputs of this process should be presented.</p>
<div class="admonition question">
<p class="admonition-title">Adding publishDir Directive</p>
<p><em>Add</em> a publishDir directive for the process with the path <code>"./outputs"</code>.</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-70-1"><a id="__codelineno-70-1" name="__codelineno-70-1" href="#__codelineno-70-1"></a>process LIMMA_VOOM {
</span><span id="__span-70-2"><a id="__codelineno-70-2" name="__codelineno-70-2" href="#__codelineno-70-2"></a>
</span><span id="__span-70-3"><a id="__codelineno-70-3" name="__codelineno-70-3" href="#__codelineno-70-3"></a>    container &quot;quay.io/biocontainers/janis-translate-limma-voom-3.34.9.9&quot;
</span><span id="__span-70-4"><a id="__codelineno-70-4" name="__codelineno-70-4" href="#__codelineno-70-4"></a>    publishDir &quot;./outputs&quot;                                                      &lt;-
</span><span id="__span-70-5"><a id="__codelineno-70-5" name="__codelineno-70-5" href="#__codelineno-70-5"></a>
</span><span id="__span-70-6"><a id="__codelineno-70-6" name="__codelineno-70-6" href="#__codelineno-70-6"></a>    ...
</span><span id="__span-70-7"><a id="__codelineno-70-7" name="__codelineno-70-7" href="#__codelineno-70-7"></a>
</span><span id="__span-70-8"><a id="__codelineno-70-8" name="__codelineno-70-8" href="#__codelineno-70-8"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p><strong>Modifying Inputs</strong></p>
<p>The <code>path out_report1</code> process input specifies the name of a html file which will present our results. 
We aren&rsquo;t actually supplying a file; we are providing the script a filename, which should be a string.</p>
<div class="admonition question">
<p class="admonition-title">Modifying inputs: html path</p>
<p>Change the type of the <code>out_report1</code> input from <code>path</code> to <code>val</code>.</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-71-1"><a id="__codelineno-71-1" name="__codelineno-71-1" href="#__codelineno-71-1"></a>process LIMMA_VOOM {
</span><span id="__span-71-2"><a id="__codelineno-71-2" name="__codelineno-71-2" href="#__codelineno-71-2"></a>
</span><span id="__span-71-3"><a id="__codelineno-71-3" name="__codelineno-71-3" href="#__codelineno-71-3"></a>    container &quot;quay.io/biocontainers/janis-translate-limma-voom-3.34.9.9&quot;
</span><span id="__span-71-4"><a id="__codelineno-71-4" name="__codelineno-71-4" href="#__codelineno-71-4"></a>    publishDir &quot;./outputs&quot;
</span><span id="__span-71-5"><a id="__codelineno-71-5" name="__codelineno-71-5" href="#__codelineno-71-5"></a>
</span><span id="__span-71-6"><a id="__codelineno-71-6" name="__codelineno-71-6" href="#__codelineno-71-6"></a>    input:
</span><span id="__span-71-7"><a id="__codelineno-71-7" name="__codelineno-71-7" href="#__codelineno-71-7"></a>    path anno_geneanno
</span><span id="__span-71-8"><a id="__codelineno-71-8" name="__codelineno-71-8" href="#__codelineno-71-8"></a>    path cont_cinfo
</span><span id="__span-71-9"><a id="__codelineno-71-9" name="__codelineno-71-9" href="#__codelineno-71-9"></a>    path input_counts
</span><span id="__span-71-10"><a id="__codelineno-71-10" name="__codelineno-71-10" href="#__codelineno-71-10"></a>    path input_fact_finfo
</span><span id="__span-71-11"><a id="__codelineno-71-11" name="__codelineno-71-11" href="#__codelineno-71-11"></a>    path limma_voom_script
</span><span id="__span-71-12"><a id="__codelineno-71-12" name="__codelineno-71-12" href="#__codelineno-71-12"></a>    val out_report1                         &lt;- 
</span><span id="__span-71-13"><a id="__codelineno-71-13" name="__codelineno-71-13" href="#__codelineno-71-13"></a>    val out_report_files_path
</span><span id="__span-71-14"><a id="__codelineno-71-14" name="__codelineno-71-14" href="#__codelineno-71-14"></a>    ...
</span><span id="__span-71-15"><a id="__codelineno-71-15" name="__codelineno-71-15" href="#__codelineno-71-15"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p><strong>Modifying Script</strong></p>
<p>From reading the documentation, you may have noticed that one particular argument isn&rsquo;t needed. 
The <code>-j</code> argument is only needed when we have multiple input files - but in our case we are using a single counts file.  </p>
<p>We will be using a single input counts file, so can remove this argument. </p>
<div class="admonition question">
<p class="admonition-title">Modifying script: <code>-j</code> argument</p>
<p><em>Remove</em> the <code>-j</code> argument and its value from the script section. </p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-72-1"><a id="__codelineno-72-1" name="__codelineno-72-1" href="#__codelineno-72-1"></a>process LIMMA_VOOM {
</span><span id="__span-72-2"><a id="__codelineno-72-2" name="__codelineno-72-2" href="#__codelineno-72-2"></a>    ...
</span><span id="__span-72-3"><a id="__codelineno-72-3" name="__codelineno-72-3" href="#__codelineno-72-3"></a>
</span><span id="__span-72-4"><a id="__codelineno-72-4" name="__codelineno-72-4" href="#__codelineno-72-4"></a>    script:
</span><span id="__span-72-5"><a id="__codelineno-72-5" name="__codelineno-72-5" href="#__codelineno-72-5"></a>    &quot;&quot;&quot;
</span><span id="__span-72-6"><a id="__codelineno-72-6" name="__codelineno-72-6" href="#__codelineno-72-6"></a>    Rscript \
</span><span id="__span-72-7"><a id="__codelineno-72-7" name="__codelineno-72-7" href="#__codelineno-72-7"></a>    ${limma_voom_script} \
</span><span id="__span-72-8"><a id="__codelineno-72-8" name="__codelineno-72-8" href="#__codelineno-72-8"></a>    -C ${cont_cinfo} \
</span><span id="__span-72-9"><a id="__codelineno-72-9" name="__codelineno-72-9" href="#__codelineno-72-9"></a>    -R ${out_report1} \
</span><span id="__span-72-10"><a id="__codelineno-72-10" name="__codelineno-72-10" href="#__codelineno-72-10"></a>    -a ${anno_geneanno} \
</span><span id="__span-72-11"><a id="__codelineno-72-11" name="__codelineno-72-11" href="#__codelineno-72-11"></a>    -f ${input_fact_finfo} \
</span><span id="__span-72-12"><a id="__codelineno-72-12" name="__codelineno-72-12" href="#__codelineno-72-12"></a>    -m ${input_counts} \
</span><span id="__span-72-13"><a id="__codelineno-72-13" name="__codelineno-72-13" href="#__codelineno-72-13"></a>    -G 10 \
</span><span id="__span-72-14"><a id="__codelineno-72-14" name="__codelineno-72-14" href="#__codelineno-72-14"></a>    -P &quot;i&quot; \
</span><span id="__span-72-15"><a id="__codelineno-72-15" name="__codelineno-72-15" href="#__codelineno-72-15"></a>    -c 1 \
</span><span id="__span-72-16"><a id="__codelineno-72-16" name="__codelineno-72-16" href="#__codelineno-72-16"></a>    -d &quot;BH&quot; \                   
</span><span id="__span-72-17"><a id="__codelineno-72-17" name="__codelineno-72-17" href="#__codelineno-72-17"></a>    -l 0 \                              &lt;- `-j &quot;&quot;` removed
</span><span id="__span-72-18"><a id="__codelineno-72-18" name="__codelineno-72-18" href="#__codelineno-72-18"></a>    -n &quot;TMM&quot; \
</span><span id="__span-72-19"><a id="__codelineno-72-19" name="__codelineno-72-19" href="#__codelineno-72-19"></a>    -o ${out_report_files_path} \
</span><span id="__span-72-20"><a id="__codelineno-72-20" name="__codelineno-72-20" href="#__codelineno-72-20"></a>    -p 0.05 \
</span><span id="__span-72-21"><a id="__codelineno-72-21" name="__codelineno-72-21" href="#__codelineno-72-21"></a>    -s 0 \
</span><span id="__span-72-22"><a id="__codelineno-72-22" name="__codelineno-72-22" href="#__codelineno-72-22"></a>    -t 3 \
</span><span id="__span-72-23"><a id="__codelineno-72-23" name="__codelineno-72-23" href="#__codelineno-72-23"></a>    -z 0
</span><span id="__span-72-24"><a id="__codelineno-72-24" name="__codelineno-72-24" href="#__codelineno-72-24"></a>    &quot;&quot;&quot;
</span><span id="__span-72-25"><a id="__codelineno-72-25" name="__codelineno-72-25" href="#__codelineno-72-25"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p><strong>Modifying Outputs</strong></p>
<p>Galaxy Tool Wrappers often allow you to generate extra outputs based on what the user wants.</p>
<p>For this tutorial, we&rsquo;re not interested in any of the optional outputs - just the single <code>outReport2</code> output.</p>
<div class="admonition question">
<p class="admonition-title">Modifying outputs</p>
<p><em>Remove</em> all the process outputs except <code>outReport2</code>.</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-73-1"><a id="__codelineno-73-1" name="__codelineno-73-1" href="#__codelineno-73-1"></a>process LIMMA_VOOM {
</span><span id="__span-73-2"><a id="__codelineno-73-2" name="__codelineno-73-2" href="#__codelineno-73-2"></a>
</span><span id="__span-73-3"><a id="__codelineno-73-3" name="__codelineno-73-3" href="#__codelineno-73-3"></a>    container &quot;quay.io/biocontainers/janis-translate-limma-voom-3.34.9.9&quot;
</span><span id="__span-73-4"><a id="__codelineno-73-4" name="__codelineno-73-4" href="#__codelineno-73-4"></a>
</span><span id="__span-73-5"><a id="__codelineno-73-5" name="__codelineno-73-5" href="#__codelineno-73-5"></a>    input:
</span><span id="__span-73-6"><a id="__codelineno-73-6" name="__codelineno-73-6" href="#__codelineno-73-6"></a>    path anno_geneanno
</span><span id="__span-73-7"><a id="__codelineno-73-7" name="__codelineno-73-7" href="#__codelineno-73-7"></a>    path cont_cinfo
</span><span id="__span-73-8"><a id="__codelineno-73-8" name="__codelineno-73-8" href="#__codelineno-73-8"></a>    path input_counts
</span><span id="__span-73-9"><a id="__codelineno-73-9" name="__codelineno-73-9" href="#__codelineno-73-9"></a>    path input_fact_finfo
</span><span id="__span-73-10"><a id="__codelineno-73-10" name="__codelineno-73-10" href="#__codelineno-73-10"></a>    path limma_voom_script
</span><span id="__span-73-11"><a id="__codelineno-73-11" name="__codelineno-73-11" href="#__codelineno-73-11"></a>    path out_report1
</span><span id="__span-73-12"><a id="__codelineno-73-12" name="__codelineno-73-12" href="#__codelineno-73-12"></a>    val out_report_files_path
</span><span id="__span-73-13"><a id="__codelineno-73-13" name="__codelineno-73-13" href="#__codelineno-73-13"></a>
</span><span id="__span-73-14"><a id="__codelineno-73-14" name="__codelineno-73-14" href="#__codelineno-73-14"></a>    output:
</span><span id="__span-73-15"><a id="__codelineno-73-15" name="__codelineno-73-15" href="#__codelineno-73-15"></a>    path &quot;outReport.html&quot;, emit: outReport2         &lt;- other outputs removed
</span><span id="__span-73-16"><a id="__codelineno-73-16" name="__codelineno-73-16" href="#__codelineno-73-16"></a>
</span><span id="__span-73-17"><a id="__codelineno-73-17" name="__codelineno-73-17" href="#__codelineno-73-17"></a>    ...
</span><span id="__span-73-18"><a id="__codelineno-73-18" name="__codelineno-73-18" href="#__codelineno-73-18"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p>Your Nextflow process should now look similar to the following: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-74-1"><a id="__codelineno-74-1" name="__codelineno-74-1" href="#__codelineno-74-1"></a>process LIMMA_VOOM {
</span><span id="__span-74-2"><a id="__codelineno-74-2" name="__codelineno-74-2" href="#__codelineno-74-2"></a>
</span><span id="__span-74-3"><a id="__codelineno-74-3" name="__codelineno-74-3" href="#__codelineno-74-3"></a>    container &quot;quay.io/grace_hall1/limma-voom:3.50.1&quot;
</span><span id="__span-74-4"><a id="__codelineno-74-4" name="__codelineno-74-4" href="#__codelineno-74-4"></a>    publishDir &quot;./outputs&quot;
</span><span id="__span-74-5"><a id="__codelineno-74-5" name="__codelineno-74-5" href="#__codelineno-74-5"></a>
</span><span id="__span-74-6"><a id="__codelineno-74-6" name="__codelineno-74-6" href="#__codelineno-74-6"></a>    input:
</span><span id="__span-74-7"><a id="__codelineno-74-7" name="__codelineno-74-7" href="#__codelineno-74-7"></a>    path anno_geneanno
</span><span id="__span-74-8"><a id="__codelineno-74-8" name="__codelineno-74-8" href="#__codelineno-74-8"></a>    path cont_cinfo
</span><span id="__span-74-9"><a id="__codelineno-74-9" name="__codelineno-74-9" href="#__codelineno-74-9"></a>    path input_counts
</span><span id="__span-74-10"><a id="__codelineno-74-10" name="__codelineno-74-10" href="#__codelineno-74-10"></a>    path input_fact_finfo
</span><span id="__span-74-11"><a id="__codelineno-74-11" name="__codelineno-74-11" href="#__codelineno-74-11"></a>    path limma_voom_script
</span><span id="__span-74-12"><a id="__codelineno-74-12" name="__codelineno-74-12" href="#__codelineno-74-12"></a>    val out_report1
</span><span id="__span-74-13"><a id="__codelineno-74-13" name="__codelineno-74-13" href="#__codelineno-74-13"></a>    val out_report_files_path
</span><span id="__span-74-14"><a id="__codelineno-74-14" name="__codelineno-74-14" href="#__codelineno-74-14"></a>
</span><span id="__span-74-15"><a id="__codelineno-74-15" name="__codelineno-74-15" href="#__codelineno-74-15"></a>    output:
</span><span id="__span-74-16"><a id="__codelineno-74-16" name="__codelineno-74-16" href="#__codelineno-74-16"></a>    path &quot;outReport.html&quot;, emit: outReport2
</span><span id="__span-74-17"><a id="__codelineno-74-17" name="__codelineno-74-17" href="#__codelineno-74-17"></a>
</span><span id="__span-74-18"><a id="__codelineno-74-18" name="__codelineno-74-18" href="#__codelineno-74-18"></a>    script:
</span><span id="__span-74-19"><a id="__codelineno-74-19" name="__codelineno-74-19" href="#__codelineno-74-19"></a>    &quot;&quot;&quot;
</span><span id="__span-74-20"><a id="__codelineno-74-20" name="__codelineno-74-20" href="#__codelineno-74-20"></a>    Rscript \
</span><span id="__span-74-21"><a id="__codelineno-74-21" name="__codelineno-74-21" href="#__codelineno-74-21"></a>    ${limma_voom_script} \
</span><span id="__span-74-22"><a id="__codelineno-74-22" name="__codelineno-74-22" href="#__codelineno-74-22"></a>    -C ${cont_cinfo} \
</span><span id="__span-74-23"><a id="__codelineno-74-23" name="__codelineno-74-23" href="#__codelineno-74-23"></a>    -R ${out_report1} \
</span><span id="__span-74-24"><a id="__codelineno-74-24" name="__codelineno-74-24" href="#__codelineno-74-24"></a>    -a ${anno_geneanno} \
</span><span id="__span-74-25"><a id="__codelineno-74-25" name="__codelineno-74-25" href="#__codelineno-74-25"></a>    -f ${input_fact_finfo} \
</span><span id="__span-74-26"><a id="__codelineno-74-26" name="__codelineno-74-26" href="#__codelineno-74-26"></a>    -m ${input_counts} \
</span><span id="__span-74-27"><a id="__codelineno-74-27" name="__codelineno-74-27" href="#__codelineno-74-27"></a>    -G 10 \
</span><span id="__span-74-28"><a id="__codelineno-74-28" name="__codelineno-74-28" href="#__codelineno-74-28"></a>    -P &quot;i&quot; \
</span><span id="__span-74-29"><a id="__codelineno-74-29" name="__codelineno-74-29" href="#__codelineno-74-29"></a>    -c 1 \
</span><span id="__span-74-30"><a id="__codelineno-74-30" name="__codelineno-74-30" href="#__codelineno-74-30"></a>    -d &quot;BH&quot; \
</span><span id="__span-74-31"><a id="__codelineno-74-31" name="__codelineno-74-31" href="#__codelineno-74-31"></a>    -l 0 \
</span><span id="__span-74-32"><a id="__codelineno-74-32" name="__codelineno-74-32" href="#__codelineno-74-32"></a>    -n &quot;TMM&quot; \
</span><span id="__span-74-33"><a id="__codelineno-74-33" name="__codelineno-74-33" href="#__codelineno-74-33"></a>    -o ${out_report_files_path} \
</span><span id="__span-74-34"><a id="__codelineno-74-34" name="__codelineno-74-34" href="#__codelineno-74-34"></a>    -p 0.05 \
</span><span id="__span-74-35"><a id="__codelineno-74-35" name="__codelineno-74-35" href="#__codelineno-74-35"></a>    -s 0 \
</span><span id="__span-74-36"><a id="__codelineno-74-36" name="__codelineno-74-36" href="#__codelineno-74-36"></a>    -t 3 \
</span><span id="__span-74-37"><a id="__codelineno-74-37" name="__codelineno-74-37" href="#__codelineno-74-37"></a>    -z 0
</span><span id="__span-74-38"><a id="__codelineno-74-38" name="__codelineno-74-38" href="#__codelineno-74-38"></a>    &quot;&quot;&quot;
</span><span id="__span-74-39"><a id="__codelineno-74-39" name="__codelineno-74-39" href="#__codelineno-74-39"></a>
</span><span id="__span-74-40"><a id="__codelineno-74-40" name="__codelineno-74-40" href="#__codelineno-74-40"></a>}
</span></code></pre></div></p>
<p>Now that we have fixed up the process definition, we can set up <code>nextflow.config</code> and run the process with sample data to test. </p>
<p><br></p>
<h3 id="running-limma-voom-as-a-workflow">Running Limma Voom as a Workflow<a class="headerlink" href="#running-limma-voom-as-a-workflow" title="Permanent link">&para;</a></h3>
<p>In this section we will run our translated <code>LIMMA_VOOM</code> process. </p>
<p>We will set up a workflow with a single <code>LIMMA_VOOM</code> task, and will supply inputs to this task using <code>nextflow.config</code>. </p>
<p><br></p>
<p><strong>Setting up nextflow.config</strong></p>
<p>To run this process, we will set up a <code>nextflow.config</code> file to supply inputs and other config. </p>
<p>Create a new file called <code>nextflow.config</code> in the <code>limma_voom/</code> folder alongside <code>limma_voom.nf</code>. </p>
<p>Copy and paste the following code into your <code>nextflow.config</code> file. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-75-1"><a id="__codelineno-75-1" name="__codelineno-75-1" href="#__codelineno-75-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-75-2"><a id="__codelineno-75-2" name="__codelineno-75-2" href="#__codelineno-75-2"></a>singularity.enabled = true
</span><span id="__span-75-3"><a id="__codelineno-75-3" name="__codelineno-75-3" href="#__codelineno-75-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-75-4"><a id="__codelineno-75-4" name="__codelineno-75-4" href="#__codelineno-75-4"></a>
</span><span id="__span-75-5"><a id="__codelineno-75-5" name="__codelineno-75-5" href="#__codelineno-75-5"></a>params {
</span><span id="__span-75-6"><a id="__codelineno-75-6" name="__codelineno-75-6" href="#__codelineno-75-6"></a>    limma_voom_script = &quot;/home2/training/limma_voom/limma_voom.R&quot;
</span><span id="__span-75-7"><a id="__codelineno-75-7" name="__codelineno-75-7" href="#__codelineno-75-7"></a>    annotation_file   = &quot;/home2/training/data/sample_data/galaxy/limma_voom/anno.txt&quot;
</span><span id="__span-75-8"><a id="__codelineno-75-8" name="__codelineno-75-8" href="#__codelineno-75-8"></a>    contrast_file     = &quot;/home2/training/data/sample_data/galaxy/limma_voom/contrasts.txt&quot;
</span><span id="__span-75-9"><a id="__codelineno-75-9" name="__codelineno-75-9" href="#__codelineno-75-9"></a>    factor_file       = &quot;/home2/training/data/sample_data/galaxy/limma_voom/factorinfo.txt&quot;
</span><span id="__span-75-10"><a id="__codelineno-75-10" name="__codelineno-75-10" href="#__codelineno-75-10"></a>    matrix_file       = &quot;/home2/training/data/sample_data/galaxy/limma_voom/matrix.txt&quot;
</span><span id="__span-75-11"><a id="__codelineno-75-11" name="__codelineno-75-11" href="#__codelineno-75-11"></a>    html_path         = &quot;outReport.html&quot;
</span><span id="__span-75-12"><a id="__codelineno-75-12" name="__codelineno-75-12" href="#__codelineno-75-12"></a>    output_path       = &quot;.&quot;
</span><span id="__span-75-13"><a id="__codelineno-75-13" name="__codelineno-75-13" href="#__codelineno-75-13"></a>}
</span></code></pre></div></p>
<p>This tells nextflow to use singularity, and sets up input parameters for our sample data.</p>
<p><br></p>
<p><strong>Creating Workflow &amp; Passing Data</strong> </p>
<p>Now we have input data set up using the <code>params</code> global variable, we will add some lines to the top of <code>limma_voom.nf</code> to turn it into a workflow.</p>
<p>Copy and paste the following lines at the top of <code>limma_voom.nf</code>:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-76-1"><a id="__codelineno-76-1" name="__codelineno-76-1" href="#__codelineno-76-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-76-2"><a id="__codelineno-76-2" name="__codelineno-76-2" href="#__codelineno-76-2"></a>
</span><span id="__span-76-3"><a id="__codelineno-76-3" name="__codelineno-76-3" href="#__codelineno-76-3"></a>workflow {
</span><span id="__span-76-4"><a id="__codelineno-76-4" name="__codelineno-76-4" href="#__codelineno-76-4"></a>    LIMMA_VOOM(
</span><span id="__span-76-5"><a id="__codelineno-76-5" name="__codelineno-76-5" href="#__codelineno-76-5"></a>        params.annotation_file,     // anno_geneanno
</span><span id="__span-76-6"><a id="__codelineno-76-6" name="__codelineno-76-6" href="#__codelineno-76-6"></a>        params.contrast_file,       // cont_cinfo
</span><span id="__span-76-7"><a id="__codelineno-76-7" name="__codelineno-76-7" href="#__codelineno-76-7"></a>        params.matrix_file,         // input_counts
</span><span id="__span-76-8"><a id="__codelineno-76-8" name="__codelineno-76-8" href="#__codelineno-76-8"></a>        params.factor_file,         // input_fact_finfo
</span><span id="__span-76-9"><a id="__codelineno-76-9" name="__codelineno-76-9" href="#__codelineno-76-9"></a>        params.limma_voom_script,   // limma_voom_script
</span><span id="__span-76-10"><a id="__codelineno-76-10" name="__codelineno-76-10" href="#__codelineno-76-10"></a>        params.html_path,           // out_report1
</span><span id="__span-76-11"><a id="__codelineno-76-11" name="__codelineno-76-11" href="#__codelineno-76-11"></a>        params.output_path,         // out_report_files_path
</span><span id="__span-76-12"><a id="__codelineno-76-12" name="__codelineno-76-12" href="#__codelineno-76-12"></a>    )
</span><span id="__span-76-13"><a id="__codelineno-76-13" name="__codelineno-76-13" href="#__codelineno-76-13"></a>}
</span></code></pre></div>
<p>The new <code>workflow {}</code> section declares the main workflow entry point. <br>
When we run <code>limma_voom.nf</code>, nextflow will look for this section and run the workflow contained within. </p>
<p>In our case, the workflow only contains a single task, which runs the <code>LIMMA_VOOM</code> process defined below the workflow section. We are passing the inputs we set up as <code>params</code> in <code>nextflow.config</code> to feed values to the process inputs.</p>
<p><br></p>
<p><strong>Running Our Workflow</strong></p>
<p>Ensure you are in the <code>limma_voom/</code> working directory, where <code>nextflow.config</code> and <code>limma_voom.nf</code> reside. </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-77-1"><a id="__codelineno-77-1" name="__codelineno-77-1" href="#__codelineno-77-1"></a>cd limma_voom
</span></code></pre></div>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-78-1"><a id="__codelineno-78-1" name="__codelineno-78-1" href="#__codelineno-78-1"></a>nextflow run limma_voom.nf
</span></code></pre></div></p>
<p>Nextflow will automatically check if there is a <code>nextflow.config</code> file in the working directory, and if so will use that to configure itself. Our inputs are supplied in <code>nextflow.config</code> alongside the dsl2 &amp; singularity config, so it should run without issue. </p>
<p><br></p>
<p><strong>Viewing Results</strong></p>
<p>If everything went well, we should see a new folder created called <code>outputs/</code> which has a single <code>outReport.html</code> file inside. </p>
<p>If needed, you can check the <code>data/final/galaxy/limma_voom</code> folder as a reference.</p>
<p><code>outReport.html</code> will be a symlink to the work folder where <code>LIMMA_VOOM</code> ran.</p>
<p>To see the symlink path, use the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-79-1"><a id="__codelineno-79-1" name="__codelineno-79-1" href="#__codelineno-79-1"></a>ls -lah outputs/outReport.html
</span></code></pre></div></p>
<p>The output should be similar to the following:
<div class="language-text highlight"><pre><span></span><code><span id="__span-80-1"><a id="__codelineno-80-1" name="__codelineno-80-1" href="#__codelineno-80-1"></a>lrwxrwxrwx 1 training users 80 Jun 18 11:06 outputs/outReport.html -&gt; 
</span><span id="__span-80-2"><a id="__codelineno-80-2" name="__codelineno-80-2" href="#__codelineno-80-2"></a>/home2/training/limma_voom/work/02/8cd12271346eb884083815bfe0cbfa/outReport.html
</span></code></pre></div></p>
<p>The actual file location is displayed after the <code>-&gt;</code>, as this denotes a symlink. </p>
<p>If using VSC, open this directory to see the various data and pdfs produced by <code>LIMMA_VOOM</code>.<br>
If using CLI, use the <code>cd</code> command instead, then <code>ls -lah</code> to print the contents of the directory. </p>
<p>You should see the following files &amp; directories: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-81-1"><a id="__codelineno-81-1" name="__codelineno-81-1" href="#__codelineno-81-1"></a># directories
</span><span id="__span-81-2"><a id="__codelineno-81-2" name="__codelineno-81-2" href="#__codelineno-81-2"></a>- glimma_MDS
</span><span id="__span-81-3"><a id="__codelineno-81-3" name="__codelineno-81-3" href="#__codelineno-81-3"></a>- glimma_Mut-WT
</span><span id="__span-81-4"><a id="__codelineno-81-4" name="__codelineno-81-4" href="#__codelineno-81-4"></a>- glimma_volcano_Mut-WT
</span><span id="__span-81-5"><a id="__codelineno-81-5" name="__codelineno-81-5" href="#__codelineno-81-5"></a>- glimma_WT-Mut
</span><span id="__span-81-6"><a id="__codelineno-81-6" name="__codelineno-81-6" href="#__codelineno-81-6"></a>- glimma_volcano_WT-Mut
</span><span id="__span-81-7"><a id="__codelineno-81-7" name="__codelineno-81-7" href="#__codelineno-81-7"></a>- glimma_Mut-WT-WT-Mut
</span><span id="__span-81-8"><a id="__codelineno-81-8" name="__codelineno-81-8" href="#__codelineno-81-8"></a>- glimma_volcano_Mut-WT-WT-Mut
</span><span id="__span-81-9"><a id="__codelineno-81-9" name="__codelineno-81-9" href="#__codelineno-81-9"></a>
</span><span id="__span-81-10"><a id="__codelineno-81-10" name="__codelineno-81-10" href="#__codelineno-81-10"></a># files
</span><span id="__span-81-11"><a id="__codelineno-81-11" name="__codelineno-81-11" href="#__codelineno-81-11"></a>- mdsscree.png
</span><span id="__span-81-12"><a id="__codelineno-81-12" name="__codelineno-81-12" href="#__codelineno-81-12"></a>- mdsscree.pdf
</span><span id="__span-81-13"><a id="__codelineno-81-13" name="__codelineno-81-13" href="#__codelineno-81-13"></a>- saplot.png
</span><span id="__span-81-14"><a id="__codelineno-81-14" name="__codelineno-81-14" href="#__codelineno-81-14"></a>- saplot.pdf
</span><span id="__span-81-15"><a id="__codelineno-81-15" name="__codelineno-81-15" href="#__codelineno-81-15"></a>- limma-trend_Mut-WT.tsv
</span><span id="__span-81-16"><a id="__codelineno-81-16" name="__codelineno-81-16" href="#__codelineno-81-16"></a>- mdplot_Mut-WT.pdf
</span><span id="__span-81-17"><a id="__codelineno-81-17" name="__codelineno-81-17" href="#__codelineno-81-17"></a>- volplot_Mut-WT.pdf
</span><span id="__span-81-18"><a id="__codelineno-81-18" name="__codelineno-81-18" href="#__codelineno-81-18"></a>- mdvolplot_Mut-WT.png
</span><span id="__span-81-19"><a id="__codelineno-81-19" name="__codelineno-81-19" href="#__codelineno-81-19"></a>- mdplot_WT-Mut.pdf
</span><span id="__span-81-20"><a id="__codelineno-81-20" name="__codelineno-81-20" href="#__codelineno-81-20"></a>- limma-trend_WT-Mut.tsv
</span><span id="__span-81-21"><a id="__codelineno-81-21" name="__codelineno-81-21" href="#__codelineno-81-21"></a>- volplot_WT-Mut.pdf
</span><span id="__span-81-22"><a id="__codelineno-81-22" name="__codelineno-81-22" href="#__codelineno-81-22"></a>- mdvolplot_WT-Mut.png
</span><span id="__span-81-23"><a id="__codelineno-81-23" name="__codelineno-81-23" href="#__codelineno-81-23"></a>- limma-trend_Mut-WT-WT-Mut.tsv
</span><span id="__span-81-24"><a id="__codelineno-81-24" name="__codelineno-81-24" href="#__codelineno-81-24"></a>- mdplot_Mut-WT-WT-Mut.pdf
</span><span id="__span-81-25"><a id="__codelineno-81-25" name="__codelineno-81-25" href="#__codelineno-81-25"></a>- volplot_Mut-WT-WT-Mut.pdf
</span><span id="__span-81-26"><a id="__codelineno-81-26" name="__codelineno-81-26" href="#__codelineno-81-26"></a>- mdvolplot_Mut-WT-WT-Mut.png
</span><span id="__span-81-27"><a id="__codelineno-81-27" name="__codelineno-81-27" href="#__codelineno-81-27"></a>- session_info.txt
</span><span id="__span-81-28"><a id="__codelineno-81-28" name="__codelineno-81-28" href="#__codelineno-81-28"></a>- outReport.html
</span></code></pre></div></p>
<p>If you&rsquo;re using VSC, you can download this folder. </p>
<p>Right-click the folder, then select &ldquo;Download&hellip;&rdquo;. <br>
Save it wherever you like on your local PC.</p>
<p>You can then open <code>outReport.html</code> inside the downloaded folder to view the <code>limma_voom</code> DE analysis results. </p>
<p>Here is an example of how <code>outReport.html</code> should look: </p>
<p><img alt="alt text" src="../media/outReport1.PNG" />
<img alt="alt text" src="../media/outReport2.PNG" />
<img alt="alt text" src="../media/outReport3.PNG" />
<img alt="alt text" src="../media/outReport4.PNG" /></p>
<p><br></p>
<h3 id="conclusion_4">Conclusion<a class="headerlink" href="#conclusion_4" title="Permanent link">&para;</a></h3>
<p>In this section we explored how to translate a challenging Galaxy Tool Wrapper to a Nextflow process. </p>
<p>In the final section - section 2.3 - we will translate a challenging Galaxy Workflow into Nextflow. </p>
<p><br></p>
<hr />
<h2 id="section-23-rna-seq-reads-to-counts-workflow-galaxy">Section 2.3: RNA-Seq Reads to Counts Workflow (Galaxy)<a class="headerlink" href="#section-23-rna-seq-reads-to-counts-workflow-galaxy" title="Permanent link">&para;</a></h2>
<h3 id="introduction_5">Introduction<a class="headerlink" href="#introduction_5" title="Permanent link">&para;</a></h3>
<p>The final section demonstrates translation of a challenging Galaxy workflow to Nextflow using <code>janis translate</code>. </p>
<p>The workflow we will translate in this section accepts raw RNA-seq reads as input, and produces gene counts for further analysis (eg. differential expression).</p>
<p>Many of the Galaxy Tool Wrappers used in this workflow are challenging. They may:</p>
<ul>
<li>Contain complex execution logic</li>
<li>Perform multiple commands</li>
<li>Use inbuilt Galaxy data for reference genomes, indexes, and annotation information.</li>
</ul>
<p>In these situations, the translated files produced by <code>janis translate</code> will need manual invertention to run. </p>
<p><br></p>
<p>The workflow used in this section is taken from the <a href="https://training.galaxyproject.org/training-material/">Galaxy Training Network (GTN)</a> resource which provides tutorials on how to use Galaxy for bioinformatic analysis. </p>
<p>The GTN has over 100 tutorials demonstrating how to use the Galaxy platform to analyse data, and is definitely worth having a look! </p>
<p>The specific workflow we use today is from the <a href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-reads-to-counts/tutorial.html">RNA-Seq reads to counts</a> page, which provides detailed instruction on how to turn raw RNA-seq reads into gene counts.</p>
<p><br></p>
<h3 id="janis-translate_5">Janis Translate<a class="headerlink" href="#janis-translate_5" title="Permanent link">&para;</a></h3>
<p><strong>Obtain the Workflow</strong></p>
<p>Galaxy workflows can be found in a number of places. </p>
<p>Today, we will use the workflow provided from the GTN RNA-Seq reads to counts tutorial. <br>
The workflow can be downloaded using <a href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-reads-to-counts/workflows/rna-seq-reads-to-counts.ga">this link</a>.</p>
<p>The downloaded <code>.ga</code> workflow is also included in the local <code>data/source/galaxy/</code> as <code>rna_seq_reads_to_counts.ga</code>. </p>
<p><br></p>
<p><em>(for your information)</em></p>
<p>Galaxy servers each have <a href="https://usegalaxy.org.au/workflows/list_published">shared workflows</a> where you can search for a community workflow which suits your needs. The link above is to the Galaxy Australia server shared workflows page, which contains hundreds of workflows which users have made public. </p>
<p><img alt="alt-text" src="../media/galaxy_shared_workflows_download.PNG" /></p>
<p>The <a href="https://training.galaxyproject.org/training-material/">Galaxy Training Network (GTN)</a> also provides workflows for their training tutorials. These are useful because the topics covered are common analyses users wish to perform. </p>
<p><img alt="alt-text" src="../media/rnaseq_reads_to_counts_download.PNG" /></p>
<p>Additionally, you can create your own workflow using the <a href="https://usegalaxy.org.au/workflows/list">Galaxy Workflow Editor</a> then download your creation. </p>
<p><img alt="alt-text" src="../media/galaxy_workflow_editor_download.PNG" /></p>
<p><br></p>
<p><strong>Run Janis Translate</strong></p>
<p>To translate a workflow, we follow the same pattern as in the previous sections. </p>
<p>For this section we have included the <code>.ga</code> file in the <code>./source</code> folder so you won&rsquo;t need to download it yourself. </p>
<p>To start, let&rsquo;s make sure we&rsquo;re in the right directory:
<div class="language-text highlight"><pre><span></span><code><span id="__span-82-1"><a id="__codelineno-82-1" name="__codelineno-82-1" href="#__codelineno-82-1"></a>cd /home2/training
</span></code></pre></div></p>
<p>As usual, we will add <code>-o rnaseq_reads_to_counts</code> to our translate command to set the output directory. </p>
<p>To translate the Galaxy workflow, run the following command:
<div class="language-text highlight"><pre><span></span><code><span id="__span-83-1"><a id="__codelineno-83-1" name="__codelineno-83-1" href="#__codelineno-83-1"></a>singularity exec ~/janis.sif janis translate -o rnaseq_reads_to_counts --from galaxy --to nextflow data/source/galaxy/rna_seq_reads_to_counts.ga
</span></code></pre></div></p>
<p>This may take up to 5 mins as janis-translate will make many requests to the Galaxy Toolshed API and quay.io. </p>
<p>Janis-translate uses the Galaxy Toolshed API to look up and download wrappers specified in the workflow.
Janis-translate also makes requests to quay.io to find the most suitable container for each Galaxy Tool Wrapper using the requirements listed in the wrapper. </p>
<p><br></p>
<p><strong>Translation Output</strong></p>
<p>The output translation will contain multiple files and directories.<br>
You will see a folder called <code>rnaseq_reads_to_counts/</code> appear - inside this folder, we should see the following structure:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-84-1"><a id="__codelineno-84-1" name="__codelineno-84-1" href="#__codelineno-84-1"></a>rnaseq_reads_to_counts
</span><span id="__span-84-2"><a id="__codelineno-84-2" name="__codelineno-84-2" href="#__codelineno-84-2"></a> main.nf                             # main workflow (rna_seq_reads_to_counts)
</span><span id="__span-84-3"><a id="__codelineno-84-3" name="__codelineno-84-3" href="#__codelineno-84-3"></a> modules                             # folder containing nextflow processes
</span><span id="__span-84-4"><a id="__codelineno-84-4" name="__codelineno-84-4" href="#__codelineno-84-4"></a>    collection_column_join.nf
</span><span id="__span-84-5"><a id="__codelineno-84-5" name="__codelineno-84-5" href="#__codelineno-84-5"></a>    cutadapt.nf
</span><span id="__span-84-6"><a id="__codelineno-84-6" name="__codelineno-84-6" href="#__codelineno-84-6"></a>    fastqc.nf
</span><span id="__span-84-7"><a id="__codelineno-84-7" name="__codelineno-84-7" href="#__codelineno-84-7"></a>    featurecounts.nf
</span><span id="__span-84-8"><a id="__codelineno-84-8" name="__codelineno-84-8" href="#__codelineno-84-8"></a>    hisat2.nf
</span><span id="__span-84-9"><a id="__codelineno-84-9" name="__codelineno-84-9" href="#__codelineno-84-9"></a>    multiqc.nf
</span><span id="__span-84-10"><a id="__codelineno-84-10" name="__codelineno-84-10" href="#__codelineno-84-10"></a>    picard_mark_duplicates.nf
</span><span id="__span-84-11"><a id="__codelineno-84-11" name="__codelineno-84-11" href="#__codelineno-84-11"></a>    rseqc_gene_body_coverage.nf
</span><span id="__span-84-12"><a id="__codelineno-84-12" name="__codelineno-84-12" href="#__codelineno-84-12"></a>    rseqc_infer_experiment.nf
</span><span id="__span-84-13"><a id="__codelineno-84-13" name="__codelineno-84-13" href="#__codelineno-84-13"></a>    rseqc_read_distribution.nf
</span><span id="__span-84-14"><a id="__codelineno-84-14" name="__codelineno-84-14" href="#__codelineno-84-14"></a>    samtools_idxstats.nf
</span><span id="__span-84-15"><a id="__codelineno-84-15" name="__codelineno-84-15" href="#__codelineno-84-15"></a> nextflow.config                     # config file to supply input information
</span><span id="__span-84-16"><a id="__codelineno-84-16" name="__codelineno-84-16" href="#__codelineno-84-16"></a> source                              # folder containing galaxy tool wrappers used to translate tools
</span><span id="__span-84-17"><a id="__codelineno-84-17" name="__codelineno-84-17" href="#__codelineno-84-17"></a> subworkflows                        # folder containing nextflow subworkflows
</span><span id="__span-84-18"><a id="__codelineno-84-18" name="__codelineno-84-18" href="#__codelineno-84-18"></a> templates                           # folder containing any scripts used by processes
</span><span id="__span-84-19"><a id="__codelineno-84-19" name="__codelineno-84-19" href="#__codelineno-84-19"></a>     collection_column_join_script
</span><span id="__span-84-20"><a id="__codelineno-84-20" name="__codelineno-84-20" href="#__codelineno-84-20"></a>     multiqc_config
</span></code></pre></div>
<p>Now that we have performed translation using <code>janis translate</code>, we need to check the translated workflow for correctness.  </p>
<p><br></p>
<h3 id="running-the-translated-workflow_1">Running the Translated Workflow<a class="headerlink" href="#running-the-translated-workflow_1" title="Permanent link">&para;</a></h3>
<p><strong>Swapping containers</strong></p>
<p>As mentioned in section 2.2, some Galaxy Tool Wrappers have multiple requirements. </p>
<p>As we are not using <code>janis translate --build-galaxy-tool-images ...</code>, some of the 
translated nextflow processes will not run. </p>
<p>Without <code>--build-galaxy-tool-images</code>, janis-translate selects the best single container from the software requirements.
This container sometimes does not have all the required software needed to run the process script. </p>
<p>We have prepared pre-built containers which contain all software requirements for the 2 affected tools. </p>
<div class="admonition question">
<p class="admonition-title">Swapping HISAT2 container directive</p>
<p>Change the HISAT2 process container directive from: <br>
<code>quay.io/biocontainers/hisat2:2.2.1--h87f3376_5</code></p>
<p>To: <br>
<code>quay.io/grace_hall1/hisat2:2.2.1</code></p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/hisat2.nf
<div class="language-text highlight"><pre><span></span><code><span id="__span-85-1"><a id="__codelineno-85-1" name="__codelineno-85-1" href="#__codelineno-85-1"></a>process HISAT2 {
</span><span id="__span-85-2"><a id="__codelineno-85-2" name="__codelineno-85-2" href="#__codelineno-85-2"></a>
</span><span id="__span-85-3"><a id="__codelineno-85-3" name="__codelineno-85-3" href="#__codelineno-85-3"></a>    container &quot;quay.io/grace_hall1/hisat2:2.2.1&quot;            &lt;-
</span><span id="__span-85-4"><a id="__codelineno-85-4" name="__codelineno-85-4" href="#__codelineno-85-4"></a>    publishDir &quot;${params.outdir}/hisat2&quot;
</span><span id="__span-85-5"><a id="__codelineno-85-5" name="__codelineno-85-5" href="#__codelineno-85-5"></a>
</span><span id="__span-85-6"><a id="__codelineno-85-6" name="__codelineno-85-6" href="#__codelineno-85-6"></a>    ...
</span><span id="__span-85-7"><a id="__codelineno-85-7" name="__codelineno-85-7" href="#__codelineno-85-7"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Swapping FEATURECOUNTS container directive</p>
<p>Change the FEATURECOUNTS process container directive from: <br>
<code>quay.io/biocontainers/coreutils:8.31--h14c3975_0</code></p>
<p>To: <br>
<code>quay.io/grace_hall1/featurecounts:2.0.1</code></p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-86-1"><a id="__codelineno-86-1" name="__codelineno-86-1" href="#__codelineno-86-1"></a>process FEATURECOUNTS {
</span><span id="__span-86-2"><a id="__codelineno-86-2" name="__codelineno-86-2" href="#__codelineno-86-2"></a>
</span><span id="__span-86-3"><a id="__codelineno-86-3" name="__codelineno-86-3" href="#__codelineno-86-3"></a>    container &quot;quay.io/grace_hall1/featurecounts:2.0.1&quot;     &lt;-
</span><span id="__span-86-4"><a id="__codelineno-86-4" name="__codelineno-86-4" href="#__codelineno-86-4"></a>    publishDir &quot;${params.outdir}/featurecounts&quot;
</span><span id="__span-86-5"><a id="__codelineno-86-5" name="__codelineno-86-5" href="#__codelineno-86-5"></a>
</span><span id="__span-86-6"><a id="__codelineno-86-6" name="__codelineno-86-6" href="#__codelineno-86-6"></a>    ...
</span><span id="__span-86-7"><a id="__codelineno-86-7" name="__codelineno-86-7" href="#__codelineno-86-7"></a>}
</span></code></pre></div>
</details>
</div>
<p><br></p>
<p><strong>Inspect main.nf</strong></p>
<p>The main workflow translation appears as <code>main.nf</code> in the <code>rnaseq_reads_to_counts/</code> folder. <br></p>
<p><br></p>
<blockquote>
<p>NOTE: <br>
Before continuing, feel free to have a look at the other nextflow files which have been generated during translation.<br>
Each Galaxy tool appears as a nextflow process in the <code>rnaseq_reads_to_counts/modules/</code> directory. <br>
Each script used by nextflow processes appear in the <code>rnaseq_reads_to_counts/templates/</code> directory.<br>
Each Galaxy Tool Wrapper used during translation appears in the <code>rnaseq_reads_to_counts/source/</code> directory. </p>
</blockquote>
<p><br></p>
<p>In <code>main.nf</code> we see imports for nextflow processes called by the main workflow:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-87-1"><a id="__codelineno-87-1" name="__codelineno-87-1" href="#__codelineno-87-1"></a>include { FASTQC as FASTQC1 } from &#39;./modules/fastqc&#39;
</span><span id="__span-87-2"><a id="__codelineno-87-2" name="__codelineno-87-2" href="#__codelineno-87-2"></a>include { CUTADAPT } from &#39;./modules/cutadapt&#39;
</span><span id="__span-87-3"><a id="__codelineno-87-3" name="__codelineno-87-3" href="#__codelineno-87-3"></a>include { FASTQC as FASTQC2 } from &#39;./modules/fastqc&#39;
</span><span id="__span-87-4"><a id="__codelineno-87-4" name="__codelineno-87-4" href="#__codelineno-87-4"></a>include { HISAT2 } from &#39;./modules/hisat2&#39;
</span><span id="__span-87-5"><a id="__codelineno-87-5" name="__codelineno-87-5" href="#__codelineno-87-5"></a>include { FEATURECOUNTS } from &#39;./modules/featurecounts&#39;
</span><span id="__span-87-6"><a id="__codelineno-87-6" name="__codelineno-87-6" href="#__codelineno-87-6"></a>include { PICARD_MARK_DUPLICATES } from &#39;./modules/picard_mark_duplicates&#39;
</span><span id="__span-87-7"><a id="__codelineno-87-7" name="__codelineno-87-7" href="#__codelineno-87-7"></a>include { SAMTOOLS_IDXSTATS } from &#39;./modules/samtools_idxstats&#39;
</span><span id="__span-87-8"><a id="__codelineno-87-8" name="__codelineno-87-8" href="#__codelineno-87-8"></a>include { RSEQC_GENE_BODY_COVERAGE } from &#39;./modules/rseqc_gene_body_coverage&#39;
</span><span id="__span-87-9"><a id="__codelineno-87-9" name="__codelineno-87-9" href="#__codelineno-87-9"></a>include { RSEQC_INFER_EXPERIMENT } from &#39;./modules/rseqc_infer_experiment&#39;
</span><span id="__span-87-10"><a id="__codelineno-87-10" name="__codelineno-87-10" href="#__codelineno-87-10"></a>include { RSEQC_READ_DISTRIBUTION } from &#39;./modules/rseqc_read_distribution&#39;
</span><span id="__span-87-11"><a id="__codelineno-87-11" name="__codelineno-87-11" href="#__codelineno-87-11"></a>include { COLLECTION_COLUMN_JOIN } from &#39;./modules/collection_column_join&#39;
</span><span id="__span-87-12"><a id="__codelineno-87-12" name="__codelineno-87-12" href="#__codelineno-87-12"></a>include { MULTIQC } from &#39;./modules/multiqc&#39;
</span></code></pre></div>
<p>We also see that some nextflow <code>Channels</code> and other variables have been set up. </p>
<p>Channels are used to supply data according to nextflow&rsquo;s adoption of the <em>dataflow</em> programming model.
The other variable definitions ensure that file input parameters are being treated as <em>files</em> by nextflow, so that they are staged properly in the process working directory. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-88-1"><a id="__codelineno-88-1" name="__codelineno-88-1" href="#__codelineno-88-1"></a>// data which will be passed as channels
</span><span id="__span-88-2"><a id="__codelineno-88-2" name="__codelineno-88-2" href="#__codelineno-88-2"></a>ch_in_input_fastqs_collection  = Channel.fromPath( params.in_input_fastqs_collection ).toList()
</span><span id="__span-88-3"><a id="__codelineno-88-3" name="__codelineno-88-3" href="#__codelineno-88-3"></a>
</span><span id="__span-88-4"><a id="__codelineno-88-4" name="__codelineno-88-4" href="#__codelineno-88-4"></a>// data which will be passed as variables
</span><span id="__span-88-5"><a id="__codelineno-88-5" name="__codelineno-88-5" href="#__codelineno-88-5"></a>collection_column_join_script  = file( params.collection_column_join_script )
</span><span id="__span-88-6"><a id="__codelineno-88-6" name="__codelineno-88-6" href="#__codelineno-88-6"></a>fastqc1_contaminants           = file( params.fastqc1_contaminants )
</span><span id="__span-88-7"><a id="__codelineno-88-7" name="__codelineno-88-7" href="#__codelineno-88-7"></a>fastqc1_limits                 = file( params.fastqc1_limits )
</span><span id="__span-88-8"><a id="__codelineno-88-8" name="__codelineno-88-8" href="#__codelineno-88-8"></a>fastqc2_contaminants           = file( params.fastqc2_contaminants )
</span><span id="__span-88-9"><a id="__codelineno-88-9" name="__codelineno-88-9" href="#__codelineno-88-9"></a>fastqc2_limits                 = file( params.fastqc2_limits )
</span><span id="__span-88-10"><a id="__codelineno-88-10" name="__codelineno-88-10" href="#__codelineno-88-10"></a>in_input_reference_gene_bed    = file( params.in_input_reference_gene_bed )
</span><span id="__span-88-11"><a id="__codelineno-88-11" name="__codelineno-88-11" href="#__codelineno-88-11"></a>multiqc_config                 = file( params.multiqc_config )
</span></code></pre></div></p>
<p><br></p>
<p>Now that we have looked at the imports and dataflow, we can look at the main <code>workflow</code> section. </p>
<p>The main workflow is captured in the <code>workflow {}</code> section and has 12 tasks. 
Each task has been supplied values according to the source workflow. 
Comments display the name of the process input which is being fed a particular value. </p>
<p>To get a sense of how the processes relate to each other in the workflow execution, compare the main <code>workflow {}</code> section with the following visualisation of our Galaxy workflow (using the Galaxy workflow editor).</p>
<p>main.nf
<div class="language-text highlight"><pre><span></span><code><span id="__span-89-1"><a id="__codelineno-89-1" name="__codelineno-89-1" href="#__codelineno-89-1"></a>workflow {
</span><span id="__span-89-2"><a id="__codelineno-89-2" name="__codelineno-89-2" href="#__codelineno-89-2"></a>
</span><span id="__span-89-3"><a id="__codelineno-89-3" name="__codelineno-89-3" href="#__codelineno-89-3"></a>    FASTQC1(
</span><span id="__span-89-4"><a id="__codelineno-89-4" name="__codelineno-89-4" href="#__codelineno-89-4"></a>        ch_in_input_fastqs_collection.flatten(),  // input_file
</span><span id="__span-89-5"><a id="__codelineno-89-5" name="__codelineno-89-5" href="#__codelineno-89-5"></a>        fastqc1_contaminants,                     // contaminants
</span><span id="__span-89-6"><a id="__codelineno-89-6" name="__codelineno-89-6" href="#__codelineno-89-6"></a>        fastqc1_limits,                           // limits
</span><span id="__span-89-7"><a id="__codelineno-89-7" name="__codelineno-89-7" href="#__codelineno-89-7"></a>        params.fastqc1_format_string,             // format_string
</span><span id="__span-89-8"><a id="__codelineno-89-8" name="__codelineno-89-8" href="#__codelineno-89-8"></a>        params.fastqc1_outdir                     // outdir
</span><span id="__span-89-9"><a id="__codelineno-89-9" name="__codelineno-89-9" href="#__codelineno-89-9"></a>    )
</span><span id="__span-89-10"><a id="__codelineno-89-10" name="__codelineno-89-10" href="#__codelineno-89-10"></a>
</span><span id="__span-89-11"><a id="__codelineno-89-11" name="__codelineno-89-11" href="#__codelineno-89-11"></a>    CUTADAPT(
</span><span id="__span-89-12"><a id="__codelineno-89-12" name="__codelineno-89-12" href="#__codelineno-89-12"></a>        ch_in_input_fastqs_collection.flatten()  // library_input_1
</span><span id="__span-89-13"><a id="__codelineno-89-13" name="__codelineno-89-13" href="#__codelineno-89-13"></a>    )
</span><span id="__span-89-14"><a id="__codelineno-89-14" name="__codelineno-89-14" href="#__codelineno-89-14"></a>
</span><span id="__span-89-15"><a id="__codelineno-89-15" name="__codelineno-89-15" href="#__codelineno-89-15"></a>    FASTQC2(
</span><span id="__span-89-16"><a id="__codelineno-89-16" name="__codelineno-89-16" href="#__codelineno-89-16"></a>        CUTADAPT.out.out12,            // input_file
</span><span id="__span-89-17"><a id="__codelineno-89-17" name="__codelineno-89-17" href="#__codelineno-89-17"></a>        fastqc2_contaminants,          // contaminants
</span><span id="__span-89-18"><a id="__codelineno-89-18" name="__codelineno-89-18" href="#__codelineno-89-18"></a>        fastqc2_limits,                // limits
</span><span id="__span-89-19"><a id="__codelineno-89-19" name="__codelineno-89-19" href="#__codelineno-89-19"></a>        params.fastqc2_format_string,  // format_string
</span><span id="__span-89-20"><a id="__codelineno-89-20" name="__codelineno-89-20" href="#__codelineno-89-20"></a>        params.fastqc2_outdir          // outdir
</span><span id="__span-89-21"><a id="__codelineno-89-21" name="__codelineno-89-21" href="#__codelineno-89-21"></a>    )
</span><span id="__span-89-22"><a id="__codelineno-89-22" name="__codelineno-89-22" href="#__codelineno-89-22"></a>
</span><span id="__span-89-23"><a id="__codelineno-89-23" name="__codelineno-89-23" href="#__codelineno-89-23"></a>    HISAT2(
</span><span id="__span-89-24"><a id="__codelineno-89-24" name="__codelineno-89-24" href="#__codelineno-89-24"></a>        CUTADAPT.out.out12,        // library_input_1
</span><span id="__span-89-25"><a id="__codelineno-89-25" name="__codelineno-89-25" href="#__codelineno-89-25"></a>        params.hisat2_index_path,  // index_path
</span><span id="__span-89-26"><a id="__codelineno-89-26" name="__codelineno-89-26" href="#__codelineno-89-26"></a>        params.hisat2_read11       // read11
</span><span id="__span-89-27"><a id="__codelineno-89-27" name="__codelineno-89-27" href="#__codelineno-89-27"></a>    )
</span><span id="__span-89-28"><a id="__codelineno-89-28" name="__codelineno-89-28" href="#__codelineno-89-28"></a>
</span><span id="__span-89-29"><a id="__codelineno-89-29" name="__codelineno-89-29" href="#__codelineno-89-29"></a>    FEATURECOUNTS(
</span><span id="__span-89-30"><a id="__codelineno-89-30" name="__codelineno-89-30" href="#__codelineno-89-30"></a>        HISAT2.out.output_alignments  // alignment
</span><span id="__span-89-31"><a id="__codelineno-89-31" name="__codelineno-89-31" href="#__codelineno-89-31"></a>    )
</span><span id="__span-89-32"><a id="__codelineno-89-32" name="__codelineno-89-32" href="#__codelineno-89-32"></a>
</span><span id="__span-89-33"><a id="__codelineno-89-33" name="__codelineno-89-33" href="#__codelineno-89-33"></a>    PICARD_MARK_DUPLICATES(
</span><span id="__span-89-34"><a id="__codelineno-89-34" name="__codelineno-89-34" href="#__codelineno-89-34"></a>        HISAT2.out.output_alignments  // input_file
</span><span id="__span-89-35"><a id="__codelineno-89-35" name="__codelineno-89-35" href="#__codelineno-89-35"></a>    )
</span><span id="__span-89-36"><a id="__codelineno-89-36" name="__codelineno-89-36" href="#__codelineno-89-36"></a>
</span><span id="__span-89-37"><a id="__codelineno-89-37" name="__codelineno-89-37" href="#__codelineno-89-37"></a>    SAMTOOLS_IDXSTATS(
</span><span id="__span-89-38"><a id="__codelineno-89-38" name="__codelineno-89-38" href="#__codelineno-89-38"></a>        HISAT2.out.output_alignments,        // input_bam
</span><span id="__span-89-39"><a id="__codelineno-89-39" name="__codelineno-89-39" href="#__codelineno-89-39"></a>        params.samtools_idxstats_addthreads  // addthreads
</span><span id="__span-89-40"><a id="__codelineno-89-40" name="__codelineno-89-40" href="#__codelineno-89-40"></a>    )
</span><span id="__span-89-41"><a id="__codelineno-89-41" name="__codelineno-89-41" href="#__codelineno-89-41"></a>
</span><span id="__span-89-42"><a id="__codelineno-89-42" name="__codelineno-89-42" href="#__codelineno-89-42"></a>    RSEQC_GENE_BODY_COVERAGE(
</span><span id="__span-89-43"><a id="__codelineno-89-43" name="__codelineno-89-43" href="#__codelineno-89-43"></a>        HISAT2.out.output_alignments,             // batch_mode_input
</span><span id="__span-89-44"><a id="__codelineno-89-44" name="__codelineno-89-44" href="#__codelineno-89-44"></a>        in_input_reference_gene_bed,              // refgene
</span><span id="__span-89-45"><a id="__codelineno-89-45" name="__codelineno-89-45" href="#__codelineno-89-45"></a>        params.rseqc_gene_body_coverage_safename  // safename
</span><span id="__span-89-46"><a id="__codelineno-89-46" name="__codelineno-89-46" href="#__codelineno-89-46"></a>    )
</span><span id="__span-89-47"><a id="__codelineno-89-47" name="__codelineno-89-47" href="#__codelineno-89-47"></a>
</span><span id="__span-89-48"><a id="__codelineno-89-48" name="__codelineno-89-48" href="#__codelineno-89-48"></a>    RSEQC_INFER_EXPERIMENT(
</span><span id="__span-89-49"><a id="__codelineno-89-49" name="__codelineno-89-49" href="#__codelineno-89-49"></a>        HISAT2.out.output_alignments,  // input_bam
</span><span id="__span-89-50"><a id="__codelineno-89-50" name="__codelineno-89-50" href="#__codelineno-89-50"></a>        in_input_reference_gene_bed    // refgene
</span><span id="__span-89-51"><a id="__codelineno-89-51" name="__codelineno-89-51" href="#__codelineno-89-51"></a>    )
</span><span id="__span-89-52"><a id="__codelineno-89-52" name="__codelineno-89-52" href="#__codelineno-89-52"></a>
</span><span id="__span-89-53"><a id="__codelineno-89-53" name="__codelineno-89-53" href="#__codelineno-89-53"></a>    RSEQC_READ_DISTRIBUTION(
</span><span id="__span-89-54"><a id="__codelineno-89-54" name="__codelineno-89-54" href="#__codelineno-89-54"></a>        HISAT2.out.output_alignments,  // input_bam
</span><span id="__span-89-55"><a id="__codelineno-89-55" name="__codelineno-89-55" href="#__codelineno-89-55"></a>        in_input_reference_gene_bed    // refgene
</span><span id="__span-89-56"><a id="__codelineno-89-56" name="__codelineno-89-56" href="#__codelineno-89-56"></a>    )
</span><span id="__span-89-57"><a id="__codelineno-89-57" name="__codelineno-89-57" href="#__codelineno-89-57"></a>
</span><span id="__span-89-58"><a id="__codelineno-89-58" name="__codelineno-89-58" href="#__codelineno-89-58"></a>    COLLECTION_COLUMN_JOIN(
</span><span id="__span-89-59"><a id="__codelineno-89-59" name="__codelineno-89-59" href="#__codelineno-89-59"></a>        FEATURECOUNTS.out.output_short.toList(),  // input_tabular
</span><span id="__span-89-60"><a id="__codelineno-89-60" name="__codelineno-89-60" href="#__codelineno-89-60"></a>        collection_column_join_script             // collection_column_join_script
</span><span id="__span-89-61"><a id="__codelineno-89-61" name="__codelineno-89-61" href="#__codelineno-89-61"></a>    )
</span><span id="__span-89-62"><a id="__codelineno-89-62" name="__codelineno-89-62" href="#__codelineno-89-62"></a>
</span><span id="__span-89-63"><a id="__codelineno-89-63" name="__codelineno-89-63" href="#__codelineno-89-63"></a>    MULTIQC(
</span><span id="__span-89-64"><a id="__codelineno-89-64" name="__codelineno-89-64" href="#__codelineno-89-64"></a>        multiqc_config,                               // multiqc_config
</span><span id="__span-89-65"><a id="__codelineno-89-65" name="__codelineno-89-65" href="#__codelineno-89-65"></a>        FASTQC2.out.out_text_file,                    // unknown1
</span><span id="__span-89-66"><a id="__codelineno-89-66" name="__codelineno-89-66" href="#__codelineno-89-66"></a>        CUTADAPT.out.out_report,                      // unknown2
</span><span id="__span-89-67"><a id="__codelineno-89-67" name="__codelineno-89-67" href="#__codelineno-89-67"></a>        RSEQC_INFER_EXPERIMENT.out.outputTextfile,    // unknown3
</span><span id="__span-89-68"><a id="__codelineno-89-68" name="__codelineno-89-68" href="#__codelineno-89-68"></a>        PICARD_MARK_DUPLICATES.out.out_metrics_file,  // unknown4
</span><span id="__span-89-69"><a id="__codelineno-89-69" name="__codelineno-89-69" href="#__codelineno-89-69"></a>        SAMTOOLS_IDXSTATS.out.outputTabular,          // unknown5
</span><span id="__span-89-70"><a id="__codelineno-89-70" name="__codelineno-89-70" href="#__codelineno-89-70"></a>        RSEQC_GENE_BODY_COVERAGE.out.outputtxt,       // unknown6
</span><span id="__span-89-71"><a id="__codelineno-89-71" name="__codelineno-89-71" href="#__codelineno-89-71"></a>        RSEQC_READ_DISTRIBUTION.out.outputTextfile,   // unknown7
</span><span id="__span-89-72"><a id="__codelineno-89-72" name="__codelineno-89-72" href="#__codelineno-89-72"></a>        FEATURECOUNTS.out.output_summary,             // unknown8
</span><span id="__span-89-73"><a id="__codelineno-89-73" name="__codelineno-89-73" href="#__codelineno-89-73"></a>        HISAT2.out.out_summary_file                   // unknown9
</span><span id="__span-89-74"><a id="__codelineno-89-74" name="__codelineno-89-74" href="#__codelineno-89-74"></a>    )
</span><span id="__span-89-75"><a id="__codelineno-89-75" name="__codelineno-89-75" href="#__codelineno-89-75"></a>
</span><span id="__span-89-76"><a id="__codelineno-89-76" name="__codelineno-89-76" href="#__codelineno-89-76"></a>}
</span></code></pre></div></p>
<p>Galaxy Workflow</p>
<p><img alt="alt-text" src="../media/rnaseq_reads_to_counts_wf.PNG" /></p>
<p><br></p>
<p>One of the tasks in this workflow is not needed for our analysis, as relates to the way Galaxy handles collections of files. </p>
<p>For this workshop, we will ignore the <code>COLLECTION_COLUMN_JOIN</code> task.</p>
<div class="admonition question">
<p class="admonition-title">Removing the <code>COLLECTION_COLUMN_JOIN</code> process call</p>
<p>Cut or comment out the <code>COLLECTION_COLUMN_JOIN</code> process call in <code>main.nf</code>. </p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-90-1"><a id="__codelineno-90-1" name="__codelineno-90-1" href="#__codelineno-90-1"></a>workflow {
</span><span id="__span-90-2"><a id="__codelineno-90-2" name="__codelineno-90-2" href="#__codelineno-90-2"></a>    ...
</span><span id="__span-90-3"><a id="__codelineno-90-3" name="__codelineno-90-3" href="#__codelineno-90-3"></a>
</span><span id="__span-90-4"><a id="__codelineno-90-4" name="__codelineno-90-4" href="#__codelineno-90-4"></a>    // commented out!
</span><span id="__span-90-5"><a id="__codelineno-90-5" name="__codelineno-90-5" href="#__codelineno-90-5"></a>    // COLLECTION_COLUMN_JOIN(
</span><span id="__span-90-6"><a id="__codelineno-90-6" name="__codelineno-90-6" href="#__codelineno-90-6"></a>    //     FEATURECOUNTS.out.output_short.toList(),  // input_tabular
</span><span id="__span-90-7"><a id="__codelineno-90-7" name="__codelineno-90-7" href="#__codelineno-90-7"></a>    //     collection_column_join_script             // collection_column_join_script
</span><span id="__span-90-8"><a id="__codelineno-90-8" name="__codelineno-90-8" href="#__codelineno-90-8"></a>    // )
</span><span id="__span-90-9"><a id="__codelineno-90-9" name="__codelineno-90-9" href="#__codelineno-90-9"></a>
</span><span id="__span-90-10"><a id="__codelineno-90-10" name="__codelineno-90-10" href="#__codelineno-90-10"></a>    ...
</span><span id="__span-90-11"><a id="__codelineno-90-11" name="__codelineno-90-11" href="#__codelineno-90-11"></a>}
</span></code></pre></div>
</details>
</div>
<p>Now we have removed the collection join, we will need to pass all results produced for each sample to <code>MULTIQC</code>. </p>
<div class="admonition question">
<p class="admonition-title">Passing all results to <code>MULTIQC</code></p>
<p>For each input to the <code>MULTIQC</code> process call in <code>main.nf</code> (except <code>multiqc_config</code>), add the <code>.toList()</code> operator.</p>
<p>This will collect each result produced by previous steps (ie one for each of the 12 samples).</p>
<details class="hint">
<summary>Show Change</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-91-1"><a id="__codelineno-91-1" name="__codelineno-91-1" href="#__codelineno-91-1"></a>workflow {
</span><span id="__span-91-2"><a id="__codelineno-91-2" name="__codelineno-91-2" href="#__codelineno-91-2"></a>    ...
</span><span id="__span-91-3"><a id="__codelineno-91-3" name="__codelineno-91-3" href="#__codelineno-91-3"></a>
</span><span id="__span-91-4"><a id="__codelineno-91-4" name="__codelineno-91-4" href="#__codelineno-91-4"></a>    MULTIQC(
</span><span id="__span-91-5"><a id="__codelineno-91-5" name="__codelineno-91-5" href="#__codelineno-91-5"></a>           multiqc_config,                                        // multiqc_config
</span><span id="__span-91-6"><a id="__codelineno-91-6" name="__codelineno-91-6" href="#__codelineno-91-6"></a>           FASTQC2.out.out_text_file.toList(),                    // unknown1       &lt;-
</span><span id="__span-91-7"><a id="__codelineno-91-7" name="__codelineno-91-7" href="#__codelineno-91-7"></a>           CUTADAPT.out.out_report.toList(),                      // unknown2       &lt;-
</span><span id="__span-91-8"><a id="__codelineno-91-8" name="__codelineno-91-8" href="#__codelineno-91-8"></a>           RSEQC_INFER_EXPERIMENT.out.outputTextfile.toList(),    // unknown3       &lt;-
</span><span id="__span-91-9"><a id="__codelineno-91-9" name="__codelineno-91-9" href="#__codelineno-91-9"></a>           PICARD_MARK_DUPLICATES.out.out_metrics_file.toList(),  // unknown4       &lt;-
</span><span id="__span-91-10"><a id="__codelineno-91-10" name="__codelineno-91-10" href="#__codelineno-91-10"></a>           SAMTOOLS_IDXSTATS.out.outputTabular.toList(),          // unknown5       &lt;-
</span><span id="__span-91-11"><a id="__codelineno-91-11" name="__codelineno-91-11" href="#__codelineno-91-11"></a>           RSEQC_GENE_BODY_COVERAGE.out.outputtxt.toList(),       // unknown6       &lt;-
</span><span id="__span-91-12"><a id="__codelineno-91-12" name="__codelineno-91-12" href="#__codelineno-91-12"></a>           RSEQC_READ_DISTRIBUTION.out.outputTextfile.toList(),   // unknown7       &lt;-
</span><span id="__span-91-13"><a id="__codelineno-91-13" name="__codelineno-91-13" href="#__codelineno-91-13"></a>           FEATURECOUNTS.out.output_summary.toList(),             // unknown8       &lt;-
</span><span id="__span-91-14"><a id="__codelineno-91-14" name="__codelineno-91-14" href="#__codelineno-91-14"></a>           HISAT2.out.out_summary_file.toList()                   // unknown9       &lt;-
</span><span id="__span-91-15"><a id="__codelineno-91-15" name="__codelineno-91-15" href="#__codelineno-91-15"></a>    )
</span><span id="__span-91-16"><a id="__codelineno-91-16" name="__codelineno-91-16" href="#__codelineno-91-16"></a>    ```
</span><span id="__span-91-17"><a id="__codelineno-91-17" name="__codelineno-91-17" href="#__codelineno-91-17"></a>
</span><span id="__span-91-18"><a id="__codelineno-91-18" name="__codelineno-91-18" href="#__codelineno-91-18"></a>    ...
</span><span id="__span-91-19"><a id="__codelineno-91-19" name="__codelineno-91-19" href="#__codelineno-91-19"></a>}
</span></code></pre></div>
</details>
</div>
<p>Now that we have covered <code>main.nf</code>, let&rsquo;s set up inputs to run our translated Nextflow workflow. 
To do this we will supply sample data using the global <code>params</code> variable in <code>nextflow.config</code>.</p>
<p><br></p>
<p><strong>Inspect nextflow.config</strong></p>
<p>To test the translated workflow, we will first set up workflow inputs in <code>nextflow.config</code>. </p>
<p>Inside the <code>rnaseq_reads_to_counts/</code> folder you will see that <code>nextflow.config</code> is already provided. </p>
<p>Janis translate creates this file to provide clarity about the necessary workflow inputs, and to set some other config variables. </p>
<p>Open <code>nextflow.config</code> and have a look at the contents. It should look similar to the following: </p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-92-1"><a id="__codelineno-92-1" name="__codelineno-92-1" href="#__codelineno-92-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-92-2"><a id="__codelineno-92-2" name="__codelineno-92-2" href="#__codelineno-92-2"></a>singularity.enabled = true
</span><span id="__span-92-3"><a id="__codelineno-92-3" name="__codelineno-92-3" href="#__codelineno-92-3"></a>singularity.cacheDir = &quot;$HOME/.singularity/cache&quot;
</span><span id="__span-92-4"><a id="__codelineno-92-4" name="__codelineno-92-4" href="#__codelineno-92-4"></a>
</span><span id="__span-92-5"><a id="__codelineno-92-5" name="__codelineno-92-5" href="#__codelineno-92-5"></a>params {
</span><span id="__span-92-6"><a id="__codelineno-92-6" name="__codelineno-92-6" href="#__codelineno-92-6"></a>
</span><span id="__span-92-7"><a id="__codelineno-92-7" name="__codelineno-92-7" href="#__codelineno-92-7"></a>    // Placeholder for null values.
</span><span id="__span-92-8"><a id="__codelineno-92-8" name="__codelineno-92-8" href="#__codelineno-92-8"></a>    // Do not alter unless you know what you are doing.
</span><span id="__span-92-9"><a id="__codelineno-92-9" name="__codelineno-92-9" href="#__codelineno-92-9"></a>    NULL_VALUE = &#39;NULL&#39;
</span><span id="__span-92-10"><a id="__codelineno-92-10" name="__codelineno-92-10" href="#__codelineno-92-10"></a>
</span><span id="__span-92-11"><a id="__codelineno-92-11" name="__codelineno-92-11" href="#__codelineno-92-11"></a>    // WORKFLOW OUTPUT DIRECTORY
</span><span id="__span-92-12"><a id="__codelineno-92-12" name="__codelineno-92-12" href="#__codelineno-92-12"></a>    outdir  = &#39;./outputs&#39;
</span><span id="__span-92-13"><a id="__codelineno-92-13" name="__codelineno-92-13" href="#__codelineno-92-13"></a>
</span><span id="__span-92-14"><a id="__codelineno-92-14" name="__codelineno-92-14" href="#__codelineno-92-14"></a>    // INPUTS (MANDATORY)
</span><span id="__span-92-15"><a id="__codelineno-92-15" name="__codelineno-92-15" href="#__codelineno-92-15"></a>    collection_column_join_script      = NULL_VALUE  // (MANDATORY generic file)  
</span><span id="__span-92-16"><a id="__codelineno-92-16" name="__codelineno-92-16" href="#__codelineno-92-16"></a>    in_input_fastqs_collection         = []          // (MANDATORY array)         eg. [file1, ...]
</span><span id="__span-92-17"><a id="__codelineno-92-17" name="__codelineno-92-17" href="#__codelineno-92-17"></a>    in_input_reference_gene_bed        = NULL_VALUE  // (MANDATORY generic file)  
</span><span id="__span-92-18"><a id="__codelineno-92-18" name="__codelineno-92-18" href="#__codelineno-92-18"></a>    multiqc_config                     = NULL_VALUE  // (MANDATORY generic file)  
</span><span id="__span-92-19"><a id="__codelineno-92-19" name="__codelineno-92-19" href="#__codelineno-92-19"></a>    fastqc1_format_string              = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-20"><a id="__codelineno-92-20" name="__codelineno-92-20" href="#__codelineno-92-20"></a>    fastqc1_outdir                     = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-21"><a id="__codelineno-92-21" name="__codelineno-92-21" href="#__codelineno-92-21"></a>    fastqc2_format_string              = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-22"><a id="__codelineno-92-22" name="__codelineno-92-22" href="#__codelineno-92-22"></a>    fastqc2_outdir                     = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-23"><a id="__codelineno-92-23" name="__codelineno-92-23" href="#__codelineno-92-23"></a>    hisat2_index_path                  = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-24"><a id="__codelineno-92-24" name="__codelineno-92-24" href="#__codelineno-92-24"></a>    hisat2_read11                      = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-25"><a id="__codelineno-92-25" name="__codelineno-92-25" href="#__codelineno-92-25"></a>    rseqc_gene_body_coverage_safename  = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-26"><a id="__codelineno-92-26" name="__codelineno-92-26" href="#__codelineno-92-26"></a>    samtools_idxstats_addthreads       = NULL_VALUE  // (MANDATORY string)        
</span><span id="__span-92-27"><a id="__codelineno-92-27" name="__codelineno-92-27" href="#__codelineno-92-27"></a>
</span><span id="__span-92-28"><a id="__codelineno-92-28" name="__codelineno-92-28" href="#__codelineno-92-28"></a>    // INPUTS (OPTIONAL)
</span><span id="__span-92-29"><a id="__codelineno-92-29" name="__codelineno-92-29" href="#__codelineno-92-29"></a>    fastqc1_contaminants  = NULL_VALUE  // (optional tabular)   
</span><span id="__span-92-30"><a id="__codelineno-92-30" name="__codelineno-92-30" href="#__codelineno-92-30"></a>    fastqc1_limits        = NULL_VALUE  // (optional textfile)  
</span><span id="__span-92-31"><a id="__codelineno-92-31" name="__codelineno-92-31" href="#__codelineno-92-31"></a>    fastqc2_contaminants  = NULL_VALUE  // (optional tabular)   
</span><span id="__span-92-32"><a id="__codelineno-92-32" name="__codelineno-92-32" href="#__codelineno-92-32"></a>    fastqc2_limits        = NULL_VALUE  // (optional textfile)  
</span><span id="__span-92-33"><a id="__codelineno-92-33" name="__codelineno-92-33" href="#__codelineno-92-33"></a>
</span><span id="__span-92-34"><a id="__codelineno-92-34" name="__codelineno-92-34" href="#__codelineno-92-34"></a>}
</span></code></pre></div>
<p><br></p>
<p><strong>Setting up Workflow Inputs</strong></p>
<p>To run the workflow, we will supply values for the mandatory workflow inputs, and leave the optional inputs as-is. </p>
<p>The auto-generated <code>nextflow.config</code> splits up workflow inputs using some headings.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-93-1"><a id="__codelineno-93-1" name="__codelineno-93-1" href="#__codelineno-93-1"></a># Workflow inputs which are mandatory
</span><span id="__span-93-2"><a id="__codelineno-93-2" name="__codelineno-93-2" href="#__codelineno-93-2"></a>// INPUTS (MANDATORY)
</span><span id="__span-93-3"><a id="__codelineno-93-3" name="__codelineno-93-3" href="#__codelineno-93-3"></a>
</span><span id="__span-93-4"><a id="__codelineno-93-4" name="__codelineno-93-4" href="#__codelineno-93-4"></a># Workflow inputs which are optional
</span><span id="__span-93-5"><a id="__codelineno-93-5" name="__codelineno-93-5" href="#__codelineno-93-5"></a>// INPUTS (OPTIONAL)
</span></code></pre></div>
<p>We will need to supply values for those under the <code>// INPUTS (MANDATORY)</code> heading. </p>
<p>Here is an explanation of the role of these inputs:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-94-1"><a id="__codelineno-94-1" name="__codelineno-94-1" href="#__codelineno-94-1"></a>collection_column_join_script      # path to script needed by COLLECTION_COLUMN_JOIN
</span><span id="__span-94-2"><a id="__codelineno-94-2" name="__codelineno-94-2" href="#__codelineno-94-2"></a>in_input_fastqs_collection         # .fastq format input reads (.gz ok)
</span><span id="__span-94-3"><a id="__codelineno-94-3" name="__codelineno-94-3" href="#__codelineno-94-3"></a>in_input_reference_gene_bed        # .bed format gene annotations
</span><span id="__span-94-4"><a id="__codelineno-94-4" name="__codelineno-94-4" href="#__codelineno-94-4"></a>multiqc_config                     # path to script needed by MULTIQC
</span><span id="__span-94-5"><a id="__codelineno-94-5" name="__codelineno-94-5" href="#__codelineno-94-5"></a>fastqc1_format_string              # string to supply the input format -f to FASTQC1
</span><span id="__span-94-6"><a id="__codelineno-94-6" name="__codelineno-94-6" href="#__codelineno-94-6"></a>fastqc1_outdir                     # string to supply the outputs folder for FASTQC1
</span><span id="__span-94-7"><a id="__codelineno-94-7" name="__codelineno-94-7" href="#__codelineno-94-7"></a>fastqc2_format_string              # string to supply the input format -f to FASTQC2
</span><span id="__span-94-8"><a id="__codelineno-94-8" name="__codelineno-94-8" href="#__codelineno-94-8"></a>fastqc2_outdir                     # string to supply the outputs folder for FASTQC2
</span><span id="__span-94-9"><a id="__codelineno-94-9" name="__codelineno-94-9" href="#__codelineno-94-9"></a>hisat2_index_path                  # path to hisat2 index for HISAT2
</span><span id="__span-94-10"><a id="__codelineno-94-10" name="__codelineno-94-10" href="#__codelineno-94-10"></a>hisat2_read11                      # ?
</span><span id="__span-94-11"><a id="__codelineno-94-11" name="__codelineno-94-11" href="#__codelineno-94-11"></a>rseqc_gene_body_coverage_safename  # ?
</span><span id="__span-94-12"><a id="__codelineno-94-12" name="__codelineno-94-12" href="#__codelineno-94-12"></a>samtools_idxstats_addthreads       # value for the -@ threads argument for SAMTOOLS_IDXSTATS
</span></code></pre></div>
<p>Sample data for the <code>.fastq</code> and <code>.bed</code> inputs have been provided in the <code>data/sample_data/galaxy/rnaseq_reads_to_counts</code> folder we obtained by downloading &amp; extracting the zenodo archive.</p>
<p>The scripts will have been copied by <code>janis translate</code> and placed in the <code>rnaseq_reads_to_counts/templates</code> folder. </p>
<p><br></p>
<p>Replace the <code>// INPUTS (MANDATORY)</code> section of <code>nextflow.config</code> with the following text to set up our sample data as workflow inputs:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-95-1"><a id="__codelineno-95-1" name="__codelineno-95-1" href="#__codelineno-95-1"></a>// INPUTS (MANDATORY)
</span><span id="__span-95-2"><a id="__codelineno-95-2" name="__codelineno-95-2" href="#__codelineno-95-2"></a>collection_column_join_script      = NULL_VALUE
</span><span id="__span-95-3"><a id="__codelineno-95-3" name="__codelineno-95-3" href="#__codelineno-95-3"></a>in_input_fastqs_collection         = [
</span><span id="__span-95-4"><a id="__codelineno-95-4" name="__codelineno-95-4" href="#__codelineno-95-4"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DG-basalvirgin.fastq.gz&quot;,
</span><span id="__span-95-5"><a id="__codelineno-95-5" name="__codelineno-95-5" href="#__codelineno-95-5"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DH-basalvirgin.fastq.gz&quot;,
</span><span id="__span-95-6"><a id="__codelineno-95-6" name="__codelineno-95-6" href="#__codelineno-95-6"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DI-basalpregnant.fastq.gz&quot;,
</span><span id="__span-95-7"><a id="__codelineno-95-7" name="__codelineno-95-7" href="#__codelineno-95-7"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DJ-basalpregnant.fq.gz&quot;,
</span><span id="__span-95-8"><a id="__codelineno-95-8" name="__codelineno-95-8" href="#__codelineno-95-8"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DK-basallactate.fastq.gz&quot;,
</span><span id="__span-95-9"><a id="__codelineno-95-9" name="__codelineno-95-9" href="#__codelineno-95-9"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-DL-basallactate.fastq.gz&quot;,
</span><span id="__span-95-10"><a id="__codelineno-95-10" name="__codelineno-95-10" href="#__codelineno-95-10"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LA-luminalvirgin.fastq.gz&quot;,
</span><span id="__span-95-11"><a id="__codelineno-95-11" name="__codelineno-95-11" href="#__codelineno-95-11"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LB-luminalvirgin.fastq.gz&quot;,
</span><span id="__span-95-12"><a id="__codelineno-95-12" name="__codelineno-95-12" href="#__codelineno-95-12"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LC-luminalpregnant.fastq.gz&quot;,
</span><span id="__span-95-13"><a id="__codelineno-95-13" name="__codelineno-95-13" href="#__codelineno-95-13"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LD-luminalpregnant.fastq.gz&quot;,
</span><span id="__span-95-14"><a id="__codelineno-95-14" name="__codelineno-95-14" href="#__codelineno-95-14"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LE-luminallactate.fastq.gz&quot;,
</span><span id="__span-95-15"><a id="__codelineno-95-15" name="__codelineno-95-15" href="#__codelineno-95-15"></a>    &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/MCL1-LF-luminallactate.fastq.gz&quot;,
</span><span id="__span-95-16"><a id="__codelineno-95-16" name="__codelineno-95-16" href="#__codelineno-95-16"></a>]
</span><span id="__span-95-17"><a id="__codelineno-95-17" name="__codelineno-95-17" href="#__codelineno-95-17"></a>in_input_reference_gene_bed        = &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/ucsc/mm10_RefSeq.bed&quot;
</span><span id="__span-95-18"><a id="__codelineno-95-18" name="__codelineno-95-18" href="#__codelineno-95-18"></a>multiqc_config                     = &quot;templates/multiqc_config&quot;
</span><span id="__span-95-19"><a id="__codelineno-95-19" name="__codelineno-95-19" href="#__codelineno-95-19"></a>fastqc1_format_string              = &quot;fastq&quot;
</span><span id="__span-95-20"><a id="__codelineno-95-20" name="__codelineno-95-20" href="#__codelineno-95-20"></a>fastqc1_outdir                     = &quot;outdir&quot;
</span><span id="__span-95-21"><a id="__codelineno-95-21" name="__codelineno-95-21" href="#__codelineno-95-21"></a>fastqc2_format_string              = &quot;fastq&quot;
</span><span id="__span-95-22"><a id="__codelineno-95-22" name="__codelineno-95-22" href="#__codelineno-95-22"></a>fastqc2_outdir                     = &quot;outdir&quot;
</span><span id="__span-95-23"><a id="__codelineno-95-23" name="__codelineno-95-23" href="#__codelineno-95-23"></a>hisat2_index_path                  = &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/hisat2_index/*.ht2&quot;
</span><span id="__span-95-24"><a id="__codelineno-95-24" name="__codelineno-95-24" href="#__codelineno-95-24"></a>hisat2_read11                      = NULL_VALUE  // (MANDATORY string)  
</span><span id="__span-95-25"><a id="__codelineno-95-25" name="__codelineno-95-25" href="#__codelineno-95-25"></a>rseqc_gene_body_coverage_safename  = NULL_VALUE  // (MANDATORY string)    
</span><span id="__span-95-26"><a id="__codelineno-95-26" name="__codelineno-95-26" href="#__codelineno-95-26"></a>samtools_idxstats_addthreads       = &quot;2&quot;       
</span></code></pre></div>
<p><br></p>
<p>Before we supply values, lets look at the inputs which were marked <code>?</code> in the explanation of inputs. Specifically, the <code>hisat2_read11</code> and <code>rseqc_gene_body_coverage_safename</code> params. </p>
<p>These both seem a bit strange. Hisat2 will receive reads as output from CUTADAPT so why do we need to supply a value? Rseqc is similar - there is no safename argument for the software, so we are unsure what this does. </p>
<p><br></p>
<p><strong>Investigating hisat2_reads11 parameter</strong></p>
<p>Open <code>modules/hisat2.nf</code> to look at its process definition. It should look similar to the following: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-96-1"><a id="__codelineno-96-1" name="__codelineno-96-1" href="#__codelineno-96-1"></a>process HISAT2 {
</span><span id="__span-96-2"><a id="__codelineno-96-2" name="__codelineno-96-2" href="#__codelineno-96-2"></a>
</span><span id="__span-96-3"><a id="__codelineno-96-3" name="__codelineno-96-3" href="#__codelineno-96-3"></a>    container &quot;quay.io/grace_hall1/hisat2:2.2.1&quot; 
</span><span id="__span-96-4"><a id="__codelineno-96-4" name="__codelineno-96-4" href="#__codelineno-96-4"></a>    publishDir &quot;${params.outdir}/hisat2&quot;
</span><span id="__span-96-5"><a id="__codelineno-96-5" name="__codelineno-96-5" href="#__codelineno-96-5"></a>
</span><span id="__span-96-6"><a id="__codelineno-96-6" name="__codelineno-96-6" href="#__codelineno-96-6"></a>    input:
</span><span id="__span-96-7"><a id="__codelineno-96-7" name="__codelineno-96-7" href="#__codelineno-96-7"></a>    path library_input_1
</span><span id="__span-96-8"><a id="__codelineno-96-8" name="__codelineno-96-8" href="#__codelineno-96-8"></a>    val index_path
</span><span id="__span-96-9"><a id="__codelineno-96-9" name="__codelineno-96-9" href="#__codelineno-96-9"></a>    val read11
</span><span id="__span-96-10"><a id="__codelineno-96-10" name="__codelineno-96-10" href="#__codelineno-96-10"></a>
</span><span id="__span-96-11"><a id="__codelineno-96-11" name="__codelineno-96-11" href="#__codelineno-96-11"></a>    output:
</span><span id="__span-96-12"><a id="__codelineno-96-12" name="__codelineno-96-12" href="#__codelineno-96-12"></a>    path &quot;summary.txt&quot;, emit: out_summary_file
</span><span id="__span-96-13"><a id="__codelineno-96-13" name="__codelineno-96-13" href="#__codelineno-96-13"></a>    path &quot;unknown_collection_pattern&quot;, emit: output_alignments
</span><span id="__span-96-14"><a id="__codelineno-96-14" name="__codelineno-96-14" href="#__codelineno-96-14"></a>
</span><span id="__span-96-15"><a id="__codelineno-96-15" name="__codelineno-96-15" href="#__codelineno-96-15"></a>    script:
</span><span id="__span-96-16"><a id="__codelineno-96-16" name="__codelineno-96-16" href="#__codelineno-96-16"></a>    &quot;&quot;&quot;
</span><span id="__span-96-17"><a id="__codelineno-96-17" name="__codelineno-96-17" href="#__codelineno-96-17"></a>    hisat2 \
</span><span id="__span-96-18"><a id="__codelineno-96-18" name="__codelineno-96-18" href="#__codelineno-96-18"></a>    ${library_input_1} \
</span><span id="__span-96-19"><a id="__codelineno-96-19" name="__codelineno-96-19" href="#__codelineno-96-19"></a>    -U ${read11} \
</span><span id="__span-96-20"><a id="__codelineno-96-20" name="__codelineno-96-20" href="#__codelineno-96-20"></a>    -x ${index_path}
</span><span id="__span-96-21"><a id="__codelineno-96-21" name="__codelineno-96-21" href="#__codelineno-96-21"></a>    &quot;&quot;&quot;
</span><span id="__span-96-22"><a id="__codelineno-96-22" name="__codelineno-96-22" href="#__codelineno-96-22"></a>
</span><span id="__span-96-23"><a id="__codelineno-96-23" name="__codelineno-96-23" href="#__codelineno-96-23"></a>}
</span></code></pre></div></p>
<p>It&rsquo;s clear that <code>janis translate</code> has gotten a bit confused here.</p>
<p>From looking at <code>main.nf</code>, we see that the output of CUTADAPT is being fed to the <code>library_input_1</code> process input of HISAT2. 
The input reads are therefore arriving as <code>library_input_1</code> rather than <code>read11</code>. </p>
<p>In the <code>script:</code> section of HISAT2 we see what is going on.</p>
<p>The <code>read11</code> input is being fed to <code>-U</code>, but this should instead be <code>library_input_1</code>. 
<code>read11</code> appears to have the same role as <code>library_input_1</code> and looks like a temporary variable the source Galaxy Tool Wrapper used. </p>
<p>To fix this, we will remove the <code>read11</code> HISAT2 process input &amp; associated references.<br>
We will then make changes to the HISAT2 process script &amp; outputs.</p>
<p><br></p>
<blockquote>
<p>Note:<br>
Why does this happen?<br>
Galaxy often uses temporary variables to configure / sanitise values.<br>
It&rsquo;s common for Galaxy tools to reassign an input to a temporary value, then pass that temp value to the actual software command. <br>
Here the Galaxy hisat2 tool has created a temporary variable (read11) then passes that to hisat2 rather than the original input. </p>
</blockquote>
<p><br></p>
<p>To begin, let&rsquo;s remove the process input itself, as well as the <code>nextflow.config</code> param, and the value supplied in the <code>HISAT2</code> process call. </p>
<div class="admonition question">
<p class="admonition-title">Removing process input</p>
<p>Remove the <code>val read11</code> input in the <code>HISAT2</code> process. </p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-97-1"><a id="__codelineno-97-1" name="__codelineno-97-1" href="#__codelineno-97-1"></a>process HISAT2 {
</span><span id="__span-97-2"><a id="__codelineno-97-2" name="__codelineno-97-2" href="#__codelineno-97-2"></a>    ...
</span><span id="__span-97-3"><a id="__codelineno-97-3" name="__codelineno-97-3" href="#__codelineno-97-3"></a>
</span><span id="__span-97-4"><a id="__codelineno-97-4" name="__codelineno-97-4" href="#__codelineno-97-4"></a>    input:
</span><span id="__span-97-5"><a id="__codelineno-97-5" name="__codelineno-97-5" href="#__codelineno-97-5"></a>    path library_input_1
</span><span id="__span-97-6"><a id="__codelineno-97-6" name="__codelineno-97-6" href="#__codelineno-97-6"></a>    val index_path              
</span><span id="__span-97-7"><a id="__codelineno-97-7" name="__codelineno-97-7" href="#__codelineno-97-7"></a>                                    &lt;- val read11 removed
</span><span id="__span-97-8"><a id="__codelineno-97-8" name="__codelineno-97-8" href="#__codelineno-97-8"></a>
</span><span id="__span-97-9"><a id="__codelineno-97-9" name="__codelineno-97-9" href="#__codelineno-97-9"></a>    ...
</span><span id="__span-97-10"><a id="__codelineno-97-10" name="__codelineno-97-10" href="#__codelineno-97-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Removing value in process call</p>
<p>Cut or comment out the value for <code>read11</code> supplied in the <code>HISAT2</code> process call in <code>main.nf</code>. </p>
<details class="hint">
<summary>Show Change</summary>
<p>main.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-98-1"><a id="__codelineno-98-1" name="__codelineno-98-1" href="#__codelineno-98-1"></a>workflow {
</span><span id="__span-98-2"><a id="__codelineno-98-2" name="__codelineno-98-2" href="#__codelineno-98-2"></a>    ...
</span><span id="__span-98-3"><a id="__codelineno-98-3" name="__codelineno-98-3" href="#__codelineno-98-3"></a>
</span><span id="__span-98-4"><a id="__codelineno-98-4" name="__codelineno-98-4" href="#__codelineno-98-4"></a>    HISAT2(
</span><span id="__span-98-5"><a id="__codelineno-98-5" name="__codelineno-98-5" href="#__codelineno-98-5"></a>        CUTADAPT.out.out12,        // library_input_1
</span><span id="__span-98-6"><a id="__codelineno-98-6" name="__codelineno-98-6" href="#__codelineno-98-6"></a>        params.hisat2_index_path,  // index_path
</span><span id="__span-98-7"><a id="__codelineno-98-7" name="__codelineno-98-7" href="#__codelineno-98-7"></a>        // params.hisat2_read11    // read11            &lt;- commented out / removed
</span><span id="__span-98-8"><a id="__codelineno-98-8" name="__codelineno-98-8" href="#__codelineno-98-8"></a>    )
</span><span id="__span-98-9"><a id="__codelineno-98-9" name="__codelineno-98-9" href="#__codelineno-98-9"></a>
</span><span id="__span-98-10"><a id="__codelineno-98-10" name="__codelineno-98-10" href="#__codelineno-98-10"></a>    ...
</span><span id="__span-98-11"><a id="__codelineno-98-11" name="__codelineno-98-11" href="#__codelineno-98-11"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Removing <code>nextflow.config</code> param</p>
<p>Cut or comment out the <code>hisat2_read11</code> param in <code>nextflow.config</code>. </p>
<details class="hint">
<summary>Show Change</summary>
<p>nextflow.config:
<div class="language-text highlight"><pre><span></span><code><span id="__span-99-1"><a id="__codelineno-99-1" name="__codelineno-99-1" href="#__codelineno-99-1"></a>params {
</span><span id="__span-99-2"><a id="__codelineno-99-2" name="__codelineno-99-2" href="#__codelineno-99-2"></a>    ...
</span><span id="__span-99-3"><a id="__codelineno-99-3" name="__codelineno-99-3" href="#__codelineno-99-3"></a>
</span><span id="__span-99-4"><a id="__codelineno-99-4" name="__codelineno-99-4" href="#__codelineno-99-4"></a>    hisat2_index_path                  = &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/hisat2_index/*.ht2&quot;
</span><span id="__span-99-5"><a id="__codelineno-99-5" name="__codelineno-99-5" href="#__codelineno-99-5"></a>    // hisat2_read11                      = NULL_VALUE  // (MANDATORY string)     &lt;-
</span><span id="__span-99-6"><a id="__codelineno-99-6" name="__codelineno-99-6" href="#__codelineno-99-6"></a>    rseqc_gene_body_coverage_safename  = NULL_VALUE  // (MANDATORY string)    
</span><span id="__span-99-7"><a id="__codelineno-99-7" name="__codelineno-99-7" href="#__codelineno-99-7"></a>    samtools_idxstats_addthreads       = &quot;2&quot;
</span><span id="__span-99-8"><a id="__codelineno-99-8" name="__codelineno-99-8" href="#__codelineno-99-8"></a>
</span><span id="__span-99-9"><a id="__codelineno-99-9" name="__codelineno-99-9" href="#__codelineno-99-9"></a>    ...
</span><span id="__span-99-10"><a id="__codelineno-99-10" name="__codelineno-99-10" href="#__codelineno-99-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Next we will swap <code>read11</code> with <code>library_input_1</code> in the script section &amp; add some extra arguments.</p>
<div class="admonition question">
<p class="admonition-title">Addressing HISAT2 script args</p>
<p>Swap <code>read11</code> with <code>library_input_1</code> in the HISAT2 script section. <br>
Remove the original line referencing <code>library_input_1</code>. <br>
Supply <code>--summary-file summary.txt</code> to hisat2 to specify an output summary textfile name.<br>
Supply <code>-S out.sam</code> to hisat2 to specify the name of the produced SAM file. </p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-100-1"><a id="__codelineno-100-1" name="__codelineno-100-1" href="#__codelineno-100-1"></a>process HISAT2 {
</span><span id="__span-100-2"><a id="__codelineno-100-2" name="__codelineno-100-2" href="#__codelineno-100-2"></a>    ...
</span><span id="__span-100-3"><a id="__codelineno-100-3" name="__codelineno-100-3" href="#__codelineno-100-3"></a>
</span><span id="__span-100-4"><a id="__codelineno-100-4" name="__codelineno-100-4" href="#__codelineno-100-4"></a>    script:
</span><span id="__span-100-5"><a id="__codelineno-100-5" name="__codelineno-100-5" href="#__codelineno-100-5"></a>    &quot;&quot;&quot;
</span><span id="__span-100-6"><a id="__codelineno-100-6" name="__codelineno-100-6" href="#__codelineno-100-6"></a>    hisat2 \
</span><span id="__span-100-7"><a id="__codelineno-100-7" name="__codelineno-100-7" href="#__codelineno-100-7"></a>    -U ${library_input_1} \             &lt;-
</span><span id="__span-100-8"><a id="__codelineno-100-8" name="__codelineno-100-8" href="#__codelineno-100-8"></a>    -x ${index_path} \
</span><span id="__span-100-9"><a id="__codelineno-100-9" name="__codelineno-100-9" href="#__codelineno-100-9"></a>    --summary-file summary.txt \        &lt;-
</span><span id="__span-100-10"><a id="__codelineno-100-10" name="__codelineno-100-10" href="#__codelineno-100-10"></a>    -S out.sam                          &lt;-
</span><span id="__span-100-11"><a id="__codelineno-100-11" name="__codelineno-100-11" href="#__codelineno-100-11"></a>    &quot;&quot;&quot;
</span><span id="__span-100-12"><a id="__codelineno-100-12" name="__codelineno-100-12" href="#__codelineno-100-12"></a>
</span><span id="__span-100-13"><a id="__codelineno-100-13" name="__codelineno-100-13" href="#__codelineno-100-13"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>While we&rsquo;re here, we will also add some samtools commands as downstream tasks required a <em>sorted &amp; indexed BAM</em>, not a SAM file which is the default output of <code>hisat2</code>. </p>
<div class="admonition question">
<p class="admonition-title">Adding Samtools commands</p>
<p>Add samtools commands at the bottom of the script section to convert the output SAM to BAM, then sort and index the file. <br>
You will need to add <code>samtools sort</code>, <code>samtools view</code>, and <code>samtools index</code> commands. </p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-101-1"><a id="__codelineno-101-1" name="__codelineno-101-1" href="#__codelineno-101-1"></a>process HISAT2 {
</span><span id="__span-101-2"><a id="__codelineno-101-2" name="__codelineno-101-2" href="#__codelineno-101-2"></a>    ...
</span><span id="__span-101-3"><a id="__codelineno-101-3" name="__codelineno-101-3" href="#__codelineno-101-3"></a>
</span><span id="__span-101-4"><a id="__codelineno-101-4" name="__codelineno-101-4" href="#__codelineno-101-4"></a>    script:
</span><span id="__span-101-5"><a id="__codelineno-101-5" name="__codelineno-101-5" href="#__codelineno-101-5"></a>    &quot;&quot;&quot;
</span><span id="__span-101-6"><a id="__codelineno-101-6" name="__codelineno-101-6" href="#__codelineno-101-6"></a>    hisat2 \
</span><span id="__span-101-7"><a id="__codelineno-101-7" name="__codelineno-101-7" href="#__codelineno-101-7"></a>    -U ${library_input_1} \
</span><span id="__span-101-8"><a id="__codelineno-101-8" name="__codelineno-101-8" href="#__codelineno-101-8"></a>    -x ${index_path} \
</span><span id="__span-101-9"><a id="__codelineno-101-9" name="__codelineno-101-9" href="#__codelineno-101-9"></a>    --summary-file summary.txt \
</span><span id="__span-101-10"><a id="__codelineno-101-10" name="__codelineno-101-10" href="#__codelineno-101-10"></a>    -S out.sam
</span><span id="__span-101-11"><a id="__codelineno-101-11" name="__codelineno-101-11" href="#__codelineno-101-11"></a>
</span><span id="__span-101-12"><a id="__codelineno-101-12" name="__codelineno-101-12" href="#__codelineno-101-12"></a>    samtools view out.sam -o out.bam                &lt;-
</span><span id="__span-101-13"><a id="__codelineno-101-13" name="__codelineno-101-13" href="#__codelineno-101-13"></a>    samtools sort out.bam -o sorted.bam             &lt;-
</span><span id="__span-101-14"><a id="__codelineno-101-14" name="__codelineno-101-14" href="#__codelineno-101-14"></a>    samtools index sorted.bam -o sorted.bam.bai     &lt;-
</span><span id="__span-101-15"><a id="__codelineno-101-15" name="__codelineno-101-15" href="#__codelineno-101-15"></a>    mv sorted.bam alignment.bam                     &lt;-
</span><span id="__span-101-16"><a id="__codelineno-101-16" name="__codelineno-101-16" href="#__codelineno-101-16"></a>    mv sorted.bam.bai alignment.bam.bai             &lt;-
</span><span id="__span-101-17"><a id="__codelineno-101-17" name="__codelineno-101-17" href="#__codelineno-101-17"></a>    &quot;&quot;&quot;
</span><span id="__span-101-18"><a id="__codelineno-101-18" name="__codelineno-101-18" href="#__codelineno-101-18"></a>
</span><span id="__span-101-19"><a id="__codelineno-101-19" name="__codelineno-101-19" href="#__codelineno-101-19"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>As a final change, we will modify the collection pattern for the <code>output_alignments</code> output in HISAT2. </p>
<p>In the example above we have named the output <code>alignment.bam</code>, so will update the collection pattern accordingly. If you have chosen a different name, make sure the correct file is being collected by <code>output_alignments</code>. </p>
<div class="admonition question">
<p class="admonition-title">Modifying output collection</p>
<p>Change the collection pattern for <code>output_alignments</code> to collect your sorted &amp; indexed BAM file.</p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-102-1"><a id="__codelineno-102-1" name="__codelineno-102-1" href="#__codelineno-102-1"></a>process HISAT2 {
</span><span id="__span-102-2"><a id="__codelineno-102-2" name="__codelineno-102-2" href="#__codelineno-102-2"></a>    ...
</span><span id="__span-102-3"><a id="__codelineno-102-3" name="__codelineno-102-3" href="#__codelineno-102-3"></a>
</span><span id="__span-102-4"><a id="__codelineno-102-4" name="__codelineno-102-4" href="#__codelineno-102-4"></a>    output:
</span><span id="__span-102-5"><a id="__codelineno-102-5" name="__codelineno-102-5" href="#__codelineno-102-5"></a>    path &quot;summary.txt&quot;, emit: out_summary_file
</span><span id="__span-102-6"><a id="__codelineno-102-6" name="__codelineno-102-6" href="#__codelineno-102-6"></a>    path &quot;alignment.bam&quot;, emit: output_alignments       &lt;-
</span><span id="__span-102-7"><a id="__codelineno-102-7" name="__codelineno-102-7" href="#__codelineno-102-7"></a>
</span><span id="__span-102-8"><a id="__codelineno-102-8" name="__codelineno-102-8" href="#__codelineno-102-8"></a>    ...
</span><span id="__span-102-9"><a id="__codelineno-102-9" name="__codelineno-102-9" href="#__codelineno-102-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p><strong>Investigating rseqc_gene_body_coverage_safename parameter</strong></p>
<p>Now that we have fixed up HISAT2 inputs and script, let&rsquo;s do the same for 
the <code>rseqc_gene_body_coverage_safename</code> param. </p>
<p>Open <code>rseqc_gene_body_coverage.nf</code> to look at the process definition. <br>
It should look similar to the following:
<div class="language-text highlight"><pre><span></span><code><span id="__span-103-1"><a id="__codelineno-103-1" name="__codelineno-103-1" href="#__codelineno-103-1"></a>process RSEQC_GENE_BODY_COVERAGE {
</span><span id="__span-103-2"><a id="__codelineno-103-2" name="__codelineno-103-2" href="#__codelineno-103-2"></a>
</span><span id="__span-103-3"><a id="__codelineno-103-3" name="__codelineno-103-3" href="#__codelineno-103-3"></a>    container &quot;quay.io/biocontainers/rseqc:2.6.4--py27hf8a1672_2&quot;
</span><span id="__span-103-4"><a id="__codelineno-103-4" name="__codelineno-103-4" href="#__codelineno-103-4"></a>    publishDir &quot;${params.outdir}/rseqc_gene_body_coverage&quot;
</span><span id="__span-103-5"><a id="__codelineno-103-5" name="__codelineno-103-5" href="#__codelineno-103-5"></a>
</span><span id="__span-103-6"><a id="__codelineno-103-6" name="__codelineno-103-6" href="#__codelineno-103-6"></a>    input:
</span><span id="__span-103-7"><a id="__codelineno-103-7" name="__codelineno-103-7" href="#__codelineno-103-7"></a>    path batch_mode_input
</span><span id="__span-103-8"><a id="__codelineno-103-8" name="__codelineno-103-8" href="#__codelineno-103-8"></a>    path refgene
</span><span id="__span-103-9"><a id="__codelineno-103-9" name="__codelineno-103-9" href="#__codelineno-103-9"></a>    val safename
</span><span id="__span-103-10"><a id="__codelineno-103-10" name="__codelineno-103-10" href="#__codelineno-103-10"></a>
</span><span id="__span-103-11"><a id="__codelineno-103-11" name="__codelineno-103-11" href="#__codelineno-103-11"></a>    output:
</span><span id="__span-103-12"><a id="__codelineno-103-12" name="__codelineno-103-12" href="#__codelineno-103-12"></a>    path &quot;output.geneBodyCoverage.curves.pdf&quot;, emit: outputcurvespdf
</span><span id="__span-103-13"><a id="__codelineno-103-13" name="__codelineno-103-13" href="#__codelineno-103-13"></a>    path &quot;output.geneBodyCoverage.txt&quot;, emit: outputtxt
</span><span id="__span-103-14"><a id="__codelineno-103-14" name="__codelineno-103-14" href="#__codelineno-103-14"></a>
</span><span id="__span-103-15"><a id="__codelineno-103-15" name="__codelineno-103-15" href="#__codelineno-103-15"></a>    script:
</span><span id="__span-103-16"><a id="__codelineno-103-16" name="__codelineno-103-16" href="#__codelineno-103-16"></a>    &quot;&quot;&quot;
</span><span id="__span-103-17"><a id="__codelineno-103-17" name="__codelineno-103-17" href="#__codelineno-103-17"></a>    geneBody_coverage.py \
</span><span id="__span-103-18"><a id="__codelineno-103-18" name="__codelineno-103-18" href="#__codelineno-103-18"></a>    ${batch_mode_input} \
</span><span id="__span-103-19"><a id="__codelineno-103-19" name="__codelineno-103-19" href="#__codelineno-103-19"></a>    -r ${refgene} \
</span><span id="__span-103-20"><a id="__codelineno-103-20" name="__codelineno-103-20" href="#__codelineno-103-20"></a>    -i ${safename} 
</span><span id="__span-103-21"><a id="__codelineno-103-21" name="__codelineno-103-21" href="#__codelineno-103-21"></a>    &quot;&quot;&quot;
</span><span id="__span-103-22"><a id="__codelineno-103-22" name="__codelineno-103-22" href="#__codelineno-103-22"></a>
</span><span id="__span-103-23"><a id="__codelineno-103-23" name="__codelineno-103-23" href="#__codelineno-103-23"></a>}
</span></code></pre></div></p>
<p>We see an identical situation as encountered for HISAT2. <br>
The <code>safename</code> process input is actually just a Galaxy temporary value, and takes the role of <code>batch_mode_input</code>. </p>
<p><br></p>
<p>Let&rsquo;s fix up <code>rseqc_gene_body_coverage_safename</code> process input in RSEQC_GENE_BODY_COVERAGE.
We will take the same approach as we did for the <code>read11</code> process input in HISAT2. </p>
<div class="admonition question">
<p class="admonition-title">Addressing <code>rseqc_gene_body_coverage_safename</code></p>
<p>Remove the <code>rseqc_gene_body_coverage_safename</code> param in <code>nextflow.config</code>. <br>
Remove the value being passed to <code>safename</code> in <code>main.nf</code>. <br>
Remove the <code>safename</code> process input from <code>RSEQC_GENE_BODY_COVERAGE</code>. <br>
Swap <code>safename</code> for <code>batch_mode_input</code> in the script, and remove the original reference to <code>batch_mode_input</code>.<br>
Additionally, supply <code>-o output</code> after the <code>-i</code> argument in the script section to set an output filename.<br><br></p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/rseqc_gene_body_coverage.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-104-1"><a id="__codelineno-104-1" name="__codelineno-104-1" href="#__codelineno-104-1"></a>process RSEQC_GENE_BODY_COVERAGE {
</span><span id="__span-104-2"><a id="__codelineno-104-2" name="__codelineno-104-2" href="#__codelineno-104-2"></a>
</span><span id="__span-104-3"><a id="__codelineno-104-3" name="__codelineno-104-3" href="#__codelineno-104-3"></a>    container &quot;quay.io/biocontainers/rseqc:2.6.4--py27hf8a1672_2&quot;
</span><span id="__span-104-4"><a id="__codelineno-104-4" name="__codelineno-104-4" href="#__codelineno-104-4"></a>    publishDir &quot;${params.outdir}/rseqc_gene_body_coverage&quot;
</span><span id="__span-104-5"><a id="__codelineno-104-5" name="__codelineno-104-5" href="#__codelineno-104-5"></a>
</span><span id="__span-104-6"><a id="__codelineno-104-6" name="__codelineno-104-6" href="#__codelineno-104-6"></a>    input:
</span><span id="__span-104-7"><a id="__codelineno-104-7" name="__codelineno-104-7" href="#__codelineno-104-7"></a>    path batch_mode_input
</span><span id="__span-104-8"><a id="__codelineno-104-8" name="__codelineno-104-8" href="#__codelineno-104-8"></a>    path refgene                                                        &lt;-
</span><span id="__span-104-9"><a id="__codelineno-104-9" name="__codelineno-104-9" href="#__codelineno-104-9"></a>
</span><span id="__span-104-10"><a id="__codelineno-104-10" name="__codelineno-104-10" href="#__codelineno-104-10"></a>    output:
</span><span id="__span-104-11"><a id="__codelineno-104-11" name="__codelineno-104-11" href="#__codelineno-104-11"></a>    path &quot;output.geneBodyCoverage.curves.pdf&quot;, emit: outputcurvespdf
</span><span id="__span-104-12"><a id="__codelineno-104-12" name="__codelineno-104-12" href="#__codelineno-104-12"></a>    path &quot;output.geneBodyCoverage.txt&quot;, emit: outputtxt
</span><span id="__span-104-13"><a id="__codelineno-104-13" name="__codelineno-104-13" href="#__codelineno-104-13"></a>
</span><span id="__span-104-14"><a id="__codelineno-104-14" name="__codelineno-104-14" href="#__codelineno-104-14"></a>    script:
</span><span id="__span-104-15"><a id="__codelineno-104-15" name="__codelineno-104-15" href="#__codelineno-104-15"></a>    &quot;&quot;&quot;
</span><span id="__span-104-16"><a id="__codelineno-104-16" name="__codelineno-104-16" href="#__codelineno-104-16"></a>    geneBody_coverage.py \
</span><span id="__span-104-17"><a id="__codelineno-104-17" name="__codelineno-104-17" href="#__codelineno-104-17"></a>    -r ${refgene} \
</span><span id="__span-104-18"><a id="__codelineno-104-18" name="__codelineno-104-18" href="#__codelineno-104-18"></a>    -i ${batch_mode_input} \                                            &lt;-
</span><span id="__span-104-19"><a id="__codelineno-104-19" name="__codelineno-104-19" href="#__codelineno-104-19"></a>    -o output                                                           &lt;-
</span><span id="__span-104-20"><a id="__codelineno-104-20" name="__codelineno-104-20" href="#__codelineno-104-20"></a>    &quot;&quot;&quot;
</span><span id="__span-104-21"><a id="__codelineno-104-21" name="__codelineno-104-21" href="#__codelineno-104-21"></a>
</span><span id="__span-104-22"><a id="__codelineno-104-22" name="__codelineno-104-22" href="#__codelineno-104-22"></a>}
</span></code></pre></div></p>
<p>main.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-105-1"><a id="__codelineno-105-1" name="__codelineno-105-1" href="#__codelineno-105-1"></a>workflow {
</span><span id="__span-105-2"><a id="__codelineno-105-2" name="__codelineno-105-2" href="#__codelineno-105-2"></a>    ...
</span><span id="__span-105-3"><a id="__codelineno-105-3" name="__codelineno-105-3" href="#__codelineno-105-3"></a>
</span><span id="__span-105-4"><a id="__codelineno-105-4" name="__codelineno-105-4" href="#__codelineno-105-4"></a>    RSEQC_GENE_BODY_COVERAGE(
</span><span id="__span-105-5"><a id="__codelineno-105-5" name="__codelineno-105-5" href="#__codelineno-105-5"></a>        HISAT2.out.output_alignments,             // batch_mode_input
</span><span id="__span-105-6"><a id="__codelineno-105-6" name="__codelineno-105-6" href="#__codelineno-105-6"></a>        in_input_reference_gene_bed,              // refgene
</span><span id="__span-105-7"><a id="__codelineno-105-7" name="__codelineno-105-7" href="#__codelineno-105-7"></a>        // params.rseqc_gene_body_coverage_safename  // safename        &lt;-
</span><span id="__span-105-8"><a id="__codelineno-105-8" name="__codelineno-105-8" href="#__codelineno-105-8"></a>    )
</span><span id="__span-105-9"><a id="__codelineno-105-9" name="__codelineno-105-9" href="#__codelineno-105-9"></a>
</span><span id="__span-105-10"><a id="__codelineno-105-10" name="__codelineno-105-10" href="#__codelineno-105-10"></a>    ...
</span><span id="__span-105-11"><a id="__codelineno-105-11" name="__codelineno-105-11" href="#__codelineno-105-11"></a>}
</span></code></pre></div></p>
<p>nextflow.config:
<div class="language-text highlight"><pre><span></span><code><span id="__span-106-1"><a id="__codelineno-106-1" name="__codelineno-106-1" href="#__codelineno-106-1"></a>params {
</span><span id="__span-106-2"><a id="__codelineno-106-2" name="__codelineno-106-2" href="#__codelineno-106-2"></a>    ...
</span><span id="__span-106-3"><a id="__codelineno-106-3" name="__codelineno-106-3" href="#__codelineno-106-3"></a>
</span><span id="__span-106-4"><a id="__codelineno-106-4" name="__codelineno-106-4" href="#__codelineno-106-4"></a>    hisat2_index_path                  = &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/hisat2_index/*.ht2&quot;
</span><span id="__span-106-5"><a id="__codelineno-106-5" name="__codelineno-106-5" href="#__codelineno-106-5"></a>    // hisat2_read11                      = NULL_VALUE  // (MANDATORY string)     
</span><span id="__span-106-6"><a id="__codelineno-106-6" name="__codelineno-106-6" href="#__codelineno-106-6"></a>    // rseqc_gene_body_coverage_safename  = NULL_VALUE  // (MANDATORY string)     &lt;-
</span><span id="__span-106-7"><a id="__codelineno-106-7" name="__codelineno-106-7" href="#__codelineno-106-7"></a>    samtools_idxstats_addthreads       = &quot;2&quot;
</span><span id="__span-106-8"><a id="__codelineno-106-8" name="__codelineno-106-8" href="#__codelineno-106-8"></a>
</span><span id="__span-106-9"><a id="__codelineno-106-9" name="__codelineno-106-9" href="#__codelineno-106-9"></a>    ...
</span><span id="__span-106-10"><a id="__codelineno-106-10" name="__codelineno-106-10" href="#__codelineno-106-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p><strong>Run the Workflow</strong></p>
<p>Now that we have set up our sample data and made a few changes to the workflow, lets run it to see where we stand. </p>
<p>Ensure you are in the <code>rnaseq_reads_to_counts/</code> working directory, where <code>nextflow.config</code> and <code>main.nf</code> reside. </p>
<p>If not, use the following to change directory. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-107-1"><a id="__codelineno-107-1" name="__codelineno-107-1" href="#__codelineno-107-1"></a>cd rnaseq_reads_to_counts
</span></code></pre></div></p>
<p>To run the workflow using our sample data, we can now write the following command: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-108-1"><a id="__codelineno-108-1" name="__codelineno-108-1" href="#__codelineno-108-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p>While the workflow runs, you will encounter one of the following errors:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-109-1"><a id="__codelineno-109-1" name="__codelineno-109-1" href="#__codelineno-109-1"></a>Caused by:
</span><span id="__span-109-2"><a id="__codelineno-109-2" name="__codelineno-109-2" href="#__codelineno-109-2"></a>  Process `FASTQC1 (9)` terminated with an error exit status (2)
</span><span id="__span-109-3"><a id="__codelineno-109-3" name="__codelineno-109-3" href="#__codelineno-109-3"></a>
</span><span id="__span-109-4"><a id="__codelineno-109-4" name="__codelineno-109-4" href="#__codelineno-109-4"></a>Command executed:
</span><span id="__span-109-5"><a id="__codelineno-109-5" name="__codelineno-109-5" href="#__codelineno-109-5"></a>  fastqc               --outdir outdir     -f fastq     MCL1-LC-luminalpregnant.fastq.gz
</span><span id="__span-109-6"><a id="__codelineno-109-6" name="__codelineno-109-6" href="#__codelineno-109-6"></a>
</span><span id="__span-109-7"><a id="__codelineno-109-7" name="__codelineno-109-7" href="#__codelineno-109-7"></a>Command error:
</span><span id="__span-109-8"><a id="__codelineno-109-8" name="__codelineno-109-8" href="#__codelineno-109-8"></a>  Specified output directory &#39;outdir&#39; does not exist
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code><span id="__span-110-1"><a id="__codelineno-110-1" name="__codelineno-110-1" href="#__codelineno-110-1"></a>ERROR ~ Error executing process &gt; &#39;CUTADAPT (3)&#39;
</span><span id="__span-110-2"><a id="__codelineno-110-2" name="__codelineno-110-2" href="#__codelineno-110-2"></a>
</span><span id="__span-110-3"><a id="__codelineno-110-3" name="__codelineno-110-3" href="#__codelineno-110-3"></a>Caused by:
</span><span id="__span-110-4"><a id="__codelineno-110-4" name="__codelineno-110-4" href="#__codelineno-110-4"></a>  Process `CUTADAPT (3)` terminated with an error exit status (255)
</span><span id="__span-110-5"><a id="__codelineno-110-5" name="__codelineno-110-5" href="#__codelineno-110-5"></a>
</span><span id="__span-110-6"><a id="__codelineno-110-6" name="__codelineno-110-6" href="#__codelineno-110-6"></a>Command executed:
</span><span id="__span-110-7"><a id="__codelineno-110-7" name="__codelineno-110-7" href="#__codelineno-110-7"></a>
</span><span id="__span-110-8"><a id="__codelineno-110-8" name="__codelineno-110-8" href="#__codelineno-110-8"></a>  cutadapt     MCL1-DI-basalpregnant.fastq.gz
</span></code></pre></div>
<p>It could be either of the above, as the dataflow paradigm adopted by nextflow doesn&rsquo;t guarantee the order which processes are executed. </p>
<p>Errors are somewhat expected when translating Galaxy workflows. Janis translate doesn&rsquo;t produce perfect translations - just the best it can do. </p>
<p>This is the first of multiple errors we will encounter and fix while making this workflow runnable. </p>
<p><br></p>
<h3 id="manual-adjustments_5">Manual Adjustments<a class="headerlink" href="#manual-adjustments_5" title="Permanent link">&para;</a></h3>
<p>In the remainer of this workshop we will fix up 7 errors we encounter along the way. </p>
<p>During the process we will discuss some differences between Galaxy and Nextflow, and will get some practise with Nextflow debugging.</p>
<p>We have split each error into parts: the <strong><em>Error message</em></strong>, <strong><em>Diagnosing the error</em></strong>, and the <strong><em>solution</em></strong>. </p>
<p>If your error message is different to the one being discussed, do not worry. You can still follow the order listed below to create the runnable workflow. </p>
<p><br></p>
<h3 id="error-1-fastqc-outdir">Error 1: FASTQC outdir<a class="headerlink" href="#error-1-fastqc-outdir" title="Permanent link">&para;</a></h3>
<p>The first issue we need to address is caused by the FASTQC process translation. </p>
<p><strong>Error message</strong>
<div class="language-text highlight"><pre><span></span><code><span id="__span-111-1"><a id="__codelineno-111-1" name="__codelineno-111-1" href="#__codelineno-111-1"></a>Caused by:
</span><span id="__span-111-2"><a id="__codelineno-111-2" name="__codelineno-111-2" href="#__codelineno-111-2"></a>  Process `FASTQC1 (9)` terminated with an error exit status (2)
</span><span id="__span-111-3"><a id="__codelineno-111-3" name="__codelineno-111-3" href="#__codelineno-111-3"></a>
</span><span id="__span-111-4"><a id="__codelineno-111-4" name="__codelineno-111-4" href="#__codelineno-111-4"></a>Command executed:
</span><span id="__span-111-5"><a id="__codelineno-111-5" name="__codelineno-111-5" href="#__codelineno-111-5"></a>  fastqc               --outdir outdir     -f fastq     MCL1-LC-luminalpregnant.fastq.gz
</span><span id="__span-111-6"><a id="__codelineno-111-6" name="__codelineno-111-6" href="#__codelineno-111-6"></a>
</span><span id="__span-111-7"><a id="__codelineno-111-7" name="__codelineno-111-7" href="#__codelineno-111-7"></a>Command error:
</span><span id="__span-111-8"><a id="__codelineno-111-8" name="__codelineno-111-8" href="#__codelineno-111-8"></a>  Specified output directory &#39;outdir&#39; does not exist
</span></code></pre></div></p>
<p><br></p>
<p><strong>Diagnosing the error</strong></p>
<p>This message tells us that FastQC was trying to place files in <code>outdir/</code> folder of the process working directory, but no such folder existed. </p>
<p>Nextflow task errors can be broken down accordingly:
<div class="language-text highlight"><pre><span></span><code><span id="__span-112-1"><a id="__codelineno-112-1" name="__codelineno-112-1" href="#__codelineno-112-1"></a>Error executing process &gt; &#39;FASTQC1 (9)&#39;     # Main error heading
</span><span id="__span-112-2"><a id="__codelineno-112-2" name="__codelineno-112-2" href="#__codelineno-112-2"></a>Caused by:                                  # Task which experienced the error
</span><span id="__span-112-3"><a id="__codelineno-112-3" name="__codelineno-112-3" href="#__codelineno-112-3"></a>Command executed:                           # Final shell command which was executed
</span><span id="__span-112-4"><a id="__codelineno-112-4" name="__codelineno-112-4" href="#__codelineno-112-4"></a>Command exit status:                        # Exit code for the executed command
</span><span id="__span-112-5"><a id="__codelineno-112-5" name="__codelineno-112-5" href="#__codelineno-112-5"></a>Command output:                             # Stdout for the executed command
</span><span id="__span-112-6"><a id="__codelineno-112-6" name="__codelineno-112-6" href="#__codelineno-112-6"></a>Command error:                              # Stderr for the executed command
</span><span id="__span-112-7"><a id="__codelineno-112-7" name="__codelineno-112-7" href="#__codelineno-112-7"></a>Work dir:                                   # The process working directory
</span></code></pre></div></p>
<p>By re-reading the error message provided by Nextflow, we see that FASTQC1 experienced an error. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-113-1"><a id="__codelineno-113-1" name="__codelineno-113-1" href="#__codelineno-113-1"></a>Process `FASTQC1 (9)` terminated with an error exit status (2)
</span></code></pre></div></p>
<p>We can also see the exact shell command which was executed (yours may have a different fastq): 
<div class="language-text highlight"><pre><span></span><code><span id="__span-114-1"><a id="__codelineno-114-1" name="__codelineno-114-1" href="#__codelineno-114-1"></a>fastqc --outdir outdir -f fastq MCL1-LC-luminalpregnant.fastq.gz
</span></code></pre></div></p>
<p>Finally we see that the FastQC software threw the error, rather than Nextflow:
<div class="language-text highlight"><pre><span></span><code><span id="__span-115-1"><a id="__codelineno-115-1" name="__codelineno-115-1" href="#__codelineno-115-1"></a>Command error:
</span><span id="__span-115-2"><a id="__codelineno-115-2" name="__codelineno-115-2" href="#__codelineno-115-2"></a>Specified output directory &#39;outdir&#39; does not exist
</span></code></pre></div></p>
<p>From the above, we can understand that the folder we specify using <code>--outdir</code> must exist before FastQC runs. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>Let&rsquo;s update the FASTQC process definition in <code>modules/fastqc.nf</code> so this directory is created before fastqc runs. Since we pass the name of this directory as the <code>outdir</code> process input, we will use a reference.</p>
<div class="admonition question">
<p class="admonition-title">Creating FASTQC output directory</p>
<p>Add <code>mkdir ${outdir}</code> to the script section of the FASTQC process. <br>
Make sure it appears before the main <code>fastqc</code> command. <br>
As it is a separate statement, make sure you <strong><em>don&rsquo;t</em></strong> add <code>\</code> after this statement.</p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/fastqc.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-116-1"><a id="__codelineno-116-1" name="__codelineno-116-1" href="#__codelineno-116-1"></a>process FASTQC {
</span><span id="__span-116-2"><a id="__codelineno-116-2" name="__codelineno-116-2" href="#__codelineno-116-2"></a>    ...
</span><span id="__span-116-3"><a id="__codelineno-116-3" name="__codelineno-116-3" href="#__codelineno-116-3"></a>
</span><span id="__span-116-4"><a id="__codelineno-116-4" name="__codelineno-116-4" href="#__codelineno-116-4"></a>    script:
</span><span id="__span-116-5"><a id="__codelineno-116-5" name="__codelineno-116-5" href="#__codelineno-116-5"></a>    def contaminants = contaminants.simpleName != params.NULL_VALUE ? &quot;--contaminants ${contaminants}&quot; : &quot;&quot;
</span><span id="__span-116-6"><a id="__codelineno-116-6" name="__codelineno-116-6" href="#__codelineno-116-6"></a>    def limits = limits.simpleName != params.NULL_VALUE ? &quot;--limits ${limits}&quot; : &quot;&quot;
</span><span id="__span-116-7"><a id="__codelineno-116-7" name="__codelineno-116-7" href="#__codelineno-116-7"></a>    &quot;&quot;&quot;
</span><span id="__span-116-8"><a id="__codelineno-116-8" name="__codelineno-116-8" href="#__codelineno-116-8"></a>    mkdir ${outdir}                 &lt;-
</span><span id="__span-116-9"><a id="__codelineno-116-9" name="__codelineno-116-9" href="#__codelineno-116-9"></a>    fastqc \
</span><span id="__span-116-10"><a id="__codelineno-116-10" name="__codelineno-116-10" href="#__codelineno-116-10"></a>    ${contaminants} \
</span><span id="__span-116-11"><a id="__codelineno-116-11" name="__codelineno-116-11" href="#__codelineno-116-11"></a>    ${limits} \
</span><span id="__span-116-12"><a id="__codelineno-116-12" name="__codelineno-116-12" href="#__codelineno-116-12"></a>    --outdir ${outdir} \
</span><span id="__span-116-13"><a id="__codelineno-116-13" name="__codelineno-116-13" href="#__codelineno-116-13"></a>    -f ${format_string} \
</span><span id="__span-116-14"><a id="__codelineno-116-14" name="__codelineno-116-14" href="#__codelineno-116-14"></a>    ${input_file} 
</span><span id="__span-116-15"><a id="__codelineno-116-15" name="__codelineno-116-15" href="#__codelineno-116-15"></a>    &quot;&quot;&quot;
</span><span id="__span-116-16"><a id="__codelineno-116-16" name="__codelineno-116-16" href="#__codelineno-116-16"></a>
</span><span id="__span-116-17"><a id="__codelineno-116-17" name="__codelineno-116-17" href="#__codelineno-116-17"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>While we are here, let&rsquo;s also adjust the output collection patterns. </p>
<p>As we are supplying <code>--outdir ${outdir}</code> to the fastqc command, output files will be in a folder determined by the <code>val outdir</code> process input. 
Fastqc also adds <code>_fastqc.ext</code> to each output file, where <code>ext</code> is the file extension.<br></p>
<div class="admonition question">
<p class="admonition-title">Modifying output collection patterns</p>
<p>Adjust the output collection pattern of the FASTQC <code>out_html_file</code> and <code>out_text_file</code> outputs. <br>
Collect each <code>.html</code> and <code>.zip</code> in the folder supplied via the <code>val outdir</code> process input. </p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/fastqc.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-117-1"><a id="__codelineno-117-1" name="__codelineno-117-1" href="#__codelineno-117-1"></a>process FASTQC {
</span><span id="__span-117-2"><a id="__codelineno-117-2" name="__codelineno-117-2" href="#__codelineno-117-2"></a>    ...
</span><span id="__span-117-3"><a id="__codelineno-117-3" name="__codelineno-117-3" href="#__codelineno-117-3"></a>
</span><span id="__span-117-4"><a id="__codelineno-117-4" name="__codelineno-117-4" href="#__codelineno-117-4"></a>    output:
</span><span id="__span-117-5"><a id="__codelineno-117-5" name="__codelineno-117-5" href="#__codelineno-117-5"></a>    path &quot;${outdir}/*_fastqc.html&quot;, emit: out_html_file     &lt;-
</span><span id="__span-117-6"><a id="__codelineno-117-6" name="__codelineno-117-6" href="#__codelineno-117-6"></a>    path &quot;${outdir}/*_fastqc.zip&quot;, emit: out_text_file      &lt;-
</span><span id="__span-117-7"><a id="__codelineno-117-7" name="__codelineno-117-7" href="#__codelineno-117-7"></a>
</span><span id="__span-117-8"><a id="__codelineno-117-8" name="__codelineno-117-8" href="#__codelineno-117-8"></a>    ...
</span><span id="__span-117-9"><a id="__codelineno-117-9" name="__codelineno-117-9" href="#__codelineno-117-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<blockquote>
<p>Note<br>
Nextflow supports <strong><em>wildcards</em></strong> wherever paths are used. <br>
Here, <code>${outdir}/*_fastqc.html</code> will collect any file which ends in _fastqc.html with <code>out1</code>, as <code>*</code> matches any text. </p>
</blockquote>
<p><br></p>
<p>After you have made the above changes, re-run the workflow using the same command as before:
<div class="language-text highlight"><pre><span></span><code><span id="__span-118-1"><a id="__codelineno-118-1" name="__codelineno-118-1" href="#__codelineno-118-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p><br></p>
<h3 id="error-2-cutadapt-output-collection">Error 2: CUTADAPT output collection<a class="headerlink" href="#error-2-cutadapt-output-collection" title="Permanent link">&para;</a></h3>
<p>The next error you will experience is due to output collection in the CUTADAPT process. </p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-119-1"><a id="__codelineno-119-1" name="__codelineno-119-1" href="#__codelineno-119-1"></a>Error executing process &gt; &#39;CUTADAPT (11)&#39;
</span><span id="__span-119-2"><a id="__codelineno-119-2" name="__codelineno-119-2" href="#__codelineno-119-2"></a>
</span><span id="__span-119-3"><a id="__codelineno-119-3" name="__codelineno-119-3" href="#__codelineno-119-3"></a>Caused by:
</span><span id="__span-119-4"><a id="__codelineno-119-4" name="__codelineno-119-4" href="#__codelineno-119-4"></a>  Missing output file(s) `out1*` expected by process `CUTADAPT (11)`
</span><span id="__span-119-5"><a id="__codelineno-119-5" name="__codelineno-119-5" href="#__codelineno-119-5"></a>
</span><span id="__span-119-6"><a id="__codelineno-119-6" name="__codelineno-119-6" href="#__codelineno-119-6"></a>Command executed:
</span><span id="__span-119-7"><a id="__codelineno-119-7" name="__codelineno-119-7" href="#__codelineno-119-7"></a>  cutadapt     MCL1-LE-luminallactate.fastq.gz
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the Error</strong></p>
<p>The error message above informs us that Nextflow had a problem collecting the outputs of a CUTADAPT task. Cutadapt seems to have run without issue, but Nextflow expected a file starting with <em>&ldquo;out1&rdquo;</em> to be present in the process working directory to collect as an output. </p>
<p>Looking at the <code>Command output:</code> section of the error message, you will also see what looks like the contents of a <code>.fastq</code> file.  As the <code>Command output:</code> component of Nextflow error messages shows us <code>stdout</code> during execution, we can gather that the cutadapt output is being sent to stdout, rather than being sent to a file. </p>
<p>Open <code>modules/cutadapt.nf</code> to look at the process definition. <br>
You should see something similar to the following:
<div class="language-text highlight"><pre><span></span><code><span id="__span-120-1"><a id="__codelineno-120-1" name="__codelineno-120-1" href="#__codelineno-120-1"></a>process CUTADAPT {
</span><span id="__span-120-2"><a id="__codelineno-120-2" name="__codelineno-120-2" href="#__codelineno-120-2"></a>
</span><span id="__span-120-3"><a id="__codelineno-120-3" name="__codelineno-120-3" href="#__codelineno-120-3"></a>    container &quot;quay.io/biocontainers/cutadapt:1.16--py35_2&quot;
</span><span id="__span-120-4"><a id="__codelineno-120-4" name="__codelineno-120-4" href="#__codelineno-120-4"></a>    publishDir &quot;${params.outdir}/cutadapt&quot;
</span><span id="__span-120-5"><a id="__codelineno-120-5" name="__codelineno-120-5" href="#__codelineno-120-5"></a>
</span><span id="__span-120-6"><a id="__codelineno-120-6" name="__codelineno-120-6" href="#__codelineno-120-6"></a>    input:
</span><span id="__span-120-7"><a id="__codelineno-120-7" name="__codelineno-120-7" href="#__codelineno-120-7"></a>    path library_input_1
</span><span id="__span-120-8"><a id="__codelineno-120-8" name="__codelineno-120-8" href="#__codelineno-120-8"></a>
</span><span id="__span-120-9"><a id="__codelineno-120-9" name="__codelineno-120-9" href="#__codelineno-120-9"></a>    output:
</span><span id="__span-120-10"><a id="__codelineno-120-10" name="__codelineno-120-10" href="#__codelineno-120-10"></a>    path &quot;out1*&quot;, emit: out12
</span><span id="__span-120-11"><a id="__codelineno-120-11" name="__codelineno-120-11" href="#__codelineno-120-11"></a>    path &quot;report.txt&quot;, emit: out_report
</span><span id="__span-120-12"><a id="__codelineno-120-12" name="__codelineno-120-12" href="#__codelineno-120-12"></a>
</span><span id="__span-120-13"><a id="__codelineno-120-13" name="__codelineno-120-13" href="#__codelineno-120-13"></a>    script:
</span><span id="__span-120-14"><a id="__codelineno-120-14" name="__codelineno-120-14" href="#__codelineno-120-14"></a>    &quot;&quot;&quot;
</span><span id="__span-120-15"><a id="__codelineno-120-15" name="__codelineno-120-15" href="#__codelineno-120-15"></a>    cutadapt \
</span><span id="__span-120-16"><a id="__codelineno-120-16" name="__codelineno-120-16" href="#__codelineno-120-16"></a>    ${library_input_1} 
</span><span id="__span-120-17"><a id="__codelineno-120-17" name="__codelineno-120-17" href="#__codelineno-120-17"></a>    &quot;&quot;&quot;
</span><span id="__span-120-18"><a id="__codelineno-120-18" name="__codelineno-120-18" href="#__codelineno-120-18"></a>
</span><span id="__span-120-19"><a id="__codelineno-120-19" name="__codelineno-120-19" href="#__codelineno-120-19"></a>}
</span></code></pre></div></p>
<p>We can see that <code>cutadapt</code> is being called using <code>library_input_1</code> as input reads. After <code>cutadapt</code> has run, the <code>out12</code> output is trying to collect a file starting with &ldquo;out1&rdquo;, and the <code>out_report</code> output is trying to collect &ldquo;report.txt&rdquo;. </p>
<p>If you are familiar with cutadapt, you will also see that we are not supplying any adapter sequence to remove from reads. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>Let&rsquo;s update the <code>input:</code> section to add an adapters input, the <code>script:</code> section to format the command correctly, and the <code>output:</code> section to collect our outputs. </p>
<p>To begin, we will address supplying an adapter sequence to the CUTADAPT process. </p>
<div class="admonition question">
<p class="admonition-title">Adding CUTADAPT process input</p>
<p>Add a new <code>val</code> process input called <code>adapter</code> to the CUTADAPT process.</p>
<details class="hint">
<summary>Show Change</summary>
<p>modules/cutadapt.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-121-1"><a id="__codelineno-121-1" name="__codelineno-121-1" href="#__codelineno-121-1"></a>CUTADAPT {
</span><span id="__span-121-2"><a id="__codelineno-121-2" name="__codelineno-121-2" href="#__codelineno-121-2"></a>    ...
</span><span id="__span-121-3"><a id="__codelineno-121-3" name="__codelineno-121-3" href="#__codelineno-121-3"></a>
</span><span id="__span-121-4"><a id="__codelineno-121-4" name="__codelineno-121-4" href="#__codelineno-121-4"></a>    input:
</span><span id="__span-121-5"><a id="__codelineno-121-5" name="__codelineno-121-5" href="#__codelineno-121-5"></a>    path library_input_1
</span><span id="__span-121-6"><a id="__codelineno-121-6" name="__codelineno-121-6" href="#__codelineno-121-6"></a>    val adapter                     &lt;-
</span><span id="__span-121-7"><a id="__codelineno-121-7" name="__codelineno-121-7" href="#__codelineno-121-7"></a>
</span><span id="__span-121-8"><a id="__codelineno-121-8" name="__codelineno-121-8" href="#__codelineno-121-8"></a>    ...
</span><span id="__span-121-9"><a id="__codelineno-121-9" name="__codelineno-121-9" href="#__codelineno-121-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Adding new param in nextflow config</p>
<p>Add a new <code>cutadapt_adapter</code> param in <code>nextflow.config</code> with the value &ldquo;AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC&rdquo;.<br></p>
<details class="hint">
<summary>Show Change</summary>
<p>nextflow.config:
<div class="language-text highlight"><pre><span></span><code><span id="__span-122-1"><a id="__codelineno-122-1" name="__codelineno-122-1" href="#__codelineno-122-1"></a>params {
</span><span id="__span-122-2"><a id="__codelineno-122-2" name="__codelineno-122-2" href="#__codelineno-122-2"></a>    ...
</span><span id="__span-122-3"><a id="__codelineno-122-3" name="__codelineno-122-3" href="#__codelineno-122-3"></a>
</span><span id="__span-122-4"><a id="__codelineno-122-4" name="__codelineno-122-4" href="#__codelineno-122-4"></a>    samtools_idxstats_addthreads       = &quot;2&quot;   
</span><span id="__span-122-5"><a id="__codelineno-122-5" name="__codelineno-122-5" href="#__codelineno-122-5"></a>    cutadapt_adapter                   = &quot;AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC&quot;     &lt;-
</span><span id="__span-122-6"><a id="__codelineno-122-6" name="__codelineno-122-6" href="#__codelineno-122-6"></a>
</span><span id="__span-122-7"><a id="__codelineno-122-7" name="__codelineno-122-7" href="#__codelineno-122-7"></a>    ...
</span><span id="__span-122-8"><a id="__codelineno-122-8" name="__codelineno-122-8" href="#__codelineno-122-8"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Suppling new param to CUTADAPT call</p>
<p>Supply the new <code>params.cutadapt_adapter</code> as a value to the CUTADAPT process call in the main workflow.<br>
Make sure you add a comma after <code>ch_in_input_fastqs_collection.flatten()</code>!</p>
<details class="hint">
<summary>Show Change</summary>
<p>main.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-123-1"><a id="__codelineno-123-1" name="__codelineno-123-1" href="#__codelineno-123-1"></a>workflow {
</span><span id="__span-123-2"><a id="__codelineno-123-2" name="__codelineno-123-2" href="#__codelineno-123-2"></a>    ...
</span><span id="__span-123-3"><a id="__codelineno-123-3" name="__codelineno-123-3" href="#__codelineno-123-3"></a>
</span><span id="__span-123-4"><a id="__codelineno-123-4" name="__codelineno-123-4" href="#__codelineno-123-4"></a>    CUTADAPT(
</span><span id="__span-123-5"><a id="__codelineno-123-5" name="__codelineno-123-5" href="#__codelineno-123-5"></a>        ch_in_input_fastqs_collection.flatten(),  // library_input_1    &lt;-
</span><span id="__span-123-6"><a id="__codelineno-123-6" name="__codelineno-123-6" href="#__codelineno-123-6"></a>        params.cutadapt_adapter                                         &lt;-
</span><span id="__span-123-7"><a id="__codelineno-123-7" name="__codelineno-123-7" href="#__codelineno-123-7"></a>    )
</span><span id="__span-123-8"><a id="__codelineno-123-8" name="__codelineno-123-8" href="#__codelineno-123-8"></a>
</span><span id="__span-123-9"><a id="__codelineno-123-9" name="__codelineno-123-9" href="#__codelineno-123-9"></a>    ...
</span><span id="__span-123-10"><a id="__codelineno-123-10" name="__codelineno-123-10" href="#__codelineno-123-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Next, we will formulate the script section of CUTADAPT so that the correct command gets executed. <br>
The correct format to run cutadapt is as follows:
<div class="language-text highlight"><pre><span></span><code><span id="__span-124-1"><a id="__codelineno-124-1" name="__codelineno-124-1" href="#__codelineno-124-1"></a>cutadapt -a AACCGGTT -o out.fastq in.fastq &gt; report.txt
</span></code></pre></div>
When used in this manner, the output fastq is sent to <code>-o</code>, and the summary report is sent to <code>stdout</code> (collected as &ldquo;report.txt&rdquo; above).</p>
<div class="admonition question">
<p class="admonition-title">Modifying CUTADAPT script</p>
<p>Add the <code>-a</code> argument to the process script section. Supply a reference to the new <code>adapters</code> process input as its value.<br>
Add the <code>-o</code> argument to the process script section. Supply a filename as its value based on the basename of our input reads (<code>library_input_1</code>) using <code>.simpleName</code>.<br>
Send <code>stdout</code> to a file using a <code>&gt;</code> redirect. Supply a filename based on the basename of our input reads as above. <br></p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/cutadapt.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-125-1"><a id="__codelineno-125-1" name="__codelineno-125-1" href="#__codelineno-125-1"></a>process CUTADAPT {
</span><span id="__span-125-2"><a id="__codelineno-125-2" name="__codelineno-125-2" href="#__codelineno-125-2"></a>    ...
</span><span id="__span-125-3"><a id="__codelineno-125-3" name="__codelineno-125-3" href="#__codelineno-125-3"></a>
</span><span id="__span-125-4"><a id="__codelineno-125-4" name="__codelineno-125-4" href="#__codelineno-125-4"></a>    script:
</span><span id="__span-125-5"><a id="__codelineno-125-5" name="__codelineno-125-5" href="#__codelineno-125-5"></a>    &quot;&quot;&quot;
</span><span id="__span-125-6"><a id="__codelineno-125-6" name="__codelineno-125-6" href="#__codelineno-125-6"></a>    cutadapt \
</span><span id="__span-125-7"><a id="__codelineno-125-7" name="__codelineno-125-7" href="#__codelineno-125-7"></a>    -a ${adapter} \                                             &lt;-
</span><span id="__span-125-8"><a id="__codelineno-125-8" name="__codelineno-125-8" href="#__codelineno-125-8"></a>    -o ${library_input_1.simpleName}.cutadapt.fastq.gz \        &lt;-
</span><span id="__span-125-9"><a id="__codelineno-125-9" name="__codelineno-125-9" href="#__codelineno-125-9"></a>    ${library_input_1} \    
</span><span id="__span-125-10"><a id="__codelineno-125-10" name="__codelineno-125-10" href="#__codelineno-125-10"></a>    &gt; ${library_input_1.simpleName}.cutadapt.txt                &lt;-
</span><span id="__span-125-11"><a id="__codelineno-125-11" name="__codelineno-125-11" href="#__codelineno-125-11"></a>    &quot;&quot;&quot;
</span><span id="__span-125-12"><a id="__codelineno-125-12" name="__codelineno-125-12" href="#__codelineno-125-12"></a>
</span><span id="__span-125-13"><a id="__codelineno-125-13" name="__codelineno-125-13" href="#__codelineno-125-13"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Now that we have fixed the script section of CUTADAPT, we will change the collection pattern for our outputs. </p>
<div class="admonition question">
<p class="admonition-title">Modifying CUTADAPT output collection</p>
<p>For the <code>out12</code> output, collect the same filename you supplied in the script section with <code>o-</code>.<br>
For the <code>out_report</code> output, collect the same filename you supplied in the script section as stdout redirect. <br></p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/cutadapt.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-126-1"><a id="__codelineno-126-1" name="__codelineno-126-1" href="#__codelineno-126-1"></a>process CUTADAPT {
</span><span id="__span-126-2"><a id="__codelineno-126-2" name="__codelineno-126-2" href="#__codelineno-126-2"></a>    ...
</span><span id="__span-126-3"><a id="__codelineno-126-3" name="__codelineno-126-3" href="#__codelineno-126-3"></a>
</span><span id="__span-126-4"><a id="__codelineno-126-4" name="__codelineno-126-4" href="#__codelineno-126-4"></a>    output:
</span><span id="__span-126-5"><a id="__codelineno-126-5" name="__codelineno-126-5" href="#__codelineno-126-5"></a>    path &quot;${library_input_1.simpleName}.cutadapt.fastq.gz&quot;, emit: out12     &lt;-
</span><span id="__span-126-6"><a id="__codelineno-126-6" name="__codelineno-126-6" href="#__codelineno-126-6"></a>    path &quot;${library_input_1.simpleName}.cutadapt.txt&quot;, emit: out_report     &lt;-
</span><span id="__span-126-7"><a id="__codelineno-126-7" name="__codelineno-126-7" href="#__codelineno-126-7"></a>
</span><span id="__span-126-8"><a id="__codelineno-126-8" name="__codelineno-126-8" href="#__codelineno-126-8"></a>    ...
</span><span id="__span-126-9"><a id="__codelineno-126-9" name="__codelineno-126-9" href="#__codelineno-126-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After making the changes above, re-run the Nextflow workflow:
<div class="language-text highlight"><pre><span></span><code><span id="__span-127-1"><a id="__codelineno-127-1" name="__codelineno-127-1" href="#__codelineno-127-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p><br></p>
<blockquote>
<p>NOTE <br>
Why were these <code>cutadapt</code> arguments missed? <br><br>
The default setting for <code>janis translate</code> is to translate tools according to their usage in the workflow. <br>
This is because Galaxy Tool Wrappers are complex and often have many inputs - most of which aren&rsquo;t relevant to the workflow being translated. <br><br>
To get the extended tool translation we can use <code>--mode full</code> when running janis translate:
<div class="language-text highlight"><pre><span></span><code><span id="__span-128-1"><a id="__codelineno-128-1" name="__codelineno-128-1" href="#__codelineno-128-1"></a>janis translate --from galaxy --to nextflow source/rna_seq_reads_to_counts.ga --mode full
</span></code></pre></div></p>
<p>For the <code>CUTADAPT</code> process, it will now look similar to the following: 
<div class="language-text highlight"><pre><span></span><code><span id="__span-129-1"><a id="__codelineno-129-1" name="__codelineno-129-1" href="#__codelineno-129-1"></a>nextflow.enable.dsl = 2
</span><span id="__span-129-2"><a id="__codelineno-129-2" name="__codelineno-129-2" href="#__codelineno-129-2"></a>
</span><span id="__span-129-3"><a id="__codelineno-129-3" name="__codelineno-129-3" href="#__codelineno-129-3"></a>process CUTADAPT {
</span><span id="__span-129-4"><a id="__codelineno-129-4" name="__codelineno-129-4" href="#__codelineno-129-4"></a>    
</span><span id="__span-129-5"><a id="__codelineno-129-5" name="__codelineno-129-5" href="#__codelineno-129-5"></a>    container &quot;quay.io/biocontainers/cutadapt:1.16--py35_2&quot;
</span><span id="__span-129-6"><a id="__codelineno-129-6" name="__codelineno-129-6" href="#__codelineno-129-6"></a>    publishDir &quot;${params.outdir}/cutadapt&quot;
</span><span id="__span-129-7"><a id="__codelineno-129-7" name="__codelineno-129-7" href="#__codelineno-129-7"></a>
</span><span id="__span-129-8"><a id="__codelineno-129-8" name="__codelineno-129-8" href="#__codelineno-129-8"></a>    input:
</span><span id="__span-129-9"><a id="__codelineno-129-9" name="__codelineno-129-9" href="#__codelineno-129-9"></a>    path library_input_1
</span><span id="__span-129-10"><a id="__codelineno-129-10" name="__codelineno-129-10" href="#__codelineno-129-10"></a>    path info_file
</span><span id="__span-129-11"><a id="__codelineno-129-11" name="__codelineno-129-11" href="#__codelineno-129-11"></a>    path output_file
</span><span id="__span-129-12"><a id="__codelineno-129-12" name="__codelineno-129-12" href="#__codelineno-129-12"></a>    path paired_output
</span><span id="__span-129-13"><a id="__codelineno-129-13" name="__codelineno-129-13" href="#__codelineno-129-13"></a>    path rest_file
</span><span id="__span-129-14"><a id="__codelineno-129-14" name="__codelineno-129-14" href="#__codelineno-129-14"></a>    path too_long_output
</span><span id="__span-129-15"><a id="__codelineno-129-15" name="__codelineno-129-15" href="#__codelineno-129-15"></a>    path too_long_paired_output
</span><span id="__span-129-16"><a id="__codelineno-129-16" name="__codelineno-129-16" href="#__codelineno-129-16"></a>    path too_short_output
</span><span id="__span-129-17"><a id="__codelineno-129-17" name="__codelineno-129-17" href="#__codelineno-129-17"></a>    path too_short_paired_output
</span><span id="__span-129-18"><a id="__codelineno-129-18" name="__codelineno-129-18" href="#__codelineno-129-18"></a>    path untrimmed_output
</span><span id="__span-129-19"><a id="__codelineno-129-19" name="__codelineno-129-19" href="#__codelineno-129-19"></a>    path untrimmed_paired_output
</span><span id="__span-129-20"><a id="__codelineno-129-20" name="__codelineno-129-20" href="#__codelineno-129-20"></a>    path wildcard_file
</span><span id="__span-129-21"><a id="__codelineno-129-21" name="__codelineno-129-21" href="#__codelineno-129-21"></a>    val discard
</span><span id="__span-129-22"><a id="__codelineno-129-22" name="__codelineno-129-22" href="#__codelineno-129-22"></a>    val discard_untrimmed
</span><span id="__span-129-23"><a id="__codelineno-129-23" name="__codelineno-129-23" href="#__codelineno-129-23"></a>    val mask_adapter
</span><span id="__span-129-24"><a id="__codelineno-129-24" name="__codelineno-129-24" href="#__codelineno-129-24"></a>    val match_read_wildcards
</span><span id="__span-129-25"><a id="__codelineno-129-25" name="__codelineno-129-25" href="#__codelineno-129-25"></a>    val no_indels
</span><span id="__span-129-26"><a id="__codelineno-129-26" name="__codelineno-129-26" href="#__codelineno-129-26"></a>    val no_trim
</span><span id="__span-129-27"><a id="__codelineno-129-27" name="__codelineno-129-27" href="#__codelineno-129-27"></a>    val trim_n
</span><span id="__span-129-28"><a id="__codelineno-129-28" name="__codelineno-129-28" href="#__codelineno-129-28"></a>    val error_rate
</span><span id="__span-129-29"><a id="__codelineno-129-29" name="__codelineno-129-29" href="#__codelineno-129-29"></a>    val length
</span><span id="__span-129-30"><a id="__codelineno-129-30" name="__codelineno-129-30" href="#__codelineno-129-30"></a>    val length_tag
</span><span id="__span-129-31"><a id="__codelineno-129-31" name="__codelineno-129-31" href="#__codelineno-129-31"></a>    val max_n
</span><span id="__span-129-32"><a id="__codelineno-129-32" name="__codelineno-129-32" href="#__codelineno-129-32"></a>    val maximum_length
</span><span id="__span-129-33"><a id="__codelineno-129-33" name="__codelineno-129-33" href="#__codelineno-129-33"></a>    val minimum_length
</span><span id="__span-129-34"><a id="__codelineno-129-34" name="__codelineno-129-34" href="#__codelineno-129-34"></a>    val nextseq_trim
</span><span id="__span-129-35"><a id="__codelineno-129-35" name="__codelineno-129-35" href="#__codelineno-129-35"></a>    val option_j
</span><span id="__span-129-36"><a id="__codelineno-129-36" name="__codelineno-129-36" href="#__codelineno-129-36"></a>    val option_u
</span><span id="__span-129-37"><a id="__codelineno-129-37" name="__codelineno-129-37" href="#__codelineno-129-37"></a>    val option_u
</span><span id="__span-129-38"><a id="__codelineno-129-38" name="__codelineno-129-38" href="#__codelineno-129-38"></a>    val overlap
</span><span id="__span-129-39"><a id="__codelineno-129-39" name="__codelineno-129-39" href="#__codelineno-129-39"></a>    val pair_filter
</span><span id="__span-129-40"><a id="__codelineno-129-40" name="__codelineno-129-40" href="#__codelineno-129-40"></a>    val prefix
</span><span id="__span-129-41"><a id="__codelineno-129-41" name="__codelineno-129-41" href="#__codelineno-129-41"></a>    val quality_cutoff
</span><span id="__span-129-42"><a id="__codelineno-129-42" name="__codelineno-129-42" href="#__codelineno-129-42"></a>    val read1
</span><span id="__span-129-43"><a id="__codelineno-129-43" name="__codelineno-129-43" href="#__codelineno-129-43"></a>    val read2
</span><span id="__span-129-44"><a id="__codelineno-129-44" name="__codelineno-129-44" href="#__codelineno-129-44"></a>    val suffix
</span><span id="__span-129-45"><a id="__codelineno-129-45" name="__codelineno-129-45" href="#__codelineno-129-45"></a>    val times
</span><span id="__span-129-46"><a id="__codelineno-129-46" name="__codelineno-129-46" href="#__codelineno-129-46"></a>
</span><span id="__span-129-47"><a id="__codelineno-129-47" name="__codelineno-129-47" href="#__codelineno-129-47"></a>    output:
</span><span id="__span-129-48"><a id="__codelineno-129-48" name="__codelineno-129-48" href="#__codelineno-129-48"></a>    path &quot;out1*&quot;, emit: out1
</span><span id="__span-129-49"><a id="__codelineno-129-49" name="__codelineno-129-49" href="#__codelineno-129-49"></a>    path &quot;out2*&quot;, emit: out2
</span><span id="__span-129-50"><a id="__codelineno-129-50" name="__codelineno-129-50" href="#__codelineno-129-50"></a>    path &quot;info_file.txt&quot;, emit: out_info_file
</span><span id="__span-129-51"><a id="__codelineno-129-51" name="__codelineno-129-51" href="#__codelineno-129-51"></a>    path &quot;report.txt&quot;, emit: out_report
</span><span id="__span-129-52"><a id="__codelineno-129-52" name="__codelineno-129-52" href="#__codelineno-129-52"></a>    path &quot;rest_output.fastqsanger&quot;, emit: out_rest_output
</span><span id="__span-129-53"><a id="__codelineno-129-53" name="__codelineno-129-53" href="#__codelineno-129-53"></a>    path &quot;too_long_output.fastqsanger&quot;, emit: out_too_long_output
</span><span id="__span-129-54"><a id="__codelineno-129-54" name="__codelineno-129-54" href="#__codelineno-129-54"></a>    path &quot;too_long_paired_output.fastqsanger&quot;, emit: out_too_long_paired_output
</span><span id="__span-129-55"><a id="__codelineno-129-55" name="__codelineno-129-55" href="#__codelineno-129-55"></a>    path &quot;too_short_output.fastqsanger&quot;, emit: out_too_short_output
</span><span id="__span-129-56"><a id="__codelineno-129-56" name="__codelineno-129-56" href="#__codelineno-129-56"></a>    path &quot;too_short_paired_output.fastqsanger&quot;, emit: out_too_short_paired_output
</span><span id="__span-129-57"><a id="__codelineno-129-57" name="__codelineno-129-57" href="#__codelineno-129-57"></a>    path &quot;untrimmed_output.fastqsanger&quot;, emit: out_untrimmed_output
</span><span id="__span-129-58"><a id="__codelineno-129-58" name="__codelineno-129-58" href="#__codelineno-129-58"></a>    path &quot;untrimmed_paired_output.fastqsanger&quot;, emit: out_untrimmed_paired_output
</span><span id="__span-129-59"><a id="__codelineno-129-59" name="__codelineno-129-59" href="#__codelineno-129-59"></a>    path &quot;wild_output.txt&quot;, emit: out_wild_output
</span><span id="__span-129-60"><a id="__codelineno-129-60" name="__codelineno-129-60" href="#__codelineno-129-60"></a>
</span><span id="__span-129-61"><a id="__codelineno-129-61" name="__codelineno-129-61" href="#__codelineno-129-61"></a>    script:
</span><span id="__span-129-62"><a id="__codelineno-129-62" name="__codelineno-129-62" href="#__codelineno-129-62"></a>    def info_file = info_file.simpleName != params.NULL_VALUE ? &quot;--info-file=${info_file}&quot; : &quot;&quot;
</span><span id="__span-129-63"><a id="__codelineno-129-63" name="__codelineno-129-63" href="#__codelineno-129-63"></a>    def output_file = output_file.simpleName != params.NULL_VALUE ? &quot;--output=${output_file}&quot; : &quot;&quot;
</span><span id="__span-129-64"><a id="__codelineno-129-64" name="__codelineno-129-64" href="#__codelineno-129-64"></a>    def paired_output = paired_output.simpleName != params.NULL_VALUE ? &quot;--paired-output=${paired_output}&quot; : &quot;&quot;
</span><span id="__span-129-65"><a id="__codelineno-129-65" name="__codelineno-129-65" href="#__codelineno-129-65"></a>    def rest_file = rest_file.simpleName != params.NULL_VALUE ? &quot;--rest-file=${rest_file}&quot; : &quot;&quot;
</span><span id="__span-129-66"><a id="__codelineno-129-66" name="__codelineno-129-66" href="#__codelineno-129-66"></a>    def too_long_output = too_long_output.simpleName != params.NULL_VALUE ? &quot;--too-long-output=${too_long_output}&quot; : &quot;&quot;
</span><span id="__span-129-67"><a id="__codelineno-129-67" name="__codelineno-129-67" href="#__codelineno-129-67"></a>    def too_long_paired_output = too_long_paired_output.simpleName != params.NULL_VALUE ? &quot;--too-long-paired-output=${too_long_paired_output}&quot; : &quot;&quot;
</span><span id="__span-129-68"><a id="__codelineno-129-68" name="__codelineno-129-68" href="#__codelineno-129-68"></a>    def too_short_output = too_short_output.simpleName != params.NULL_VALUE ? &quot;--too-short-output=${too_short_output}&quot; : &quot;&quot;
</span><span id="__span-129-69"><a id="__codelineno-129-69" name="__codelineno-129-69" href="#__codelineno-129-69"></a>    def too_short_paired_output = too_short_paired_output.simpleName != params.NULL_VALUE ? &quot;--too-short-paired-output=${too_short_paired_output}&quot; : &quot;&quot;
</span><span id="__span-129-70"><a id="__codelineno-129-70" name="__codelineno-129-70" href="#__codelineno-129-70"></a>    def untrimmed_output = untrimmed_output.simpleName != params.NULL_VALUE ? &quot;--untrimmed-output=${untrimmed_output}&quot; : &quot;&quot;
</span><span id="__span-129-71"><a id="__codelineno-129-71" name="__codelineno-129-71" href="#__codelineno-129-71"></a>    def untrimmed_paired_output = untrimmed_paired_output.simpleName != params.NULL_VALUE ? &quot;--untrimmed-paired-output=${untrimmed_paired_output}&quot; : &quot;&quot;
</span><span id="__span-129-72"><a id="__codelineno-129-72" name="__codelineno-129-72" href="#__codelineno-129-72"></a>    def wildcard_file = wildcard_file.simpleName != params.NULL_VALUE ? &quot;--wildcard-file=${wildcard_file}&quot; : &quot;&quot;
</span><span id="__span-129-73"><a id="__codelineno-129-73" name="__codelineno-129-73" href="#__codelineno-129-73"></a>    def discard = discard ? &quot;--discard&quot; : &quot;&quot;
</span><span id="__span-129-74"><a id="__codelineno-129-74" name="__codelineno-129-74" href="#__codelineno-129-74"></a>    def discard_untrimmed = discard_untrimmed ? &quot;--discard-untrimmed&quot; : &quot;&quot;
</span><span id="__span-129-75"><a id="__codelineno-129-75" name="__codelineno-129-75" href="#__codelineno-129-75"></a>    def mask_adapter = mask_adapter ? &quot;--mask-adapter&quot; : &quot;&quot;
</span><span id="__span-129-76"><a id="__codelineno-129-76" name="__codelineno-129-76" href="#__codelineno-129-76"></a>    def match_read_wildcards = match_read_wildcards ? &quot;--match-read-wildcards&quot; : &quot;&quot;
</span><span id="__span-129-77"><a id="__codelineno-129-77" name="__codelineno-129-77" href="#__codelineno-129-77"></a>    def no_indels = no_indels ? &quot;--no-indels&quot; : &quot;&quot;
</span><span id="__span-129-78"><a id="__codelineno-129-78" name="__codelineno-129-78" href="#__codelineno-129-78"></a>    def no_trim = no_trim ? &quot;--no-trim&quot; : &quot;&quot;
</span><span id="__span-129-79"><a id="__codelineno-129-79" name="__codelineno-129-79" href="#__codelineno-129-79"></a>    def trim_n = trim_n ? &quot;--trim-n&quot; : &quot;&quot;
</span><span id="__span-129-80"><a id="__codelineno-129-80" name="__codelineno-129-80" href="#__codelineno-129-80"></a>    def error_rate = error_rate != params.NULL_VALUE ? error_rate : 0.1
</span><span id="__span-129-81"><a id="__codelineno-129-81" name="__codelineno-129-81" href="#__codelineno-129-81"></a>    def length = length != params.NULL_VALUE ? length : 0
</span><span id="__span-129-82"><a id="__codelineno-129-82" name="__codelineno-129-82" href="#__codelineno-129-82"></a>    def length_tag = length_tag != params.NULL_VALUE ? &quot;--length-tag ${length_tag}&quot; : &quot;&quot;
</span><span id="__span-129-83"><a id="__codelineno-129-83" name="__codelineno-129-83" href="#__codelineno-129-83"></a>    def max_n = max_n != params.NULL_VALUE ? &quot;--max-n ${max_n}&quot; : &quot;&quot;
</span><span id="__span-129-84"><a id="__codelineno-129-84" name="__codelineno-129-84" href="#__codelineno-129-84"></a>    def maximum_length = maximum_length != params.NULL_VALUE ? maximum_length : 0
</span><span id="__span-129-85"><a id="__codelineno-129-85" name="__codelineno-129-85" href="#__codelineno-129-85"></a>    def minimum_length = minimum_length != params.NULL_VALUE ? minimum_length : 0
</span><span id="__span-129-86"><a id="__codelineno-129-86" name="__codelineno-129-86" href="#__codelineno-129-86"></a>    def nextseq_trim = nextseq_trim != params.NULL_VALUE ? nextseq_trim : 0
</span><span id="__span-129-87"><a id="__codelineno-129-87" name="__codelineno-129-87" href="#__codelineno-129-87"></a>    def option_u = option_u != params.NULL_VALUE ? option_u : 0
</span><span id="__span-129-88"><a id="__codelineno-129-88" name="__codelineno-129-88" href="#__codelineno-129-88"></a>    def option_j = option_j != params.NULL_VALUE ? option_j : 1
</span><span id="__span-129-89"><a id="__codelineno-129-89" name="__codelineno-129-89" href="#__codelineno-129-89"></a>    def option_u = option_u != params.NULL_VALUE ? option_u : 0
</span><span id="__span-129-90"><a id="__codelineno-129-90" name="__codelineno-129-90" href="#__codelineno-129-90"></a>    def overlap = overlap != params.NULL_VALUE ? overlap : 3
</span><span id="__span-129-91"><a id="__codelineno-129-91" name="__codelineno-129-91" href="#__codelineno-129-91"></a>    def pair_filter = pair_filter != params.NULL_VALUE ? pair_filter : &quot;any&quot;
</span><span id="__span-129-92"><a id="__codelineno-129-92" name="__codelineno-129-92" href="#__codelineno-129-92"></a>    def prefix = prefix != params.NULL_VALUE ? &quot;--prefix ${prefix}&quot; : &quot;&quot;
</span><span id="__span-129-93"><a id="__codelineno-129-93" name="__codelineno-129-93" href="#__codelineno-129-93"></a>    def quality_cutoff = quality_cutoff != params.NULL_VALUE ? quality_cutoff : &quot;0&quot;
</span><span id="__span-129-94"><a id="__codelineno-129-94" name="__codelineno-129-94" href="#__codelineno-129-94"></a>    def read1 = read1 != params.NULL_VALUE ? read1 : &quot;&quot;
</span><span id="__span-129-95"><a id="__codelineno-129-95" name="__codelineno-129-95" href="#__codelineno-129-95"></a>    def read2 = read2 != params.NULL_VALUE ? read2 : &quot;&quot;
</span><span id="__span-129-96"><a id="__codelineno-129-96" name="__codelineno-129-96" href="#__codelineno-129-96"></a>    def suffix = suffix != params.NULL_VALUE ? &quot;--suffix ${suffix}&quot; : &quot;&quot;
</span><span id="__span-129-97"><a id="__codelineno-129-97" name="__codelineno-129-97" href="#__codelineno-129-97"></a>    def times = times != params.NULL_VALUE ? times : 1
</span><span id="__span-129-98"><a id="__codelineno-129-98" name="__codelineno-129-98" href="#__codelineno-129-98"></a>    &quot;&quot;&quot;
</span><span id="__span-129-99"><a id="__codelineno-129-99" name="__codelineno-129-99" href="#__codelineno-129-99"></a>    cutadapt \
</span><span id="__span-129-100"><a id="__codelineno-129-100" name="__codelineno-129-100" href="#__codelineno-129-100"></a>    ${library_input_1} \
</span><span id="__span-129-101"><a id="__codelineno-129-101" name="__codelineno-129-101" href="#__codelineno-129-101"></a>    ${info_file} \
</span><span id="__span-129-102"><a id="__codelineno-129-102" name="__codelineno-129-102" href="#__codelineno-129-102"></a>    ${output_file} \
</span><span id="__span-129-103"><a id="__codelineno-129-103" name="__codelineno-129-103" href="#__codelineno-129-103"></a>    ${paired_output} \
</span><span id="__span-129-104"><a id="__codelineno-129-104" name="__codelineno-129-104" href="#__codelineno-129-104"></a>    ${rest_file} \
</span><span id="__span-129-105"><a id="__codelineno-129-105" name="__codelineno-129-105" href="#__codelineno-129-105"></a>    ${too_long_output} \
</span><span id="__span-129-106"><a id="__codelineno-129-106" name="__codelineno-129-106" href="#__codelineno-129-106"></a>    ${too_long_paired_output} \
</span><span id="__span-129-107"><a id="__codelineno-129-107" name="__codelineno-129-107" href="#__codelineno-129-107"></a>    ${too_short_output} \
</span><span id="__span-129-108"><a id="__codelineno-129-108" name="__codelineno-129-108" href="#__codelineno-129-108"></a>    ${too_short_paired_output} \
</span><span id="__span-129-109"><a id="__codelineno-129-109" name="__codelineno-129-109" href="#__codelineno-129-109"></a>    ${untrimmed_output} \
</span><span id="__span-129-110"><a id="__codelineno-129-110" name="__codelineno-129-110" href="#__codelineno-129-110"></a>    ${untrimmed_paired_output} \
</span><span id="__span-129-111"><a id="__codelineno-129-111" name="__codelineno-129-111" href="#__codelineno-129-111"></a>    ${wildcard_file} \
</span><span id="__span-129-112"><a id="__codelineno-129-112" name="__codelineno-129-112" href="#__codelineno-129-112"></a>    --error-rate ${error_rate} \
</span><span id="__span-129-113"><a id="__codelineno-129-113" name="__codelineno-129-113" href="#__codelineno-129-113"></a>    --length ${length} \
</span><span id="__span-129-114"><a id="__codelineno-129-114" name="__codelineno-129-114" href="#__codelineno-129-114"></a>    ${length_tag} \
</span><span id="__span-129-115"><a id="__codelineno-129-115" name="__codelineno-129-115" href="#__codelineno-129-115"></a>    ${max_n} \
</span><span id="__span-129-116"><a id="__codelineno-129-116" name="__codelineno-129-116" href="#__codelineno-129-116"></a>    --maximum-length ${maximum_length} \
</span><span id="__span-129-117"><a id="__codelineno-129-117" name="__codelineno-129-117" href="#__codelineno-129-117"></a>    --minimum-length ${minimum_length} \
</span><span id="__span-129-118"><a id="__codelineno-129-118" name="__codelineno-129-118" href="#__codelineno-129-118"></a>    --nextseq-trim ${nextseq_trim} \
</span><span id="__span-129-119"><a id="__codelineno-129-119" name="__codelineno-129-119" href="#__codelineno-129-119"></a>    --overlap ${overlap} \
</span><span id="__span-129-120"><a id="__codelineno-129-120" name="__codelineno-129-120" href="#__codelineno-129-120"></a>    --pair-filter ${pair_filter} \
</span><span id="__span-129-121"><a id="__codelineno-129-121" name="__codelineno-129-121" href="#__codelineno-129-121"></a>    ${prefix} \
</span><span id="__span-129-122"><a id="__codelineno-129-122" name="__codelineno-129-122" href="#__codelineno-129-122"></a>    --quality-cutoff ${quality_cutoff} \
</span><span id="__span-129-123"><a id="__codelineno-129-123" name="__codelineno-129-123" href="#__codelineno-129-123"></a>    ${suffix} \
</span><span id="__span-129-124"><a id="__codelineno-129-124" name="__codelineno-129-124" href="#__codelineno-129-124"></a>    --times ${times} \
</span><span id="__span-129-125"><a id="__codelineno-129-125" name="__codelineno-129-125" href="#__codelineno-129-125"></a>    -U ${option_u} \
</span><span id="__span-129-126"><a id="__codelineno-129-126" name="__codelineno-129-126" href="#__codelineno-129-126"></a>    -j ${option_j} \
</span><span id="__span-129-127"><a id="__codelineno-129-127" name="__codelineno-129-127" href="#__codelineno-129-127"></a>    -u ${option_u} \
</span><span id="__span-129-128"><a id="__codelineno-129-128" name="__codelineno-129-128" href="#__codelineno-129-128"></a>    ${discard} \
</span><span id="__span-129-129"><a id="__codelineno-129-129" name="__codelineno-129-129" href="#__codelineno-129-129"></a>    ${discard_untrimmed} \
</span><span id="__span-129-130"><a id="__codelineno-129-130" name="__codelineno-129-130" href="#__codelineno-129-130"></a>    ${mask_adapter} \
</span><span id="__span-129-131"><a id="__codelineno-129-131" name="__codelineno-129-131" href="#__codelineno-129-131"></a>    ${match_read_wildcards} \
</span><span id="__span-129-132"><a id="__codelineno-129-132" name="__codelineno-129-132" href="#__codelineno-129-132"></a>    ${no_indels} \
</span><span id="__span-129-133"><a id="__codelineno-129-133" name="__codelineno-129-133" href="#__codelineno-129-133"></a>    ${no_trim} \
</span><span id="__span-129-134"><a id="__codelineno-129-134" name="__codelineno-129-134" href="#__codelineno-129-134"></a>    ${trim_n} \
</span><span id="__span-129-135"><a id="__codelineno-129-135" name="__codelineno-129-135" href="#__codelineno-129-135"></a>    ${read1} \
</span><span id="__span-129-136"><a id="__codelineno-129-136" name="__codelineno-129-136" href="#__codelineno-129-136"></a>    ${read2} \
</span><span id="__span-129-137"><a id="__codelineno-129-137" name="__codelineno-129-137" href="#__codelineno-129-137"></a>    &quot;&quot;&quot;
</span><span id="__span-129-138"><a id="__codelineno-129-138" name="__codelineno-129-138" href="#__codelineno-129-138"></a>
</span><span id="__span-129-139"><a id="__codelineno-129-139" name="__codelineno-129-139" href="#__codelineno-129-139"></a>}
</span></code></pre></div></p>
<p>In this case: 
- The <code>--output</code> option is now present as <code>--output=${output_file}</code> when supplied
- The <code>--adapter</code> option is not needed as we are using <code>--info-file=${info_file}</code> to supply adapter information</p>
</blockquote>
<p><br></p>
<h3 id="error-3-hisat2-index">Error 3: HISAT2 index<a class="headerlink" href="#error-3-hisat2-index" title="Permanent link">&para;</a></h3>
<p>The third error is due to the index not being supplied correctly to the HISAT2 process.</p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-130-1"><a id="__codelineno-130-1" name="__codelineno-130-1" href="#__codelineno-130-1"></a>Error executing process &gt; &#39;HISAT2 (1)&#39;
</span><span id="__span-130-2"><a id="__codelineno-130-2" name="__codelineno-130-2" href="#__codelineno-130-2"></a>
</span><span id="__span-130-3"><a id="__codelineno-130-3" name="__codelineno-130-3" href="#__codelineno-130-3"></a>Caused by:
</span><span id="__span-130-4"><a id="__codelineno-130-4" name="__codelineno-130-4" href="#__codelineno-130-4"></a>  Process `HISAT2 (1)` terminated with an error exit status (2)
</span><span id="__span-130-5"><a id="__codelineno-130-5" name="__codelineno-130-5" href="#__codelineno-130-5"></a>
</span><span id="__span-130-6"><a id="__codelineno-130-6" name="__codelineno-130-6" href="#__codelineno-130-6"></a>Command executed:
</span><span id="__span-130-7"><a id="__codelineno-130-7" name="__codelineno-130-7" href="#__codelineno-130-7"></a>  hisat2     -U MCL1-LD-luminalpregnant.cutadapt.fastq.gz     -x /home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/hisat2_index/*.ht2     --summary-file summary.txt     -S out.sam
</span><span id="__span-130-8"><a id="__codelineno-130-8" name="__codelineno-130-8" href="#__codelineno-130-8"></a>
</span><span id="__span-130-9"><a id="__codelineno-130-9" name="__codelineno-130-9" href="#__codelineno-130-9"></a>  samtools view out.sam -o out.bam
</span><span id="__span-130-10"><a id="__codelineno-130-10" name="__codelineno-130-10" href="#__codelineno-130-10"></a>  samtools sort out.bam -o sorted.bam
</span><span id="__span-130-11"><a id="__codelineno-130-11" name="__codelineno-130-11" href="#__codelineno-130-11"></a>  samtools index sorted.bam -o sorted.bam.bai
</span><span id="__span-130-12"><a id="__codelineno-130-12" name="__codelineno-130-12" href="#__codelineno-130-12"></a>  mv sorted.bam alignment.bam 
</span><span id="__span-130-13"><a id="__codelineno-130-13" name="__codelineno-130-13" href="#__codelineno-130-13"></a>  mv sorted.bam.bai alignment.bam.bai
</span><span id="__span-130-14"><a id="__codelineno-130-14" name="__codelineno-130-14" href="#__codelineno-130-14"></a>
</span><span id="__span-130-15"><a id="__codelineno-130-15" name="__codelineno-130-15" href="#__codelineno-130-15"></a>Command error:
</span><span id="__span-130-16"><a id="__codelineno-130-16" name="__codelineno-130-16" href="#__codelineno-130-16"></a>  (ERR): &quot;/home2/training/data/sample_data/galaxy/rnaseq_reads_to_counts/hisat2_index/*.ht2&quot; does not exist
</span><span id="__span-130-17"><a id="__codelineno-130-17" name="__codelineno-130-17" href="#__codelineno-130-17"></a>  Exiting now ...
</span><span id="__span-130-18"><a id="__codelineno-130-18" name="__codelineno-130-18" href="#__codelineno-130-18"></a>
</span><span id="__span-130-19"><a id="__codelineno-130-19" name="__codelineno-130-19" href="#__codelineno-130-19"></a>Work dir:
</span><span id="__span-130-20"><a id="__codelineno-130-20" name="__codelineno-130-20" href="#__codelineno-130-20"></a>  janis-translate-examples/rnaseq_reads_to_counts/work/1b/542e03429accaf016afc55379ad951
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the Error</strong></p>
<p>The error message above tells us that hisat2 threw an error as it could not find the index to use during read alignment. </p>
<p>For those similar with hisat2, you will know that we need to supply the basename of the index rather than the full path. </p>
<p>Let&rsquo;s look at the process working directory to check how the index files present. <br>
Open the process working directory provided as the <code>Work dir:</code> section of your error. <br>
Alternatively you can execute the following command, substituting <code>&lt;work_dir&gt;</code> for the path nextflow provided:
<div class="language-text highlight"><pre><span></span><code><span id="__span-131-1"><a id="__codelineno-131-1" name="__codelineno-131-1" href="#__codelineno-131-1"></a>ls -lah &lt;work_dir&gt;
</span></code></pre></div></p>
<p>You will see the following files:
<div class="language-text highlight"><pre><span></span><code><span id="__span-132-1"><a id="__codelineno-132-1" name="__codelineno-132-1" href="#__codelineno-132-1"></a>.command.begin
</span><span id="__span-132-2"><a id="__codelineno-132-2" name="__codelineno-132-2" href="#__codelineno-132-2"></a>.command.err
</span><span id="__span-132-3"><a id="__codelineno-132-3" name="__codelineno-132-3" href="#__codelineno-132-3"></a>.command.log
</span><span id="__span-132-4"><a id="__codelineno-132-4" name="__codelineno-132-4" href="#__codelineno-132-4"></a>.command.out
</span><span id="__span-132-5"><a id="__codelineno-132-5" name="__codelineno-132-5" href="#__codelineno-132-5"></a>.command.run
</span><span id="__span-132-6"><a id="__codelineno-132-6" name="__codelineno-132-6" href="#__codelineno-132-6"></a>.command.sh
</span><span id="__span-132-7"><a id="__codelineno-132-7" name="__codelineno-132-7" href="#__codelineno-132-7"></a>.exitcode
</span><span id="__span-132-8"><a id="__codelineno-132-8" name="__codelineno-132-8" href="#__codelineno-132-8"></a>MCL1-LD-luminalpregnant.cutadapt.fastq.gz
</span></code></pre></div></p>
<p>From looking at the contents of the working directory, you will notice that the hisat2 index is missing. </p>
<p>You may remember we provided a filepath to the index using the <code>hisat2_index_path</code> param in <code>nextflow.config</code>, and this param is provided as an input value for the HISAT2 task in <code>main.nf</code> process, so why doesn&rsquo;t it appear? </p>
<p>This demonstrates an important aspect of how Nextflow treats different datatypes. </p>
<p>For <code>val</code> process inputs, nextflow is expecting a simple value such as a <em>string</em> or <em>integer</em>. <br>
It holds this as a variable which can be used within the process. </p>
<p>For <code>path</code> process inputs, nextflow knows that the input is a file or directory, so <em>stages</em> it into the process working directory for use. </p>
<p>Looking at our HISAT2 process definition, we can see that the <code>index_path</code> input is a <code>val</code> input. <br>
We haven&rsquo;t told nextflow to treat it as a file, so the file isn&rsquo;t localised into the process working directory. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue, we will change the process input type for <code>index_path</code> to a <code>path</code> type. </p>
<p>We will also add a <code>Channel</code> declaration to <code>main.nf</code> so that nextflow provides the index correctly to the <code>index_path</code> input. </p>
<p>Finally, we will update the <code>script:</code> section to specify the index basename to hisat2, rather than the full path. </p>
<p><br></p>
<p>To start, let&rsquo;s update <code>main.nf</code> with our Channel declaration. </p>
<p>Hisat2 indexes are split into multiple chunks, so we will using <code>.toList()</code> to ensure they are moved together in the workflow. </p>
<div class="admonition question">
<p class="admonition-title">Adding a Channel declaration</p>
<p>Add a new Channel declaration at the top of <code>main.nf</code> under the <code>ch_in_input_fastqs_collection</code> channel. <br>
This channel should be called <code>ch_hisat2_index</code> and should use <code>params.hisat2_index_path</code> to supply the files. <br>
Additionally, append <code>.toList()</code> so that nextflow groups each chunk of the index together. </p>
<details class="hint">
<summary>Show Change</summary>
<p>main.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-133-1"><a id="__codelineno-133-1" name="__codelineno-133-1" href="#__codelineno-133-1"></a>...
</span><span id="__span-133-2"><a id="__codelineno-133-2" name="__codelineno-133-2" href="#__codelineno-133-2"></a>
</span><span id="__span-133-3"><a id="__codelineno-133-3" name="__codelineno-133-3" href="#__codelineno-133-3"></a>// data which will be passed as channels
</span><span id="__span-133-4"><a id="__codelineno-133-4" name="__codelineno-133-4" href="#__codelineno-133-4"></a>ch_in_input_fastqs_collection  = Channel.fromPath( params.in_input_fastqs_collection ).toList()
</span><span id="__span-133-5"><a id="__codelineno-133-5" name="__codelineno-133-5" href="#__codelineno-133-5"></a>ch_hisat2_index                = Channel.fromPath( params.hisat2_index_path ).toList()              &lt;-
</span><span id="__span-133-6"><a id="__codelineno-133-6" name="__codelineno-133-6" href="#__codelineno-133-6"></a>
</span><span id="__span-133-7"><a id="__codelineno-133-7" name="__codelineno-133-7" href="#__codelineno-133-7"></a>...
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Now that we have created a channel for the hisat2 index, we will need to replace the current value supplied to the HISAT2 task. </p>
<div class="admonition question">
<p class="admonition-title">Replacing HISAT2 process call value</p>
<p>Replace the value supplied for <code>index_path</code> in the HISAT2 task call of <code>main.nf</code>. <br>
Its new value should be <code>ch_hisat2_index</code> as this is the channel we just created to feed the index.</p>
<details class="hint">
<summary>Show Change</summary>
<p>main.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-134-1"><a id="__codelineno-134-1" name="__codelineno-134-1" href="#__codelineno-134-1"></a>workflow {
</span><span id="__span-134-2"><a id="__codelineno-134-2" name="__codelineno-134-2" href="#__codelineno-134-2"></a>    ...
</span><span id="__span-134-3"><a id="__codelineno-134-3" name="__codelineno-134-3" href="#__codelineno-134-3"></a>
</span><span id="__span-134-4"><a id="__codelineno-134-4" name="__codelineno-134-4" href="#__codelineno-134-4"></a>    HISAT2(
</span><span id="__span-134-5"><a id="__codelineno-134-5" name="__codelineno-134-5" href="#__codelineno-134-5"></a>        CUTADAPT.out.out12,        // library_input_1
</span><span id="__span-134-6"><a id="__codelineno-134-6" name="__codelineno-134-6" href="#__codelineno-134-6"></a>        ch_hisat2_index,           // index_path            &lt;-
</span><span id="__span-134-7"><a id="__codelineno-134-7" name="__codelineno-134-7" href="#__codelineno-134-7"></a>    )
</span><span id="__span-134-8"><a id="__codelineno-134-8" name="__codelineno-134-8" href="#__codelineno-134-8"></a>
</span><span id="__span-134-9"><a id="__codelineno-134-9" name="__codelineno-134-9" href="#__codelineno-134-9"></a>    ...
</span><span id="__span-134-10"><a id="__codelineno-134-10" name="__codelineno-134-10" href="#__codelineno-134-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Finally we will update the <code>index_path</code> input of HISAT2 to a <code>path</code> type, and update the <code>script:</code> section of HISAT2 so that we supply the basename of the index to HISAT2.  </p>
<p>The hisat2 index consists of multiple files in a list. <br>
As each chunk starts with the same basename we will use the first chunk basename with the <code>-x</code> argument. </p>
<div class="admonition question">
<p class="admonition-title">Modifying HISAT2 inputs and script</p>
<p>Update the type of the <code>index_path</code> HISAT2 process input from <code>val</code> to <code>path</code>.</p>
<p>Update the script section of HISAT2 to correctly supply the <code>-x</code> argument.<br>
The value should be the basename of <strong><em>the first item</em></strong> in the <code>index_path</code> input.<br></p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-135-1"><a id="__codelineno-135-1" name="__codelineno-135-1" href="#__codelineno-135-1"></a>process HISAT2 {
</span><span id="__span-135-2"><a id="__codelineno-135-2" name="__codelineno-135-2" href="#__codelineno-135-2"></a>
</span><span id="__span-135-3"><a id="__codelineno-135-3" name="__codelineno-135-3" href="#__codelineno-135-3"></a>    container &quot;quay.io/biocontainers/janis-translate-hisat2-2.2.1&quot;
</span><span id="__span-135-4"><a id="__codelineno-135-4" name="__codelineno-135-4" href="#__codelineno-135-4"></a>    publishDir &quot;${params.outdir}/hisat2&quot;
</span><span id="__span-135-5"><a id="__codelineno-135-5" name="__codelineno-135-5" href="#__codelineno-135-5"></a>
</span><span id="__span-135-6"><a id="__codelineno-135-6" name="__codelineno-135-6" href="#__codelineno-135-6"></a>    input:
</span><span id="__span-135-7"><a id="__codelineno-135-7" name="__codelineno-135-7" href="#__codelineno-135-7"></a>    path library_input_1
</span><span id="__span-135-8"><a id="__codelineno-135-8" name="__codelineno-135-8" href="#__codelineno-135-8"></a>    path index_path                                                         &lt;-
</span><span id="__span-135-9"><a id="__codelineno-135-9" name="__codelineno-135-9" href="#__codelineno-135-9"></a>
</span><span id="__span-135-10"><a id="__codelineno-135-10" name="__codelineno-135-10" href="#__codelineno-135-10"></a>    output:
</span><span id="__span-135-11"><a id="__codelineno-135-11" name="__codelineno-135-11" href="#__codelineno-135-11"></a>    path &quot;summary.txt&quot;, emit: out_summary_file
</span><span id="__span-135-12"><a id="__codelineno-135-12" name="__codelineno-135-12" href="#__codelineno-135-12"></a>    path &quot;alignment.bam&quot;, emit: output_alignments   
</span><span id="__span-135-13"><a id="__codelineno-135-13" name="__codelineno-135-13" href="#__codelineno-135-13"></a>
</span><span id="__span-135-14"><a id="__codelineno-135-14" name="__codelineno-135-14" href="#__codelineno-135-14"></a>    script:
</span><span id="__span-135-15"><a id="__codelineno-135-15" name="__codelineno-135-15" href="#__codelineno-135-15"></a>    &quot;&quot;&quot;
</span><span id="__span-135-16"><a id="__codelineno-135-16" name="__codelineno-135-16" href="#__codelineno-135-16"></a>    hisat2 \
</span><span id="__span-135-17"><a id="__codelineno-135-17" name="__codelineno-135-17" href="#__codelineno-135-17"></a>    -U ${library_input_1} \
</span><span id="__span-135-18"><a id="__codelineno-135-18" name="__codelineno-135-18" href="#__codelineno-135-18"></a>    -x ${index_path[0].simpleName} \                                        &lt;-
</span><span id="__span-135-19"><a id="__codelineno-135-19" name="__codelineno-135-19" href="#__codelineno-135-19"></a>    --summary-file ${library_input_1.simpleName}.alignment_summary.txt \
</span><span id="__span-135-20"><a id="__codelineno-135-20" name="__codelineno-135-20" href="#__codelineno-135-20"></a>    -S out.sam
</span><span id="__span-135-21"><a id="__codelineno-135-21" name="__codelineno-135-21" href="#__codelineno-135-21"></a>
</span><span id="__span-135-22"><a id="__codelineno-135-22" name="__codelineno-135-22" href="#__codelineno-135-22"></a>    samtools view out.sam -o out.bam
</span><span id="__span-135-23"><a id="__codelineno-135-23" name="__codelineno-135-23" href="#__codelineno-135-23"></a>    samtools sort out.bam -o sorted.bam
</span><span id="__span-135-24"><a id="__codelineno-135-24" name="__codelineno-135-24" href="#__codelineno-135-24"></a>    samtools index sorted.bam -o sorted.bam.bai
</span><span id="__span-135-25"><a id="__codelineno-135-25" name="__codelineno-135-25" href="#__codelineno-135-25"></a>    mv sorted.bam ${library_input_1.simpleName}.bam 
</span><span id="__span-135-26"><a id="__codelineno-135-26" name="__codelineno-135-26" href="#__codelineno-135-26"></a>    mv sorted.bam.bai ${library_input_1.simpleName}.bam.bai
</span><span id="__span-135-27"><a id="__codelineno-135-27" name="__codelineno-135-27" href="#__codelineno-135-27"></a>    &quot;&quot;&quot;
</span><span id="__span-135-28"><a id="__codelineno-135-28" name="__codelineno-135-28" href="#__codelineno-135-28"></a>
</span><span id="__span-135-29"><a id="__codelineno-135-29" name="__codelineno-135-29" href="#__codelineno-135-29"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After making these changes, re-run the workflow:
<div class="language-text highlight"><pre><span></span><code><span id="__span-136-1"><a id="__codelineno-136-1" name="__codelineno-136-1" href="#__codelineno-136-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p><br></p>
<h3 id="error-4-featurecounts-command">Error 4: FEATURECOUNTS command<a class="headerlink" href="#error-4-featurecounts-command" title="Permanent link">&para;</a></h3>
<p>The 4<sup>th</sup> error is due to incorrect formatting in the FEATURECOUNTS process script. </p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-137-1"><a id="__codelineno-137-1" name="__codelineno-137-1" href="#__codelineno-137-1"></a>Error executing process &gt; &#39;FEATURECOUNTS (1)&#39;
</span><span id="__span-137-2"><a id="__codelineno-137-2" name="__codelineno-137-2" href="#__codelineno-137-2"></a>
</span><span id="__span-137-3"><a id="__codelineno-137-3" name="__codelineno-137-3" href="#__codelineno-137-3"></a>Caused by:
</span><span id="__span-137-4"><a id="__codelineno-137-4" name="__codelineno-137-4" href="#__codelineno-137-4"></a>  Process `FEATURECOUNTS (1)` terminated with an error exit status (255)
</span><span id="__span-137-5"><a id="__codelineno-137-5" name="__codelineno-137-5" href="#__codelineno-137-5"></a>
</span><span id="__span-137-6"><a id="__codelineno-137-6" name="__codelineno-137-6" href="#__codelineno-137-6"></a>Command executed:
</span><span id="__span-137-7"><a id="__codelineno-137-7" name="__codelineno-137-7" href="#__codelineno-137-7"></a>  featureCounts     alignment.bam
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the Error</strong></p>
<p>By looking at the <code>Command error:</code> section of your nextflow error, you will see that the featureCounts helptext has been printed. </p>
<p>This indicates that the error was triggered by featureCounts, and that the reason was our command wasn&rsquo;t formulated correctly. </p>
<p>Nextflow will truncate the <code>Command output:</code> and <code>Command error:</code> sections of its error messages as these could be very long depending on the tool and the error encountered. </p>
<p>To view the full <code>stderr</code> stream of an executed process, we can navigate to the process working directory and view the <code>.command.err</code> file. </p>
<p><br></p>
<p>Open <code>.command.err</code> for the FEATURECOUNTS process which caused the error using the path nextflow provides in the <code>Work dir:</code> section of its error message. </p>
<p>At the top of <code>.command.err</code> you will see that featureCounts tells us the correct way to format a command:<br>
<div class="language-text highlight"><pre><span></span><code><span id="__span-138-1"><a id="__codelineno-138-1" name="__codelineno-138-1" href="#__codelineno-138-1"></a>Usage: featureCounts [options] -a &lt;annotation_file&gt; -o &lt;output_file&gt; input_file1 [input_file2] ... 
</span></code></pre></div></p>
<p>Below this, we see its extended options including descriptions of each. </p>
<p>For our use of featureCounts we will use the inbuilt <code>mm10</code> bed file for genomic feature annotations, and will provide an output filename to collect. </p>
<p>From reading the helptext, we can deduce that the correct way structure our command is as follows:
<div class="language-text highlight"><pre><span></span><code><span id="__span-139-1"><a id="__codelineno-139-1" name="__codelineno-139-1" href="#__codelineno-139-1"></a>featureCounts -a &lt;annotations&gt; -F &quot;SAF&quot; -o &lt;outfile&gt; &lt;infile&gt;
</span></code></pre></div></p>
<p>The <code>-F "SAF"</code> part of the command is to tell featureCounts that the annotations file is in SAF format. The inbuilt featureCounts annotation files are all in SAF format, so we will need to include this as we are using the inbuilt annotations. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue we will restructure the FEATURECOUNTS <code>script:</code> section, and will change the output collection patterns. </p>
<p>First, let&rsquo;s modify the FEATURECOUNTS script section. </p>
<div class="admonition question">
<p class="admonition-title">Modifying FEATURECOUNTS script</p>
<p>Add the <code>-a</code> argument to your command in the script section. <br>
Its value should be <code>/usr/local/annotation/mm10_RefSeq_exon.txt</code> as this is where the required mm10 annotations file exists inside the container image we are using. <br>
Add the <code>-F</code> argument to your command and supply its value as <code>"SAF"</code> <br>
Add the <code>-o</code> argument to your command and supply a <code>.txt</code> filename as its value based on the basename of the <code>alignment</code> process input. <br></p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/featurecounts.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-140-1"><a id="__codelineno-140-1" name="__codelineno-140-1" href="#__codelineno-140-1"></a>process FEATURECOUNTS {
</span><span id="__span-140-2"><a id="__codelineno-140-2" name="__codelineno-140-2" href="#__codelineno-140-2"></a>    ...
</span><span id="__span-140-3"><a id="__codelineno-140-3" name="__codelineno-140-3" href="#__codelineno-140-3"></a>
</span><span id="__span-140-4"><a id="__codelineno-140-4" name="__codelineno-140-4" href="#__codelineno-140-4"></a>    script:
</span><span id="__span-140-5"><a id="__codelineno-140-5" name="__codelineno-140-5" href="#__codelineno-140-5"></a>    &quot;&quot;&quot;
</span><span id="__span-140-6"><a id="__codelineno-140-6" name="__codelineno-140-6" href="#__codelineno-140-6"></a>    featureCounts \
</span><span id="__span-140-7"><a id="__codelineno-140-7" name="__codelineno-140-7" href="#__codelineno-140-7"></a>    -a /usr/local/annotation/mm10_RefSeq_exon.txt \         &lt;-
</span><span id="__span-140-8"><a id="__codelineno-140-8" name="__codelineno-140-8" href="#__codelineno-140-8"></a>    -F &quot;SAF&quot; \                                              &lt;-
</span><span id="__span-140-9"><a id="__codelineno-140-9" name="__codelineno-140-9" href="#__codelineno-140-9"></a>    -o ${alignment.simpleName}.txt \                        &lt;-
</span><span id="__span-140-10"><a id="__codelineno-140-10" name="__codelineno-140-10" href="#__codelineno-140-10"></a>    ${alignment} 
</span><span id="__span-140-11"><a id="__codelineno-140-11" name="__codelineno-140-11" href="#__codelineno-140-11"></a>    &quot;&quot;&quot;
</span><span id="__span-140-12"><a id="__codelineno-140-12" name="__codelineno-140-12" href="#__codelineno-140-12"></a>
</span><span id="__span-140-13"><a id="__codelineno-140-13" name="__codelineno-140-13" href="#__codelineno-140-13"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Now that the script has the correct structure, let&rsquo;s modify the output collection patterns to capture the files we need. </p>
<div class="admonition question">
<p class="admonition-title">Modifying FEATURECOUNTS output collection</p>
<p>Modify the <code>output_short</code> and <code>output_summary</code> outputs so they capture the target files.<br>
The <code>output_short</code> output should capture whatever you supplied to the <code>-o</code> argument.<br>
The <code>output_summary</code> output should capture whatever you supplied to the <code>-o</code> argument, with <code>.summary</code> appended on the end.</p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/featurecounts.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-141-1"><a id="__codelineno-141-1" name="__codelineno-141-1" href="#__codelineno-141-1"></a>process FEATURECOUNTS {
</span><span id="__span-141-2"><a id="__codelineno-141-2" name="__codelineno-141-2" href="#__codelineno-141-2"></a>    ...
</span><span id="__span-141-3"><a id="__codelineno-141-3" name="__codelineno-141-3" href="#__codelineno-141-3"></a>
</span><span id="__span-141-4"><a id="__codelineno-141-4" name="__codelineno-141-4" href="#__codelineno-141-4"></a>    output:
</span><span id="__span-141-5"><a id="__codelineno-141-5" name="__codelineno-141-5" href="#__codelineno-141-5"></a>    path &quot;${alignment.simpleName}.txt&quot;, emit: output_short              &lt;-
</span><span id="__span-141-6"><a id="__codelineno-141-6" name="__codelineno-141-6" href="#__codelineno-141-6"></a>    path &quot;${alignment.simpleName}.txt.summary&quot;, emit: output_summary    &lt;-
</span><span id="__span-141-7"><a id="__codelineno-141-7" name="__codelineno-141-7" href="#__codelineno-141-7"></a>
</span><span id="__span-141-8"><a id="__codelineno-141-8" name="__codelineno-141-8" href="#__codelineno-141-8"></a>    ...
</span><span id="__span-141-9"><a id="__codelineno-141-9" name="__codelineno-141-9" href="#__codelineno-141-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After those changes have been made, re-run the workflow. </p>
<p>This time we will add <code>-resume</code> so that previous tasks before FEATURECOUNTS aren&rsquo;t executed again. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-142-1"><a id="__codelineno-142-1" name="__codelineno-142-1" href="#__codelineno-142-1"></a>nextflow run main.nf -resume
</span></code></pre></div></p>
<p><br></p>
<h3 id="error-5-picard_mark_duplicates-command">Error 5: PICARD_MARK_DUPLICATES command<a class="headerlink" href="#error-5-picard_mark_duplicates-command" title="Permanent link">&para;</a></h3>
<p>The 5<sup>th</sup> error is due to incorrect formatting in the PICARD_MARK_DUPLICATES process script. </p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-143-1"><a id="__codelineno-143-1" name="__codelineno-143-1" href="#__codelineno-143-1"></a>Error executing process &gt; &#39;PICARD_MARK_DUPLICATES (2)&#39;
</span><span id="__span-143-2"><a id="__codelineno-143-2" name="__codelineno-143-2" href="#__codelineno-143-2"></a>
</span><span id="__span-143-3"><a id="__codelineno-143-3" name="__codelineno-143-3" href="#__codelineno-143-3"></a>Caused by:
</span><span id="__span-143-4"><a id="__codelineno-143-4" name="__codelineno-143-4" href="#__codelineno-143-4"></a>  Process `PICARD_MARK_DUPLICATES (2)` terminated with an error exit status (1)
</span><span id="__span-143-5"><a id="__codelineno-143-5" name="__codelineno-143-5" href="#__codelineno-143-5"></a>
</span><span id="__span-143-6"><a id="__codelineno-143-6" name="__codelineno-143-6" href="#__codelineno-143-6"></a>Command executed:
</span><span id="__span-143-7"><a id="__codelineno-143-7" name="__codelineno-143-7" href="#__codelineno-143-7"></a>  picard MarkDuplicates     INPUT=alignment.bam
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the Error</strong></p>
<p>This error is identical to the FEATURECOUNTS error. </p>
<p>You will see that the helptext is printed in the <code>Command error:</code> section of your nextflow error, again indicating that the command was not correct. </p>
<p>By viewing <code>.command.err</code> in the process working directory for the errored task, we see the following at the top of the file:
<div class="language-text highlight"><pre><span></span><code><span id="__span-144-1"><a id="__codelineno-144-1" name="__codelineno-144-1" href="#__codelineno-144-1"></a>ERROR: Option &#39;OUTPUT&#39; is required.
</span></code></pre></div></p>
<p>Further down, we see the following:
<div class="language-text highlight"><pre><span></span><code><span id="__span-145-1"><a id="__codelineno-145-1" name="__codelineno-145-1" href="#__codelineno-145-1"></a>Usage example:
</span><span id="__span-145-2"><a id="__codelineno-145-2" name="__codelineno-145-2" href="#__codelineno-145-2"></a>
</span><span id="__span-145-3"><a id="__codelineno-145-3" name="__codelineno-145-3" href="#__codelineno-145-3"></a>java -jar picard.jar MarkDuplicates \
</span><span id="__span-145-4"><a id="__codelineno-145-4" name="__codelineno-145-4" href="#__codelineno-145-4"></a>I=input.bam \
</span><span id="__span-145-5"><a id="__codelineno-145-5" name="__codelineno-145-5" href="#__codelineno-145-5"></a>O=marked_duplicates.bam \
</span><span id="__span-145-6"><a id="__codelineno-145-6" name="__codelineno-145-6" href="#__codelineno-145-6"></a>M=marked_dup_metrics.txt
</span></code></pre></div></p>
<p>This indicates that we need to add the <code>OUTPUT=</code> and <code>METRICS_FILE=</code> arguments to our command. We can also use the shorthand notation for our arguments if desired as seen in the usage example (<code>I=</code>, <code>O=</code>, <code>M=</code>). <br>
For this tutorial we will stick to the longform notation. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue, we will supply the <code>OUTPUT=</code> and <code>METRICS_FILE=</code> arguments to our command, and will adjust the output collection patterns to collect these files.</p>
<p>First, we will add these extra arguments to the PICARD_MARK_DUPLICATES script section.</p>
<div class="admonition question">
<p class="admonition-title">Adding arguments to PICARD_MARK_DUPLICATES script</p>
<p>Add the <code>OUTPUT=</code> argument to the script section of PICARD_MARK_DUPLICATES. <br>
Its value should be a filename derived using the basename of the <code>input_file</code> process input, and should have the <code>.bam</code> extension.</p>
<p>Add the <code>METRICS_FILE=</code> argument to the script section of PICARD_MARK_DUPLICATES. <br>
Its value should be a filename derived using the basename of the <code>input_file</code> process input, and should have the <code>.txt</code> extension.</p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/picard_mark_duplicates.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-146-1"><a id="__codelineno-146-1" name="__codelineno-146-1" href="#__codelineno-146-1"></a>process PICARD_MARK_DUPLICATES {
</span><span id="__span-146-2"><a id="__codelineno-146-2" name="__codelineno-146-2" href="#__codelineno-146-2"></a>    ...
</span><span id="__span-146-3"><a id="__codelineno-146-3" name="__codelineno-146-3" href="#__codelineno-146-3"></a>
</span><span id="__span-146-4"><a id="__codelineno-146-4" name="__codelineno-146-4" href="#__codelineno-146-4"></a>    script:
</span><span id="__span-146-5"><a id="__codelineno-146-5" name="__codelineno-146-5" href="#__codelineno-146-5"></a>    &quot;&quot;&quot;
</span><span id="__span-146-6"><a id="__codelineno-146-6" name="__codelineno-146-6" href="#__codelineno-146-6"></a>    picard MarkDuplicates \
</span><span id="__span-146-7"><a id="__codelineno-146-7" name="__codelineno-146-7" href="#__codelineno-146-7"></a>    INPUT=${input_file} \
</span><span id="__span-146-8"><a id="__codelineno-146-8" name="__codelineno-146-8" href="#__codelineno-146-8"></a>    OUTPUT=${input_file.simpleName}.markdup.bam \       &lt;-
</span><span id="__span-146-9"><a id="__codelineno-146-9" name="__codelineno-146-9" href="#__codelineno-146-9"></a>    METRICS_FILE=${input_file.simpleName}.metrics.txt   &lt;-
</span><span id="__span-146-10"><a id="__codelineno-146-10" name="__codelineno-146-10" href="#__codelineno-146-10"></a>    &quot;&quot;&quot;
</span><span id="__span-146-11"><a id="__codelineno-146-11" name="__codelineno-146-11" href="#__codelineno-146-11"></a>
</span><span id="__span-146-12"><a id="__codelineno-146-12" name="__codelineno-146-12" href="#__codelineno-146-12"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>Now that we have added the required arguments, let&rsquo;s modify the output collection patterns. </p>
<div class="admonition question">
<p class="admonition-title">Adding arguments to PICARD_MARK_DUPLICATES script</p>
<p>Modify the <code>outFile2</code> output collection pattern to collect the file you specified using the <code>OUTPUT=</code> argument. </p>
<p>Modify the <code>out_metrics_file</code> output collection pattern to collect the file you specified using the <code>METRICS_FILE=</code> argument. </p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/picard_mark_duplicates.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-147-1"><a id="__codelineno-147-1" name="__codelineno-147-1" href="#__codelineno-147-1"></a>process PICARD_MARK_DUPLICATES {
</span><span id="__span-147-2"><a id="__codelineno-147-2" name="__codelineno-147-2" href="#__codelineno-147-2"></a>    ...
</span><span id="__span-147-3"><a id="__codelineno-147-3" name="__codelineno-147-3" href="#__codelineno-147-3"></a>
</span><span id="__span-147-4"><a id="__codelineno-147-4" name="__codelineno-147-4" href="#__codelineno-147-4"></a>    output: 
</span><span id="__span-147-5"><a id="__codelineno-147-5" name="__codelineno-147-5" href="#__codelineno-147-5"></a>    path &quot;${input_file.simpleName}.markdup.bam&quot;, emit: outFile2             &lt;-
</span><span id="__span-147-6"><a id="__codelineno-147-6" name="__codelineno-147-6" href="#__codelineno-147-6"></a>    path &quot;${input_file.simpleName}.metrics.txt&quot;, emit: out_metrics_file     &lt;-
</span><span id="__span-147-7"><a id="__codelineno-147-7" name="__codelineno-147-7" href="#__codelineno-147-7"></a>
</span><span id="__span-147-8"><a id="__codelineno-147-8" name="__codelineno-147-8" href="#__codelineno-147-8"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After making these changes, re-run the workflow.
<div class="language-text highlight"><pre><span></span><code><span id="__span-148-1"><a id="__codelineno-148-1" name="__codelineno-148-1" href="#__codelineno-148-1"></a>nextflow run main.nf -resume
</span></code></pre></div></p>
<p><br></p>
<h3 id="error-6-multiqc-input-filename-collision">Error 6: MULTIQC input filename collision<a class="headerlink" href="#error-6-multiqc-input-filename-collision" title="Permanent link">&para;</a></h3>
<p>The 6<sup>th</sup> error is due to multiple input files possessing the same filename. </p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-149-1"><a id="__codelineno-149-1" name="__codelineno-149-1" href="#__codelineno-149-1"></a>Error executing process &gt; &#39;MULTIQC&#39;
</span><span id="__span-149-2"><a id="__codelineno-149-2" name="__codelineno-149-2" href="#__codelineno-149-2"></a>
</span><span id="__span-149-3"><a id="__codelineno-149-3" name="__codelineno-149-3" href="#__codelineno-149-3"></a>Caused by:
</span><span id="__span-149-4"><a id="__codelineno-149-4" name="__codelineno-149-4" href="#__codelineno-149-4"></a>  Process `MULTIQC` input file name collision -- There are multiple input files for each of the following file names: alignment.txt.summary, alignment.metrics.txt, outputTextfile.txt, outputTabular.tabular, summary.txt, output.geneBodyCoverage.txt
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the error</strong></p>
<p>The cause of this error is related to how we create &amp; capture output files in previous tasks. 
From the error message above, nextflow is telling us that multiple input files had the same filename. </p>
<p>Nextflow doesn&rsquo;t allow this behaviour, which is probably a good thing. <br>
Nextflow is forcing us ensure each <code>path</code> input provided to a process has a unique name. <br>
This is ensuring we use good practises when writing our code, so that unexpected <em>silent</em> errors don&rsquo;t occur. </p>
<p>If avoidable, we should never use hard-coded filenames for output files. 
Instead, we should generate output filenames based on input filenames.
Each output filename will be a variant of the input filename, enabling us to track each sample when running a batch workflow (like the one used in this tutorial).</p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue, we will check the outputs of previous tasks to ensure they are based on the input filename. <br>
For any that are hard-coded, we will modify them to be generated based on an input to that process. </p>
<div class="admonition question">
<p class="admonition-title">Adding arguments to PICARD_MARK_DUPLICATES script</p>
<p>Make a list of the processes which are using hard-coded output filenames. <br>
Do this by looking at each process definition in the <code>modules/</code> folder. </p>
<ul>
<li>Ignore <code>collection_column_join.nf</code> as this process is unused. <br></li>
<li>Ignore <code>multiqc.nf</code> as this process collates the results of other tasks, so a hard-coded filename is ok for this tutorial. </li>
</ul>
<details class="hint">
<summary>Show list of processes which have hard-coded output filenames</summary>
<ul>
<li>HISAT2</li>
<li>RSEQC_GENE_BODY_COVERAGE</li>
<li>RSEQC_INFER_EXPERIMENT</li>
<li>RSEQC_READ_DISTRIBUTION</li>
<li>SAMTOOLS_IDXSTATS</li>
</ul>
</details>
</div>
<p><br></p>
<blockquote>
<p>NOTE:
For a production workflow we would create a new param eg. <code>batch_name</code> and pass this to MULTIQC via a new <code>val</code> input.<br>
We would then supply this input in the MULTIQC script to create a unique name for the multiqc report.<br>
In the essence of brevity, we are leaving this of the tutorial as it is already long. </p>
</blockquote>
<p><br></p>
<p>For each of the processes in your list, we will modify the script section and output file collection to ensure unique output filenames. </p>
<div class="admonition question">
<p class="admonition-title">Modifying affected processes</p>
<p>Modify the processes in your list so that output filenames are based on input filenames and are unique. <br>
You can use <code>.simpleName</code> when referencing an input file to get the basename without its extension. </p>
<details class="hint">
<summary>Show Changes</summary>
<p>hisat2.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-150-1"><a id="__codelineno-150-1" name="__codelineno-150-1" href="#__codelineno-150-1"></a>process HISAT2 {
</span><span id="__span-150-2"><a id="__codelineno-150-2" name="__codelineno-150-2" href="#__codelineno-150-2"></a>
</span><span id="__span-150-3"><a id="__codelineno-150-3" name="__codelineno-150-3" href="#__codelineno-150-3"></a>    container &quot;quay.io/biocontainers/janis-translate-hisat2-2.2.1&quot;
</span><span id="__span-150-4"><a id="__codelineno-150-4" name="__codelineno-150-4" href="#__codelineno-150-4"></a>    publishDir &quot;${params.outdir}/hisat2&quot;
</span><span id="__span-150-5"><a id="__codelineno-150-5" name="__codelineno-150-5" href="#__codelineno-150-5"></a>
</span><span id="__span-150-6"><a id="__codelineno-150-6" name="__codelineno-150-6" href="#__codelineno-150-6"></a>    input:
</span><span id="__span-150-7"><a id="__codelineno-150-7" name="__codelineno-150-7" href="#__codelineno-150-7"></a>    path library_input_1
</span><span id="__span-150-8"><a id="__codelineno-150-8" name="__codelineno-150-8" href="#__codelineno-150-8"></a>    path index_path
</span><span id="__span-150-9"><a id="__codelineno-150-9" name="__codelineno-150-9" href="#__codelineno-150-9"></a>
</span><span id="__span-150-10"><a id="__codelineno-150-10" name="__codelineno-150-10" href="#__codelineno-150-10"></a>    output:
</span><span id="__span-150-11"><a id="__codelineno-150-11" name="__codelineno-150-11" href="#__codelineno-150-11"></a>    path &quot;${library_input_1.simpleName}.alignment_summary.txt&quot;, emit: out_summary_file  &lt;-
</span><span id="__span-150-12"><a id="__codelineno-150-12" name="__codelineno-150-12" href="#__codelineno-150-12"></a>    path &quot;${library_input_1.simpleName}.bam&quot;, emit: output_alignments                   &lt;-
</span><span id="__span-150-13"><a id="__codelineno-150-13" name="__codelineno-150-13" href="#__codelineno-150-13"></a>
</span><span id="__span-150-14"><a id="__codelineno-150-14" name="__codelineno-150-14" href="#__codelineno-150-14"></a>    script:
</span><span id="__span-150-15"><a id="__codelineno-150-15" name="__codelineno-150-15" href="#__codelineno-150-15"></a>    &quot;&quot;&quot;
</span><span id="__span-150-16"><a id="__codelineno-150-16" name="__codelineno-150-16" href="#__codelineno-150-16"></a>    hisat2 \
</span><span id="__span-150-17"><a id="__codelineno-150-17" name="__codelineno-150-17" href="#__codelineno-150-17"></a>    -U ${library_input_1} \
</span><span id="__span-150-18"><a id="__codelineno-150-18" name="__codelineno-150-18" href="#__codelineno-150-18"></a>    -x ${index_path[0].simpleName} \
</span><span id="__span-150-19"><a id="__codelineno-150-19" name="__codelineno-150-19" href="#__codelineno-150-19"></a>    --summary-file ${library_input_1.simpleName}.alignment_summary.txt \               &lt;-
</span><span id="__span-150-20"><a id="__codelineno-150-20" name="__codelineno-150-20" href="#__codelineno-150-20"></a>    -S out.sam
</span><span id="__span-150-21"><a id="__codelineno-150-21" name="__codelineno-150-21" href="#__codelineno-150-21"></a>
</span><span id="__span-150-22"><a id="__codelineno-150-22" name="__codelineno-150-22" href="#__codelineno-150-22"></a>    samtools view out.sam -o out.bam
</span><span id="__span-150-23"><a id="__codelineno-150-23" name="__codelineno-150-23" href="#__codelineno-150-23"></a>    samtools sort out.bam -o sorted.bam
</span><span id="__span-150-24"><a id="__codelineno-150-24" name="__codelineno-150-24" href="#__codelineno-150-24"></a>    samtools index sorted.bam -o sorted.bam.bai
</span><span id="__span-150-25"><a id="__codelineno-150-25" name="__codelineno-150-25" href="#__codelineno-150-25"></a>    mv sorted.bam ${library_input_1.simpleName}.bam                                    &lt;-
</span><span id="__span-150-26"><a id="__codelineno-150-26" name="__codelineno-150-26" href="#__codelineno-150-26"></a>    mv sorted.bam.bai ${library_input_1.simpleName}.bam.bai                            &lt;-
</span><span id="__span-150-27"><a id="__codelineno-150-27" name="__codelineno-150-27" href="#__codelineno-150-27"></a>    &quot;&quot;&quot;
</span><span id="__span-150-28"><a id="__codelineno-150-28" name="__codelineno-150-28" href="#__codelineno-150-28"></a>
</span><span id="__span-150-29"><a id="__codelineno-150-29" name="__codelineno-150-29" href="#__codelineno-150-29"></a>}
</span></code></pre></div></p>
<p>rseqc_gene_body_coverage.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-151-1"><a id="__codelineno-151-1" name="__codelineno-151-1" href="#__codelineno-151-1"></a>process RSEQC_GENE_BODY_COVERAGE {
</span><span id="__span-151-2"><a id="__codelineno-151-2" name="__codelineno-151-2" href="#__codelineno-151-2"></a>
</span><span id="__span-151-3"><a id="__codelineno-151-3" name="__codelineno-151-3" href="#__codelineno-151-3"></a>    container &quot;quay.io/biocontainers/rseqc:2.6.4--py27hf8a1672_2&quot;
</span><span id="__span-151-4"><a id="__codelineno-151-4" name="__codelineno-151-4" href="#__codelineno-151-4"></a>    publishDir &quot;${params.outdir}/rseqc_gene_body_coverage&quot;
</span><span id="__span-151-5"><a id="__codelineno-151-5" name="__codelineno-151-5" href="#__codelineno-151-5"></a>
</span><span id="__span-151-6"><a id="__codelineno-151-6" name="__codelineno-151-6" href="#__codelineno-151-6"></a>    input:
</span><span id="__span-151-7"><a id="__codelineno-151-7" name="__codelineno-151-7" href="#__codelineno-151-7"></a>    path batch_mode_input
</span><span id="__span-151-8"><a id="__codelineno-151-8" name="__codelineno-151-8" href="#__codelineno-151-8"></a>    path refgene
</span><span id="__span-151-9"><a id="__codelineno-151-9" name="__codelineno-151-9" href="#__codelineno-151-9"></a>
</span><span id="__span-151-10"><a id="__codelineno-151-10" name="__codelineno-151-10" href="#__codelineno-151-10"></a>    output:
</span><span id="__span-151-11"><a id="__codelineno-151-11" name="__codelineno-151-11" href="#__codelineno-151-11"></a>    path &quot;${batch_mode_input.simpleName}.geneBodyCoverage.curves.pdf&quot;, emit: outputcurvespdf   &lt;-
</span><span id="__span-151-12"><a id="__codelineno-151-12" name="__codelineno-151-12" href="#__codelineno-151-12"></a>    path &quot;${batch_mode_input.simpleName}.geneBodyCoverage.txt&quot;, emit: outputtxt                &lt;-
</span><span id="__span-151-13"><a id="__codelineno-151-13" name="__codelineno-151-13" href="#__codelineno-151-13"></a>
</span><span id="__span-151-14"><a id="__codelineno-151-14" name="__codelineno-151-14" href="#__codelineno-151-14"></a>    script:
</span><span id="__span-151-15"><a id="__codelineno-151-15" name="__codelineno-151-15" href="#__codelineno-151-15"></a>    &quot;&quot;&quot;
</span><span id="__span-151-16"><a id="__codelineno-151-16" name="__codelineno-151-16" href="#__codelineno-151-16"></a>    geneBody_coverage.py \
</span><span id="__span-151-17"><a id="__codelineno-151-17" name="__codelineno-151-17" href="#__codelineno-151-17"></a>    -r ${refgene} \
</span><span id="__span-151-18"><a id="__codelineno-151-18" name="__codelineno-151-18" href="#__codelineno-151-18"></a>    -i ${batch_mode_input} \
</span><span id="__span-151-19"><a id="__codelineno-151-19" name="__codelineno-151-19" href="#__codelineno-151-19"></a>    -o ${batch_mode_input.simpleName}                                                          &lt;-
</span><span id="__span-151-20"><a id="__codelineno-151-20" name="__codelineno-151-20" href="#__codelineno-151-20"></a>    &quot;&quot;&quot;
</span><span id="__span-151-21"><a id="__codelineno-151-21" name="__codelineno-151-21" href="#__codelineno-151-21"></a>
</span><span id="__span-151-22"><a id="__codelineno-151-22" name="__codelineno-151-22" href="#__codelineno-151-22"></a>}
</span></code></pre></div></p>
<p>rseqc_infer_experiment.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-152-1"><a id="__codelineno-152-1" name="__codelineno-152-1" href="#__codelineno-152-1"></a>process RSEQC_INFER_EXPERIMENT {
</span><span id="__span-152-2"><a id="__codelineno-152-2" name="__codelineno-152-2" href="#__codelineno-152-2"></a>
</span><span id="__span-152-3"><a id="__codelineno-152-3" name="__codelineno-152-3" href="#__codelineno-152-3"></a>    container &quot;quay.io/biocontainers/rseqc:2.6.4--py27hf8a1672_2&quot;
</span><span id="__span-152-4"><a id="__codelineno-152-4" name="__codelineno-152-4" href="#__codelineno-152-4"></a>    publishDir &quot;${params.outdir}/rseqc_infer_experiment&quot;
</span><span id="__span-152-5"><a id="__codelineno-152-5" name="__codelineno-152-5" href="#__codelineno-152-5"></a>
</span><span id="__span-152-6"><a id="__codelineno-152-6" name="__codelineno-152-6" href="#__codelineno-152-6"></a>    input:
</span><span id="__span-152-7"><a id="__codelineno-152-7" name="__codelineno-152-7" href="#__codelineno-152-7"></a>    path input_bam
</span><span id="__span-152-8"><a id="__codelineno-152-8" name="__codelineno-152-8" href="#__codelineno-152-8"></a>    path refgene
</span><span id="__span-152-9"><a id="__codelineno-152-9" name="__codelineno-152-9" href="#__codelineno-152-9"></a>
</span><span id="__span-152-10"><a id="__codelineno-152-10" name="__codelineno-152-10" href="#__codelineno-152-10"></a>    output:
</span><span id="__span-152-11"><a id="__codelineno-152-11" name="__codelineno-152-11" href="#__codelineno-152-11"></a>    path &quot;${input_bam.simpleName}.infer_experiment.txt&quot;, emit: outputTextfile  &lt;-
</span><span id="__span-152-12"><a id="__codelineno-152-12" name="__codelineno-152-12" href="#__codelineno-152-12"></a>
</span><span id="__span-152-13"><a id="__codelineno-152-13" name="__codelineno-152-13" href="#__codelineno-152-13"></a>    script:
</span><span id="__span-152-14"><a id="__codelineno-152-14" name="__codelineno-152-14" href="#__codelineno-152-14"></a>    &quot;&quot;&quot;
</span><span id="__span-152-15"><a id="__codelineno-152-15" name="__codelineno-152-15" href="#__codelineno-152-15"></a>    infer_experiment.py \
</span><span id="__span-152-16"><a id="__codelineno-152-16" name="__codelineno-152-16" href="#__codelineno-152-16"></a>    -i ${input_bam} \
</span><span id="__span-152-17"><a id="__codelineno-152-17" name="__codelineno-152-17" href="#__codelineno-152-17"></a>    -r ${refgene} \
</span><span id="__span-152-18"><a id="__codelineno-152-18" name="__codelineno-152-18" href="#__codelineno-152-18"></a>    &gt; ${input_bam.simpleName}.infer_experiment.txt                             &lt;-
</span><span id="__span-152-19"><a id="__codelineno-152-19" name="__codelineno-152-19" href="#__codelineno-152-19"></a>    &quot;&quot;&quot;
</span><span id="__span-152-20"><a id="__codelineno-152-20" name="__codelineno-152-20" href="#__codelineno-152-20"></a>
</span><span id="__span-152-21"><a id="__codelineno-152-21" name="__codelineno-152-21" href="#__codelineno-152-21"></a>}
</span></code></pre></div></p>
<p>rseqc_read_distribution.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-153-1"><a id="__codelineno-153-1" name="__codelineno-153-1" href="#__codelineno-153-1"></a>process RSEQC_READ_DISTRIBUTION {
</span><span id="__span-153-2"><a id="__codelineno-153-2" name="__codelineno-153-2" href="#__codelineno-153-2"></a>
</span><span id="__span-153-3"><a id="__codelineno-153-3" name="__codelineno-153-3" href="#__codelineno-153-3"></a>    container &quot;quay.io/biocontainers/rseqc:2.6.4--py27hf8a1672_2&quot;
</span><span id="__span-153-4"><a id="__codelineno-153-4" name="__codelineno-153-4" href="#__codelineno-153-4"></a>    publishDir &quot;${params.outdir}/rseqc_read_distribution&quot;
</span><span id="__span-153-5"><a id="__codelineno-153-5" name="__codelineno-153-5" href="#__codelineno-153-5"></a>
</span><span id="__span-153-6"><a id="__codelineno-153-6" name="__codelineno-153-6" href="#__codelineno-153-6"></a>    input:
</span><span id="__span-153-7"><a id="__codelineno-153-7" name="__codelineno-153-7" href="#__codelineno-153-7"></a>    path input_bam
</span><span id="__span-153-8"><a id="__codelineno-153-8" name="__codelineno-153-8" href="#__codelineno-153-8"></a>    path refgene
</span><span id="__span-153-9"><a id="__codelineno-153-9" name="__codelineno-153-9" href="#__codelineno-153-9"></a>
</span><span id="__span-153-10"><a id="__codelineno-153-10" name="__codelineno-153-10" href="#__codelineno-153-10"></a>    output:
</span><span id="__span-153-11"><a id="__codelineno-153-11" name="__codelineno-153-11" href="#__codelineno-153-11"></a>    path &quot;${input_bam.simpleName}.read_distribution.txt&quot;, emit: outputTextfile &lt;-
</span><span id="__span-153-12"><a id="__codelineno-153-12" name="__codelineno-153-12" href="#__codelineno-153-12"></a>
</span><span id="__span-153-13"><a id="__codelineno-153-13" name="__codelineno-153-13" href="#__codelineno-153-13"></a>    script:
</span><span id="__span-153-14"><a id="__codelineno-153-14" name="__codelineno-153-14" href="#__codelineno-153-14"></a>    &quot;&quot;&quot;
</span><span id="__span-153-15"><a id="__codelineno-153-15" name="__codelineno-153-15" href="#__codelineno-153-15"></a>    read_distribution.py \
</span><span id="__span-153-16"><a id="__codelineno-153-16" name="__codelineno-153-16" href="#__codelineno-153-16"></a>    -i ${input_bam} \
</span><span id="__span-153-17"><a id="__codelineno-153-17" name="__codelineno-153-17" href="#__codelineno-153-17"></a>    -r ${refgene} \
</span><span id="__span-153-18"><a id="__codelineno-153-18" name="__codelineno-153-18" href="#__codelineno-153-18"></a>    &gt; ${input_bam.simpleName}.read_distribution.txt                            &lt;-
</span><span id="__span-153-19"><a id="__codelineno-153-19" name="__codelineno-153-19" href="#__codelineno-153-19"></a>    &quot;&quot;&quot;
</span><span id="__span-153-20"><a id="__codelineno-153-20" name="__codelineno-153-20" href="#__codelineno-153-20"></a>
</span><span id="__span-153-21"><a id="__codelineno-153-21" name="__codelineno-153-21" href="#__codelineno-153-21"></a>}
</span></code></pre></div></p>
<p>samtools_idxstats.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-154-1"><a id="__codelineno-154-1" name="__codelineno-154-1" href="#__codelineno-154-1"></a>process SAMTOOLS_IDXSTATS {
</span><span id="__span-154-2"><a id="__codelineno-154-2" name="__codelineno-154-2" href="#__codelineno-154-2"></a>
</span><span id="__span-154-3"><a id="__codelineno-154-3" name="__codelineno-154-3" href="#__codelineno-154-3"></a>    container &quot;quay.io/biocontainers/samtools:1.9--h10a08f8_12&quot;
</span><span id="__span-154-4"><a id="__codelineno-154-4" name="__codelineno-154-4" href="#__codelineno-154-4"></a>    publishDir &quot;${params.outdir}/samtools_idxstats&quot;
</span><span id="__span-154-5"><a id="__codelineno-154-5" name="__codelineno-154-5" href="#__codelineno-154-5"></a>
</span><span id="__span-154-6"><a id="__codelineno-154-6" name="__codelineno-154-6" href="#__codelineno-154-6"></a>    input:
</span><span id="__span-154-7"><a id="__codelineno-154-7" name="__codelineno-154-7" href="#__codelineno-154-7"></a>    path input_bam
</span><span id="__span-154-8"><a id="__codelineno-154-8" name="__codelineno-154-8" href="#__codelineno-154-8"></a>    val addthreads
</span><span id="__span-154-9"><a id="__codelineno-154-9" name="__codelineno-154-9" href="#__codelineno-154-9"></a>
</span><span id="__span-154-10"><a id="__codelineno-154-10" name="__codelineno-154-10" href="#__codelineno-154-10"></a>    output:
</span><span id="__span-154-11"><a id="__codelineno-154-11" name="__codelineno-154-11" href="#__codelineno-154-11"></a>    path &quot;${input_bam.simpleName}.idxstats.tabular&quot;, emit: outputTabular   &lt;-
</span><span id="__span-154-12"><a id="__codelineno-154-12" name="__codelineno-154-12" href="#__codelineno-154-12"></a>
</span><span id="__span-154-13"><a id="__codelineno-154-13" name="__codelineno-154-13" href="#__codelineno-154-13"></a>    script:
</span><span id="__span-154-14"><a id="__codelineno-154-14" name="__codelineno-154-14" href="#__codelineno-154-14"></a>    &quot;&quot;&quot;
</span><span id="__span-154-15"><a id="__codelineno-154-15" name="__codelineno-154-15" href="#__codelineno-154-15"></a>    samtools idxstats \
</span><span id="__span-154-16"><a id="__codelineno-154-16" name="__codelineno-154-16" href="#__codelineno-154-16"></a>    -@ ${addthreads} \
</span><span id="__span-154-17"><a id="__codelineno-154-17" name="__codelineno-154-17" href="#__codelineno-154-17"></a>    ${input_bam} \
</span><span id="__span-154-18"><a id="__codelineno-154-18" name="__codelineno-154-18" href="#__codelineno-154-18"></a>    &gt; ${input_bam.simpleName}.idxstats.tabular                             &lt;-
</span><span id="__span-154-19"><a id="__codelineno-154-19" name="__codelineno-154-19" href="#__codelineno-154-19"></a>    &quot;&quot;&quot;
</span><span id="__span-154-20"><a id="__codelineno-154-20" name="__codelineno-154-20" href="#__codelineno-154-20"></a>
</span><span id="__span-154-21"><a id="__codelineno-154-21" name="__codelineno-154-21" href="#__codelineno-154-21"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After these changes have been made, re-run the workflow.</p>
<p>As the filenames will be different this run, don&rsquo;t add <code>-resume</code>.<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-155-1"><a id="__codelineno-155-1" name="__codelineno-155-1" href="#__codelineno-155-1"></a>nextflow run main.nf
</span></code></pre></div></p>
<p><br></p>
<h3 id="final-error-multiqc-script">Final Error: MULTIQC script<a class="headerlink" href="#final-error-multiqc-script" title="Permanent link">&para;</a></h3>
<p>The last error we need to fix is the MULTIQC script. </p>
<p><strong>Error message</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-156-1"><a id="__codelineno-156-1" name="__codelineno-156-1" href="#__codelineno-156-1"></a>Error executing process &gt; &#39;MULTIQC&#39;
</span><span id="__span-156-2"><a id="__codelineno-156-2" name="__codelineno-156-2" href="#__codelineno-156-2"></a>
</span><span id="__span-156-3"><a id="__codelineno-156-3" name="__codelineno-156-3" href="#__codelineno-156-3"></a>Caused by:
</span><span id="__span-156-4"><a id="__codelineno-156-4" name="__codelineno-156-4" href="#__codelineno-156-4"></a>  Process `MULTIQC` terminated with an error exit status (2)
</span><span id="__span-156-5"><a id="__codelineno-156-5" name="__codelineno-156-5" href="#__codelineno-156-5"></a>
</span><span id="__span-156-6"><a id="__codelineno-156-6" name="__codelineno-156-6" href="#__codelineno-156-6"></a>Command executed:
</span><span id="__span-156-7"><a id="__codelineno-156-7" name="__codelineno-156-7" href="#__codelineno-156-7"></a>  multiqc multiqc_WDir     --config multiqc_config
</span><span id="__span-156-8"><a id="__codelineno-156-8" name="__codelineno-156-8" href="#__codelineno-156-8"></a>
</span><span id="__span-156-9"><a id="__codelineno-156-9" name="__codelineno-156-9" href="#__codelineno-156-9"></a>Command error:
</span><span id="__span-156-10"><a id="__codelineno-156-10" name="__codelineno-156-10" href="#__codelineno-156-10"></a>  Usage: multiqc [OPTIONS] &lt;analysis directory&gt;
</span><span id="__span-156-11"><a id="__codelineno-156-11" name="__codelineno-156-11" href="#__codelineno-156-11"></a>
</span><span id="__span-156-12"><a id="__codelineno-156-12" name="__codelineno-156-12" href="#__codelineno-156-12"></a>  Error: Invalid value for &quot;analysis_dir&quot;: Path &quot;multiqc_WDir&quot; does not exist.
</span></code></pre></div>
<p><br></p>
<p><strong>Diagnosing the error</strong></p>
<p>This error is due to our MULTIQC script command. </p>
<p>Galaxy expects a folder called <code>multiqc_WDir</code> to be present, and for our input files to be staged in that directory. 
Additionally, Galaxy supplies a config file using the <code>--config</code> argument when it runs multiqc to do special formatting of the output report. </p>
<p>By looking at our MULTIQC process script, we can see these two arguments being supplied: <br>
<div class="language-text highlight"><pre><span></span><code><span id="__span-157-1"><a id="__codelineno-157-1" name="__codelineno-157-1" href="#__codelineno-157-1"></a>script:
</span><span id="__span-157-2"><a id="__codelineno-157-2" name="__codelineno-157-2" href="#__codelineno-157-2"></a>&quot;&quot;&quot;
</span><span id="__span-157-3"><a id="__codelineno-157-3" name="__codelineno-157-3" href="#__codelineno-157-3"></a>multiqc multiqc_WDir \
</span><span id="__span-157-4"><a id="__codelineno-157-4" name="__codelineno-157-4" href="#__codelineno-157-4"></a>--config ${multiqc_config} \
</span><span id="__span-157-5"><a id="__codelineno-157-5" name="__codelineno-157-5" href="#__codelineno-157-5"></a>&quot;&quot;&quot;
</span></code></pre></div></p>
<p>For our nextflow workflow, our input files will be staged into the process working directory, rather than <code>multiqc_WDir</code>. </p>
<p>We also don&rsquo;t need the <code>--config</code> argument as this is unnecessary. </p>
<p><br></p>
<p><strong>Solution</strong></p>
<p>To fix this issue, we will modify the script section of MULTIQC. </p>
<p>We will also update the output collection patterns to be multiqc defaults.</p>
<div class="admonition question">
<p class="admonition-title">Modifying MULTIQC script</p>
<p>In the script section, change <code>multiqc_WDir</code> to <code>.</code></p>
<p>In the script section, remove the <code>--config</code> argument and its value.</p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/multiqc.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-158-1"><a id="__codelineno-158-1" name="__codelineno-158-1" href="#__codelineno-158-1"></a>process MULTIQC {
</span><span id="__span-158-2"><a id="__codelineno-158-2" name="__codelineno-158-2" href="#__codelineno-158-2"></a>    ...
</span><span id="__span-158-3"><a id="__codelineno-158-3" name="__codelineno-158-3" href="#__codelineno-158-3"></a>
</span><span id="__span-158-4"><a id="__codelineno-158-4" name="__codelineno-158-4" href="#__codelineno-158-4"></a>    script:
</span><span id="__span-158-5"><a id="__codelineno-158-5" name="__codelineno-158-5" href="#__codelineno-158-5"></a>    &quot;&quot;&quot;
</span><span id="__span-158-6"><a id="__codelineno-158-6" name="__codelineno-158-6" href="#__codelineno-158-6"></a>    multiqc .           &lt;- modified line
</span><span id="__span-158-7"><a id="__codelineno-158-7" name="__codelineno-158-7" href="#__codelineno-158-7"></a>                        &lt;- removed line
</span><span id="__span-158-8"><a id="__codelineno-158-8" name="__codelineno-158-8" href="#__codelineno-158-8"></a>    &quot;&quot;&quot;
</span><span id="__span-158-9"><a id="__codelineno-158-9" name="__codelineno-158-9" href="#__codelineno-158-9"></a>
</span><span id="__span-158-10"><a id="__codelineno-158-10" name="__codelineno-158-10" href="#__codelineno-158-10"></a>}
</span></code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Modifying MULTIQC output collection patterns</p>
<p>Modify the <code>out_html_report</code> output collection pattern to <code>"multiqc_report.html"</code> (default).</p>
<p>Modify the <code>out_stats</code> output collection pattern to <code>"multiqc_data/multiqc_*.txt"</code>. <br>
This will collect each text file from the default MultiQC output directory.</p>
<details class="hint">
<summary>Show Changes</summary>
<p>modules/multiqc.nf:
<div class="language-text highlight"><pre><span></span><code><span id="__span-159-1"><a id="__codelineno-159-1" name="__codelineno-159-1" href="#__codelineno-159-1"></a>process MULTIQC {
</span><span id="__span-159-2"><a id="__codelineno-159-2" name="__codelineno-159-2" href="#__codelineno-159-2"></a>    ...
</span><span id="__span-159-3"><a id="__codelineno-159-3" name="__codelineno-159-3" href="#__codelineno-159-3"></a>
</span><span id="__span-159-4"><a id="__codelineno-159-4" name="__codelineno-159-4" href="#__codelineno-159-4"></a>    output:
</span><span id="__span-159-5"><a id="__codelineno-159-5" name="__codelineno-159-5" href="#__codelineno-159-5"></a>    path &quot;multiqc_report.html&quot;, emit: out_html_report       &lt;-
</span><span id="__span-159-6"><a id="__codelineno-159-6" name="__codelineno-159-6" href="#__codelineno-159-6"></a>    path &quot;multiqc_data/multiqc_*.txt&quot;, emit: out_stats      &lt;-
</span><span id="__span-159-7"><a id="__codelineno-159-7" name="__codelineno-159-7" href="#__codelineno-159-7"></a>
</span><span id="__span-159-8"><a id="__codelineno-159-8" name="__codelineno-159-8" href="#__codelineno-159-8"></a>    ...
</span><span id="__span-159-9"><a id="__codelineno-159-9" name="__codelineno-159-9" href="#__codelineno-159-9"></a>}
</span></code></pre></div></p>
</details>
</div>
<p><br></p>
<p>After making these changes, re-run the workflow using the <code>-resume</code> argument. 
<div class="language-text highlight"><pre><span></span><code><span id="__span-160-1"><a id="__codelineno-160-1" name="__codelineno-160-1" href="#__codelineno-160-1"></a>nextflow run main.nf -resume
</span></code></pre></div></p>
<p><br></p>
<h3 id="completed-workflow_1">Completed Workflow<a class="headerlink" href="#completed-workflow_1" title="Permanent link">&para;</a></h3>
<p>Your workflow should now run to completion. <br>
You can now use this workflow to transform raw reads into count data. </p>
<p>If you have any issues, the final adjusted workflow can be seen in the <code>final/</code> folder inside <code>janis-translate-examples/</code>.  </p>
<p>To view the outputs, check the <code>outputs/</code> folder. It will contain the results of each process. </p>
<p>This folder was specified to each process as a <code>publishDir</code> directive using the <code>outdir</code> param set in <code>nextflow.config</code>. </p>
<p>The final result is located at <code>outputs/multiqc/multiqc_report.html</code>. <br>
Reveal this file on your PC and open it with an internet browser. </p>
<p>If all went well, you should see a MultiQC report with metrics for each sample we processed. </p>
<p><br></p>
<p><img alt="alt text" src="../media/multiqc_report1.PNG" /></p>
<p><img alt="alt text" src="../media/multiqc_report2.PNG" /></p>
<p>&hellip;</p>
<h2 id="conclusion_5">Conclusion<a class="headerlink" href="#conclusion_5" title="Permanent link">&para;</a></h2>
<p>Thank you for your time! </p>
<p>We hope <code>janis translate</code> can assist you in your work. </p>
<p>If you have any bugs or issues you need help with, please raise an issue in the <a href="https://github.com/PMCC-BioinformaticsCore/janis-core">janis-core github repository</a>. </p>
<p>We look at each issue raised and aim to resolve these if possible. </p>
<p><br></p>
<p><br></p>
<p><br></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            <a href="https://www.melbournebioinformatics.org.au/">Melbourne Bioinformatics  The University of Melbourne </a>
          </div>
        
        Powered by <a href="https://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">Material for MkDocs</a>.
      </div>
      
        
<div class="md-social">
  
</div>
      
    </div>
  </div>
</footer>

      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": ["content.code.copy", "content.code.annotate"], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../../../javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>