
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../seurat-de/seurat-de/">
      
      
        <link rel="next" href="../../proteomics_basic/proteomics_basic/">
      
      
      <link rel="icon" href="../../../img/favicon-16x16.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Gene Ontology (GO) and Data Integration for scRNA-seq - MBITE Tutorials</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<script async src="https://www.googletagmanager.com/gtag/js?id=G-EMNLJD0DWP"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }
    gtag('js', new Date());
    gtag('config', 'G-EMNLJD0DWP');
</script>

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gene-ontology-analysis-and-integration-for-single-cell-rna-seq-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="MBITE Tutorials" class="md-header__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MBITE Tutorials
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Gene Ontology (GO) and Data Integration for scRNA-seq
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="MBITE Tutorials" class="md-nav__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    MBITE Tutorials
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Popular
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Popular
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Git for Researchers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/nectar-instances/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to the Nectar Cloud
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to R for Biologists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to RNA-seq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/rna-seq-counts-to-genes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq: counts to genes
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tools and skill development
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tools and skill development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/bioinfo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    What is Bioinformatics?
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/galaxy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to the Galaxy Platform
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galaxy-workflows/galaxy-workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Galaxy Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_overview/python_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Python
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Unix
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro_R_biologists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to R for Biologists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Genome_browsers/GenomeBrowsers_Intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Genome Browsers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl_2022/cwl_2022/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to CWL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq Experimental Design
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../docker/docker/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Containerized Bioinformatics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../bionitio/bionitio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command-line software development
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../git/git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Version Control with Git
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../janis_translate/janis_translate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Translation With Janis
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Genomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Genomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Variant detection
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Variant detection
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Variant detection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_galaxy_1/variant_calling_galaxy_1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant Calling part 1 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant Calling part 2 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../longread_sv_calling/longread_sv_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Long-read Structural Variant Calling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_gatk1/variant_calling_gatk1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant calling using command-line tools
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    de novo assembly
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            de novo assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to de novo assembly with Velvet
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-protocol/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to de novo genome assembly for Illumina reads
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of Illumina reads using Velvet (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of Illumina reads using Spades (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pacbio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of PacBio reads (unix)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../hybrid_assembly/nanopore_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid genome assembly - Nanopore and Illumina
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Metabarcoding
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Metabarcoding
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2/qiime2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to Metabarcoding using QIIME2: *Marine dataset*
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2_mammal/qiime2_mammal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to Metabarcoding using QIIME2: *Mammal dataset*
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Transcriptomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Transcriptomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq Experimental Design
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna-seq-lcg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq: counts to genes
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Bulk RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Bulk RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to bulk RNAseq analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq analysis using HISAT2 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tuxedo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq analysis using Tuxedo (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_advanced/rna_seq_advanced_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq differential expression tool comparision (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_in_r/rna_seq_r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq differential expression in R
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" checked>
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Single-cell RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Single-cell RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../singlecell/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    10X scRNA-seq analysis in R
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-de/seurat-de/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    scRNA-seq Integration and Differential Expression
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Gene Ontology (GO) and Data Integration for scRNA-seq
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Gene Ontology (GO) and Data Integration for scRNA-seq
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-these-packages" class="md-nav__link">
    <span class="md-ellipsis">
      Load these packages
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-datasets" class="md-nav__link">
    <span class="md-ellipsis">
      Download the datasets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-the-datasets-seurat-object" class="md-nav__link">
    <span class="md-ellipsis">
      Load the datasets (Seurat object)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing-for-single-cell-rna-seq-data-seurat-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing for single-cell RNA-seq data (Seurat pipeline)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing for single-cell RNA-seq data (Seurat pipeline)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-for-rna-data-already-performed-and-saved-in-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      QC for RNA data (already performed and saved in the dataset)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preprocess-rna-data-from-seurat-workshop" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocess RNA data (from Seurat workshop)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-and-visualise-the-rna-data" class="md-nav__link">
    <span class="md-ellipsis">
      Cluster and visualise the RNA data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes-for-each-cell-sub-type" class="md-nav__link">
    <span class="md-ellipsis">
      Find marker genes for each cell sub-type
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-ontology-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Ontology analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Gene Ontology analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preview-the-marker-genes" class="md-nav__link">
    <span class="md-ellipsis">
      Preview the marker genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes-for-cluster-2-cluster-3-and-cluster-8" class="md-nav__link">
    <span class="md-ellipsis">
      Find marker genes for Cluster 2, Cluster 3 and Cluster 8
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#convert-gene-symbols-to-entrez-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Convert gene symbols to Entrez IDs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform-go-enrichment-analysis-for-cluster-2" class="md-nav__link">
    <span class="md-ellipsis">
      Perform GO enrichment analysis for Cluster 2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Visualise the GO enrichment results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bar-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Bar plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dot-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Dot plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enrichment-map-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Enrichment map plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cnet-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Cnet plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results-for-the-other-2-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results for the other 2 clusters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#perform-gene-ontology-analysis-for-all-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Perform gene ontology analysis for all clusters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Perform gene ontology analysis for all clusters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results-for-all-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results for all clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-go-enrichment-results-for-other-two-ontology-categories" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise GO enrichment results for other two ontology categories
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#integrate-atac-seq-data-with-single-cell-rna-seq-data-n-integration" class="md-nav__link">
    <span class="md-ellipsis">
      Integrate ATAC-seq data with single-cell RNA-seq data (N Integration)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Integrate ATAC-seq data with single-cell RNA-seq data (N Integration)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-for-atac-seq-data-already-performed-and-saved-in-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      QC for ATAC-seq data (already performed and saved in the dataset)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preprocess-atac-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocess ATAC-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-atac-seq-data-with-scrna-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the ATAC-seq data with scRNA-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quantify-gene-activity" class="md-nav__link">
    <span class="md-ellipsis">
      Quantify gene activity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#integrate-rna-seq-data-with-atac-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Integrate RNA-seq data with ATAC-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-integrated-data" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the integrated data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-atac-seq-regions-around-a-gene-of-interest" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the ATAC-seq regions around a gene of interest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#additional-benchmark-study-evaluate-the-accuracy-of-the-integrated-data" class="md-nav__link">
    <span class="md-ellipsis">
      Additional benchmark study: Evaluate the accuracy of the integrated data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proteomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Proteomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proteomics_basic/proteomics_basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Identifying proteins from mass spectrometry data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Statistics and Visualisation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Statistics and Visualisation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_viz/python_viz/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data visualisation with Python
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Structural Modelling
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Structural Modelling
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_101/molecular_dynamics_101/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Molecular Dynamics - Introduction to cluster computing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_201/molecular_dynamics_201/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Molecular Dynamics - Building input files, visualising the trajectory
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workshop Delivery Mode Information
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Workshop Delivery Mode Information
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workshop_delivery_mode_info/workshops_nectar/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workshops on Nectar Instances
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
        
          
          <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Archive
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_10">
            <span class="md-nav__icon md-icon"></span>
            Archive
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix_archive_2022/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Unix-Archive
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl/cwl/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to CWL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../using_git/Using_Git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Git and Github (deprecated)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-these-packages" class="md-nav__link">
    <span class="md-ellipsis">
      Load these packages
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-datasets" class="md-nav__link">
    <span class="md-ellipsis">
      Download the datasets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-the-datasets-seurat-object" class="md-nav__link">
    <span class="md-ellipsis">
      Load the datasets (Seurat object)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing-for-single-cell-rna-seq-data-seurat-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing for single-cell RNA-seq data (Seurat pipeline)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing for single-cell RNA-seq data (Seurat pipeline)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-for-rna-data-already-performed-and-saved-in-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      QC for RNA data (already performed and saved in the dataset)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preprocess-rna-data-from-seurat-workshop" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocess RNA data (from Seurat workshop)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-and-visualise-the-rna-data" class="md-nav__link">
    <span class="md-ellipsis">
      Cluster and visualise the RNA data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes-for-each-cell-sub-type" class="md-nav__link">
    <span class="md-ellipsis">
      Find marker genes for each cell sub-type
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-ontology-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Ontology analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Gene Ontology analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preview-the-marker-genes" class="md-nav__link">
    <span class="md-ellipsis">
      Preview the marker genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes-for-cluster-2-cluster-3-and-cluster-8" class="md-nav__link">
    <span class="md-ellipsis">
      Find marker genes for Cluster 2, Cluster 3 and Cluster 8
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#convert-gene-symbols-to-entrez-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Convert gene symbols to Entrez IDs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform-go-enrichment-analysis-for-cluster-2" class="md-nav__link">
    <span class="md-ellipsis">
      Perform GO enrichment analysis for Cluster 2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Visualise the GO enrichment results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bar-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Bar plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dot-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Dot plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enrichment-map-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Enrichment map plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cnet-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Cnet plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results-for-the-other-2-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results for the other 2 clusters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#perform-gene-ontology-analysis-for-all-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Perform gene ontology analysis for all clusters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Perform gene ontology analysis for all clusters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#visualise-the-go-enrichment-results-for-all-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the GO enrichment results for all clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-go-enrichment-results-for-other-two-ontology-categories" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise GO enrichment results for other two ontology categories
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#integrate-atac-seq-data-with-single-cell-rna-seq-data-n-integration" class="md-nav__link">
    <span class="md-ellipsis">
      Integrate ATAC-seq data with single-cell RNA-seq data (N Integration)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Integrate ATAC-seq data with single-cell RNA-seq data (N Integration)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-for-atac-seq-data-already-performed-and-saved-in-the-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      QC for ATAC-seq data (already performed and saved in the dataset)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preprocess-atac-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocess ATAC-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-atac-seq-data-with-scrna-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the ATAC-seq data with scRNA-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quantify-gene-activity" class="md-nav__link">
    <span class="md-ellipsis">
      Quantify gene activity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#integrate-rna-seq-data-with-atac-seq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Integrate RNA-seq data with ATAC-seq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-integrated-data" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the integrated data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualise-the-atac-seq-regions-around-a-gene-of-interest" class="md-nav__link">
    <span class="md-ellipsis">
      Visualise the ATAC-seq regions around a gene of interest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#additional-benchmark-study-evaluate-the-accuracy-of-the-integrated-data" class="md-nav__link">
    <span class="md-ellipsis">
      Additional benchmark study: Evaluate the accuracy of the integrated data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="gene-ontology-analysis-and-integration-for-single-cell-rna-seq-data">Gene ontology analysis and integration for single-cell RNA-seq data<a class="headerlink" href="#gene-ontology-analysis-and-integration-for-single-cell-rna-seq-data" title="Permanent link">&para;</a></h1>
<p><strong>Author:</strong> Xiaochen Zhang, Lê Cao Lab, The University of Melbourne.<br />
<strong>Contributors:</strong> Vini Salazar, Melbourne Bioinformatics.</p>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p><strong>Topic</strong></p>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Genomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Transcriptomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Proteomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Metabolomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Statistics and visualisation</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Structural Modelling</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Basic skills</li>
</ul>
<p><strong>Skill level</strong></p>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Beginner  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Intermediate  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Advanced  </li>
</ul>
<p><strong>Data:</strong> 10k Human PBMCs, Multiome v1.0, Chromium X.</p>
<p><strong>Tools:</strong> R &gt;= 4.4.0 and associated packages:  </p>
<ul>
<li>Seurat  </li>
<li>SeuratData  </li>
<li>clusterProfiler  </li>
<li>Signac  </li>
<li>ggplot2  </li>
</ul>
<p><strong>Pipeline:</strong><br />
<em>Section 1:</em> Setup, Introduction, and Preprocessing.<br />
<em>Section 2:</em> Gene Ontology analysis.<br />
<em>Section 3:</em> Integration of ATAC-seq data.  </p>
<p><strong>Learning objectives:</strong>  </p>
<ul>
<li>Apply GO enrichment analysis to identify biological processes, molecular functions, and cellular components represented in single-cell RNA-seq data.  </li>
<li>Interpret GO results and identify key pathways and functional modules relevant to specific biological questions.  </li>
<li>Design a workflow for integrating various omics layers, considering the biological context and research goals.  </li>
<li>Take advantage of the Seurat, Signac and clusterProfiler R packages to perform the analysis, generate plots and interpret the results.  </li>
</ul>
<!--
!!! warning "Disclaimer"
    This tutorial is partially based on some existing material.
-->

<!-- Use this divider between sections -->
<hr />
<!-- Use this divider between sections -->

<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<p><strong>Install R packages for this workshop</strong></p>
<p>You only need to run this code once.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1"># install Packages</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">)</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="nf">setRepositories</span><span class="p">(</span><span class="n">ind</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="c1"># needed to automatically install Bioconductor dependencies</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">quietly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="w">    </span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;Rsamtools&quot;</span><span class="p">)</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
</span><span id="__span-0-10"><a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;Signac&quot;</span><span class="p">)</span>
</span><span id="__span-0-11"><a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;ggplot2&quot;</span><span class="p">)</span>
</span><span id="__span-0-12"><a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;cowplot&quot;</span><span class="p">)</span>
</span><span id="__span-0-13"><a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;devtools&quot;</span><span class="p">)</span>
</span><span id="__span-0-14"><a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="n">devtools</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&#39;satijalab/seurat-data&#39;</span><span class="p">)</span>
</span><span id="__span-0-15"><a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
</span><span id="__span-0-16"><a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;EnsDb.Hsapiens.v86&quot;</span><span class="p">)</span>
</span><span id="__span-0-17"><a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;biovizBase&quot;</span><span class="p">)</span>
</span><span id="__span-0-18"><a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;BSgenome.Hsapiens.UCSC.hg38&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;EnsDb.Hsapiens.v86&#39;</span><span class="p">))</span>
</span><span id="__span-0-19"><a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;clusterProfiler&quot;</span><span class="p">)</span>
</span><span id="__span-0-20"><a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>
</span><span id="__span-0-21"><a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;enrichplot&quot;</span><span class="p">)</span>
</span></code></pre></div>
<div class="admonition success">
<p class="admonition-title">Well done!</p>
<p>You are ready to start the workshop.</p>
</div>
<!-- Use this divider between sections -->
<hr />
<!-- Use this divider between sections -->

<h2 id="introduction">Introduction<a class="headerlink" href="#introduction" title="Permanent link">&para;</a></h2>
<h3 id="load-these-packages">Load these packages<a class="headerlink" href="#load-these-packages" title="Permanent link">&para;</a></h3>
<div class="language-r highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="c1"># load libraries</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="nf">library</span><span class="p">(</span><span class="n">Signac</span><span class="p">)</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="nf">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="nf">library</span><span class="p">(</span><span class="n">cowplot</span><span class="p">)</span>
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="nf">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span>
</span><span id="__span-1-8"><a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</span><span id="__span-1-9"><a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="nf">library</span><span class="p">(</span><span class="n">enrichplot</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="download-the-datasets">Download the datasets<a class="headerlink" href="#download-the-datasets" title="Permanent link">&para;</a></h3>
<div class="language-r highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="nf">library</span><span class="p">(</span><span class="n">SeuratData</span><span class="p">)</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="c1"># install the dataset and load requirements</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="nf">InstallData</span><span class="p">(</span><span class="s">&quot;pbmcMultiome&quot;</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="load-the-datasets-seurat-object">Load the datasets (Seurat object)<a class="headerlink" href="#load-the-datasets-seurat-object" title="Permanent link">&para;</a></h3>
<p>We will use two datasets in this workshop. One is single-cell RNA-seq data
(<code>pbmc.rna</code>), and the other is single-cell ATAC-seq data
(<code>pbmc.atac</code>). We will do a gene ontology analysis for the single-cell
RNA-seq dataset first. Afterwards, we will integrate the scRNA-seq dataset with the
scATAC-seq dataset to explain differences observed in the
scRNA-seq data.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># load both modalities</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">LoadData</span><span class="p">(</span><span class="s">&quot;pbmcMultiome&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pbmc.rna&quot;</span><span class="p">)</span>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">LoadData</span><span class="p">(</span><span class="s">&quot;pbmcMultiome&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pbmc.atac&quot;</span><span class="p">)</span>
</span></code></pre></div>
<h2 id="preprocessing-for-single-cell-rna-seq-data-seurat-pipeline">Preprocessing for single-cell RNA-seq data (Seurat pipeline)<a class="headerlink" href="#preprocessing-for-single-cell-rna-seq-data-seurat-pipeline" title="Permanent link">&para;</a></h2>
<h3 id="qc-for-rna-data-already-performed-and-saved-in-the-dataset">QC for RNA data (already performed and saved in the dataset)<a class="headerlink" href="#qc-for-rna-data-already-performed-and-saved-in-the-dataset" title="Permanent link">&para;</a></h3>
<p>In order to focus this workshop on downstream analysis, we have performed the QC
steps for the RNA data and applied the label &lsquo;filtered&rsquo;. We move forward by 
selecting only cells that are not labelled &lsquo;filtered&rsquo;.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="c1"># We use Seurat V5 object</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="n">pbmc.rna</span><span class="p">[[</span><span class="s">&quot;RNA&quot;</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">[[</span><span class="s">&quot;RNA&quot;</span><span class="p">]],</span><span class="w"> </span><span class="n">Class</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Assay5&quot;</span><span class="p">)</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="c1"># repeat QC steps performed in the WNN vignette</span>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">seurat_annotations</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s">&quot;filtered&quot;</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="preprocess-rna-data-from-seurat-workshop">Preprocess RNA data (from Seurat workshop)<a class="headerlink" href="#preprocess-rna-data-from-seurat-workshop" title="Permanent link">&para;</a></h3>
<p>We will perform standard Seurat pipeline for single-cell RNA analysis.
We will normalise the data, find variable features, scale the data, run
PCA, and run UMAP. See also our introduction to single-cell RNA-seq
workshop for more details.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="c1"># Perform standard analysis of each modality independently RNA analysis</span>
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">)</span>
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="c1">## Normalising layer: counts</span>
</span><span id="__span-5-5"><a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>
</span><span id="__span-5-6"><a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">)</span>
</span><span id="__span-5-7"><a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a>
</span><span id="__span-5-8"><a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a><span class="c1">## Finding variable features for layer counts</span>
</span><span id="__span-5-9"><a id="__codelineno-5-9" name="__codelineno-5-9" href="#__codelineno-5-9"></a>
</span><span id="__span-5-10"><a id="__codelineno-5-10" name="__codelineno-5-10" href="#__codelineno-5-10"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">)</span>
</span><span id="__span-5-11"><a id="__codelineno-5-11" name="__codelineno-5-11" href="#__codelineno-5-11"></a>
</span><span id="__span-5-12"><a id="__codelineno-5-12" name="__codelineno-5-12" href="#__codelineno-5-12"></a><span class="c1">## Centering and scaling data matrix</span>
</span><span id="__span-5-13"><a id="__codelineno-5-13" name="__codelineno-5-13" href="#__codelineno-5-13"></a>
</span><span id="__span-5-14"><a id="__codelineno-5-14" name="__codelineno-5-14" href="#__codelineno-5-14"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunPCA</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">)</span>
</span></code></pre></div>
<details class="note">
<summary>Output</summary>
<div class="language-r highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="c1">## PC_ 1 </span>
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="c1">## Positive:  PLXDC2, SLC8A1, LRMDA, FCN1, TYMP, MCTP1, JAK2, RBM47, IRAK3, NAMPT </span>
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="c1">##     DMXL2, TBXAS1, ZEB2, LYN, LRRK2, SAT1, GAB2, CYBB, TNFAIP2, CSF3R </span>
</span><span id="__span-6-4"><a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="c1">##     HCK, TLR2, CLEC7A, LYST, VCAN, DENND1A, FGD4, CD36, GRK3, FAM49A </span>
</span><span id="__span-6-5"><a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="c1">## Negative:  CD247, IL32, CAMK4, IL7R, LTB, LEF1, BACH2, INPP4B, BCL2, SYNE2 </span>
</span><span id="__span-6-6"><a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="c1">##     THEMIS, TRBC2, RORA, TXK, ANK3, CD69, TRBC1, MLLT3, APBA2, ACTG1 </span>
</span><span id="__span-6-7"><a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="c1">##     RASGRF2, NELL2, LINC01934, TAFA1, SAMD3, PCAT1, PFN1, NCALD, CTSW, KCNQ5 </span>
</span><span id="__span-6-8"><a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a><span class="c1">## PC_ 2 </span>
</span><span id="__span-6-9"><a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a><span class="c1">## Positive:  CD247, IL32, DPYD, IL7R, INPP4B, AOAH, CAMK4, PDE3B, THEMIS, TXK </span>
</span><span id="__span-6-10"><a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a><span class="c1">##     RORA, LEF1, SLCO3A1, NEAT1, FNDC3B, ARHGAP26, ANXA1, SRGN, ADGRE5, SAMD3 </span>
</span><span id="__span-6-11"><a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a><span class="c1">##     SYNE2, NCALD, TRBC1, CTSW, PLCB1, APBA2, CCL5, S100A4, TGFBR3, PRF1 </span>
</span><span id="__span-6-12"><a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a><span class="c1">## Negative:  BANK1, PAX5, MS4A1, FCRL1, NIBAN3, AFF3, IGHM, EBF1, LINC00926, OSBPL10 </span>
</span><span id="__span-6-13"><a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a><span class="c1">##     RALGPS2, CD79A, CD22, COBLL1, BLK, BLNK, AP002075.1, ADAM28, IGHD, COL19A1 </span>
</span><span id="__span-6-14"><a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a><span class="c1">##     BCL11A, CD79B, PLEKHG1, GNG7, DENND5B, WDFY4, TCF4, AC120193.1, RUBCNL, SPIB </span>
</span><span id="__span-6-15"><a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a><span class="c1">## PC_ 3 </span>
</span><span id="__span-6-16"><a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a><span class="c1">## Positive:  BACH2, LEF1, PDE3B, CAMK4, IL7R, LTB, ANK3, FHIT, INPP4B, NELL2 </span>
</span><span id="__span-6-17"><a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a><span class="c1">##     BCL2, RASGRF2, CSGALNACT1, SLC2A3, MLLT3, AC139720.1, PLCL1, PRKN, PAX5, FCRL1 </span>
</span><span id="__span-6-18"><a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a><span class="c1">##     LINC00926, TSHZ2, MS4A1, PTPRK, AL589693.1, CD79A, EBF1, ST6GALNAC3, BANK1, COL19A1 </span>
</span><span id="__span-6-19"><a id="__codelineno-6-19" name="__codelineno-6-19" href="#__codelineno-6-19"></a><span class="c1">## Negative:  GZMB, NKG7, GNLY, PRF1, KLRD1, GZMA, CST7, CLIC3, CCL5, FGFBP2 </span>
</span><span id="__span-6-20"><a id="__codelineno-6-20" name="__codelineno-6-20" href="#__codelineno-6-20"></a><span class="c1">##     SPON2, ADGRG1, KLRF1, GZMH, CCL4, TGFBR3, LINC00996, LINC01478, PDGFD, PTGDR </span>
</span><span id="__span-6-21"><a id="__codelineno-6-21" name="__codelineno-6-21" href="#__codelineno-6-21"></a><span class="c1">##     C1orf21, FCRL6, CLEC4C, SLAMF7, BNC2, CTSW, IL2RB, EPHB1, LILRA4, HOPX </span>
</span><span id="__span-6-22"><a id="__codelineno-6-22" name="__codelineno-6-22" href="#__codelineno-6-22"></a><span class="c1">## PC_ 4 </span>
</span><span id="__span-6-23"><a id="__codelineno-6-23" name="__codelineno-6-23" href="#__codelineno-6-23"></a><span class="c1">## Positive:  NKG7, GNLY, PRF1, KLRD1, GZMA, CST7, CCL5, FGFBP2, MCTP2, ADGRG1 </span>
</span><span id="__span-6-24"><a id="__codelineno-6-24" name="__codelineno-6-24" href="#__codelineno-6-24"></a><span class="c1">##     GZMH, CCL4, KLRF1, FCGR3A, SPON2, TGFBR3, PDGFD, PAX5, FCRL1, MS4A1 </span>
</span><span id="__span-6-25"><a id="__codelineno-6-25" name="__codelineno-6-25" href="#__codelineno-6-25"></a><span class="c1">##     FCRL6, PTGDR, BNC2, MTSS1, CX3CR1, LINC00926, C1orf21, IL2RB, PPP2R2B, HOPX </span>
</span><span id="__span-6-26"><a id="__codelineno-6-26" name="__codelineno-6-26" href="#__codelineno-6-26"></a><span class="c1">## Negative:  LINC01478, CUX2, EPHB1, CLEC4C, LILRA4, COL26A1, AC023590.1, PTPRS, SCN9A, LINC01374 </span>
</span><span id="__span-6-27"><a id="__codelineno-6-27" name="__codelineno-6-27" href="#__codelineno-6-27"></a><span class="c1">##     COL24A1, LINC00996, NRP1, PHEX, FAM160A1, TNFRSF21, PLXNA4, PACSIN1, SCAMP5, SLC35F3 </span>
</span><span id="__span-6-28"><a id="__codelineno-6-28" name="__codelineno-6-28" href="#__codelineno-6-28"></a><span class="c1">##     DNASE1L3, P3H2, LRRC26, PLD4, ITM2C, SCN1A-AS1, PLEKHD1, SERPINF1, SCT, PTCRA </span>
</span><span id="__span-6-29"><a id="__codelineno-6-29" name="__codelineno-6-29" href="#__codelineno-6-29"></a><span class="c1">## PC_ 5 </span>
</span><span id="__span-6-30"><a id="__codelineno-6-30" name="__codelineno-6-30" href="#__codelineno-6-30"></a><span class="c1">## Positive:  CDKN1C, IFITM3, HES4, CSF1R, SMIM25, FCGR3A, MS4A7, LST1, AIF1, LRRC25 </span>
</span><span id="__span-6-31"><a id="__codelineno-6-31" name="__codelineno-6-31" href="#__codelineno-6-31"></a><span class="c1">##     AC104809.2, MAFB, CALHM6, CFD, AC020651.2, SERPINA1, GPBAR1, SIGLEC10, CST3, HMOX1 </span>
</span><span id="__span-6-32"><a id="__codelineno-6-32" name="__codelineno-6-32" href="#__codelineno-6-32"></a><span class="c1">##     CTSL, TCF7L2, FMNL2, HLA-DPA1, CKB, MTSS1, COTL1, IFI30, CCDC26, SPRED1 </span>
</span><span id="__span-6-33"><a id="__codelineno-6-33" name="__codelineno-6-33" href="#__codelineno-6-33"></a><span class="c1">## Negative:  PLCB1, VCAN, ARHGAP24, CSF3R, DYSF, LINC00937, FNDC3B, GLT1D1, VCAN-AS1, AC020916.1 </span>
</span><span id="__span-6-34"><a id="__codelineno-6-34" name="__codelineno-6-34" href="#__codelineno-6-34"></a><span class="c1">##     PDE4D, PADI4, ARHGAP26, CREB5, TREM1, MCTP2, CD36, MEGF9, KIF13A, MICAL2 </span>
</span><span id="__span-6-35"><a id="__codelineno-6-35" name="__codelineno-6-35" href="#__codelineno-6-35"></a><span class="c1">##     FCAR, ACSL1, CLMN, JUN, ANXA1, MIR646HG, RAB11FIP1, CRISPLD2, GNLY, ZBTB16</span>
</span></code></pre></div>
</details>
<div class="language-r highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunUMAP</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="c1">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="c1">## To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;</span>
</span><span id="__span-7-5"><a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="c1">## This message will be shown once per session</span>
</span><span id="__span-7-6"><a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="c1">## 11:47:36 UMAP embedding parameters a = 0.9922 b = 1.112</span>
</span><span id="__span-7-7"><a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="c1">## 11:47:36 Read 10412 rows and found 30 numeric columns</span>
</span><span id="__span-7-8"><a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a><span class="c1">## 11:47:36 Using Annoy for neighbor search, n_neighbors = 30</span>
</span><span id="__span-7-9"><a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a><span class="c1">## 11:47:36 Building Annoy index with metric = cosine, n_trees = 50</span>
</span><span id="__span-7-10"><a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a><span class="c1">## 0%   10   20   30   40   50   60   70   80   90   100%</span>
</span><span id="__span-7-11"><a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a><span class="c1">## [----|----|----|----|----|----|----|----|----|----|</span>
</span><span id="__span-7-12"><a id="__codelineno-7-12" name="__codelineno-7-12" href="#__codelineno-7-12"></a><span class="c1">## **************************************************|</span>
</span><span id="__span-7-13"><a id="__codelineno-7-13" name="__codelineno-7-13" href="#__codelineno-7-13"></a><span class="c1">## 11:47:36 Writing NN index file to temp file /var/folders/pv/fvynh7953flggrfb49p2lqsc0000gn/T//RtmpuPPPuo/file4dc32cd5de65</span>
</span><span id="__span-7-14"><a id="__codelineno-7-14" name="__codelineno-7-14" href="#__codelineno-7-14"></a><span class="c1">## 11:47:36 Searching Annoy index using 1 thread, search_k = 3000</span>
</span><span id="__span-7-15"><a id="__codelineno-7-15" name="__codelineno-7-15" href="#__codelineno-7-15"></a><span class="c1">## 11:47:38 Annoy recall = 100%</span>
</span><span id="__span-7-16"><a id="__codelineno-7-16" name="__codelineno-7-16" href="#__codelineno-7-16"></a><span class="c1">## 11:47:38 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span>
</span><span id="__span-7-17"><a id="__codelineno-7-17" name="__codelineno-7-17" href="#__codelineno-7-17"></a><span class="c1">## 11:47:39 Initializing from normalized Laplacian + noise (using RSpectra)</span>
</span><span id="__span-7-18"><a id="__codelineno-7-18" name="__codelineno-7-18" href="#__codelineno-7-18"></a><span class="c1">## 11:47:39 Commencing optimization for 200 epochs, with 450622 positive edges</span>
</span><span id="__span-7-19"><a id="__codelineno-7-19" name="__codelineno-7-19" href="#__codelineno-7-19"></a><span class="c1">## 11:47:42 Optimization finished</span>
</span></code></pre></div>
<h3 id="cluster-and-visualise-the-rna-data">Cluster and visualise the RNA data<a class="headerlink" href="#cluster-and-visualise-the-rna-data" title="Permanent link">&para;</a></h3>
<p>Next, we obtain an overview of the RNA data by visualising clusters.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="c1"># Clustering</span>
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="c1">## Computing nearest neighbor graph</span>
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a><span class="c1">## Computing SNN</span>
</span><span id="__span-8-6"><a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>
</span><span id="__span-8-7"><a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a><span class="n">pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindClusters</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">resolution</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span>
</span><span id="__span-8-8"><a id="__codelineno-8-8" name="__codelineno-8-8" href="#__codelineno-8-8"></a>
</span><span id="__span-8-9"><a id="__codelineno-8-9" name="__codelineno-8-9" href="#__codelineno-8-9"></a><span class="c1">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
</span><span id="__span-8-10"><a id="__codelineno-8-10" name="__codelineno-8-10" href="#__codelineno-8-10"></a><span class="c1">## </span>
</span><span id="__span-8-11"><a id="__codelineno-8-11" name="__codelineno-8-11" href="#__codelineno-8-11"></a><span class="c1">## Number of nodes: 10412</span>
</span><span id="__span-8-12"><a id="__codelineno-8-12" name="__codelineno-8-12" href="#__codelineno-8-12"></a><span class="c1">## Number of edges: 412246</span>
</span><span id="__span-8-13"><a id="__codelineno-8-13" name="__codelineno-8-13" href="#__codelineno-8-13"></a><span class="c1">## </span>
</span><span id="__span-8-14"><a id="__codelineno-8-14" name="__codelineno-8-14" href="#__codelineno-8-14"></a><span class="c1">## Running Louvain algorithm...</span>
</span><span id="__span-8-15"><a id="__codelineno-8-15" name="__codelineno-8-15" href="#__codelineno-8-15"></a><span class="c1">## Maximum modularity in 10 random starts: 0.9154</span>
</span><span id="__span-8-16"><a id="__codelineno-8-16" name="__codelineno-8-16" href="#__codelineno-8-16"></a><span class="c1">## Number of communities: 17</span>
</span><span id="__span-8-17"><a id="__codelineno-8-17" name="__codelineno-8-17" href="#__codelineno-8-17"></a><span class="c1">## Elapsed time: 0 seconds</span>
</span><span id="__span-8-18"><a id="__codelineno-8-18" name="__codelineno-8-18" href="#__codelineno-8-18"></a>
</span><span id="__span-8-19"><a id="__codelineno-8-19" name="__codelineno-8-19" href="#__codelineno-8-19"></a><span class="c1"># Visualise the RNA data</span>
</span><span id="__span-8-20"><a id="__codelineno-8-20" name="__codelineno-8-20" href="#__codelineno-8-20"></a><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_clusters&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-6-1.png" /></p>
<h3 id="find-marker-genes-for-each-cell-sub-type">Find marker genes for each cell sub-type<a class="headerlink" href="#find-marker-genes-for-each-cell-sub-type" title="Permanent link">&para;</a></h3>
<p>We found that clusters 2, 3, and 8 likely have different functions,
forming distinct clusters. We aim to find marker
genes for each cluster to further explore the differences.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="c1"># Find marker genes for each cluster</span>
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;seurat_clusters&quot;</span>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="n">marker.genes.pbmc.rna</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">only.pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="w">                                        </span><span class="n">min.pct</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">logfc.threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span>
</span><span id="__span-9-5"><a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a>
</span><span id="__span-9-6"><a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="c1">## Calculating cluster 0</span>
</span><span id="__span-9-7"><a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a><span class="c1">## For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span>
</span><span id="__span-9-8"><a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a><span class="c1">## (default method for FindMarkers) please install the presto package</span>
</span><span id="__span-9-9"><a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a><span class="c1">## --------------------------------------------</span>
</span><span id="__span-9-10"><a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a><span class="c1">## install.packages(&#39;devtools&#39;)</span>
</span><span id="__span-9-11"><a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a><span class="c1">## devtools::install_github(&#39;immunogenomics/presto&#39;)</span>
</span><span id="__span-9-12"><a id="__codelineno-9-12" name="__codelineno-9-12" href="#__codelineno-9-12"></a><span class="c1">## --------------------------------------------</span>
</span><span id="__span-9-13"><a id="__codelineno-9-13" name="__codelineno-9-13" href="#__codelineno-9-13"></a><span class="c1">## After installation of presto, Seurat will automatically use the more </span>
</span><span id="__span-9-14"><a id="__codelineno-9-14" name="__codelineno-9-14" href="#__codelineno-9-14"></a><span class="c1">## efficient implementation (no further action necessary).</span>
</span><span id="__span-9-15"><a id="__codelineno-9-15" name="__codelineno-9-15" href="#__codelineno-9-15"></a><span class="c1">## This message will be shown once per session</span>
</span><span id="__span-9-16"><a id="__codelineno-9-16" name="__codelineno-9-16" href="#__codelineno-9-16"></a><span class="c1">## Calculating cluster 1</span>
</span><span id="__span-9-17"><a id="__codelineno-9-17" name="__codelineno-9-17" href="#__codelineno-9-17"></a><span class="c1">## Calculating cluster 2</span>
</span><span id="__span-9-18"><a id="__codelineno-9-18" name="__codelineno-9-18" href="#__codelineno-9-18"></a><span class="c1">## Calculating cluster 3</span>
</span><span id="__span-9-19"><a id="__codelineno-9-19" name="__codelineno-9-19" href="#__codelineno-9-19"></a><span class="c1">## Calculating cluster 4</span>
</span><span id="__span-9-20"><a id="__codelineno-9-20" name="__codelineno-9-20" href="#__codelineno-9-20"></a><span class="c1">## Calculating cluster 5</span>
</span><span id="__span-9-21"><a id="__codelineno-9-21" name="__codelineno-9-21" href="#__codelineno-9-21"></a><span class="c1">## Calculating cluster 6</span>
</span><span id="__span-9-22"><a id="__codelineno-9-22" name="__codelineno-9-22" href="#__codelineno-9-22"></a><span class="c1">## Calculating cluster 7</span>
</span><span id="__span-9-23"><a id="__codelineno-9-23" name="__codelineno-9-23" href="#__codelineno-9-23"></a><span class="c1">## Calculating cluster 8</span>
</span><span id="__span-9-24"><a id="__codelineno-9-24" name="__codelineno-9-24" href="#__codelineno-9-24"></a><span class="c1">## Calculating cluster 9</span>
</span><span id="__span-9-25"><a id="__codelineno-9-25" name="__codelineno-9-25" href="#__codelineno-9-25"></a><span class="c1">## Calculating cluster 10</span>
</span><span id="__span-9-26"><a id="__codelineno-9-26" name="__codelineno-9-26" href="#__codelineno-9-26"></a><span class="c1">## Calculating cluster 11</span>
</span><span id="__span-9-27"><a id="__codelineno-9-27" name="__codelineno-9-27" href="#__codelineno-9-27"></a><span class="c1">## Calculating cluster 12</span>
</span><span id="__span-9-28"><a id="__codelineno-9-28" name="__codelineno-9-28" href="#__codelineno-9-28"></a><span class="c1">## Calculating cluster 13</span>
</span><span id="__span-9-29"><a id="__codelineno-9-29" name="__codelineno-9-29" href="#__codelineno-9-29"></a><span class="c1">## Calculating cluster 14</span>
</span><span id="__span-9-30"><a id="__codelineno-9-30" name="__codelineno-9-30" href="#__codelineno-9-30"></a><span class="c1">## Calculating cluster 15</span>
</span><span id="__span-9-31"><a id="__codelineno-9-31" name="__codelineno-9-31" href="#__codelineno-9-31"></a><span class="c1">## Calculating cluster 16</span>
</span></code></pre></div>
<h2 id="gene-ontology-analysis">Gene Ontology analysis<a class="headerlink" href="#gene-ontology-analysis" title="Permanent link">&para;</a></h2>
<p>We now have a marker gene list for each cluster. But it can be difficult
to interpret a cluster based on gene names alone. A good method to
interpret a list of genes is gene ontology analysis. We will perform gene
ontology analysis to understand the biological functions of these marker
genes.</p>
<h3 id="preview-the-marker-genes">Preview the marker genes<a class="headerlink" href="#preview-the-marker-genes" title="Permanent link">&para;</a></h3>
<p>Let&rsquo;s have a look the data structure of the marker gene list.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">marker.genes.pbmc.rna</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
</span></code></pre></div>
<table>
<thead>
<tr>
<th></th>
<th style="text-align: right;">p_val &lt;dbl&gt;</th>
<th style="text-align: right;">avg_log2FC &lt;dbl&gt;</th>
<th style="text-align: right;">pct.1 &lt;dbl&gt;</th>
<th style="text-align: right;">pct.2 &lt;dbl&gt;</th>
<th style="text-align: right;">p_val_adj &lt;dbl&gt;</th>
<th>cluster &lt;fct&gt;</th>
<th>gene &lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr>
<td>LRMDA</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.907970</td>
<td style="text-align: right;">0.976</td>
<td style="text-align: right;">0.142</td>
<td style="text-align: right;">0</td>
<td>0</td>
<td>LRMDA</td>
</tr>
<tr>
<td>PLXDC2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.948556</td>
<td style="text-align: right;">0.994</td>
<td style="text-align: right;">0.171</td>
<td style="text-align: right;">0</td>
<td>0</td>
<td>PLXDC2</td>
</tr>
<tr>
<td>SLC8A1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.389460</td>
<td style="text-align: right;">0.977</td>
<td style="text-align: right;">0.158</td>
<td style="text-align: right;">0</td>
<td>0</td>
<td>SLC8A1</td>
</tr>
<tr>
<td>CSF3R</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4.401877</td>
<td style="text-align: right;">0.903</td>
<td style="text-align: right;">0.092</td>
<td style="text-align: right;">0</td>
<td>0</td>
<td>CSF3R</td>
</tr>
<tr>
<td>RBM47</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.747440</td>
<td style="text-align: right;">0.935</td>
<td style="text-align: right;">0.153</td>
<td style="text-align: right;">0</td>
<td>0</td>
<td>RBM47</td>
</tr>
</tbody>
</table>
<h3 id="find-marker-genes-for-cluster-2-cluster-3-and-cluster-8">Find marker genes for Cluster 2, Cluster 3 and Cluster 8<a class="headerlink" href="#find-marker-genes-for-cluster-2-cluster-3-and-cluster-8" title="Permanent link">&para;</a></h3>
<p>Because the gene list is long, we will focus on Cluster 2, Cluster 3
and Cluster 8 marker genes with significant p-values (adjust p-value &lt; 0.05).</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="n">Cluster2.markers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="p">[</span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">cluster</span><span class="o">==</span><span class="m">2</span><span class="w"> </span>
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="w">                                           </span><span class="o">&amp;</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">p_val_adj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
</span><span id="__span-11-3"><a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="n">Cluster3.markers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="p">[</span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">cluster</span><span class="o">==</span><span class="m">3</span><span class="w"> </span>
</span><span id="__span-11-4"><a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="w">                                           </span><span class="o">&amp;</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">p_val_adj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
</span><span id="__span-11-5"><a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="n">Cluster8.markers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="p">[</span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">cluster</span><span class="o">==</span><span class="m">8</span>
</span><span id="__span-11-6"><a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="w">                                           </span><span class="o">&amp;</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">p_val_adj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
</span></code></pre></div>
<h3 id="convert-gene-symbols-to-entrez-ids">Convert gene symbols to Entrez IDs<a class="headerlink" href="#convert-gene-symbols-to-entrez-ids" title="Permanent link">&para;</a></h3>
<p>Gene symbols are useful, because they can be easily understand by humans. But gene
symbols are not unique, so we need to convert them to Entrez IDs for gene
ontology analysis.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="n">Cluster2.gene_ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">bitr</span><span class="p">(</span><span class="n">Cluster2.markers</span><span class="o">$</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">fromType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span>
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="w">                                   </span><span class="n">toType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ENTREZID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</span><span id="__span-12-3"><a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>
</span><span id="__span-12-4"><a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-12-5"><a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="c1">## Warning in bitr(Cluster2.markers$gene, fromType = &quot;SYMBOL&quot;, toType =</span>
</span><span id="__span-12-6"><a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="c1">## &quot;ENTREZID&quot;, : 0.41% of input gene IDs are fail to map...</span>
</span><span id="__span-12-7"><a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a>
</span><span id="__span-12-8"><a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="n">Cluster3.gene_ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">bitr</span><span class="p">(</span><span class="n">Cluster3.markers</span><span class="o">$</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">fromType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span>
</span><span id="__span-12-9"><a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a><span class="w">                                   </span><span class="n">toType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ENTREZID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</span><span id="__span-12-10"><a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a>
</span><span id="__span-12-11"><a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-12-12"><a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a><span class="c1">## Warning in bitr(Cluster3.markers$gene, fromType = &quot;SYMBOL&quot;, toType =</span>
</span><span id="__span-12-13"><a id="__codelineno-12-13" name="__codelineno-12-13" href="#__codelineno-12-13"></a><span class="c1">## &quot;ENTREZID&quot;, : 1.01% of input gene IDs are fail to map...</span>
</span><span id="__span-12-14"><a id="__codelineno-12-14" name="__codelineno-12-14" href="#__codelineno-12-14"></a>
</span><span id="__span-12-15"><a id="__codelineno-12-15" name="__codelineno-12-15" href="#__codelineno-12-15"></a><span class="n">Cluster8.gene_ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">bitr</span><span class="p">(</span><span class="n">Cluster8.markers</span><span class="o">$</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">fromType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span>
</span><span id="__span-12-16"><a id="__codelineno-12-16" name="__codelineno-12-16" href="#__codelineno-12-16"></a><span class="w">                                   </span><span class="n">toType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ENTREZID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</span><span id="__span-12-17"><a id="__codelineno-12-17" name="__codelineno-12-17" href="#__codelineno-12-17"></a>
</span><span id="__span-12-18"><a id="__codelineno-12-18" name="__codelineno-12-18" href="#__codelineno-12-18"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-12-19"><a id="__codelineno-12-19" name="__codelineno-12-19" href="#__codelineno-12-19"></a><span class="c1">## Warning in bitr(Cluster8.markers$gene, fromType = &quot;SYMBOL&quot;, toType =</span>
</span><span id="__span-12-20"><a id="__codelineno-12-20" name="__codelineno-12-20" href="#__codelineno-12-20"></a><span class="c1">## &quot;ENTREZID&quot;, : 2.94% of input gene IDs are fail to map...</span>
</span></code></pre></div>
<h3 id="perform-go-enrichment-analysis-for-cluster-2">Perform GO enrichment analysis for Cluster 2<a class="headerlink" href="#perform-go-enrichment-analysis-for-cluster-2" title="Permanent link">&para;</a></h3>
<p>Let&rsquo;s begin with Cluster 2. This section demonstrates how to perform gene ontology
analysis for Cluster 2. You can follow the same steps to do the analysis
for Cluster 3 and Cluster 8.</p>
<p>We will use the <code>enrichGO()</code> function from <code>clusterProfiler</code> package to
perform gene ontology analysis.</p>
<p>Firstly, we need to load the gene ontology database. We use the
<code>org.Hs.eg.db</code> database for human gene ontology analysis. If you want to
do gene ontology analysis for other species, you can find the
corresponding database via Bioconductor.</p>
<p>We will first focus on biological process (BP) ontology, so
<code>ont = "BP"</code>. There are also another two databases (CC:
Cellular Component and MF: Molecular Function), you can try them later.</p>
<p>We will set the p-value cutoff and q-value cutoff to 0.05 to control the
false positive rate and false negative rate respectively. We will also
set the p-value adjustment method to &ldquo;BH&rdquo; (Benjamini &amp; Hochberg). If you
want to know more about p-value adjustment methods (why we need them and
what are they), see also:
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099145/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099145/</a>.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="n">Cluster2.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Cluster2.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="w">                </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-13-3"><a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="w">                </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BP&quot;</span><span class="p">,</span><span class="w"> </span><span class="c1"># biological process</span>
</span><span id="__span-13-4"><a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="w">                </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-13-5"><a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="w">                </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-13-6"><a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="w">                </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-13-7"><a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="w">                </span><span class="n">readable</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="visualise-the-go-enrichment-results">Visualise the GO enrichment results<a class="headerlink" href="#visualise-the-go-enrichment-results" title="Permanent link">&para;</a></h3>
<h4 id="bar-plot">Bar plot<a class="headerlink" href="#bar-plot" title="Permanent link">&para;</a></h4>
<p>A standard way to visualise GO enrichment results is to use a bar plot.
It can tell you how many genes in the marker gene list overlap with each GO term,
and how significant the enrichment is.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="nf">barplot</span><span class="p">(</span><span class="n">Cluster2.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 2&quot;</span><span class="p">)</span>
</span></code></pre></div>
<p><img alt="" src="../media/unnamed-chunk-12-1.png" /></p>
<h4 id="dot-plot">Dot plot<a class="headerlink" href="#dot-plot" title="Permanent link">&para;</a></h4>
<p>A dot plot is another way to visualise the GO enrichment results with more
information. It can show you the proportion of the overlapping genes, the 
overlapping gene counts, and the p-value for each GO term. You can also set
the number of GO terms you want to show.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">Cluster2.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 2&quot;</span><span class="p">)</span>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">))</span>
</span></code></pre></div>
<p><img alt="" src="../media/unnamed-chunk-13-1.png" /></p>
<h4 id="enrichment-map-plot">Enrichment map plot<a class="headerlink" href="#enrichment-map-plot" title="Permanent link">&para;</a></h4>
<p>GO terms are correlated with each other. The enrichment map plot can
show you the correlation between GO terms. It can help you to understand
the relationship between different biological processes.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="n">Cluster2.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">Cluster2.ego</span><span class="p">)</span>
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="nf">emapplot</span><span class="p">(</span><span class="n">Cluster2.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
</span></code></pre></div>
<p><img alt="" src="../media/unnamed-chunk-14-1.png" /></p>
<h4 id="cnet-plot">Cnet plot<a class="headerlink" href="#cnet-plot" title="Permanent link">&para;</a></h4>
<p>A Cnet plot can show you the relationship between GO terms and genes.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="nf">cnetplot</span><span class="p">(</span><span class="n">Cluster2.ego</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a><span class="w">         </span><span class="n">foldChange</span><span class="o">=</span><span class="n">Cluster2.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
</span></code></pre></div>
<p><img alt="" src="../media/unnamed-chunk-15-1.png" /></p>
<div class="admonition question">
<p class="admonition-title">Challenge</p>
<p>Perform GO enrichment analysis for the other 2 cluster.</p>
<details class="info">
<summary>Solution</summary>
<div class="language-r highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="n">Cluster3.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Cluster3.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="w">                </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a><span class="w">                </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BP&quot;</span><span class="p">,</span>
</span><span id="__span-18-4"><a id="__codelineno-18-4" name="__codelineno-18-4" href="#__codelineno-18-4"></a><span class="w">                </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-5"><a id="__codelineno-18-5" name="__codelineno-18-5" href="#__codelineno-18-5"></a><span class="w">                </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-6"><a id="__codelineno-18-6" name="__codelineno-18-6" href="#__codelineno-18-6"></a><span class="w">                </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-7"><a id="__codelineno-18-7" name="__codelineno-18-7" href="#__codelineno-18-7"></a><span class="w">                </span><span class="n">readable</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</span><span id="__span-18-8"><a id="__codelineno-18-8" name="__codelineno-18-8" href="#__codelineno-18-8"></a>
</span><span id="__span-18-9"><a id="__codelineno-18-9" name="__codelineno-18-9" href="#__codelineno-18-9"></a><span class="n">Cluster8.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Cluster8.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span>
</span><span id="__span-18-10"><a id="__codelineno-18-10" name="__codelineno-18-10" href="#__codelineno-18-10"></a><span class="w">                </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-11"><a id="__codelineno-18-11" name="__codelineno-18-11" href="#__codelineno-18-11"></a><span class="w">                </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BP&quot;</span><span class="p">,</span>
</span><span id="__span-18-12"><a id="__codelineno-18-12" name="__codelineno-18-12" href="#__codelineno-18-12"></a><span class="w">                </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-13"><a id="__codelineno-18-13" name="__codelineno-18-13" href="#__codelineno-18-13"></a><span class="w">                </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-14"><a id="__codelineno-18-14" name="__codelineno-18-14" href="#__codelineno-18-14"></a><span class="w">                </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-18-15"><a id="__codelineno-18-15" name="__codelineno-18-15" href="#__codelineno-18-15"></a><span class="w">                </span><span class="n">readable</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</span></code></pre></div>
</details>
</div>
<h3 id="visualise-the-go-enrichment-results-for-the-other-2-clusters">Visualise the GO enrichment results for the other 2 clusters<a class="headerlink" href="#visualise-the-go-enrichment-results-for-the-other-2-clusters" title="Permanent link">&para;</a></h3>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a><span class="nf">barplot</span><span class="p">(</span><span class="n">Cluster3.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-19-2"><a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 3&quot;</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-17-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a><span class="n">p_cluster3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">Cluster3.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-20-2"><a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 3&quot;</span><span class="p">)</span>
</span><span id="__span-20-3"><a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a><span class="n">p_cluster3</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">))</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-18-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a><span class="n">Cluster3.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">Cluster3.ego</span><span class="p">)</span>
</span><span id="__span-21-2"><a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a><span class="nf">emapplot</span><span class="p">(</span><span class="n">Cluster3.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-19-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="nf">cnetplot</span><span class="p">(</span><span class="n">Cluster3.ego</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-22-2"><a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a><span class="w">         </span><span class="n">foldChange</span><span class="o">=</span><span class="n">Cluster3.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
</span><span id="__span-22-3"><a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a>
</span><span id="__span-22-4"><a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a><span class="c1">## Warning in cnetplot.enrichResult(x, ...): Use &#39;color.params = list(foldChange = your_value)&#39; instead of &#39;foldChange&#39;.</span>
</span><span id="__span-22-5"><a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a><span class="c1">##  The foldChange parameter will be removed in the next version.</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-20-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a><span class="nf">barplot</span><span class="p">(</span><span class="n">Cluster8.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 8&quot;</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-21-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a><span class="n">p_cluster8</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">Cluster8.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-24-2"><a id="__codelineno-24-2" name="__codelineno-24-2" href="#__codelineno-24-2"></a><span class="w">        </span><span class="n">title</span><span class="o">=</span><span class="s">&quot;GO Enrichment Analysis for Cluster 8&quot;</span><span class="p">)</span>
</span><span id="__span-24-3"><a id="__codelineno-24-3" name="__codelineno-24-3" href="#__codelineno-24-3"></a><span class="n">p_cluster8</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">))</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-22-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a><span class="n">Cluster8.ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">Cluster8.ego</span><span class="p">)</span>
</span><span id="__span-25-2"><a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a><span class="nf">emapplot</span><span class="p">(</span><span class="n">Cluster8.ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-23-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a><span class="nf">cnetplot</span><span class="p">(</span><span class="n">Cluster8.ego</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-26-2"><a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a><span class="w">         </span><span class="n">foldChange</span><span class="o">=</span><span class="n">Cluster8.gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
</span><span id="__span-26-3"><a id="__codelineno-26-3" name="__codelineno-26-3" href="#__codelineno-26-3"></a>
</span><span id="__span-26-4"><a id="__codelineno-26-4" name="__codelineno-26-4" href="#__codelineno-26-4"></a><span class="c1">## Warning in cnetplot.enrichResult(x, ...): Use &#39;color.params = list(foldChange = your_value)&#39; instead of &#39;foldChange&#39;.</span>
</span><span id="__span-26-5"><a id="__codelineno-26-5" name="__codelineno-26-5" href="#__codelineno-26-5"></a><span class="c1">##  The foldChange parameter will be removed in the next version.</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-24-1.png" /></p>
<div class="admonition question">
<p class="admonition-title">Discussion</p>
<ul>
<li>What are the differences between the three clusters?</li>
<li>Which plots are most informative for you?</li>
<li>What are the most interesting GO terms you found?</li>
</ul>
</div>
<h2 id="perform-gene-ontology-analysis-for-all-clusters">Perform gene ontology analysis for all clusters<a class="headerlink" href="#perform-gene-ontology-analysis-for-all-clusters" title="Permanent link">&para;</a></h2>
<p>We have completed gene ontology analysis for each cluster. But we can also perform
gene ontology analysis for all clusters together. We can combine all
marker genes from all clusters and do gene ontology analysis for them.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a><span class="n">cluster_gene_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">split</span><span class="p">(</span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">marker.genes.pbmc.rna</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
</span><span id="__span-27-2"><a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a>
</span><span id="__span-27-3"><a id="__codelineno-27-3" name="__codelineno-27-3" href="#__codelineno-27-3"></a><span class="n">cluster_gene_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">cluster_gene_list</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
</span><span id="__span-27-4"><a id="__codelineno-27-4" name="__codelineno-27-4" href="#__codelineno-27-4"></a><span class="w">  </span><span class="n">gene_ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">bitr</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span><span class="w"> </span><span class="n">fromType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">toType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ENTREZID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</span><span id="__span-27-5"><a id="__codelineno-27-5" name="__codelineno-27-5" href="#__codelineno-27-5"></a><span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">gene_ids</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">)</span>
</span><span id="__span-27-6"><a id="__codelineno-27-6" name="__codelineno-27-6" href="#__codelineno-27-6"></a><span class="p">})</span>
</span><span id="__span-27-7"><a id="__codelineno-27-7" name="__codelineno-27-7" href="#__codelineno-27-7"></a>
</span><span id="__span-27-8"><a id="__codelineno-27-8" name="__codelineno-27-8" href="#__codelineno-27-8"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-9"><a id="__codelineno-27-9" name="__codelineno-27-9" href="#__codelineno-27-9"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-10"><a id="__codelineno-27-10" name="__codelineno-27-10" href="#__codelineno-27-10"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-11"><a id="__codelineno-27-11" name="__codelineno-27-11" href="#__codelineno-27-11"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-12"><a id="__codelineno-27-12" name="__codelineno-27-12" href="#__codelineno-27-12"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-13"><a id="__codelineno-27-13" name="__codelineno-27-13" href="#__codelineno-27-13"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-14"><a id="__codelineno-27-14" name="__codelineno-27-14" href="#__codelineno-27-14"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-15"><a id="__codelineno-27-15" name="__codelineno-27-15" href="#__codelineno-27-15"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-16"><a id="__codelineno-27-16" name="__codelineno-27-16" href="#__codelineno-27-16"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-17"><a id="__codelineno-27-17" name="__codelineno-27-17" href="#__codelineno-27-17"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-18"><a id="__codelineno-27-18" name="__codelineno-27-18" href="#__codelineno-27-18"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-19"><a id="__codelineno-27-19" name="__codelineno-27-19" href="#__codelineno-27-19"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-20"><a id="__codelineno-27-20" name="__codelineno-27-20" href="#__codelineno-27-20"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-21"><a id="__codelineno-27-21" name="__codelineno-27-21" href="#__codelineno-27-21"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-22"><a id="__codelineno-27-22" name="__codelineno-27-22" href="#__codelineno-27-22"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span><span id="__span-27-23"><a id="__codelineno-27-23" name="__codelineno-27-23" href="#__codelineno-27-23"></a><span class="c1">## &#39;select()&#39; returned 1:many mapping between keys and columns</span>
</span><span id="__span-27-24"><a id="__codelineno-27-24" name="__codelineno-27-24" href="#__codelineno-27-24"></a><span class="c1">## &#39;select()&#39; returned 1:1 mapping between keys and columns</span>
</span></code></pre></div>
<p>We can use the <code>compareCluster()</code> function from <code>clusterProfiler</code>
package to perform gene ontology analysis for all clusters together.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a><span class="c1"># GO enrichment analysis for BP</span>
</span><span id="__span-28-2"><a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a><span class="n">go_compare_bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cluster_gene_list</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-28-3"><a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a><span class="w">                             </span><span class="n">fun</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;enrichGO&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-28-4"><a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a><span class="w">                             </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-28-5"><a id="__codelineno-28-5" name="__codelineno-28-5" href="#__codelineno-28-5"></a><span class="w">                             </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BP&quot;</span><span class="p">,</span><span class="w">  </span><span class="c1"># &quot;BP&quot;、&quot;MF&quot; or &quot;CC&quot;</span>
</span><span id="__span-28-6"><a id="__codelineno-28-6" name="__codelineno-28-6" href="#__codelineno-28-6"></a><span class="w">                             </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span>
</span><span id="__span-28-7"><a id="__codelineno-28-7" name="__codelineno-28-7" href="#__codelineno-28-7"></a><span class="w">                             </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span>
</span><span id="__span-28-8"><a id="__codelineno-28-8" name="__codelineno-28-8" href="#__codelineno-28-8"></a><span class="w">                             </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="visualise-the-go-enrichment-results-for-all-clusters">Visualise the GO enrichment results for all clusters<a class="headerlink" href="#visualise-the-go-enrichment-results-for-all-clusters" title="Permanent link">&para;</a></h3>
<p>A dot plot is a good way to visualise the GO enrichment results for all
clusters. It can show you the proportion of overlap and p-value for each
GO term, for all clusters in a single plot.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a><span class="n">cluster.p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">go_compare_bp</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span>
</span><span id="__span-29-2"><a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a><span class="w">              </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;GO Enrichment (Biological Process)&quot;</span><span class="p">)</span>
</span><span id="__span-29-3"><a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a><span class="n">cluster.p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">))</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-27-1.png" /></p>
<p>But dot plots are not good for showing the relationship between GO terms.
We can select and cluster GO terms first, and then use a tree plot to
show the correlations between GO terms.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a><span class="n">trim_bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">go_compare_bp</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
</span><span id="__span-30-2"><a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a><span class="nf">treeplot</span><span class="p">(</span><span class="n">trim_bp</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>
</span><span id="__span-30-3"><a id="__codelineno-30-3" name="__codelineno-30-3" href="#__codelineno-30-3"></a>
</span><span id="__span-30-4"><a id="__codelineno-30-4" name="__codelineno-30-4" href="#__codelineno-30-4"></a><span class="c1">## Scale for fill is already present.</span>
</span><span id="__span-30-5"><a id="__codelineno-30-5" name="__codelineno-30-5" href="#__codelineno-30-5"></a><span class="c1">## Adding another scale for fill, which will replace the existing scale.</span>
</span><span id="__span-30-6"><a id="__codelineno-30-6" name="__codelineno-30-6" href="#__codelineno-30-6"></a><span class="c1">## Warning in (function (..., type = getOption(&quot;ggplot2.continuous.fill&quot;)) : log-10 transformation introduced infinite values.</span>
</span><span id="__span-30-7"><a id="__codelineno-30-7" name="__codelineno-30-7" href="#__codelineno-30-7"></a><span class="c1">## log-10 transformation introduced infinite values.</span>
</span></code></pre></div>
<p><img alt="" src="../media/unnamed-chunk-28-1.png" /></p>
<div class="admonition question">
<p class="admonition-title">Challenge</p>
<p>Perform GO enrichment analysis for other two ontologies.</p>
<details class="info">
<summary>Solution</summary>
<div class="language-r highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a><span class="c1"># GO enrichment analysis for CC</span>
</span><span id="__span-31-2"><a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a><span class="n">go_compare_cc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cluster_gene_list</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-3"><a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a><span class="w">                            </span><span class="n">fun</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;enrichGO&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-4"><a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a><span class="w">                            </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-5"><a id="__codelineno-31-5" name="__codelineno-31-5" href="#__codelineno-31-5"></a><span class="w">                            </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CC&quot;</span><span class="p">,</span><span class="w">  </span><span class="c1"># &quot;BP&quot;、&quot;MF&quot; or &quot;CC&quot;</span>
</span><span id="__span-31-6"><a id="__codelineno-31-6" name="__codelineno-31-6" href="#__codelineno-31-6"></a><span class="w">                            </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span>
</span><span id="__span-31-7"><a id="__codelineno-31-7" name="__codelineno-31-7" href="#__codelineno-31-7"></a><span class="w">                            </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span>
</span><span id="__span-31-8"><a id="__codelineno-31-8" name="__codelineno-31-8" href="#__codelineno-31-8"></a><span class="w">                            </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span>
</span><span id="__span-31-9"><a id="__codelineno-31-9" name="__codelineno-31-9" href="#__codelineno-31-9"></a>
</span><span id="__span-31-10"><a id="__codelineno-31-10" name="__codelineno-31-10" href="#__codelineno-31-10"></a><span class="c1"># GO enrichment analysis for MF</span>
</span><span id="__span-31-11"><a id="__codelineno-31-11" name="__codelineno-31-11" href="#__codelineno-31-11"></a><span class="n">go_compare_mf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cluster_gene_list</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-12"><a id="__codelineno-31-12" name="__codelineno-31-12" href="#__codelineno-31-12"></a><span class="w">                            </span><span class="n">fun</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;enrichGO&quot;</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-13"><a id="__codelineno-31-13" name="__codelineno-31-13" href="#__codelineno-31-13"></a><span class="w">                            </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-31-14"><a id="__codelineno-31-14" name="__codelineno-31-14" href="#__codelineno-31-14"></a><span class="w">                            </span><span class="n">ont</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;MF&quot;</span><span class="p">,</span><span class="w">  </span><span class="c1"># &quot;BP&quot;、&quot;MF&quot; or &quot;CC&quot;</span>
</span><span id="__span-31-15"><a id="__codelineno-31-15" name="__codelineno-31-15" href="#__codelineno-31-15"></a><span class="w">                            </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;BH&quot;</span><span class="p">,</span>
</span><span id="__span-31-16"><a id="__codelineno-31-16" name="__codelineno-31-16" href="#__codelineno-31-16"></a><span class="w">                            </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span>
</span><span id="__span-31-17"><a id="__codelineno-31-17" name="__codelineno-31-17" href="#__codelineno-31-17"></a><span class="w">                            </span><span class="n">qvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span>
</span></code></pre></div>
</details>
</div>
<h3 id="visualise-go-enrichment-results-for-other-two-ontology-categories">Visualise GO enrichment results for other two ontology categories<a class="headerlink" href="#visualise-go-enrichment-results-for-other-two-ontology-categories" title="Permanent link">&para;</a></h3>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a><span class="n">cluster.p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">go_compare_cc</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span>
</span><span id="__span-32-2"><a id="__codelineno-32-2" name="__codelineno-32-2" href="#__codelineno-32-2"></a><span class="w">              </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;GO Enrichment (Cellular Component)&quot;</span><span class="p">)</span>
</span><span id="__span-32-3"><a id="__codelineno-32-3" name="__codelineno-32-3" href="#__codelineno-32-3"></a>
</span><span id="__span-32-4"><a id="__codelineno-32-4" name="__codelineno-32-4" href="#__codelineno-32-4"></a><span class="n">cluster.p2</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">))</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-30-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a><span class="n">trim_cc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">go_compare_cc</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
</span><span id="__span-33-2"><a id="__codelineno-33-2" name="__codelineno-33-2" href="#__codelineno-33-2"></a><span class="nf">treeplot</span><span class="p">(</span><span class="n">trim_cc</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-31-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-34-1"><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a><span class="n">cluster.p3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="n">go_compare_mf</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span>
</span><span id="__span-34-2"><a id="__codelineno-34-2" name="__codelineno-34-2" href="#__codelineno-34-2"></a><span class="w">              </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;GO Enrichment (Molecular Function)&quot;</span><span class="p">)</span>
</span><span id="__span-34-3"><a id="__codelineno-34-3" name="__codelineno-34-3" href="#__codelineno-34-3"></a><span class="n">cluster.p3</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">))</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-32-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-35-1"><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a><span class="n">trim_mf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">go_compare_mf</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
</span><span id="__span-35-2"><a id="__codelineno-35-2" name="__codelineno-35-2" href="#__codelineno-35-2"></a><span class="nf">treeplot</span><span class="p">(</span><span class="n">trim_mf</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-33-1.png" /></p>
<h2 id="integrate-atac-seq-data-with-single-cell-rna-seq-data-n-integration">Integrate ATAC-seq data with single-cell RNA-seq data (N Integration)<a class="headerlink" href="#integrate-atac-seq-data-with-single-cell-rna-seq-data-n-integration" title="Permanent link">&para;</a></h2>
<p>From the gene ontology analysis, we found that the three clusters have
different biological functions. But we still don&rsquo;t know why they have
different functions. One possible reason is that they have different
gene activity. We can use the ATAC-seq data to quantify the gene
activity and integrate the ATAC-seq data with the single-cell RNA-seq
data to explain the differences.</p>
<h3 id="qc-for-atac-seq-data-already-performed-and-saved-in-the-dataset">QC for ATAC-seq data (already performed and saved in the dataset)<a class="headerlink" href="#qc-for-atac-seq-data-already-performed-and-saved-in-the-dataset" title="Permanent link">&para;</a></h3>
<p>Similar to the RNA data, we have completed the QC for the ATAC-seq data
and applied the label &ldquo;filtered&rdquo;. We move forward by 
selecting only cells that are not labelled &lsquo;filtered&rsquo;.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-36-1"><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">seurat_annotations</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s">&quot;filtered&quot;</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="preprocess-atac-seq-data">Preprocess ATAC-seq data<a class="headerlink" href="#preprocess-atac-seq-data" title="Permanent link">&para;</a></h3>
<p>We will perform the standard Seurat pipeline for single-cell ATAC analysis.
We will normalise the data, find top features, run SVD, and run UMAP.
See <code>Signac</code> (also provided by Satija&rsquo;s group) tutorial for more
details.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-37-1"><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a><span class="c1"># Add gene annotation information</span>
</span><span id="__span-37-2"><a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a><span class="n">annotations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetGRangesFromEnsDb</span><span class="p">(</span><span class="n">ensdb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
</span><span id="__span-37-3"><a id="__codelineno-37-3" name="__codelineno-37-3" href="#__codelineno-37-3"></a><span class="nf">seqlevelsStyle</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;UCSC&quot;</span>
</span><span id="__span-37-4"><a id="__codelineno-37-4" name="__codelineno-37-4" href="#__codelineno-37-4"></a><span class="nf">genome</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;hg38&quot;</span>
</span><span id="__span-37-5"><a id="__codelineno-37-5" name="__codelineno-37-5" href="#__codelineno-37-5"></a><span class="nf">Annotation</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">annotations</span>
</span><span id="__span-37-6"><a id="__codelineno-37-6" name="__codelineno-37-6" href="#__codelineno-37-6"></a>
</span><span id="__span-37-7"><a id="__codelineno-37-7" name="__codelineno-37-7" href="#__codelineno-37-7"></a><span class="c1"># We exclude the first dimension as this is typically correlated with sequencing depth</span>
</span><span id="__span-37-8"><a id="__codelineno-37-8" name="__codelineno-37-8" href="#__codelineno-37-8"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunTFIDF</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span>
</span><span id="__span-37-9"><a id="__codelineno-37-9" name="__codelineno-37-9" href="#__codelineno-37-9"></a>
</span><span id="__span-37-10"><a id="__codelineno-37-10" name="__codelineno-37-10" href="#__codelineno-37-10"></a><span class="c1">## Performing TF-IDF normalisation</span>
</span><span id="__span-37-11"><a id="__codelineno-37-11" name="__codelineno-37-11" href="#__codelineno-37-11"></a>
</span><span id="__span-37-12"><a id="__codelineno-37-12" name="__codelineno-37-12" href="#__codelineno-37-12"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindTopFeatures</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">min.cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;q0&quot;</span><span class="p">)</span>
</span><span id="__span-37-13"><a id="__codelineno-37-13" name="__codelineno-37-13" href="#__codelineno-37-13"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunSVD</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span>
</span><span id="__span-37-14"><a id="__codelineno-37-14" name="__codelineno-37-14" href="#__codelineno-37-14"></a>
</span><span id="__span-37-15"><a id="__codelineno-37-15" name="__codelineno-37-15" href="#__codelineno-37-15"></a><span class="c1">## Running SVD</span>
</span><span id="__span-37-16"><a id="__codelineno-37-16" name="__codelineno-37-16" href="#__codelineno-37-16"></a><span class="c1">## Scaling cell embeddings</span>
</span><span id="__span-37-17"><a id="__codelineno-37-17" name="__codelineno-37-17" href="#__codelineno-37-17"></a>
</span><span id="__span-37-18"><a id="__codelineno-37-18" name="__codelineno-37-18" href="#__codelineno-37-18"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunUMAP</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lsi&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="m">30</span><span class="p">,</span><span class="w"> </span>
</span><span id="__span-37-19"><a id="__codelineno-37-19" name="__codelineno-37-19" href="#__codelineno-37-19"></a><span class="w">                     </span><span class="n">reduction.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap.atac&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">reduction.key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;atacUMAP_&quot;</span><span class="p">)</span>
</span><span id="__span-37-20"><a id="__codelineno-37-20" name="__codelineno-37-20" href="#__codelineno-37-20"></a>
</span><span id="__span-37-21"><a id="__codelineno-37-21" name="__codelineno-37-21" href="#__codelineno-37-21"></a><span class="c1">## 11:52:35 UMAP embedding parameters a = 0.9922 b = 1.112</span>
</span><span id="__span-37-22"><a id="__codelineno-37-22" name="__codelineno-37-22" href="#__codelineno-37-22"></a><span class="c1">## 11:52:36 Read 10412 rows and found 29 numeric columns</span>
</span><span id="__span-37-23"><a id="__codelineno-37-23" name="__codelineno-37-23" href="#__codelineno-37-23"></a><span class="c1">## 11:52:36 Using Annoy for neighbor search, n_neighbors = 30</span>
</span><span id="__span-37-24"><a id="__codelineno-37-24" name="__codelineno-37-24" href="#__codelineno-37-24"></a><span class="c1">## 11:52:36 Building Annoy index with metric = cosine, n_trees = 50</span>
</span><span id="__span-37-25"><a id="__codelineno-37-25" name="__codelineno-37-25" href="#__codelineno-37-25"></a><span class="c1">## 0%   10   20   30   40   50   60   70   80   90   100%</span>
</span><span id="__span-37-26"><a id="__codelineno-37-26" name="__codelineno-37-26" href="#__codelineno-37-26"></a><span class="c1">## [----|----|----|----|----|----|----|----|----|----|</span>
</span><span id="__span-37-27"><a id="__codelineno-37-27" name="__codelineno-37-27" href="#__codelineno-37-27"></a><span class="c1">## **************************************************|</span>
</span><span id="__span-37-28"><a id="__codelineno-37-28" name="__codelineno-37-28" href="#__codelineno-37-28"></a><span class="c1">## 11:52:36 Writing NN index file to temp file /var/folders/pv/fvynh7953flggrfb49p2lqsc0000gn/T//RtmpuPPPuo/file4dc37680c360</span>
</span><span id="__span-37-29"><a id="__codelineno-37-29" name="__codelineno-37-29" href="#__codelineno-37-29"></a><span class="c1">## 11:52:36 Searching Annoy index using 1 thread, search_k = 3000</span>
</span><span id="__span-37-30"><a id="__codelineno-37-30" name="__codelineno-37-30" href="#__codelineno-37-30"></a><span class="c1">## 11:52:38 Annoy recall = 100%</span>
</span><span id="__span-37-31"><a id="__codelineno-37-31" name="__codelineno-37-31" href="#__codelineno-37-31"></a><span class="c1">## 11:52:39 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span>
</span><span id="__span-37-32"><a id="__codelineno-37-32" name="__codelineno-37-32" href="#__codelineno-37-32"></a><span class="c1">## 11:52:40 Initializing from normalized Laplacian + noise (using RSpectra)</span>
</span><span id="__span-37-33"><a id="__codelineno-37-33" name="__codelineno-37-33" href="#__codelineno-37-33"></a><span class="c1">## 11:52:40 Commencing optimization for 200 epochs, with 414208 positive edges</span>
</span><span id="__span-37-34"><a id="__codelineno-37-34" name="__codelineno-37-34" href="#__codelineno-37-34"></a><span class="c1">## 11:52:43 Optimization finished</span>
</span></code></pre></div>
<h3 id="visualise-the-atac-seq-data-with-scrna-seq-data">Visualise the ATAC-seq data with scRNA-seq data<a class="headerlink" href="#visualise-the-atac-seq-data-with-scrna-seq-data" title="Permanent link">&para;</a></h3>
<p>Next, we want a side-by-side overview of the single-cell ATAC-seq 
and RNA-seq datasets.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-38-1"><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a><span class="c1"># Visualise two datasets together</span>
</span><span id="__span-38-2"><a id="__codelineno-38-2" name="__codelineno-38-2" href="#__codelineno-38-2"></a><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_clusters&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">NoLegend</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;RNA&quot;</span><span class="p">)</span>
</span><span id="__span-38-3"><a id="__codelineno-38-3" name="__codelineno-38-3" href="#__codelineno-38-3"></a><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;orig.ident&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">NoLegend</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;ATAC&quot;</span><span class="p">)</span>
</span><span id="__span-38-4"><a id="__codelineno-38-4" name="__codelineno-38-4" href="#__codelineno-38-4"></a><span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p2</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-36-1.png" /></p>
<p>From the visualisation, we have the clustering information from the RNA
data, but we know nothing about the ATAC data.</p>
<p>We can see that the two datasets have different clustering patterns. We
will integrate the two datasets to explain the ATAC-seq clusters.</p>
<h3 id="quantify-gene-activity">Quantify gene activity<a class="headerlink" href="#quantify-gene-activity" title="Permanent link">&para;</a></h3>
<p>Before we integrate the two datasets, we need to quantify the gene
activity from the ATAC-seq data. We will use the <code>GeneActivity()</code>
function from <code>Signac</code> package to quantify the gene activity.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-39-1"><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a><span class="c1"># quantify gene activity</span>
</span><span id="__span-39-2"><a id="__codelineno-39-2" name="__codelineno-39-2" href="#__codelineno-39-2"></a><span class="n">gene.activities</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GeneActivity</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">pbmc.rna</span><span class="p">))</span>
</span><span id="__span-39-3"><a id="__codelineno-39-3" name="__codelineno-39-3" href="#__codelineno-39-3"></a>
</span><span id="__span-39-4"><a id="__codelineno-39-4" name="__codelineno-39-4" href="#__codelineno-39-4"></a><span class="c1">## Extracting gene coordinates</span>
</span><span id="__span-39-5"><a id="__codelineno-39-5" name="__codelineno-39-5" href="#__codelineno-39-5"></a><span class="c1">## Extracting reads overlapping genomic regions</span>
</span><span id="__span-39-6"><a id="__codelineno-39-6" name="__codelineno-39-6" href="#__codelineno-39-6"></a>
</span><span id="__span-39-7"><a id="__codelineno-39-7" name="__codelineno-39-7" href="#__codelineno-39-7"></a><span class="c1"># add gene activities as a new assay</span>
</span><span id="__span-39-8"><a id="__codelineno-39-8" name="__codelineno-39-8" href="#__codelineno-39-8"></a><span class="n">pbmc.atac</span><span class="p">[[</span><span class="s">&quot;ACTIVITY&quot;</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateAssayObject</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene.activities</span><span class="p">)</span>
</span><span id="__span-39-9"><a id="__codelineno-39-9" name="__codelineno-39-9" href="#__codelineno-39-9"></a>
</span><span id="__span-39-10"><a id="__codelineno-39-10" name="__codelineno-39-10" href="#__codelineno-39-10"></a><span class="c1"># normalise gene activities</span>
</span><span id="__span-39-11"><a id="__codelineno-39-11" name="__codelineno-39-11" href="#__codelineno-39-11"></a><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;ACTIVITY&quot;</span>
</span><span id="__span-39-12"><a id="__codelineno-39-12" name="__codelineno-39-12" href="#__codelineno-39-12"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span>
</span><span id="__span-39-13"><a id="__codelineno-39-13" name="__codelineno-39-13" href="#__codelineno-39-13"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">))</span>
</span><span id="__span-39-14"><a id="__codelineno-39-14" name="__codelineno-39-14" href="#__codelineno-39-14"></a>
</span><span id="__span-39-15"><a id="__codelineno-39-15" name="__codelineno-39-15" href="#__codelineno-39-15"></a><span class="c1">## Centering and scaling data matrix</span>
</span></code></pre></div>
<h3 id="integrate-rna-seq-data-with-atac-seq-data">Integrate RNA-seq data with ATAC-seq data<a class="headerlink" href="#integrate-rna-seq-data-with-atac-seq-data" title="Permanent link">&para;</a></h3>
<p>Similar to integrating two RNA-seq data, we will use the
<code>FindTransferAnchors()</code> function from the Seurat package to integrate the
RNA-seq data with the ATAC-seq data. We will use the CCA method to
integrate the two datasets. Then, we can transfer the clustering
information from the RNA-seq data to the ATAC-seq data.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-40-1"><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a><span class="c1"># Identify anchors</span>
</span><span id="__span-40-2"><a id="__codelineno-40-2" name="__codelineno-40-2" href="#__codelineno-40-2"></a><span class="n">transfer.anchors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindTransferAnchors</span><span class="p">(</span><span class="n">reference</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.rna</span><span class="p">,</span><span class="w"> </span><span class="n">query</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.rna</span><span class="p">),</span>
</span><span id="__span-40-3"><a id="__codelineno-40-3" name="__codelineno-40-3" href="#__codelineno-40-3"></a><span class="w">                                        </span><span class="n">reference.assay</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">query.assay</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ACTIVITY&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;cca&quot;</span><span class="p">)</span>
</span><span id="__span-40-4"><a id="__codelineno-40-4" name="__codelineno-40-4" href="#__codelineno-40-4"></a>
</span><span id="__span-40-5"><a id="__codelineno-40-5" name="__codelineno-40-5" href="#__codelineno-40-5"></a><span class="c1">## Warning: 714 features of the features specified were not present in both the reference query assays. </span>
</span><span id="__span-40-6"><a id="__codelineno-40-6" name="__codelineno-40-6" href="#__codelineno-40-6"></a><span class="c1">## Continuing with remaining 1286 features.</span>
</span><span id="__span-40-7"><a id="__codelineno-40-7" name="__codelineno-40-7" href="#__codelineno-40-7"></a><span class="c1">## Running CCA</span>
</span><span id="__span-40-8"><a id="__codelineno-40-8" name="__codelineno-40-8" href="#__codelineno-40-8"></a><span class="c1">## Merging objects</span>
</span><span id="__span-40-9"><a id="__codelineno-40-9" name="__codelineno-40-9" href="#__codelineno-40-9"></a><span class="c1">## Finding neighborhoods</span>
</span><span id="__span-40-10"><a id="__codelineno-40-10" name="__codelineno-40-10" href="#__codelineno-40-10"></a><span class="c1">## Finding anchors</span>
</span><span id="__span-40-11"><a id="__codelineno-40-11" name="__codelineno-40-11" href="#__codelineno-40-11"></a><span class="c1">##  Found 23956 anchors</span>
</span><span id="__span-40-12"><a id="__codelineno-40-12" name="__codelineno-40-12" href="#__codelineno-40-12"></a>
</span><span id="__span-40-13"><a id="__codelineno-40-13" name="__codelineno-40-13" href="#__codelineno-40-13"></a><span class="n">cluster.predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">TransferData</span><span class="p">(</span><span class="n">anchorset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transfer.anchors</span><span class="p">,</span><span class="w"> </span><span class="n">refdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.rna</span><span class="o">$</span><span class="n">seurat_clusters</span><span class="p">,</span>
</span><span id="__span-40-14"><a id="__codelineno-40-14" name="__codelineno-40-14" href="#__codelineno-40-14"></a><span class="w">                                     </span><span class="n">weight.reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="p">[[</span><span class="s">&quot;lsi&quot;</span><span class="p">]],</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span><span id="__span-40-15"><a id="__codelineno-40-15" name="__codelineno-40-15" href="#__codelineno-40-15"></a>
</span><span id="__span-40-16"><a id="__codelineno-40-16" name="__codelineno-40-16" href="#__codelineno-40-16"></a><span class="c1">## Finding integration vectors</span>
</span><span id="__span-40-17"><a id="__codelineno-40-17" name="__codelineno-40-17" href="#__codelineno-40-17"></a><span class="c1">## Finding integration vector weights</span>
</span><span id="__span-40-18"><a id="__codelineno-40-18" name="__codelineno-40-18" href="#__codelineno-40-18"></a><span class="c1">## Predicting cell labels</span>
</span><span id="__span-40-19"><a id="__codelineno-40-19" name="__codelineno-40-19" href="#__codelineno-40-19"></a>
</span><span id="__span-40-20"><a id="__codelineno-40-20" name="__codelineno-40-20" href="#__codelineno-40-20"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AddMetaData</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cluster.predictions</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="visualise-the-integrated-data">Visualise the integrated data<a class="headerlink" href="#visualise-the-integrated-data" title="Permanent link">&para;</a></h3>
<p>We are now able to visualise the integrated data. We can see the
predicted cluster annotation of the scATAC-seq data on the UMAP plot.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-41-1"><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a><span class="c1"># Visualise the integrated data</span>
</span><span id="__span-41-2"><a id="__codelineno-41-2" name="__codelineno-41-2" href="#__codelineno-41-2"></a><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">NoLegend</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Predicted clusters annotation&quot;</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-39-1.png" /></p>
<h3 id="visualise-the-atac-seq-regions-around-a-gene-of-interest">Visualise the ATAC-seq regions around a gene of interest<a class="headerlink" href="#visualise-the-atac-seq-regions-around-a-gene-of-interest" title="Permanent link">&para;</a></h3>
<p>We can also visualise the ATAC-seq regions around a gene of interest. We
will use two marker genes of T cells (CD4 and CD8). We can see the
ATAC-seq regions around the CD4 and CD8A genes differ for different cell types.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-42-1"><a id="__codelineno-42-1" name="__codelineno-42-1" href="#__codelineno-42-1"></a><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;ATAC&quot;</span>
</span><span id="__span-42-2"><a id="__codelineno-42-2" name="__codelineno-42-2" href="#__codelineno-42-2"></a><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span>
</span><span id="__span-42-3"><a id="__codelineno-42-3" name="__codelineno-42-3" href="#__codelineno-42-3"></a><span class="nf">CoveragePlot</span><span class="p">(</span>
</span><span id="__span-42-4"><a id="__codelineno-42-4" name="__codelineno-42-4" href="#__codelineno-42-4"></a><span class="w">  </span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span>
</span><span id="__span-42-5"><a id="__codelineno-42-5" name="__codelineno-42-5" href="#__codelineno-42-5"></a><span class="w">  </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;8&quot;</span><span class="p">),</span>
</span><span id="__span-42-6"><a id="__codelineno-42-6" name="__codelineno-42-6" href="#__codelineno-42-6"></a><span class="w">  </span><span class="n">extend.downstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="n">extend.upstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">)</span>
</span><span id="__span-42-7"><a id="__codelineno-42-7" name="__codelineno-42-7" href="#__codelineno-42-7"></a>
</span><span id="__span-42-8"><a id="__codelineno-42-8" name="__codelineno-42-8" href="#__codelineno-42-8"></a><span class="c1">## Warning: Removed 10 rows containing missing values or values outside the scale range</span>
</span><span id="__span-42-9"><a id="__codelineno-42-9" name="__codelineno-42-9" href="#__codelineno-42-9"></a><span class="c1">## (`geom_segment()`).</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-40-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-43-1"><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;ATAC&quot;</span>
</span><span id="__span-43-2"><a id="__codelineno-43-2" name="__codelineno-43-2" href="#__codelineno-43-2"></a><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span>
</span><span id="__span-43-3"><a id="__codelineno-43-3" name="__codelineno-43-3" href="#__codelineno-43-3"></a><span class="nf">CoveragePlot</span><span class="p">(</span>
</span><span id="__span-43-4"><a id="__codelineno-43-4" name="__codelineno-43-4" href="#__codelineno-43-4"></a><span class="w">  </span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD8A&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span>
</span><span id="__span-43-5"><a id="__codelineno-43-5" name="__codelineno-43-5" href="#__codelineno-43-5"></a><span class="w">  </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;8&quot;</span><span class="p">),</span>
</span><span id="__span-43-6"><a id="__codelineno-43-6" name="__codelineno-43-6" href="#__codelineno-43-6"></a><span class="w">  </span><span class="n">extend.downstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="n">extend.upstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-41-1.png" /></p>
<h3 id="additional-benchmark-study-evaluate-the-accuracy-of-the-integrated-data">Additional benchmark study: Evaluate the accuracy of the integrated data<a class="headerlink" href="#additional-benchmark-study-evaluate-the-accuracy-of-the-integrated-data" title="Permanent link">&para;</a></h3>
<p>We predicted the cell-type labels for the scATAC-seq data from the
scRNA-seq data. So, it is useful to know how accurate the predictions are. We can
evaluate the accuracy of the integrated data by comparing the predicted
labels with the ground-truth labels.</p>
<div class="language-r highlight"><pre><span></span><code><span id="__span-44-1"><a id="__codelineno-44-1" name="__codelineno-44-1" href="#__codelineno-44-1"></a><span class="n">celltype.predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">TransferData</span><span class="p">(</span><span class="n">anchorset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transfer.anchors</span><span class="p">,</span><span class="w"> </span><span class="n">refdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.rna</span><span class="o">$</span><span class="n">seurat_annotations</span><span class="p">,</span>
</span><span id="__span-44-2"><a id="__codelineno-44-2" name="__codelineno-44-2" href="#__codelineno-44-2"></a><span class="w">                                     </span><span class="n">weight.reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="p">[[</span><span class="s">&quot;lsi&quot;</span><span class="p">]],</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span><span id="__span-44-3"><a id="__codelineno-44-3" name="__codelineno-44-3" href="#__codelineno-44-3"></a>
</span><span id="__span-44-4"><a id="__codelineno-44-4" name="__codelineno-44-4" href="#__codelineno-44-4"></a><span class="c1">## Finding integration vectors</span>
</span><span id="__span-44-5"><a id="__codelineno-44-5" name="__codelineno-44-5" href="#__codelineno-44-5"></a><span class="c1">## Finding integration vector weights</span>
</span><span id="__span-44-6"><a id="__codelineno-44-6" name="__codelineno-44-6" href="#__codelineno-44-6"></a><span class="c1">## Predicting cell labels</span>
</span><span id="__span-44-7"><a id="__codelineno-44-7" name="__codelineno-44-7" href="#__codelineno-44-7"></a>
</span><span id="__span-44-8"><a id="__codelineno-44-8" name="__codelineno-44-8" href="#__codelineno-44-8"></a><span class="n">pbmc.atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AddMetaData</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">celltype.predictions</span><span class="p">)</span>
</span></code></pre></div>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-45-1"><a id="__codelineno-45-1" name="__codelineno-45-1" href="#__codelineno-45-1"></a><span class="c1"># Visualise the integrated data</span>
</span><span id="__span-45-2"><a id="__codelineno-45-2" name="__codelineno-45-2" href="#__codelineno-45-2"></a><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">annotation_correct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">predicted.id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">seurat_annotations</span>
</span><span id="__span-45-3"><a id="__codelineno-45-3" name="__codelineno-45-3" href="#__codelineno-45-3"></a><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">NoLegend</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Predicted annotation&quot;</span><span class="p">)</span>
</span><span id="__span-45-4"><a id="__codelineno-45-4" name="__codelineno-45-4" href="#__codelineno-45-4"></a><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_annotations&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">NoLegend</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Ground-truth annotation&quot;</span><span class="p">)</span>
</span><span id="__span-45-5"><a id="__codelineno-45-5" name="__codelineno-45-5" href="#__codelineno-45-5"></a><span class="n">p1</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">p2</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-43-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-46-1"><a id="__codelineno-46-1" name="__codelineno-46-1" href="#__codelineno-46-1"></a><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;ATAC&quot;</span>
</span><span id="__span-46-2"><a id="__codelineno-46-2" name="__codelineno-46-2" href="#__codelineno-46-2"></a><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span>
</span><span id="__span-46-3"><a id="__codelineno-46-3" name="__codelineno-46-3" href="#__codelineno-46-3"></a><span class="nf">CoveragePlot</span><span class="p">(</span>
</span><span id="__span-46-4"><a id="__codelineno-46-4" name="__codelineno-46-4" href="#__codelineno-46-4"></a><span class="w">  </span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span>
</span><span id="__span-46-5"><a id="__codelineno-46-5" name="__codelineno-46-5" href="#__codelineno-46-5"></a><span class="w">  </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD8 Naive&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8 TEM_1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8 TEM_2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 Naive&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 TCM&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 TEM&quot;</span><span class="p">),</span>
</span><span id="__span-46-6"><a id="__codelineno-46-6" name="__codelineno-46-6" href="#__codelineno-46-6"></a><span class="w">  </span><span class="n">extend.downstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="n">extend.upstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">)</span>
</span><span id="__span-46-7"><a id="__codelineno-46-7" name="__codelineno-46-7" href="#__codelineno-46-7"></a>
</span><span id="__span-46-8"><a id="__codelineno-46-8" name="__codelineno-46-8" href="#__codelineno-46-8"></a><span class="c1">## Warning: Removed 10 rows containing missing values or values outside the scale range</span>
</span><span id="__span-46-9"><a id="__codelineno-46-9" name="__codelineno-46-9" href="#__codelineno-46-9"></a><span class="c1">## (`geom_segment()`).</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-44-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-47-1"><a id="__codelineno-47-1" name="__codelineno-47-1" href="#__codelineno-47-1"></a><span class="nf">CoveragePlot</span><span class="p">(</span>
</span><span id="__span-47-2"><a id="__codelineno-47-2" name="__codelineno-47-2" href="#__codelineno-47-2"></a><span class="w">  </span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD8A&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;predicted.id&quot;</span><span class="p">,</span>
</span><span id="__span-47-3"><a id="__codelineno-47-3" name="__codelineno-47-3" href="#__codelineno-47-3"></a><span class="w">  </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD8 Naive&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8 TEM_1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8 TEM_2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 Naive&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 TCM&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 TEM&quot;</span><span class="p">),</span>
</span><span id="__span-47-4"><a id="__codelineno-47-4" name="__codelineno-47-4" href="#__codelineno-47-4"></a><span class="w">  </span><span class="n">extend.downstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="n">extend.upstream</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5000</span><span class="p">)</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-45-1.png" /></p>
<p>Overall, the visualisation of the integrated data shows that the
predicted labels are consistent with the ground-truth labels. The
coverage plots also show that the ATAC-seq regions around the CD4 and
CD8A genes are different for different cell types.</p>
<p>We can also use some quantitative metrics to evaluate the accuracy of
the integrated data. We can calculate the accuracy of the predicted
labels by comparing them with the ground-truth labels.</p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-48-1"><a id="__codelineno-48-1" name="__codelineno-48-1" href="#__codelineno-48-1"></a><span class="c1"># Evaluate the accuracy of the integrated data</span>
</span><span id="__span-48-2"><a id="__codelineno-48-2" name="__codelineno-48-2" href="#__codelineno-48-2"></a><span class="n">predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">table</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">seurat_annotations</span><span class="p">,</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">predicted.id</span><span class="p">)</span>
</span><span id="__span-48-3"><a id="__codelineno-48-3" name="__codelineno-48-3" href="#__codelineno-48-3"></a><span class="n">predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predictions</span><span class="o">/</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span><span class="w">  </span><span class="c1"># normalize for number of cells in each cell type</span>
</span><span id="__span-48-4"><a id="__codelineno-48-4" name="__codelineno-48-4" href="#__codelineno-48-4"></a><span class="n">predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>
</span><span id="__span-48-5"><a id="__codelineno-48-5" name="__codelineno-48-5" href="#__codelineno-48-5"></a><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">Var1</span><span class="p">,</span><span class="w"> </span><span class="n">Var2</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Freq</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_tile</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">scale_fill_gradient</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Fraction of cells&quot;</span><span class="p">,</span>
</span><span id="__span-48-6"><a id="__codelineno-48-6" name="__codelineno-48-6" href="#__codelineno-48-6"></a><span class="w">                                                                                            </span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#ffffc8&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#7d0025&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Cell type annotation (RNA)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;Predicted cell type label (ATAC)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
</span><span id="__span-48-7"><a id="__codelineno-48-7" name="__codelineno-48-7" href="#__codelineno-48-7"></a><span class="w">  </span><span class="nf">theme_cowplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
</span><span id="__span-48-8"><a id="__codelineno-48-8" name="__codelineno-48-8" href="#__codelineno-48-8"></a>
</span><span id="__span-48-9"><a id="__codelineno-48-9" name="__codelineno-48-9" href="#__codelineno-48-9"></a><span class="n">correct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">seurat_annotations</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">predicted.id</span><span class="p">))</span>
</span><span id="__span-48-10"><a id="__codelineno-48-10" name="__codelineno-48-10" href="#__codelineno-48-10"></a><span class="n">incorrect</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">seurat_annotations</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">pbmc.atac</span><span class="o">$</span><span class="n">predicted.id</span><span class="p">))</span>
</span><span id="__span-48-11"><a id="__codelineno-48-11" name="__codelineno-48-11" href="#__codelineno-48-11"></a><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FetchData</span><span class="p">(</span><span class="n">pbmc.atac</span><span class="p">,</span><span class="w"> </span><span class="n">vars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;prediction.score.max&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;annotation_correct&quot;</span><span class="p">))</span>
</span><span id="__span-48-12"><a id="__codelineno-48-12" name="__codelineno-48-12" href="#__codelineno-48-12"></a><span class="n">p1</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-46-1.png" /></p>
<p><div class="language-r highlight"><pre><span></span><code><span id="__span-49-1"><a id="__codelineno-49-1" name="__codelineno-49-1" href="#__codelineno-49-1"></a><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">prediction.score.max</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotation_correct</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotation_correct</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
</span><span id="__span-49-2"><a id="__codelineno-49-2" name="__codelineno-49-2" href="#__codelineno-49-2"></a><span class="w">  </span><span class="nf">geom_density</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme_cowplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">scale_fill_discrete</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Annotation Correct&quot;</span><span class="p">,</span>
</span><span id="__span-49-3"><a id="__codelineno-49-3" name="__codelineno-49-3" href="#__codelineno-49-3"></a><span class="w">                                                                    </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;FALSE (n = &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">incorrect</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">),</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;TRUE (n = &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">correct</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">scale_color_discrete</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="s">&quot;Annotation Correct&quot;</span><span class="p">,</span>
</span><span id="__span-49-4"><a id="__codelineno-49-4" name="__codelineno-49-4" href="#__codelineno-49-4"></a><span class="w">                                                                                                                                                                                  </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;FALSE (n = &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">incorrect</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">),</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;TRUE (n = &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">correct</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Prediction Score&quot;</span><span class="p">)</span>
</span><span id="__span-49-5"><a id="__codelineno-49-5" name="__codelineno-49-5" href="#__codelineno-49-5"></a>
</span><span id="__span-49-6"><a id="__codelineno-49-6" name="__codelineno-49-6" href="#__codelineno-49-6"></a><span class="n">p2</span>
</span></code></pre></div>
<img alt="" src="../media/unnamed-chunk-47-1.png" /></p>
<p>In this example, annotations for an scATAC-seq profile are correctly
predicted via scRNA-seq integration ~90% of the time. In addition, the
prediction.score.max field quantifies the uncertainty associated with
our predicted annotations. We can see that cells that are correctly
annotated are typically associated with high prediction scores (&gt;90%),
while cells that are incorrectly annotated are associated with sharply
lower prediction scores (&lt;50%). Incorrect assignments also tend to
reflect closely related cell types (i.e. Intermediate vs. Naive B
cells).</p>
<!--
## Glossary

A glossary at the end of the page is always helpful. This way you can refer to [terms](./template.md#glossary) throughout the lesson.

**Terms:** are written in bold in the glossary.
-->












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            <a href="https://www.melbournebioinformatics.org.au/">Melbourne Bioinformatics – The University of Melbourne ©</a>
          </div>
        
        Powered by <a href="https://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">Material for MkDocs</a>.
      </div>
      
        
<div class="md-social">
  
</div>
      
    </div>
  </div>
</footer>

      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["content.code.copy", "content.code.annotate"], "search": "../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../../javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>