
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../janis_translate/janis_translate/">
      
      
        <link rel="next" href="../../variant_calling_galaxy_1/variant_calling_galaxy_1/">
      
      
      <link rel="icon" href="../../../img/favicon-16x16.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Introduction to Variant detection - MBITE Tutorials</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<script async src="https://www.googletagmanager.com/gtag/js?id=G-EMNLJD0DWP"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }
    gtag('js', new Date());
    gtag('config', 'G-EMNLJD0DWP');
</script>

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction-to-variant-detection" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="MBITE Tutorials" class="md-header__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MBITE Tutorials
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Introduction to Variant detection
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="MBITE Tutorials" class="md-nav__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    MBITE Tutorials
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Popular
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Popular
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Git for Researchers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/nectar-instances/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to the Nectar Cloud
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to R for Biologists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to RNA-seq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/rna-seq-counts-to-genes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq: counts to genes
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tools and skill development
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tools and skill development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/bioinfo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    What is Bioinformatics?
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/galaxy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to the Galaxy Platform
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galaxy-workflows/galaxy-workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Galaxy Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_overview/python_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Python
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Unix
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro_R_biologists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to R for Biologists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Genome_browsers/GenomeBrowsers_Intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Genome Browsers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl_2022/cwl_2022/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to CWL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq Experimental Design
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../docker/docker/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Containerized Bioinformatics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../bionitio/bionitio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command-line software development
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../git/git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Version Control with Git
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../janis_translate/janis_translate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Translation With Janis
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Genomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Genomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Variant detection
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Variant detection
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Introduction to Variant detection
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Introduction to Variant detection
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#background" class="md-nav__link">
    <span class="md-ellipsis">
      Background
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Background">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-calling-vs-genotyping" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling vs genotyping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#process-of-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Process of variant calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#homozygous-or-heterozygous-mutations" class="md-nav__link">
    <span class="md-ellipsis">
      Homozygous or Heterozygous mutations:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-calling-software" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling Software:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variant Calling Software:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-samtools-mpileup-bcftools" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using Samtools (Mpileup + bcftools)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-gatk-unified-genotyper" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using GATK-Unified Genotyper
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-freebayes" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using FreeBayes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#evaluation-of-detected-variants-using-variant-eval" class="md-nav__link">
    <span class="md-ellipsis">
      Evaluation of detected variants using Variant Eval
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Matrix:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#local-realignment" class="md-nav__link">
    <span class="md-ellipsis">
      Local realignment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-galaxy-workflow-platform" class="md-nav__link">
    <span class="md-ellipsis">
      The Galaxy workflow platform
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-format-used-in-the-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      Data Format used in the tutorial
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Format used in the tutorial">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequence-alignment-map-format" class="md-nav__link">
    <span class="md-ellipsis">
      Sequence Alignment Map format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binary-sequence-alignment-map-format" class="md-nav__link">
    <span class="md-ellipsis">
      Binary Sequence Alignment Map format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      VCF file format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-format-in-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      VCF format in Galaxy:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bcf-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      Bcf file format:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_galaxy_1/variant_calling_galaxy_1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant Calling part 1 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../var_detect_advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant Calling part 2 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../longread_sv_calling/longread_sv_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Long-read Structural Variant Calling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_gatk1/variant_calling_gatk1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Variant calling using command-line tools
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    de novo assembly
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            de novo assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to de novo assembly with Velvet
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-protocol/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to de novo genome assembly for Illumina reads
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of Illumina reads using Velvet (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of Illumina reads using Spades (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pacbio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    de novo assembly of PacBio reads (unix)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../hybrid_assembly/nanopore_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid genome assembly - Nanopore and Illumina
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Metabarcoding
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Metabarcoding
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2/qiime2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to Metabarcoding using QIIME2: *Marine dataset*
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2_mammal/qiime2_mammal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro to Metabarcoding using QIIME2: *Mammal dataset*
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Transcriptomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Transcriptomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq Experimental Design
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna-seq-lcg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNA-seq: counts to genes
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Bulk RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Bulk RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to bulk RNAseq analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq analysis using HISAT2 (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tuxedo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq analysis using Tuxedo (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_advanced/rna_seq_advanced_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq differential expression tool comparision (Galaxy)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_in_r/rna_seq_r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RNAseq differential expression in R
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Single-cell RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Single-cell RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../singlecell/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    10X scRNA-seq analysis in R
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-de/seurat-de/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    scRNA-seq Integration and Differential Expression
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-go/seurat-go/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Gene Ontology (GO) and Data Integration for scRNA-seq
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proteomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Proteomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proteomics_basic/proteomics_basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Identifying proteins from mass spectrometry data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Statistics and Visualisation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Statistics and Visualisation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_viz/python_viz/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data visualisation with Python
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Structural Modelling
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Structural Modelling
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_101/molecular_dynamics_101/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Molecular Dynamics - Introduction to cluster computing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_201/molecular_dynamics_201/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Molecular Dynamics - Building input files, visualising the trajectory
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workshop Delivery Mode Information
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Workshop Delivery Mode Information
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workshop_delivery_mode_info/workshops_nectar/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workshops on Nectar Instances
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
        
          
          <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Archive
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_10">
            <span class="md-nav__icon md-icon"></span>
            Archive
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix_archive_2022/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Unix-Archive
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl/cwl/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to CWL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../using_git/Using_Git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Git and Github (deprecated)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#background" class="md-nav__link">
    <span class="md-ellipsis">
      Background
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Background">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-calling-vs-genotyping" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling vs genotyping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#process-of-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Process of variant calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#homozygous-or-heterozygous-mutations" class="md-nav__link">
    <span class="md-ellipsis">
      Homozygous or Heterozygous mutations:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-calling-software" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling Software:
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variant Calling Software:">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-samtools-mpileup-bcftools" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using Samtools (Mpileup + bcftools)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-gatk-unified-genotyper" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using GATK-Unified Genotyper
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-using-freebayes" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling using FreeBayes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#evaluation-of-detected-variants-using-variant-eval" class="md-nav__link">
    <span class="md-ellipsis">
      Evaluation of detected variants using Variant Eval
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quality-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Matrix:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#local-realignment" class="md-nav__link">
    <span class="md-ellipsis">
      Local realignment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-galaxy-workflow-platform" class="md-nav__link">
    <span class="md-ellipsis">
      The Galaxy workflow platform
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-format-used-in-the-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      Data Format used in the tutorial
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Format used in the tutorial">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequence-alignment-map-format" class="md-nav__link">
    <span class="md-ellipsis">
      Sequence Alignment Map format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binary-sequence-alignment-map-format" class="md-nav__link">
    <span class="md-ellipsis">
      Binary Sequence Alignment Map format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      VCF file format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vcf-format-in-galaxy" class="md-nav__link">
    <span class="md-ellipsis">
      VCF format in Galaxy:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bcf-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      Bcf file format:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="introduction-to-variant-detection">Introduction to Variant detection<a class="headerlink" href="#introduction-to-variant-detection" title="Permanent link">&para;</a></h1>
<h2 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h2>
<p>A variant is something that is different from a standard or type.</p>
<p>The aim of variation detection is to detect how many bases out of the total are different to a reference genome.</p>
<p>In Craig Venter’s genome 4.1 million DNA variants were reported.</p>
<p>What sort of variation could we find in the DNA sequencing?</p>
<ul>
<li>Single nucleotide variations (SNVs)</li>
<li>Single nucleotide polymorphisms (SNPs)</li>
<li>Small insertions and deletions (INDELs)</li>
<li>Large Chromosome rearrangements-Structural variations (SV)</li>
<li>Copy number variations (CNV)</li>
</ul>
<h3 id="variant-calling-vs-genotyping">Variant Calling vs genotyping<a class="headerlink" href="#variant-calling-vs-genotyping" title="Permanent link">&para;</a></h3>
<p>Variant calling is concerned with whether there is evidence of variant in a particular locus whereas genotyping talks about what the sets of alleles in that locus are and their frequencies. In haploid organisms variant calling and genotyping are equivalent whereas the same rule does not apply to other organisms.</p>
<p>Variant callers estimate the probability of a particular genotype given the observed data.</p>
<p>The question one would be asking is what possible genotypes would be possible for a sample. The remaining question is, given that our variant calling process calls a variant, does that mean that there is truly a variant in this locus and also given that the variant caller doesn’t detect a variant in a position does that mean there is no variant in that position.</p>
<p>The result of variant calling is a list of probable variants.</p>
<h3 id="process-of-variant-calling">Process of variant calling<a class="headerlink" href="#process-of-variant-calling" title="Permanent link">&para;</a></h3>
<p>Sample DNA -&gt; Sequencing -&gt; Read alignment -&gt; BAM file of aligned reads against reference genome -&gt; Genotyper -&gt; Variant list</p>
<p>The number of reads that stack up on each other is called <em>read coverage</em>. The data is converted into positional information of the reference with the read counts that have piled up under each position. Variant calling will look at how many bases out of the total number of bases is different to the reference at any position.</p>
<h3 id="homozygous-or-heterozygous-mutations">Homozygous or Heterozygous mutations:<a class="headerlink" href="#homozygous-or-heterozygous-mutations" title="Permanent link">&para;</a></h3>
<p>What should be noted about variants is that they are rare events and homozygous variants are even rarer than heterozygous events.</p>
<h2 id="variant-calling-software">Variant Calling Software:<a class="headerlink" href="#variant-calling-software" title="Permanent link">&para;</a></h2>
<p>There a number of software available for variant calling some of which are as follows:</p>
<ol>
<li>SAMtools (mpileup and bcftools): Li 2009 Bioinformatics</li>
<li>GATK: McKenna et al. 2010 Genome Res</li>
<li>FreeBayes: MIT</li>
<li>DiBayes: SOLiD software <a href="http://www.lifetechnologies.com">http://www.lifetechnologies.com</a></li>
<li>InGAP: Qi J, Zhao F, Buboltz A, Schuster SC.. 2009. <em>Bioinformatics</em></li>
<li>MAQGene: Bigelow H, Doitsidou M, Sarin S, Hobert O. 2009. <em>Nature Methods</em></li>
</ol>
<h3 id="variant-calling-using-samtools-mpileup-bcftools">Variant Calling using Samtools (Mpileup + bcftools)<a class="headerlink" href="#variant-calling-using-samtools-mpileup-bcftools" title="Permanent link">&para;</a></h3>
<p>Samtools calculates the genotype likelihoods. We then pipe the output to bcftools, which does our SNP calling based on those likelihoods.</p>
<p>Mpileup: Input: BAM file Output: Pileuped up reads under the reference</p>
<p>bcftools: Input: Pileup output from Mpileup Output: VCF file with sites and genotypes</p>
<p><a href="http://samtools.sourceforge.net/mpileup.shtml">Further information</a></p>
<h3 id="variant-calling-using-gatk-unified-genotyper">Variant Calling using GATK-Unified Genotyper<a class="headerlink" href="#variant-calling-using-gatk-unified-genotyper" title="Permanent link">&para;</a></h3>
<p>GATK is a programming framework based on the philosophy of MapReduce for developing NGS tools in a distributed or shared memory parallelized form.</p>
<p>GATK unified genotyper uses a Bayesian probabilistic model to calculate genotype likelihoods.</p>
<p>Inputs: BAM file</p>
<p>Output: VCF file with sites and genotypes.</p>
<p>The probability of a variant genotype for a given sequence of data is calculated using the <a href="http://en.wikipedia.org/wiki/Bayes%27_theorem">Bayes Theorem</a> as follows:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>P(Genotype | Data) =  (P(Data | Genotype) * P(Genotype)) / P(Data)
</span></code></pre></div>
<p>P(Genotype) is the overall probability of that genotype being present in a sequence. This is called the prior probability of a Genotype.</p>
<p>P(Data | Genotype) is the probability of the data (the reads) given the genotype</p>
<p>P(Data) is the probability of seeing the reads.</p>
<p>GATK unified genotyper is not very good in dealing with INDELs and thus we would only calculate SNPs throughout this tutorial. GATK is setup to work with diploid genomes but can be used on haploids as well.</p>
<p><a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_genotyper_UnifiedGenotyper.html">Further information</a></p>
<h3 id="variant-calling-using-freebayes">Variant Calling using FreeBayes<a class="headerlink" href="#variant-calling-using-freebayes" title="Permanent link">&para;</a></h3>
<p>FreeBayes is a high performance, flexible variant caller which uses the open source Freebayes tool to detect genetic variations based high throughput sequencing data (BAM files).</p>
<p><a href="http://bioinformatics.bc.edu/marthlab/FreeBayes">Further information</a></p>
<h3 id="evaluation-of-detected-variants-using-variant-eval">Evaluation of detected variants using Variant Eval<a class="headerlink" href="#evaluation-of-detected-variants-using-variant-eval" title="Permanent link">&para;</a></h3>
<p>The identified variation can further be evaluated against known variations such as common dbSNPs. The result can be checked for high concordance to the common SNPs or a known set of SNPs, the truth set.</p>
<p>The results will have:</p>
<ul>
<li>True Positives (TP): The variants called by the software which are also a known variant in the known variants file.</li>
<li>False Positives (FP): The Variants called by the software which are not known to be variants in the known variants file.</li>
<li>True Negatives (TN): The variants not called by the software which are not known to be variants in the known variants file.</li>
<li>False Negatives (FN): The variants not called by the software which are known as variants in the known variants file.</li>
</ul>
<h3 id="quality-matrix">Quality Matrix:<a class="headerlink" href="#quality-matrix" title="Permanent link">&para;</a></h3>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>TP | FP
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>---|----
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>TN | FN
</span></code></pre></div>
<p>Sensitivity: TP/(TP+FN)</p>
<p>Specificity: TN/(TN+FP)</p>
<p><em>Note:</em></p>
<p>Although software methods available can find variants in unique regions reliably, the short NGS read length prevent them from detecting variations in repetitive regions with comparable sensitivity.</p>
<p>DNA substitution mutations are of two types: Transitions and Transversions. The Ti/Tv ratio (Transitions/Transversions) is also an indicator of how well the model has performed for genotyping.</p>
<ul>
<li>
<p>Transition: a point mutation in which a purine nucleotide is changed to another purine nucleotide. (A\&lt;-&gt;G) or a pyrimidine nucleotide to another pyrimidine. Approximately 2 out of 3 SNPs are Transitions.</p>
</li>
<li>
<p>Transversion: a substitute of a purine for a pyrimidine.</p>
</li>
</ul>
<p>Although there are twice as many Transversions as there are Transitions because of the molecular mechanisms by which they are generated, Transition mutations occur at the higher rate than the Transversion mutations.</p>
<p>For more details on variant eval visit: <a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_varianteval_VariantEval.html">http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_varianteval_VariantEval.html</a></p>
<p><strong>Notes:</strong></p>
<p>An important thing worth noting is the more data the better the variant calling. In addition multisampling improves performance.</p>
<h3 id="local-realignment">Local realignment<a class="headerlink" href="#local-realignment" title="Permanent link">&para;</a></h3>
<p>In order to call SNPs close by INDELs correctly, local realignment is strongly recommended before variant calling when using both UnifiedGenotyper and FreeBayes. Samtools mpileup output would not however be affected since it works around this by introducing Base Alignment Quality (BAC). For more information on BAC refer to: <a href="http://samtools.sourceforge.net/mpileup.shtml">http://samtools.sourceforge.net/mpileup.shtml</a></p>
<h2 id="the-galaxy-workflow-platform">The Galaxy workflow platform<a class="headerlink" href="#the-galaxy-workflow-platform" title="Permanent link">&para;</a></h2>
<p>Galaxy is an online bioinformatics workflow management system. Essentially, you upload your files, create various analysis pipelines and run them, then visualise your results.</p>
<p>Galaxy is really an interface to the various tools that do the data processing; each of these tools could be run from the command line, outside of Galaxy. Galaxy makes it easier to link up the tools together and visualise the entire analysis pipeline.</p>
<p>Galaxy uses the concept of &lsquo;histories&rsquo;. Histories are sets of data and workflows that act on that data. The data for this workshop is available in a shared history, which you can import into your own Galaxy account</p>
<p><a href="http://wiki.g2.bx.psu.edu/">Learn more about Galaxy here</a></p>
<p><strong>The Galaxy interface. Tools on the left, data in the middle, analysis workflow on the right.</strong></p>
<p><img alt="" src="../media/image03.png" /></p>
<h2 id="data-format-used-in-the-tutorial">Data Format used in the tutorial<a class="headerlink" href="#data-format-used-in-the-tutorial" title="Permanent link">&para;</a></h2>
<h4 id="sequence-alignment-map-format"><strong>Sequence Alignment Map format</strong><a class="headerlink" href="#sequence-alignment-map-format" title="Permanent link">&para;</a></h4>
<p><strong>SAM format</strong></p>
<p>Sequence Alignment/Map format records all information relevant to how a set of reads aligns to a reference genome. A SAM file has an optional set of header lines describing the context of the alignment, then one line per read, with the following format:</p>
<ul>
<li>11 mandatory fields (+ variable number of optional fields)</li>
<li>1 QNAME: Query name of the read</li>
<li>2 FLAG</li>
<li>3 RNAME: Reference sequence name</li>
<li>4 POS: Position of alignment in reference sequence</li>
<li>5 MAPQ: Mapping quality (Phred-scaled)</li>
<li>6 CIGAR: String that describes the specifics of the alignment against the reference</li>
<li>7 MRNM</li>
<li>8 MPOS</li>
<li>9 ISIZE</li>
<li>10 SEQQuery: Sequence on the same strand as the reference</li>
<li>11 QUAL: Query quality (ASCII-33=Phred base quality)</li>
</ul>
<p><strong>SAM example</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>    SRR017937.312 16 chr20 43108717 37 76M * 0 0
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>    TGAGCCTCCGGGCTATGTGTGCTCACTGACAGAAGACCTGGTCACCAAAGCCCGGGAAGAGCTGCAGGAAAAGCCG
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>    ?,@A=A\&lt;5=,@==A:BB@=B9(.;A@B;\&gt;@ABBB@@9BB@:@5\&lt;BBBB9)\&gt;BBB2\&lt;BBB@BBB?;;BABBBBBBB@
</span></code></pre></div>
<p>For this example:</p>
<ul>
<li><strong>QNAME = SRR017937.312</strong> - this is the name of this read</li>
<li><strong>FLAG = 16</strong> - see the format description below</li>
<li><strong>RNAME = chr20</strong> - this read aligns to chromosome 20</li>
<li><strong>POS = 43108717</strong> - this read aligns the sequence on chr20 at position 43108717</li>
<li><strong>MAPQ = 37</strong> - this is quite a high quality score for the alignment (b/w 0 and 90)</li>
<li><strong>CIGAR = 76M</strong> - this read aligns to the reference segment across all bases (76 Matches means no deletions or insertions. Note that &lsquo;aligns&rsquo; can mean &lsquo;aligns with mismatches&rsquo; - mismatches that don&rsquo;t affect the alignment are not recorded in this field)</li>
<li><strong>MRNM = *</strong> - see the format description below</li>
<li><strong>MPOS = 0</strong> as there is no mate for this read - the sequenced DNA library was single ended, not mate paired*.</li>
<li><strong>ISIZE = 0</strong> as there is no mate for this read</li>
<li><strong>SEQQuery =</strong> the 76bp sequence of the reference segment</li>
<li><strong>QUAL =</strong> per-base quality scores for each position on the alignment. This is just a copy of what is in the FASTQ file</li>
</ul>
<p><strong>SAM</strong> format is described more fully <a href="http://samtools.sourceforge.net/SAM1.pdf"><em>here</em></a></p>
<p>NOTE: reads are shown mapped to the &ldquo;sense&rdquo; strand of the reference, and bases are listed in 5&rsquo; -&gt; 3&rsquo; order. This is important because an actual read might be from the other strand of DNA. The alignment tool will try to map the read as it is, and also the reverse compliment. If it was on the other strand then the reverse compliment is shown in the SAM file, rather than the original read itself</p>
<p>See<a href="http://www.illumina.com/technology/paired_end_sequencing_assay.ilmn">http://www.illumina.com/technology/paired_end_sequencing_assay.ilmn</a> for an overview of paired-end sequencing.</p>
<p><strong>SAM file in Galaxy</strong></p>
<p><img alt="" src="../media/image07.png" /></p>
<h4 id="binary-sequence-alignment-map-format">Binary Sequence Alignment Map format<a class="headerlink" href="#binary-sequence-alignment-map-format" title="Permanent link">&para;</a></h4>
<p><strong>BAM format</strong></p>
<p>SAM is a text format which is not space efficient. Binary Sequence Alignment is a compressed version of SAM.</p>
<p>Data in a BAM file is binary and therefore can&rsquo;t be visualised as text. If you try and visualise in Galaxy, it will default to downloading the file</p>
<p><strong>BAM file in IGV</strong></p>
<p><img alt="" src="../media/image01.png" /></p>
<h4 id="vcf-file-format">VCF file format<a class="headerlink" href="#vcf-file-format" title="Permanent link">&para;</a></h4>
<p><strong>What is VCF file:</strong></p>
<p>The <a href="http://vcftools.sourceforge.net/specs.html"><em>Variant Call Format</em></a> (VCF) is the emerging standard for storing variant data. Originally designed for SNPs and short INDELs, it also works for structural variations.</p>
<p>VCF consists of a header section and a data section. The header must contain a line starting with one &lsquo;#&lsquo;, showing the name of each field, and then the sample names starting at the 10<sup>th</sup> column. The data section is TAB delimited with each line consisting of at least 8 mandatory fields (the first 8 fields in the table below). The FORMAT field and sample information are allowed to be absent. We refer to the official <a href="http://vcftools.sourceforge.net/specs.html"><em>VCF spec</em></a> for a more rigorous description of the format.</p>
<table>
<thead>
<tr>
<th>Col</th>
<th>Field</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>CHROM</td>
<td>Chromosome name</td>
</tr>
<tr>
<td>2</td>
<td>POS</td>
<td>1-based position. For an indel, this is the position preceding the indel.</td>
</tr>
<tr>
<td>3</td>
<td>ID</td>
<td>Variant identifier. Usually the dbSNP rsID.</td>
</tr>
<tr>
<td>4</td>
<td>REF</td>
<td>Reference sequence at POS involved in the variant. For a SNP, it is a single base.</td>
</tr>
<tr>
<td>5</td>
<td>ALT</td>
<td>Comma delimited list of alternative sequence(s).</td>
</tr>
<tr>
<td>6</td>
<td>QUAL</td>
<td>Phred-scaled probability of all samples being homozygous reference.</td>
</tr>
<tr>
<td>7</td>
<td>FILTER</td>
<td>Semicolon delimited list of filters that the variant fails to pass.</td>
</tr>
<tr>
<td>8</td>
<td>INFO</td>
<td>Semicolon delimited list of variant information.</td>
</tr>
<tr>
<td>9</td>
<td>FORMAT</td>
<td>Colon delimited list of the format of individual genotypes in the following fields.</td>
</tr>
<tr>
<td>10+</td>
<td>Sample(s)</td>
<td>Individual genotype information defined by FORMAT.</td>
</tr>
</tbody>
</table>
<h4 id="vcf-format-in-galaxy">VCF format in Galaxy:<a class="headerlink" href="#vcf-format-in-galaxy" title="Permanent link">&para;</a></h4>
<p><img alt="" src="../media/image06.png" /></p>
<h4 id="bcf-file-format">Bcf file format:<a class="headerlink" href="#bcf-file-format" title="Permanent link">&para;</a></h4>
<h3 id="_1"><a class="headerlink" href="#_1" title="Permanent link">&para;</a></h3>
<p><strong>BCF format:</strong></p>
<p>BCF, or the binary variant call format, is the binary version of VCF. It keeps the same information in VCF, while much more efficient to process especially for many samples. The relationship between BCF and VCF is similar to that between BAM and SAM.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            <a href="https://www.melbournebioinformatics.org.au/">Melbourne Bioinformatics – The University of Melbourne ©</a>
          </div>
        
        Powered by <a href="https://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">Material for MkDocs</a>.
      </div>
      
        
<div class="md-social">
  
</div>
      
    </div>
  </div>
</footer>

      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["content.code.copy", "content.code.annotate"], "search": "../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../../javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>