
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
        
      
      
      <link rel="icon" href="../../../img/favicon-16x16.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Variant calling gatk1 print - MBITE Tutorials</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<script async src="https://www.googletagmanager.com/gtag/js?id=G-EMNLJD0DWP"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }
    gtag('js', new Date());
    gtag('config', 'G-EMNLJD0DWP');
</script>

    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#variant-calling-using-gatk4" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="MBITE Tutorials" class="md-header__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MBITE Tutorials
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Variant calling gatk1 print
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="MBITE Tutorials" class="md-nav__button md-logo" aria-label="MBITE Tutorials" data-md-component="logo">
      
  <img src="../../../img/apple-touch-icon.png" alt="logo">

    </a>
    MBITE Tutorials
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/melbournebioinformatics/MelBioInf_docs" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    melbournebioinformatics/MelBioInf_docs
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Popular
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Popular
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Git for Researchers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/nectar-instances/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to the Nectar Cloud
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to R for Biologists
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/rna-seq-counts-to-genes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq: counts to genes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-proteomics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Proteomics
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Tools and skill development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Tools and skill development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/bioinfo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    What is Bioinformatics?
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/galaxy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to the Galaxy Platform
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galaxy-workflows/galaxy-workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Galaxy Workflows
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_overview/python_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Python
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Unix
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro_R_biologists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to R for Biologists
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Genome_browsers/GenomeBrowsers_Intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Genome Browsers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl_2022/cwl_2022/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to CWL
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq Experimental Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../docker/docker/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Containerized Bioinformatics
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../bionitio/bionitio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Command-line software development
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../git/git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Version Control with Git
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../janis_translate/janis_translate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Workflow Translation With Janis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Genomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Genomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Variant detection
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Variant detection
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Variant detection
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../variant_calling_galaxy_1/variant_calling_galaxy_1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant Calling part 1 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../var_detect_advanced/var_detect_advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant Calling part 2 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../longread_sv_calling/longread_sv_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Long-read Structural Variant Calling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling_gatk1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Variant calling using command-line tools
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    de novo assembly
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to de novo assembly with Velvet
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly-protocol/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to de novo genome assembly for Illumina reads
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of Illumina reads using Velvet (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of Illumina reads using Spades (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pacbio/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    de novo assembly of PacBio reads (unix)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../hybrid_assembly/nanopore_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Hybrid genome assembly - Nanopore and Illumina
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Metabarcoding
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Metabarcoding
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2/qiime2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Metabarcoding using QIIME2: *Marine dataset*
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../qiime2_mammal/qiime2_mammal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intro to Metabarcoding using QIIME2: *Mammal dataset*
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Transcriptomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Transcriptomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="https://melbournebioinformatics.github.io/intro-to-rna-seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq Experimental Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna-seq-lcg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq: counts to genes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Bulk RNA-seq
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Bulk RNA-seq
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to bulk RNAseq analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq analysis using HISAT2 (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_basic/rna_seq_basic_tuxedo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq analysis using Tuxedo (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_advanced/rna_seq_advanced_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq differential expression tool comparision (Galaxy)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_dge_in_r/rna_seq_r/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNAseq differential expression in R
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Single-cell RNA-seq
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Single-cell RNA-seq
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../singlecell/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    10X scRNA-seq analysis in R
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-de/seurat-de/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    scRNA-seq Integration and Differential Expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../seurat-go/seurat-go/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Gene Ontology (GO) and Data Integration for scRNA-seq
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Proteomics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Proteomics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proteomics_basic/proteomics_basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Identifying proteins from mass spectrometry data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Statistics and Visualisation
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    Statistics and Visualisation
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../python_viz/python_viz/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Data visualisation with Python
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Structural Modelling
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Structural Modelling
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_101/molecular_dynamics_101/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Molecular Dynamics - Introduction to cluster computing
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../molecular_dynamics_201/molecular_dynamics_201/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Molecular Dynamics - Building input files, visualising the trajectory
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Workshop Delivery Mode Information
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Workshop Delivery Mode Information
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workshop_delivery_mode_info/workshops_nectar/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Workshops on Nectar Instances
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
        
          
          <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Archive
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_10">
            <span class="md-nav__icon md-icon"></span>
            
  
    Archive
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unix_archive_2022/unix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Unix-Archive
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cwl/cwl/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to CWL
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../using_git/Using_Git/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction to Git and Github (deprecated)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rna_seq_exp_design/rna_seq_experimental_design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA-seq Experimental Design (deprecated)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    <span class="md-ellipsis">
      
        Author Information
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    <span class="md-ellipsis">
      
        Learning Objectives
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#description" class="md-nav__link">
    <span class="md-ellipsis">
      
        Description
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements-and-preparation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Requirements and preparation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Requirements and preparation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mode-of-delivery" class="md-nav__link">
    <span class="md-ellipsis">
      
        Mode of Delivery
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#byobu-screen" class="md-nav__link">
    <span class="md-ellipsis">
      
        Byobu-screen
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-setting" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial setting
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tutorial setting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-genome-analysis-toolkit-gatk" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Genome Analysis Toolkit (GATK)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-this-tutorial-works" class="md-nav__link">
    <span class="md-ellipsis">
      
        How this tutorial works
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-contents-table" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial contents table
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-1-map-raw-mapped-reads-to-reference-genome" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 1: Map raw mapped reads to reference genome
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 1: Map raw mapped reads to reference genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-preparation-and-data-import" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Preparation and data import
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-align-genome" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Align genome
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-2-prepare-analysis-ready-reads" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 2: Prepare analysis ready reads
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2: Prepare analysis ready reads">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-sort-sambam" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Sort SAM/BAM
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-mark-duplicate-reads" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Mark duplicate reads
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-base-quality-recalibration" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Base quality recalibration
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-3-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 3: Variant calling
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 3: Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-apply-haplotypecaller" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Apply HaplotypeCaller
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-apply-combinegvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Apply CombineGVCFs
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-apply-genotypegvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Apply GenotypeGVCFs
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-4-filter-and-prepare-analysis-ready-variants" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 4: Filter and prepare analysis ready variants
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 4: Filter and prepare analysis ready variants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variant-quality-score-recalibration" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Variant Quality Score Recalibration
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-additional-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Additional filtering
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-final-analysis-ready-vcf-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Final analysis ready VCF file
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-5-exporting-variant-data-and-visualisation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Section 5: Exporting variant data and visualisation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 5: Exporting variant data and visualisation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variantstotable" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. VariantsToTable
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-html-report" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. HTML report
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<p><img alt="melbioinf_logo" src="../../media/melbioinf_logo.png" style="width:350px; padding-right:50px" />       <img alt="unimelb_logo" src="../../media/PRIMARY_A_Vertical_Housed_RGB.png" style="width:150px" /></p>
<!-- pre code {
  white-space: pre-wrap;
} -->

<h1 id="variant-calling-using-gatk4">Variant calling using GATK4<a class="headerlink" href="#variant-calling-using-gatk4" title="Permanent link">&para;</a></h1>
<!-- <mark>This workshop is under development and testing and may contain bugs.</mark> -->
<!-- {â€‹==**This workshop is under development and testing and may contain bugs.**==} -->

<p>Anticipated workshop duration when delivered to a group of participants is <strong>4 hours</strong>.</p>
<p>For queries relating to this workshop, contact Melbourne Bioinformatics (<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#98;&#105;&#111;&#105;&#110;&#102;&#111;&#114;&#109;&#97;&#116;&#105;&#99;&#115;&#45;&#116;&#114;&#97;&#105;&#110;&#105;&#110;&#103;&#64;&#117;&#110;&#105;&#109;&#101;&#108;&#98;&#46;&#101;&#100;&#117;&#46;&#97;&#117;">&#98;&#105;&#111;&#105;&#110;&#102;&#111;&#114;&#109;&#97;&#116;&#105;&#99;&#115;&#45;&#116;&#114;&#97;&#105;&#110;&#105;&#110;&#103;&#64;&#117;&#110;&#105;&#109;&#101;&#108;&#98;&#46;&#101;&#100;&#117;&#46;&#97;&#117;</a>).</p>
<h2 id="author-information">Author Information<a class="headerlink" href="#author-information" title="Permanent link">&para;</a></h2>
<p>Khalid Mahmood<br />
Melbourne Bioinformatics, University of Melbourne<br />
Developed: September 2021<br />
Reviewed: June 2023</p>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p><em>Topic</em></p>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Genomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Transcriptomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Proteomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Metabolomics</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Statistics and visualisation</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Structural Modelling</li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Basic skills</li>
</ul>
<p><em>Skill level</em></p>
<ul class="task-list">
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Beginner  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled checked/><span class="task-list-indicator"></span></label> Intermediate  </li>
<li class="task-list-item"><label class="task-list-control"><input type="checkbox" disabled/><span class="task-list-indicator"></span></label> Advanced  </li>
</ul>
<p><br>
This workshop is designed for participants with some command-line knowledge. You will need to be able to <code>ssh</code> into a remote machine, navigate the directory structure and <code>scp</code> files from a remote computer to your local computer.</p>
<h2 id="learning-objectives">Learning Objectives<a class="headerlink" href="#learning-objectives" title="Permanent link">&para;</a></h2>
<p>At the end of this workshop, you will be able to:</p>
<ul>
<li>Take raw DNA sequencing reads and perform variant calling to produce a variant list using GATK4.</li>
<li>Perform basic exploration of variants.</li>
</ul>
<h2 id="description">Description<a class="headerlink" href="#description" title="Permanent link">&para;</a></h2>
<p>This tutorial runs through the GATK4 best practices workflow for variant calling. The workflow starts and a pair of sequencing reads and performs a series of steps to determine a set to genetic variants.</p>
<p><strong>Data:</strong> Illumina HiSeq paired-end (2Ã—100 bp) reads in FASTQ format.<br />
<strong>Tools:</strong> GATK4, Picard, Bcftools and jigv<br />
<strong>Reference data:</strong> GATK4 hg38 reference bundle and hg38 refGene annotation (hg38.refGene.gtf.gz)</p>
<h2 id="requirements-and-preparation">Requirements and preparation<a class="headerlink" href="#requirements-and-preparation" title="Permanent link">&para;</a></h2>
<div class="admonition attention">
<p class="admonition-title">Important</p>
<p><strong>Attendees are required to use their own laptop computers.</strong>  </p>
<p>At least one week before the workshop, if required, participants should install the software below.  This should provide sufficient time for participants to liaise with their own IT support should they encounter any IT problems.  </p>
</div>
<p><strong>Software</strong>  </p>
<p><em>Mac Users:</em> No additional software needs to be installed for this workshop.</p>
<p><em>Windows Users:</em><br />
1. A terminal emulator such as <a href="https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html">PuTTY</a>(free and open-source) will need to be downloaded.<br />
2. Software for file transfers between a local computer and remote server such as <a href="https://winscp.net/eng/index.php">WinSCP</a> or <a href="https://filezilla-project.org/">FileZilla</a>.</p>
<p><strong>Data</strong></p>
<ul>
<li>All required data will be made available on the workshop virtual machines (hosted at the Melbourne Research Cloud). Login details will be provided closer to the tutorial data.</li>
</ul>
<!-- * If you wish to analyse the data independently at a later stage, it can be downloaded from [here](https://github.com/melbournebioinformatics/MelBioInf_docs/blob/29266406cb16cf376ef5f3d48e9bf8ac3578f1b0/docs/tutorials/qiime2/raw_data.zip). This zipped folder contains both the FASTQs and associated metadata file.     -->

<h3 id="mode-of-delivery">Mode of Delivery<a class="headerlink" href="#mode-of-delivery" title="Permanent link">&para;</a></h3>
<p>This workshop will be run on a <a href="https://cloud.nectar.org.au/">Nectar</a> Instance. An â€œInstanceâ€ is Nectar terminology for a virtual machine running on the Nectar Cloud OpenStack infrastructure. An â€œInstanceâ€ runs on a â€œcompute nodeâ€; i.e. a physical computer populated with processor chips, memory chips and so on.</p>
<p>You will be given an individual username, IP address and password to log on to using the SSH client tool on your computer (Terminal on Mac or PuTTY on Windows).</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>ssh<span class="w"> </span>username@ip-address
</span></code></pre></div>
<!-- <br>
Should you wish to do this tutorial at a later stage independently, it is possible to apply for your own instance directly through a [Nectar allocation](https://support.ehelp.edu.au/support/solutions/articles/6000068044-managing-an-allocation). There are also many helpful [Nectar Research Cloud tutorials](https://tutorials.rc.nectar.org.au/). -->

<h3 id="byobu-screen">Byobu-screen<a class="headerlink" href="#byobu-screen" title="Permanent link">&para;</a></h3>
<p>Some of the commands in this tutorial take a while to run. Should your ssh connection drop and the SSH session on Nectar terminates, any commands that are running will terminate too. To mitigate this, once logged on to the Nectar Instance, we&rsquo;ll run <code>byobu-screen</code> (an enhancement for the <code>screen</code> terminal multiplexer) which allows us to resume a session. In other words, processes running in <code>byobu-screen</code> will continue to run when their window is not visible, even if you get disconnected.</p>
<p>On Nectar, to start a <code>byobu-screen</code> session called <code>workshop</code>, type  </p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>byobu-screen<span class="w"> </span>-S<span class="w"> </span>workshop
</span></code></pre></div>
<p>You can then proceed to run the commands in the workshop as normal.</p>
<details class="example">
<summary>Help: Byobu basics</summary>
<p>The screen should look like
<img alt="Byobu" src="../media/byobu.png" />
<br>
Should your SSH session on Nectar terminate, once you log back in to your Nectar instance, list running sessions/screens:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>byobu-screen<span class="w"> </span>-ls
</span></code></pre></div>
<p>If it says (Detached) next to the <code>workshop</code> session in the list, reattach to <code>workshop</code> by:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>byobu-screen<span class="w"> </span>-r<span class="w"> </span>workshop
</span></code></pre></div>
<p>If it says (Attached) next to the <code>workshop</code> session in the list, you can access the <code>workshop</code> session which is already attached by:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>byobu-screen<span class="w"> </span>-r<span class="w"> </span>-d<span class="w"> </span>workshop
</span></code></pre></div>
<p><br>
Some other useful <code>byobu-screen</code> commands:</p>
<ul>
<li>
<p>To detach from <code>workshop</code>, type <code>ctrl-a ctrl-d</code> while inside the <code>workshop</code> session.
(You will need to configure Byobu&rsquo;s ctrl-a behaviour if it hasn&rsquo;t already been configured (text will appear on the screen telling you this). Follow the information on the screen and select <code>1</code> for Screen mode).</p>
</li>
<li>
<p>To terminate <code>workshop</code>, type <code>ctrl-d</code> while inside the <code>workshop</code> session.</p>
</li>
</ul>
</details>
<h2 id="tutorial-setting">Tutorial setting<a class="headerlink" href="#tutorial-setting" title="Permanent link">&para;</a></h2>
<p>The setting for this tutorial is as follows: you receive some germline sequencing data from an individual <a href="https://www.internationalgenome.org/data-portal/sample/NA12878">NA12878</a> (chr20 from sample NA12878 is sourced from the International Genome Sample Resource). Your aim is to identify small genetic variants from this data.</p>
<h3 id="the-genome-analysis-toolkit-gatk">The Genome Analysis Toolkit (GATK)<a class="headerlink" href="#the-genome-analysis-toolkit-gatk" title="Permanent link">&para;</a></h3>
<p>GATK is the industry standard toolkit for analysis of germline DNA to identify SNVs and indels. The GATK tool is mainly designed fo the human whole genome and exome analysis. The current version (GATK4) has expanded scope now and includes more complex analysis such copy number (CNV), structural variant (SV) and somatic variants. The current version of GATK also includes several utility functions for processing alignment files, VCF files and other complex processing workflows. The Picard toolkit is also now incorporated with GATK4.
Van der Auwera GA &amp; O&rsquo;Connor BD. (2020). Genomics in the Cloud: Using Docker, GATK, and WDL in Terra (1<sup>st</sup> Edition). O&rsquo;Reilly Media.</p>
<h3 id="how-this-tutorial-works">How this tutorial works<a class="headerlink" href="#how-this-tutorial-works" title="Permanent link">&para;</a></h3>
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>These grey coloured boxes are code blocks. The rectangular boxes in the top
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>right hand corner of this code block/grey box can be used to copy the code to
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>the clipboard.
</span></code></pre></div>
<details class="example">
<summary><strong>Click here</strong> </summary>
<p>REVEALED!
Coloured boxes like these when clicked will reveal their content.</p>
</details>
<p>To copy data from the remote workshop computer to your local computer.</p>
<p><strong>Command line/Mac Users</strong></p>
<p>The syntax to do this depends on whether you are running the copying command on your local computer, or on the remote computer (Nectar cloud).</p>
<ol>
<li>
<p>When running the command from your local computer, the syntax for copying a file <em>from</em> Nectar is:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>scp<span class="w"> </span>username@nectar_IP_address:FILENAME<span class="w"> </span>/PATH/TO/TARGET/FOLDER/
</span></code></pre></div>
</li>
<li>
<p>Running the command on the remote computer, the syntax for copying a file <em>to</em> your local computer is:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>scp<span class="w"> </span>FILENAME<span class="w"> </span>username@your_IP_address:/PATH/TO/TARGET/FOLDER/
</span></code></pre></div></p>
</li>
</ol>
<p><strong>SFTP clients</strong></p>
<p>Using WinSCP or FileZilla or CyberDuck you will need the following details to connect and browse.</p>
<ol>
<li>Host: The IP address of the Nectar instance</li>
<li>Username:</li>
<li>Port:</li>
</ol>
<h2 id="tutorial-contents-table">Tutorial contents table<a class="headerlink" href="#tutorial-contents-table" title="Permanent link">&para;</a></h2>
<ul>
<li><a href="#section-1-map-raw-mapped-reads-to-reference-genome">Section 1: Map raw mapped reads to reference genome</a></li>
<li><a href="#section-2-prepare-analysis-ready-reads">Section 2: Prepare analysis ready reads</a></li>
<li><a href="#section-3-variant-calling">Section 3: Variant calling</a></li>
<li><a href="#section-4-filter-and-prepare-analysis-ready-variants">Section 4: Filter and prepare analysis ready variants</a></li>
<li><a href="#section-5-exporting-variant-data-and-visualisation">Section 5: Exporting variant data and visualisation</a></li>
</ul>
<h2 id="section-1-map-raw-mapped-reads-to-reference-genome">Section 1: Map raw mapped reads to reference genome<a class="headerlink" href="#section-1-map-raw-mapped-reads-to-reference-genome" title="Permanent link">&para;</a></h2>
<h3 id="1-preparation-and-data-import">1. Preparation and data import<a class="headerlink" href="#1-preparation-and-data-import" title="Permanent link">&para;</a></h3>
<p>Lets start by setting up the directory structure for this analysis.
The <code class="language-bash highlight">data</code> directory is where the raw sequencing reads will be. All output files will be directed to the <code class="language-bash highlight">output</code> directory. All reference files will be pointed to the <code class="language-bash highlight">reference</code> directory. The command-line scripts are stored in simple bash script format in the <code class="language-bash highlight">scripts</code> directory. For those interested, equivalent slurm scripts to run on Spartan are available in the <code class="language-bash highlight">slurm_scripts</code> directory. Although all tools are installed on the server, we will create a <code class="language-bash highlight">tools</code> directory.    </p>
<p>Lets begin by creating a byobu-screen session (see above sections for more help):
<div class="language-Bash highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="nb">cd</span>
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>byobu-screen<span class="w"> </span>-S<span class="w"> </span>workshop
</span></code></pre></div></p>
<p>Create workshop directories:
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>mkdir data
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>mkdir output
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>mkdir reference
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>mkdir reference/hg38
</span><span id="__span-9-5"><a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a>mkdir scripts
</span><span id="__span-9-6"><a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a>mkdir slurm_scripts
</span><span id="__span-9-7"><a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a>mkdir temp
</span><span id="__span-9-8"><a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a>mkdir tools
</span></code></pre></div></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All analysis is being carried out in the home directory (the directory you log in to).</p>
</div>
<p>The data for this tutorial is sourced from the <a href="https://www.internationalgenome.org/data-portal/sample/NA12878">International Genome Sample Resources</a>. Raw sequencing reads from chromosome 20 are used in this tutorial. We have prepared the files which can be copied as follows:</p>
<div class="language-Bash highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>cp<span class="w"> </span>-p<span class="w"> </span>/mnt/shared_data/NA12878.chr20.region_1.fastq.gz<span class="w"> </span>data/.
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>cp<span class="w"> </span>-p<span class="w"> </span>/mnt/shared_data/NA12878.chr20.region_2.fastq.gz<span class="w"> </span>data/.
</span></code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To perform quality control checks on the raw fastq data use the tool <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. Another useful QC tool output aggregator is the <a href="https://multiqc.info/">MultiQC</a> tool. MultiQC aggregates the output from several tools and outputs a single QC report for all samples. We will have a look at some of the QC data later in this section.</p>
</div>
<p>Next, we need to prepare the reference data. Luckily, we have downloaded the data and all we need to do is to create a <a href="https://kb.iu.edu/d/abbe">symbolic link</a> to the data folder as follows:</p>
<div class="language-Bash highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>ln<span class="w"> </span>-s<span class="w"> </span>/mnt/shared_data/*<span class="w"> </span>reference/hg38/.
</span></code></pre></div>
<p>There are several files in the reference directory. These included the GATK bundle of reference files downloaded from (<a href="ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/hg38/">ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/hg38/</a>). Additional files include in the directory are the BWA index files generated for the reference genome.</p>
<h3 id="2-align-genome">2. Align genome<a class="headerlink" href="#2-align-genome" title="Permanent link">&para;</a></h3>
<p>Run the command below to map the raw sequencing data to the Homo sapiens (human) genome assembly GRCh38 (hg38). We are using the <a href="https://github.com/lh3/bwa">BWA-MEM</a> algorithms for mapping DNA sequences against large reference genomes. Note that we have already run the created the BWA index files by running the command <code class="language-bash highlight">bwa<span class="w"> </span>index<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta</code>.</p>
<p>Run BWA as follows, but first navigate to the scripts folder:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>bwa<span class="w"> </span>mem<span class="w"> </span>-M<span class="w"> </span>-t<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>-R<span class="w"> </span><span class="s2">&quot;@RG\\tID:SRR622461.7\\tSM:NA12878\\tLB:ERR194147\\tPL:ILLUMINA&quot;</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-12-3"><a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-12-4"><a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a>data/NA12878.chr20.region_1.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-12-5"><a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a>data/NA12878.chr20.region_2.fastq.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-12-6"><a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a>samtools<span class="w"> </span>view<span class="w"> </span>-b<span class="w"> </span>-h<span class="w"> </span>-o<span class="w"> </span>output/NA12878.bam<span class="w"> </span>-
</span></code></pre></div>
<p>There are two parts to the command here. The first part uses BWA to perform the alignment and the second part take the output from BWA and uses Samtools to convert the output to the BAM format.</p>
<p>At the end of this step you should have a file called <code class="language-bash highlight">NA12878.bam</code> in the <code class="language-bash highlight">output</code> directory.</p>
<h2 id="section-2-prepare-analysis-ready-reads">Section 2: Prepare analysis ready reads<a class="headerlink" href="#section-2-prepare-analysis-ready-reads" title="Permanent link">&para;</a></h2>
<h3 id="1-sort-sambam">1. Sort SAM/BAM<a class="headerlink" href="#1-sort-sambam" title="Permanent link">&para;</a></h3>
<p>The alignment file <code class="language-bash highlight">NA12878.bam</code> is not sorted. Before proceeding, we should sort the BAM file using the <a href="https://broadinstitute.github.io/picard/">Picard</a> tools.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>picard<span class="w"> </span>-Xmx7g<span class="w"> </span>SortSam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="nv">I</span><span class="o">=</span>output/NA12878.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-3"><a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="nv">O</span><span class="o">=</span>output/NA12878.sort.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-4"><a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="nv">VALIDATION_STRINGENCY</span><span class="o">=</span>LENIENT<span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-5"><a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="nv">SORT_ORDER</span><span class="o">=</span>coordinate<span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-6"><a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="nv">MAX_RECORDS_IN_RAM</span><span class="o">=</span><span class="m">3000000</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-13-7"><a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="nv">CREATE_INDEX</span><span class="o">=</span>True<span class="w">    </span>
</span></code></pre></div>
<p>The above command will create a coordinate sorted BAM file and an index (<code class="language-bash highlight">.bai</code>) file.</p>
<div class="admonition alignment statistics">
<p class="admonition-title">Alignment</p>
<p>Given we now have a sorted BAM file, we can now generate some useful statistics. To do so we can use the <code class="language-bash highlight">samtools<span class="w"> </span>flagstat</code> command. More details are available <a href="http://www.htslib.org/doc/samtools-flagstat.html">here</a>.</p>
</div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="c1"># lets go to the home directory</span>
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="nb">cd</span>
</span><span id="__span-14-3"><a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>
</span><span id="__span-14-4"><a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a>samtools<span class="w"> </span>flagstat<span class="w"> </span>output/NA12878.sort.bam
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="c1"># output</span>
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="m">2032568</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="k">in</span><span class="w"> </span>total<span class="w"> </span><span class="o">(</span>QC-passed<span class="w"> </span>reads<span class="w"> </span>+<span class="w"> </span>QC-failed<span class="w"> </span>reads<span class="o">)</span>
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="m">2030516</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary
</span><span id="__span-15-5"><a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a><span class="m">2052</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>secondary
</span><span id="__span-15-6"><a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a><span class="m">0</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>supplementary
</span><span id="__span-15-7"><a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a><span class="m">0</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>duplicates
</span><span id="__span-15-8"><a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a><span class="m">0</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary<span class="w"> </span>duplicates
</span><span id="__span-15-9"><a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a><span class="m">2032563</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>mapped<span class="w"> </span><span class="o">(</span><span class="m">100</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-15-10"><a id="__codelineno-15-10" name="__codelineno-15-10" href="#__codelineno-15-10"></a><span class="m">2030511</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary<span class="w"> </span>mapped<span class="w"> </span><span class="o">(</span><span class="m">100</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-15-11"><a id="__codelineno-15-11" name="__codelineno-15-11" href="#__codelineno-15-11"></a><span class="m">2030516</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>paired<span class="w"> </span><span class="k">in</span><span class="w"> </span>sequencing
</span><span id="__span-15-12"><a id="__codelineno-15-12" name="__codelineno-15-12" href="#__codelineno-15-12"></a><span class="m">1015258</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>read1
</span><span id="__span-15-13"><a id="__codelineno-15-13" name="__codelineno-15-13" href="#__codelineno-15-13"></a><span class="m">1015258</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>read2
</span><span id="__span-15-14"><a id="__codelineno-15-14" name="__codelineno-15-14" href="#__codelineno-15-14"></a><span class="m">2030298</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>properly<span class="w"> </span>paired<span class="w"> </span><span class="o">(</span><span class="m">99</span>.99%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-15-15"><a id="__codelineno-15-15" name="__codelineno-15-15" href="#__codelineno-15-15"></a><span class="m">2030510</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>itself<span class="w"> </span>and<span class="w"> </span>mate<span class="w"> </span>mapped
</span><span id="__span-15-16"><a id="__codelineno-15-16" name="__codelineno-15-16" href="#__codelineno-15-16"></a><span class="m">1</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>singletons<span class="w"> </span><span class="o">(</span><span class="m">0</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-15-17"><a id="__codelineno-15-17" name="__codelineno-15-17" href="#__codelineno-15-17"></a><span class="m">182</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>mate<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>a<span class="w"> </span>different<span class="w"> </span>chr
</span><span id="__span-15-18"><a id="__codelineno-15-18" name="__codelineno-15-18" href="#__codelineno-15-18"></a><span class="m">124</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>mate<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>a<span class="w"> </span>different<span class="w"> </span>chr<span class="w"> </span><span class="o">(</span>mapQ&gt;<span class="o">=</span><span class="m">5</span><span class="o">)</span>
</span></code></pre></div>
<h3 id="2-mark-duplicate-reads">2. Mark duplicate reads<a class="headerlink" href="#2-mark-duplicate-reads" title="Permanent link">&para;</a></h3>
<p>The aim of this step is to locate and tag duplicate reads in the BAM file. Duplicate reads can arise due to several reasons, for more details go to <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">MarkDuplicates</a>.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>picard<span class="w"> </span>-Xmx7g<span class="w"> </span>MarkDuplicates<span class="w"> </span><span class="se">\</span>
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="w">    </span><span class="nv">I</span><span class="o">=</span>output/NA12878.sort.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-16-3"><a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="w">    </span><span class="nv">O</span><span class="o">=</span>output/NA12878.sort.dup.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-16-4"><a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a><span class="w">    </span><span class="nv">METRICS_FILE</span><span class="o">=</span>output/marked_dup_metrics.txt
</span></code></pre></div>
<div class="admonition question">
<p class="admonition-title">Question: How many duplicate reads are in the duplicate marked BAM file?</p>
<details class="answer">
<summary>Answer</summary>
<p><div class="language-bash highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>samtools<span class="w"> </span>flagstat<span class="w"> </span>output/NA12878.sort.dup.bam
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="m">2032568</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="k">in</span><span class="w"> </span>total<span class="w"> </span><span class="o">(</span>QC-passed<span class="w"> </span>reads<span class="w"> </span>+<span class="w"> </span>QC-failed<span class="w"> </span>reads<span class="o">)</span>
</span><span id="__span-17-4"><a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a><span class="m">2030516</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary
</span><span id="__span-17-5"><a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a><span class="m">2052</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>secondary
</span><span id="__span-17-6"><a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a><span class="m">0</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>supplementary
</span><span id="__span-17-7"><a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a><span class="m">7207</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>duplicates
</span><span id="__span-17-8"><a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a><span class="m">7207</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary<span class="w"> </span>duplicates
</span><span id="__span-17-9"><a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a><span class="m">2032563</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>mapped<span class="w"> </span><span class="o">(</span><span class="m">100</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-17-10"><a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a><span class="m">2030511</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>primary<span class="w"> </span>mapped<span class="w"> </span><span class="o">(</span><span class="m">100</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-17-11"><a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a><span class="m">2030516</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>paired<span class="w"> </span><span class="k">in</span><span class="w"> </span>sequencing
</span><span id="__span-17-12"><a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a><span class="m">1015258</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>read1
</span><span id="__span-17-13"><a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a><span class="m">1015258</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>read2
</span><span id="__span-17-14"><a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a><span class="m">2030298</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>properly<span class="w"> </span>paired<span class="w"> </span><span class="o">(</span><span class="m">99</span>.99%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-17-15"><a id="__codelineno-17-15" name="__codelineno-17-15" href="#__codelineno-17-15"></a><span class="m">2030510</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>itself<span class="w"> </span>and<span class="w"> </span>mate<span class="w"> </span>mapped
</span><span id="__span-17-16"><a id="__codelineno-17-16" name="__codelineno-17-16" href="#__codelineno-17-16"></a><span class="m">1</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>singletons<span class="w"> </span><span class="o">(</span><span class="m">0</span>.00%<span class="w"> </span>:<span class="w"> </span>N/A<span class="o">)</span>
</span><span id="__span-17-17"><a id="__codelineno-17-17" name="__codelineno-17-17" href="#__codelineno-17-17"></a><span class="m">182</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>mate<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>a<span class="w"> </span>different<span class="w"> </span>chr
</span><span id="__span-17-18"><a id="__codelineno-17-18" name="__codelineno-17-18" href="#__codelineno-17-18"></a><span class="m">124</span><span class="w"> </span>+<span class="w"> </span><span class="m">0</span><span class="w"> </span>with<span class="w"> </span>mate<span class="w"> </span>mapped<span class="w"> </span>to<span class="w"> </span>a<span class="w"> </span>different<span class="w"> </span>chr<span class="w"> </span><span class="o">(</span>mapQ&gt;<span class="o">=</span><span class="m">5</span><span class="o">)</span>
</span></code></pre></div>
Looks like there are <strong>7207</strong> duplicate reads.</p>
</details>
</div>
<h3 id="3-base-quality-recalibration">3. Base quality recalibration<a class="headerlink" href="#3-base-quality-recalibration" title="Permanent link">&para;</a></h3>
<p>The last step of pre-processing mapped reads is the base quality score recalibration (BQSR) stage. The GATK tools detects systematic errors made by the sequencing machine while estimating the accuracy of each base. The systematic errors can be have various sources ranging from technical machine errors to the variability in the sequencing chemical reactions. The two step BQSR process applies machine learning to model the possible errors and adjust the base quality scores accordingly. More details <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890531-Base-Quality-Score-Recalibration-BQSR-">here</a>.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="c1"># lets go to the home directory again</span>
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="nb">cd</span>
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a>
</span><span id="__span-18-4"><a id="__codelineno-18-4" name="__codelineno-18-4" href="#__codelineno-18-4"></a><span class="c1"># step 1  - Build the model</span>
</span><span id="__span-18-5"><a id="__codelineno-18-5" name="__codelineno-18-5" href="#__codelineno-18-5"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>BaseRecalibrator<span class="w"> </span><span class="se">\</span>
</span><span id="__span-18-6"><a id="__codelineno-18-6" name="__codelineno-18-6" href="#__codelineno-18-6"></a><span class="w">    </span>-I<span class="w"> </span>output/NA12878.sort.dup.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-18-7"><a id="__codelineno-18-7" name="__codelineno-18-7" href="#__codelineno-18-7"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-18-8"><a id="__codelineno-18-8" name="__codelineno-18-8" href="#__codelineno-18-8"></a><span class="w">    </span>--known-sites<span class="w"> </span>reference/hg38/dbsnp_146.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-18-9"><a id="__codelineno-18-9" name="__codelineno-18-9" href="#__codelineno-18-9"></a><span class="w">    </span>-O<span class="w"> </span>output/recal_data.table
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a><span class="c1"># step 2: Apply the model to adjust the base quality scores</span>
</span><span id="__span-19-2"><a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>ApplyBQSR<span class="w"> </span><span class="se">\</span>
</span><span id="__span-19-3"><a id="__codelineno-19-3" name="__codelineno-19-3" href="#__codelineno-19-3"></a><span class="w">    </span>-I<span class="w"> </span>output/NA12878.sort.dup.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-19-4"><a id="__codelineno-19-4" name="__codelineno-19-4" href="#__codelineno-19-4"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-19-5"><a id="__codelineno-19-5" name="__codelineno-19-5" href="#__codelineno-19-5"></a><span class="w">    </span>--bqsr-recal-file<span class="w"> </span>output/recal_data.table<span class="w"> </span><span class="se">\</span>
</span><span id="__span-19-6"><a id="__codelineno-19-6" name="__codelineno-19-6" href="#__codelineno-19-6"></a><span class="w">    </span>-O<span class="w"> </span>output/NA12878.sort.dup.bqsr.bam
</span></code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In a workflow such as this it is a good practice to give the output files an appropriate name. In this case, we are appending the workflow step details to the filenames. For example, append <code>dup</code> after running the mark duplicates step.</p>
</div>
<p>We now have a pre-processed BAM file (<code class="language-bash highlight">NA12878.sort.dup.bqsr.bam</code>) ready for variant calling.</p>
<p>But before we proceed, let&rsquo;s take a detour and run some summary statistics of the alignment data and QC.</p>
<details class="example">
<summary><strong>BAM statistics and QC</strong> </summary>
<p>The commands below uses FastQC and Picard to generate QC metrics followed by multiQC tools then aggregating the data to produce an HTML report.    </p>
<p><div class="language-bash highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a><span class="c1"># FastQC</span>
</span><span id="__span-20-2"><a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a>fastqc<span class="w"> </span>data/NA12878.chr20.region_1.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-20-3"><a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a>data/NA12878.chr20.region_2.fastq.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-20-4"><a id="__codelineno-20-4" name="__codelineno-20-4" href="#__codelineno-20-4"></a>-o<span class="w"> </span>output/
</span><span id="__span-20-5"><a id="__codelineno-20-5" name="__codelineno-20-5" href="#__codelineno-20-5"></a>
</span><span id="__span-20-6"><a id="__codelineno-20-6" name="__codelineno-20-6" href="#__codelineno-20-6"></a><span class="c1"># CollectInsertSizeMetrics</span>
</span><span id="__span-20-7"><a id="__codelineno-20-7" name="__codelineno-20-7" href="#__codelineno-20-7"></a>picard<span class="w"> </span>CollectMultipleMetrics<span class="w"> </span><span class="se">\</span>
</span><span id="__span-20-8"><a id="__codelineno-20-8" name="__codelineno-20-8" href="#__codelineno-20-8"></a><span class="nv">R</span><span class="o">=</span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-20-9"><a id="__codelineno-20-9" name="__codelineno-20-9" href="#__codelineno-20-9"></a><span class="nv">I</span><span class="o">=</span>output/NA12878.sort.dup.bqsr.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-20-10"><a id="__codelineno-20-10" name="__codelineno-20-10" href="#__codelineno-20-10"></a><span class="nv">O</span><span class="o">=</span>output/NA12878.sort.dup.bqsr.CollectMultipleMetrics
</span><span id="__span-20-11"><a id="__codelineno-20-11" name="__codelineno-20-11" href="#__codelineno-20-11"></a>
</span><span id="__span-20-12"><a id="__codelineno-20-12" name="__codelineno-20-12" href="#__codelineno-20-12"></a><span class="c1"># MultiQC</span>
</span><span id="__span-20-13"><a id="__codelineno-20-13" name="__codelineno-20-13" href="#__codelineno-20-13"></a>multiqc<span class="w"> </span>output/.<span class="w"> </span>-o<span class="w"> </span>output/.<span class="w">    </span>
</span></code></pre></div>
We have precomputed this and the resulting MultiQC report is <a href="../files/multiqc_report.html" target="_blank">here</a>.</p>
</details>
<h2 id="section-3-variant-calling">Section 3: Variant calling<a class="headerlink" href="#section-3-variant-calling" title="Permanent link">&para;</a></h2>
<p>The next step in the GATK best practices workflow is to proceed with the variant calling.</p>
<p>There are a couple of workflows to call variants using GATK4. Here we will follow the Genomic Variant Call Format (GVCF) workflow which is more suited for scalable variant calling i.e. allows incremental addition of samples for joint genotyping.</p>
<h3 id="1-apply-haplotypecaller">1. Apply HaplotypeCaller<a class="headerlink" href="#1-apply-haplotypecaller" title="Permanent link">&para;</a></h3>
<p>HaplotypeCaller is the focal tool within GATK4 to simultaneously call germline SNVs and small Indels using local de-novo assembly of haplotype regions.</p>
<div class="admonition algorithm">
<p class="admonition-title">Algorithm</p>
<p>Briefly, the HaplotypeCaller works by:
1. Identify active regions or regions with evidence of variations.
2. Re-asssemble the active regions.
3. Re-align active region reads to the assembled regions to identify allele.
More details about the HaplotypeCaller can be found <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller">here</a>.</p>
</div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-2"><a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a><span class="w">    </span>-I<span class="w"> </span>output/NA12878.sort.dup.bqsr.bam<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-3"><a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-4"><a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a><span class="w">    </span>-ERC<span class="w"> </span>GVCF<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-5"><a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a><span class="w">    </span>-L<span class="w"> </span>chr20<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-6"><a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a><span class="w">    </span>-O<span class="w"> </span>output/NA12878.g.vcf.gz
</span></code></pre></div>
<p>The output of this step is a GVCF file. The format for the GVCF file is similar to a VCF file. The key difference is that the GVCF file contains records for each sequenced genomic coordinate. The <code class="language-bash highlight">--emit-ref-confidence</code> or <code class="language-bash highlight">-ERC</code> parameter lets you select a method to summarise confidence in the genomic site being homozygous-reference. The option <code class="language-bash highlight">-ERC<span class="w"> </span>GVCF</code> is more efficient and recommended for large samples and therefore more scalable.</p>
<h3 id="2-apply-combinegvcfs">2. Apply CombineGVCFs<a class="headerlink" href="#2-apply-combinegvcfs" title="Permanent link">&para;</a></h3>
<p>The CombineGVCFs tool is applied to combine multiple single sample GVCF files to merge these in to a single multi-sample GVCF file.</p>
<p>We have pre-processed two additional samples (NA12891 and NA12892) up to the HaplotypeCaller step (above). Lets first copy the gvcf files to the output directory.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="c1">#lets make sure that we are in the apropriate directory</span>
</span><span id="__span-22-2"><a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a><span class="nb">cd</span>
</span><span id="__span-22-3"><a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a>
</span><span id="__span-22-4"><a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a>cp<span class="w"> </span>/mnt/shared_data/NA12891.g.vcf.gz*<span class="w"> </span>output/.
</span><span id="__span-22-5"><a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a>cp<span class="w"> </span>/mnt/shared_data/NA12892.g.vcf.gz*<span class="w"> </span>output/.
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>CombineGVCFs<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-3"><a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a><span class="w">    </span>-V<span class="w"> </span>output/NA12878.g.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-4"><a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="w">    </span>-V<span class="w"> </span>output/NA12891.g.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-5"><a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a><span class="w">    </span>-V<span class="w"> </span>output/NA12892.g.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-6"><a id="__codelineno-23-6" name="__codelineno-23-6" href="#__codelineno-23-6"></a><span class="w">    </span>-L<span class="w"> </span>chr20<span class="w"> </span><span class="se">\</span>
</span><span id="__span-23-7"><a id="__codelineno-23-7" name="__codelineno-23-7" href="#__codelineno-23-7"></a><span class="w">    </span>-O<span class="w"> </span>output/cohort.g.vcf.gz
</span></code></pre></div>
<details class="example">
<summary>Lets look at the combined gvcf file</summary>
<p><div class="language-bash highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>less<span class="w"> </span>output/cohort.g.vcf.gz
</span></code></pre></div>
Work your way down to the variant records? How many samples do you see in the VCF file?
Hint: look at the header row.</p>
</details>
<p>Now that we have a merged GVCF file, we are ready to perform genotyping.</p>
<h3 id="3-apply-genotypegvcfs">3. Apply GenotypeGVCFs<a class="headerlink" href="#3-apply-genotypegvcfs" title="Permanent link">&para;</a></h3>
<p>GenotypeGVCFs</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>GenotypeGVCFs<span class="w"> </span><span class="se">\</span>
</span><span id="__span-25-2"><a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-25-3"><a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a><span class="w">    </span>-V<span class="w"> </span>output/cohort.g.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-25-4"><a id="__codelineno-25-4" name="__codelineno-25-4" href="#__codelineno-25-4"></a><span class="w">    </span>-L<span class="w"> </span>chr20<span class="w"> </span><span class="se">\</span>
</span><span id="__span-25-5"><a id="__codelineno-25-5" name="__codelineno-25-5" href="#__codelineno-25-5"></a><span class="w">    </span>-O<span class="w"> </span>output/output.vcf.gz
</span></code></pre></div>
<details class="information">
<summary>Information</summary>
<p>An alternative to CombineGVCFs is <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360057439331-GenomicsDBImport">GenomicsDBImport</a>, which is more efficient for sample numbers and stores the content in a a GenomicsDB data store. Therefore, CombineGVCFs could be slow and inefficient for more than a few samples. A possible work around is to split up the tasks per interval regions such as chromosomes.</p>
</details>
<details class="example">
<summary>Visualisations: VCF file</summary>
<p>Screenshot from output.vcf.gz
<img alt="fig1" src="../media/fig1.png" /></p>
</details>
<h2 id="section-4-filter-and-prepare-analysis-ready-variants">Section 4: Filter and prepare analysis ready variants<a class="headerlink" href="#section-4-filter-and-prepare-analysis-ready-variants" title="Permanent link">&para;</a></h2>
<h3 id="1-variant-quality-score-recalibration">1. Variant Quality Score Recalibration<a class="headerlink" href="#1-variant-quality-score-recalibration" title="Permanent link">&para;</a></h3>
<p>The raw VCF file from the previous step (<code class="language-bash highlight">output.vcf.gz</code>) contains 10467 variants. Not all of these are real, therefore, the aim of this step is to filter out artifacts or false positive variants. GATK has provided different workflows for variant filtering. Here we will walk through the Variant Quality Score Recalibration or the VQSR strategy. VQSR is a two step process (1) the first step builds a model that describes how variant metric or quality measures co-vary with the known variants in the training set. (2) The second step then ranks each variant according to the target sensitivity cutoff and applies a filter expression.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a><span class="c1">#Step 1 - VariantRecalibrator</span>
</span><span id="__span-26-2"><a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>VariantRecalibrator<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-3"><a id="__codelineno-26-3" name="__codelineno-26-3" href="#__codelineno-26-3"></a><span class="w">    </span>-V<span class="w"> </span>output/output.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-4"><a id="__codelineno-26-4" name="__codelineno-26-4" href="#__codelineno-26-4"></a><span class="w">    </span>--trust-all-polymorphic<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-5"><a id="__codelineno-26-5" name="__codelineno-26-5" href="#__codelineno-26-5"></a><span class="w">    </span>-mode<span class="w"> </span>SNP<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-6"><a id="__codelineno-26-6" name="__codelineno-26-6" href="#__codelineno-26-6"></a><span class="w">    </span>--max-gaussians<span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-7"><a id="__codelineno-26-7" name="__codelineno-26-7" href="#__codelineno-26-7"></a><span class="w">    </span>--resource:hapmap,known<span class="o">=</span>false,training<span class="o">=</span>true,truth<span class="o">=</span>true,prior<span class="o">=</span><span class="m">15</span><span class="w"> </span>reference/hg38/hapmap_3.3.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-8"><a id="__codelineno-26-8" name="__codelineno-26-8" href="#__codelineno-26-8"></a><span class="w">    </span>--resource:omni,known<span class="o">=</span>false,training<span class="o">=</span>true,truth<span class="o">=</span>true,prior<span class="o">=</span><span class="m">12</span><span class="w"> </span>reference/hg38/1000G_omni2.5.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-9"><a id="__codelineno-26-9" name="__codelineno-26-9" href="#__codelineno-26-9"></a><span class="w">    </span>--resource:1000G,known<span class="o">=</span>false,training<span class="o">=</span>true,truth<span class="o">=</span>false,prior<span class="o">=</span><span class="m">10</span><span class="w"> </span>reference/hg38/1000G_phase1.snps.high_confidence.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-10"><a id="__codelineno-26-10" name="__codelineno-26-10" href="#__codelineno-26-10"></a><span class="w">    </span>--resource:dbsnp,known<span class="o">=</span>true,training<span class="o">=</span>false,truth<span class="o">=</span>false,prior<span class="o">=</span><span class="m">7</span><span class="w"> </span>reference/hg38/dbsnp_138.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-11"><a id="__codelineno-26-11" name="__codelineno-26-11" href="#__codelineno-26-11"></a><span class="w">    </span>-an<span class="w"> </span>QD<span class="w"> </span>-an<span class="w"> </span>MQRankSum<span class="w"> </span>-an<span class="w"> </span>ReadPosRankSum<span class="w"> </span>-an<span class="w"> </span>FS<span class="w"> </span>-an<span class="w"> </span>MQ<span class="w"> </span>-an<span class="w"> </span>SOR<span class="w"> </span>-an<span class="w"> </span>DP<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-12"><a id="__codelineno-26-12" name="__codelineno-26-12" href="#__codelineno-26-12"></a><span class="w">    </span>-O<span class="w"> </span>output/cohort_snps.recal<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-13"><a id="__codelineno-26-13" name="__codelineno-26-13" href="#__codelineno-26-13"></a><span class="w">    </span>--tranches-file<span class="w"> </span>output/cohort_snps.tranches
</span><span id="__span-26-14"><a id="__codelineno-26-14" name="__codelineno-26-14" href="#__codelineno-26-14"></a>
</span><span id="__span-26-15"><a id="__codelineno-26-15" name="__codelineno-26-15" href="#__codelineno-26-15"></a><span class="c1">#Step 2 - ApplyVQSR</span>
</span><span id="__span-26-16"><a id="__codelineno-26-16" name="__codelineno-26-16" href="#__codelineno-26-16"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>ApplyVQSR<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-17"><a id="__codelineno-26-17" name="__codelineno-26-17" href="#__codelineno-26-17"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-18"><a id="__codelineno-26-18" name="__codelineno-26-18" href="#__codelineno-26-18"></a><span class="w">    </span>-V<span class="w"> </span>output/output.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-19"><a id="__codelineno-26-19" name="__codelineno-26-19" href="#__codelineno-26-19"></a><span class="w">    </span>-O<span class="w"> </span>output/output.vqsr.vcf<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-20"><a id="__codelineno-26-20" name="__codelineno-26-20" href="#__codelineno-26-20"></a><span class="w">    </span>--truth-sensitivity-filter-level<span class="w"> </span><span class="m">99</span>.0<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-21"><a id="__codelineno-26-21" name="__codelineno-26-21" href="#__codelineno-26-21"></a><span class="w">    </span>--tranches-file<span class="w"> </span>output/cohort_snps.tranches<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-22"><a id="__codelineno-26-22" name="__codelineno-26-22" href="#__codelineno-26-22"></a><span class="w">    </span>--recal-file<span class="w"> </span>output/cohort_snps.recal<span class="w"> </span><span class="se">\</span>
</span><span id="__span-26-23"><a id="__codelineno-26-23" name="__codelineno-26-23" href="#__codelineno-26-23"></a><span class="w">    </span>-mode<span class="w"> </span>SNP
</span></code></pre></div>
<div class="admonition countvariants">
<p class="admonition-title">Countvariants</p>
<p>There are number of ways to count the variants in a VCF file. A very straight forward way using the GATK4 tools is as follows:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>gatk<span class="w"> </span>CountVariants<span class="w"> </span>-V<span class="w"> </span>output/output.vqsr.vcf
</span></code></pre></div></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>Tool returned:
</span><span id="__span-28-2"><a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a>10467
</span></code></pre></div>
</div>
<p>There are several protocols for filtering VCF files. We have walked throught the VQSR strategy above and for other options please visit this <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering">link</a>.</p>
<details class="example">
<summary>Filtering strategy for a single sample VCF file</summary>
<p>Consider the following method to filter a single sample VCF file. Here we will go through the Convolutional Neural Net based protocol to annotate and filter the VCF file.</p>
<p>This is a two step process:</p>
<p>(i) CNNScoreVariants will annotate the variant with pre-computed single-sample derived model scores in the INFO field CNN_1D (the neural network performs convolutions over the reference sequence surrounding the variant and combines those features with a multilayer perceptron on the variant annotations).</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>CNNScoreVariants<span class="w">  </span><span class="se">\</span>
</span><span id="__span-29-2"><a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a><span class="w">   </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-29-3"><a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a><span class="w">   </span>-V<span class="w"> </span>output/output.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-29-4"><a id="__codelineno-29-4" name="__codelineno-29-4" href="#__codelineno-29-4"></a><span class="w">   </span>-O<span class="w"> </span>output/output.cnns.vcf
</span></code></pre></div>
<p>(ii) FilterVariantTranches takes as input the percent sensitivities (0-100) to known sites to apply the filter. Variants with scores higher than for e.g. 99<sup>th</sup> percentile of variants in the resources pass through the filter and will have <code>PASS</code> in the filter. Others will have a filter values like &lsquo;CNN_1D_INDEL_Tranche_99.40_100.00&rsquo; or &lsquo;CNN_1D_SNP_Tranche_99.95_100.00&rsquo;.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>FilterVariantTranches<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-2"><a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a><span class="w">    </span>-V<span class="w"> </span>output/output.cnns.vcf<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-3"><a id="__codelineno-30-3" name="__codelineno-30-3" href="#__codelineno-30-3"></a><span class="w">    </span>--resource<span class="w"> </span>reference/hg38/hapmap_3.3.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-4"><a id="__codelineno-30-4" name="__codelineno-30-4" href="#__codelineno-30-4"></a><span class="w">    </span>--resource<span class="w"> </span>reference/hg38/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-5"><a id="__codelineno-30-5" name="__codelineno-30-5" href="#__codelineno-30-5"></a><span class="w">    </span>--info-key<span class="w"> </span>CNN_1D<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-6"><a id="__codelineno-30-6" name="__codelineno-30-6" href="#__codelineno-30-6"></a><span class="w">    </span>--snp-tranche<span class="w"> </span><span class="m">99</span>.95<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-7"><a id="__codelineno-30-7" name="__codelineno-30-7" href="#__codelineno-30-7"></a><span class="w">    </span>--indel-tranche<span class="w"> </span><span class="m">99</span>.4<span class="w"> </span><span class="se">\</span>
</span><span id="__span-30-8"><a id="__codelineno-30-8" name="__codelineno-30-8" href="#__codelineno-30-8"></a><span class="w">    </span>-O<span class="w"> </span>output/output.cnns.cnnfilter.vcf
</span></code></pre></div>
</details>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>BCFtools is a useful tool to manipulate, filter and query VCF files. More details from <a href="https://samtools.github.io/bcftools/">BCFtools</a>. BCFtools can be combined with linux command line tools as well to summarise data. For example, the command below can used extract and print the &lsquo;FILTER&rsquo; column from the VCF file.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a><span class="w">    </span>bcftools<span class="w"> </span>query<span class="w"> </span>-f<span class="s1">&#39;%FILTER\n&#39;</span><span class="w"> </span>output/output.vqsr.vcf.gz
</span></code></pre></div>
</div>
<h3 id="2-additional-filtering">2. Additional filtering<a class="headerlink" href="#2-additional-filtering" title="Permanent link">&para;</a></h3>
<p>The VariantFiltration tools is designed for hard-filtering variant calls based on custom quality criteria such as sequencing depth, mapping quality etc. The two parameters are the filter-name and filter-expression. The parameter filter-name is the name of the filter to be used in the FILTER column if the expression in filter-expression is true. In the example below, if the sequencing depth at the variant site (VCF field DP) is less than 10, the FILTER field will be populated with the value &lsquo;Low_depth10&rsquo;. Users can add multiple filter expression/name combinations.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>gatk<span class="w"> </span>--java-options<span class="w"> </span><span class="s2">&quot;-Xmx7g&quot;</span><span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\</span>
</span><span id="__span-32-2"><a id="__codelineno-32-2" name="__codelineno-32-2" href="#__codelineno-32-2"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-32-3"><a id="__codelineno-32-3" name="__codelineno-32-3" href="#__codelineno-32-3"></a><span class="w">    </span>-V<span class="w"> </span>output/output.vqsr.vcf<span class="w"> </span><span class="se">\</span>
</span><span id="__span-32-4"><a id="__codelineno-32-4" name="__codelineno-32-4" href="#__codelineno-32-4"></a><span class="w">    </span>-O<span class="w"> </span>output/output.vqsr.varfilter.vcf<span class="w"> </span><span class="se">\</span>
</span><span id="__span-32-5"><a id="__codelineno-32-5" name="__codelineno-32-5" href="#__codelineno-32-5"></a><span class="w">    </span>--filter-name<span class="w"> </span><span class="s2">&quot;Low_depth10&quot;</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-32-6"><a id="__codelineno-32-6" name="__codelineno-32-6" href="#__codelineno-32-6"></a><span class="w">    </span>--filter-expression<span class="w"> </span><span class="s2">&quot;DP &lt; 10&quot;</span>
</span></code></pre></div>
<div class="admonition question">
<p class="admonition-title">Question: How many variants have a low sequencing depth (DP&lt;10) in the file output.vqsr.varfilter.vcf.</p>
<details class="answer">
<summary>Answer</summary>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a><span class="w">    </span>bcftools<span class="w"> </span>query<span class="w"> </span>-f<span class="s1">&#39;%FILTER\n&#39;</span><span class="w"> </span>output/output.vqsr.varfilter.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code><span id="__span-34-1"><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>  6 Low_depth10
</span><span id="__span-34-2"><a id="__codelineno-34-2" name="__codelineno-34-2" href="#__codelineno-34-2"></a>  2 Low_depth10;VQSRTrancheSNP99.00to99.90
</span><span id="__span-34-3"><a id="__codelineno-34-3" name="__codelineno-34-3" href="#__codelineno-34-3"></a>  9 Low_depth10;VQSRTrancheSNP99.90to100.00
</span><span id="__span-34-4"><a id="__codelineno-34-4" name="__codelineno-34-4" href="#__codelineno-34-4"></a>  9064 PASS
</span><span id="__span-34-5"><a id="__codelineno-34-5" name="__codelineno-34-5" href="#__codelineno-34-5"></a>  1278 VQSRTrancheSNP99.00to99.90
</span><span id="__span-34-6"><a id="__codelineno-34-6" name="__codelineno-34-6" href="#__codelineno-34-6"></a>  108 VQSRTrancheSNP99.90to100.00
</span></code></pre></div>
</details>
</div>
<h3 id="3-final-analysis-ready-vcf-file">3. Final analysis ready VCF file<a class="headerlink" href="#3-final-analysis-ready-vcf-file" title="Permanent link">&para;</a></h3>
<p>Given we have a filter annotated VCF files (), we can now create an analysis ready VCF file.</p>
<div class="admonition question">
<p class="admonition-title">Question: Create a VCF file called <code>output/output.vqsr.varfilter.pass.vcf.gz</code> that contains only PASS variants? The input VCF file is <code>output/output.vqsr.varfilter.vcf</code>.&rdquo; Hint: try using the Bcftools application.</p>
<details class="answer">
<summary>Answer</summary>
<p>Use the bcftools to filter PASS variants.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-35-1"><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a>bcftools<span class="w"> </span>view<span class="w"> </span>-f<span class="w"> </span><span class="s1">&#39;PASS,.&#39;</span><span class="w"> </span>-O<span class="w"> </span>vcf<span class="w"> </span>-o<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf<span class="w"> </span>output/output.vqsr.varfilter.vcf<span class="w">            </span>
</span></code></pre></div></p>
<p>We have now created an analysis ready version of the VCF file. It is also a good practice to compress and index the file.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-36-1"><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a>bgzip<span class="w"> </span>-c<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf<span class="w"> </span>&gt;<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf.gz
</span><span id="__span-36-2"><a id="__codelineno-36-2" name="__codelineno-36-2" href="#__codelineno-36-2"></a>tabix<span class="w"> </span>-p<span class="w"> </span>vcf<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf.gz
</span></code></pre></div>
</details>
</div>
<h2 id="section-5-exporting-variant-data-and-visualisation">Section 5: Exporting variant data and visualisation<a class="headerlink" href="#section-5-exporting-variant-data-and-visualisation" title="Permanent link">&para;</a></h2>
<p>VCF files, although in tabular format, are not user friendly. We will go through a couple of ways to share share and visualise variant data. This is important for downstream analysis as well as sharing data. First, we will convert the VCF file in to a TSV file (ready for Excel for example) in a manner where we extract data fields of interest.</p>
<h3 id="1-variantstotable">1. VariantsToTable<a class="headerlink" href="#1-variantstotable" title="Permanent link">&para;</a></h3>
<p>This GATK4 tool extracts fields of interest from each record in a VCF file. <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360056968292-VariantsToTable">VariantsToTable</a> can extract field from both the INFO and FORMAT columns in the VCF file.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>VariantsToTable, by default, only extracts PASS or . (no filtering applied) variants. Use the <code class="language-bash highlight">--show-filtered</code> parameter to show all variants.</p>
</div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-37-1"><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a>gatk<span class="w"> </span>VariantsToTable<span class="w"> </span><span class="se">\</span>
</span><span id="__span-37-2"><a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a><span class="w">    </span>-R<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-37-3"><a id="__codelineno-37-3" name="__codelineno-37-3" href="#__codelineno-37-3"></a><span class="w">    </span>-V<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-37-4"><a id="__codelineno-37-4" name="__codelineno-37-4" href="#__codelineno-37-4"></a><span class="w">    </span>-F<span class="w"> </span>CHROM<span class="w"> </span>-F<span class="w"> </span>POS<span class="w"> </span>-F<span class="w"> </span>FILTER<span class="w"> </span>-F<span class="w"> </span>TYPE<span class="w"> </span>-GF<span class="w"> </span>AD<span class="w"> </span>-GF<span class="w"> </span>DP<span class="w"> </span><span class="se">\</span>
</span><span id="__span-37-5"><a id="__codelineno-37-5" name="__codelineno-37-5" href="#__codelineno-37-5"></a><span class="w">    </span>--show-filtered<span class="w"> </span><span class="se">\</span>
</span><span id="__span-37-6"><a id="__codelineno-37-6" name="__codelineno-37-6" href="#__codelineno-37-6"></a><span class="w">    </span>-O<span class="w"> </span>output/output.vqsr.varfilter.pass.tsv
</span></code></pre></div>
<h3 id="2-html-report">2. HTML report<a class="headerlink" href="#2-html-report" title="Permanent link">&para;</a></h3>
<p>Another useful method for sharing data is an interactive HTML file. This is suited for sharing a smaller subset of variants along with sequencing data. Here we will go through a simple example using the <a href="https://github.com/brentp/jigv">jigv</a> tool.</p>
<p><img alt="fig2" src="../media/fig2.png" /></p>
<p>We will start with creating a subset of variants to report.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-38-1"><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a>bcftools<span class="w"> </span>view<span class="w"> </span>output/output.vqsr.varfilter.pass.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-38-2"><a id="__codelineno-38-2" name="__codelineno-38-2" href="#__codelineno-38-2"></a>chr20:3822018-3999324<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-38-3"><a id="__codelineno-38-3" name="__codelineno-38-3" href="#__codelineno-38-3"></a>bgzip<span class="w"> </span>-c<span class="w"> </span>&gt;<span class="w"> </span>output/subset.vcf.gz
</span><span id="__span-38-4"><a id="__codelineno-38-4" name="__codelineno-38-4" href="#__codelineno-38-4"></a>
</span><span id="__span-38-5"><a id="__codelineno-38-5" name="__codelineno-38-5" href="#__codelineno-38-5"></a>tabix<span class="w"> </span>-p<span class="w"> </span>vcf<span class="w"> </span>output/subset.vcf.gz
</span></code></pre></div>
<p>Now, we will call the jigv tool command to generate the report.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-39-1"><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a>jigv<span class="w"> </span>--sample<span class="w"> </span>NA12878<span class="w"> </span><span class="se">\</span>
</span><span id="__span-39-2"><a id="__codelineno-39-2" name="__codelineno-39-2" href="#__codelineno-39-2"></a>--sites<span class="w"> </span>output/subset.vcf.gz<span class="w"> </span><span class="se">\</span>
</span><span id="__span-39-3"><a id="__codelineno-39-3" name="__codelineno-39-3" href="#__codelineno-39-3"></a>--fasta<span class="w"> </span>reference/hg38/Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
</span><span id="__span-39-4"><a id="__codelineno-39-4" name="__codelineno-39-4" href="#__codelineno-39-4"></a>output/NA12878.sort.dup.bqsr.bam<span class="w"> </span>&gt;<span class="w"> </span>output/NA12878.jigv.html
</span></code></pre></div></p>
<p>Here is an example <a href="../files/NA12878.html" target="_blank">report</a> we created earlier.</p>
<!-- ```bash
jigv --sample NA12878 --ped family.ped --sites output/subset.vcf.gz --fasta reference/hg38/Homo_sapiens_assembly38.fasta output/*.sort.dup.bqsr.bam > output/NA12878.jigv.html
``` -->












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            <a href="https://www.melbournebioinformatics.org.au/">Melbourne Bioinformatics â€“ The University of Melbourne Â©</a>
          </div>
        
        Powered by <a href="https://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">Material for MkDocs</a>.
      </div>
      
        
<div class="md-social">
  
</div>
      
    </div>
  </div>
</footer>

      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": ["content.code.copy", "content.code.annotate"], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../../../javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>